<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Bioinformatics Introductory Course - Master Molecular Biotechnology</title>
  <meta name="description" content="<p>RMarkdowns for the Data Analysis introduction course (SoSe2024), MoBi MA</p>" />
  <meta name="generator" content="bookdown #bookdown:version# and GitBook 2.6.7" />

  <meta property="og:title" content="Bioinformatics Introductory Course - Master Molecular Biotechnology" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="<p>RMarkdowns for the Data Analysis introduction course (SoSe2024), MoBi MA</p>" />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Bioinformatics Introductory Course - Master Molecular Biotechnology" />
  
  <meta name="twitter:description" content="<p>RMarkdowns for the Data Analysis introduction course (SoSe2024), MoBi MA</p>" />
  

<meta name="author" content="Biomedical Genomics group, IPMB" />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<!--bookdown:link_prev-->
<!--bookdown:link_next-->
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



<!--bookdown:title:start-->
<div id="header">
<h1 class="title">Bioinformatics Introductory Course - Master Molecular Biotechnology</h1>
<p class="author"><em>Biomedical Genomics group, IPMB</em></p>
</div>
<!--bookdown:title:end-->

<!--bookdown:toc:start-->
  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">
<!--bookdown:toc2:start-->
<ul>
<li><a href="#introduction" id="toc-introduction">Introduction</a>
<ul>
<li><a href="#where-to-start" id="toc-where-to-start">Where to start</a>
<ul>
<li><a href="#log-into-the-server" id="toc-log-into-the-server">Log into the server</a></li>
<li><a href="#installing-and-configuring-cyberduck" id="toc-installing-and-configuring-cyberduck">Installing and configuring Cyberduck</a></li>
<li><a href="#installing-r-and-rstudio" id="toc-installing-r-and-rstudio">Installing R and RStudio</a></li>
<li><a href="#courses" id="toc-courses">Courses</a></li>
</ul></li>
<li><a href="#good-practice" id="toc-good-practice">Good practice</a></li>
<li><a href="#r-markdown" id="toc-r-markdown">R Markdown</a></li>
<li><a href="#r-basics" id="toc-r-basics">R Basics</a>
<ul>
<li><a href="#variable-assignment-in-r" id="toc-variable-assignment-in-r">Variable assignment in R</a></li>
<li><a href="#classes" id="toc-classes">Classes</a></li>
<li><a href="#logical-operations" id="toc-logical-operations">Logical operations</a></li>
<li><a href="#vectors" id="toc-vectors">Vectors</a></li>
<li><a href="#matrixes" id="toc-matrixes">Matrixes</a></li>
<li><a href="#dataframes" id="toc-dataframes">Dataframes</a></li>
<li><a href="#apply-sapply-lapply" id="toc-apply-sapply-lapply">Apply, sapply, lapply</a></li>
<li><a href="#for-loops" id="toc-for-loops">For Loops</a></li>
<li><a href="#if-statements" id="toc-if-statements">If Statements</a></li>
</ul></li>
<li><a href="#data-wrangling-with-tidyverse" id="toc-data-wrangling-with-tidyverse">Data wrangling with tidyverse</a>
<ul>
<li><a href="#introducing-tidyverse" id="toc-introducing-tidyverse">Introducing tidyverse</a></li>
<li><a href="#filter-based-on-conditions" id="toc-filter-based-on-conditions">Filter based on conditions</a></li>
<li><a href="#select-specific-columns" id="toc-select-specific-columns">Select specific columns</a></li>
<li><a href="#operations-on-groups" id="toc-operations-on-groups">Operations on groups</a></li>
<li><a href="#reshaping-data" id="toc-reshaping-data">Reshaping data</a></li>
</ul></li>
<li><a href="#plotting-using-ggplot2" id="toc-plotting-using-ggplot2">Plotting using ggplot2</a>
<ul>
<li><a href="#scatter-plot" id="toc-scatter-plot">Scatter plot</a></li>
<li><a href="#histogram" id="toc-histogram">Histogram</a></li>
<li><a href="#boxplot" id="toc-boxplot">Boxplot</a></li>
</ul></li>
<li><a href="#common-errors" id="toc-common-errors">Common errors</a></li>
<li><a href="#introductory-exercises" id="toc-introductory-exercises">Introductory exercises</a>
<ul>
<li><a href="#exercise-1-vectors" id="toc-exercise-1-vectors">Exercise 1: Vectors</a></li>
<li><a href="#exercise-2-matrices" id="toc-exercise-2-matrices">Exercise 2: Matrices</a></li>
<li><a href="#exercise-3-data-frames" id="toc-exercise-3-data-frames">Exercise 3: Data Frames</a></li>
</ul></li>
</ul></li>
<li><a href="#lecture-files" id="toc-lecture-files">Lecture files</a>
<ul>
<li><a href="#week-1---data-analysis-statistics" id="toc-week-1---data-analysis-statistics">Week 1 - data analysis &amp; statistics</a></li>
<li><a href="#week-2---sequence-analysis" id="toc-week-2---sequence-analysis">Week 2 - sequence analysis</a></li>
</ul></li>
<li><a href="#week-1-data-analysis" id="toc-week-1-data-analysis">Week 1 : data analysis</a>
<ul>
<li><a href="#day-1-descriptive-statistics-and-data-types" id="toc-day-1-descriptive-statistics-and-data-types">Day 1: Descriptive statistics and data types</a>
<ul>
<li><a href="#data-features-and-where-to-find-them" id="toc-data-features-and-where-to-find-them">Data features and where to find them</a></li>
<li><a href="#data-cleaning" id="toc-data-cleaning">Data cleaning</a></li>
<li><a href="#visualizing-data-distribution" id="toc-visualizing-data-distribution">Visualizing data distribution</a></li>
<li><a href="#correlation" id="toc-correlation">Correlation</a></li>
<li><a href="#exercises" id="toc-exercises">EXERCISES</a></li>
</ul></li>
<li><a href="#day-2-dimensionality-reduction-and-unsupervised-learning" id="toc-day-2-dimensionality-reduction-and-unsupervised-learning">Day 2: Dimensionality reduction and unsupervised learning</a>
<ul>
<li><a href="#preparing-the-data" id="toc-preparing-the-data">Preparing the data</a></li>
<li><a href="#k-means-clustering" id="toc-k-means-clustering">k-means clustering</a></li>
<li><a href="#hierarchical-clustering" id="toc-hierarchical-clustering">Hierarchical clustering</a></li>
<li><a href="#principal-component-analysis" id="toc-principal-component-analysis">Principal component analysis</a></li>
<li><a href="#exercises-1" id="toc-exercises-1">EXERCISES</a></li>
</ul></li>
<li><a href="#day-3-probability-distributions" id="toc-day-3-probability-distributions">Day 3: Probability distributions</a>
<ul>
<li><a href="#probability-distributions" id="toc-probability-distributions">Probability distributions</a></li>
<li><a href="#confidence-interval" id="toc-confidence-interval">Confidence interval</a></li>
<li><a href="#exercises-2" id="toc-exercises-2">EXERCISES</a></li>
</ul></li>
<li><a href="#day-4-hypothesis-testing" id="toc-day-4-hypothesis-testing">Day 4: Hypothesis testing</a>
<ul>
<li><a href="#tests-of-mean" id="toc-tests-of-mean">Tests of mean</a></li>
<li><a href="#proportion-tests" id="toc-proportion-tests">Proportion tests</a></li>
<li><a href="#exercises-3" id="toc-exercises-3">EXERCISES</a></li>
</ul></li>
<li><a href="#day-5-multiple-testing-and-regression" id="toc-day-5-multiple-testing-and-regression">Day 5: Multiple testing and regression</a>
<ul>
<li><a href="#multiple-testing" id="toc-multiple-testing">Multiple testing</a></li>
<li><a href="#regression" id="toc-regression">Regression</a></li>
</ul></li>
</ul></li>
<li><a href="#week-2-sequence-analysis" id="toc-week-2-sequence-analysis">Week 2 : sequence analysis</a>
<ul>
<li><a href="#day-1-annotathon-identifying-orfs" id="toc-day-1-annotathon-identifying-orfs">Day 1: Annotathon! Identifying ORFs</a>
<ul>
<li><a href="#getting-ready" id="toc-getting-ready">Getting ready</a></li>
<li><a href="#getting-started" id="toc-getting-started">Getting started</a></li>
<li><a href="#finding-orfs" id="toc-finding-orfs">Finding ORFs</a></li>
<li><a href="#conserved-protein-domains" id="toc-conserved-protein-domains">Conserved protein domains</a></li>
</ul></li>
<li><a href="#day-2-annotathon-alignments" id="toc-day-2-annotathon-alignments">Day 2: Annotathon! Alignments</a>
<ul>
<li><a href="#blast-homolog-search" id="toc-blast-homolog-search">BLAST homolog search</a></li>
<li><a href="#blast-taxonomic-report" id="toc-blast-taxonomic-report">BLAST Taxonomic report</a></li>
<li><a href="#selecting-sequences-for-multiple-alignment" id="toc-selecting-sequences-for-multiple-alignment">Selecting sequences for multiple alignment</a></li>
</ul></li>
<li><a href="#day-3-annotathon-phylogenetic-trees-and-conclusion" id="toc-day-3-annotathon-phylogenetic-trees-and-conclusion">Day 3: Annotathon! Phylogenetic trees and conclusion</a>
<ul>
<li><a href="#multiple-alignments" id="toc-multiple-alignments">Multiple alignments</a></li>
<li><a href="#phylogeny" id="toc-phylogeny">Phylogeny</a></li>
<li><a href="#conclusion" id="toc-conclusion">Conclusion</a></li>
</ul></li>
<li><a href="#day-4-data-formats-and-where-to-find-them" id="toc-day-4-data-formats-and-where-to-find-them">Day 4: Data formats and where to find them</a>
<ul>
<li><a href="#meet-igv" id="toc-meet-igv">Meet IGV</a></li>
<li><a href="#identifying-formats" id="toc-identifying-formats">Identifying formats</a></li>
<li><a href="#a-practical-guide-to-geo" id="toc-a-practical-guide-to-geo">A practical guide to GEO</a></li>
<li><a href="#exercises-4" id="toc-exercises-4">EXERCISES</a></li>
</ul></li>
<li><a href="#day-5-rna-seq---from-fastq-to-count-matrix" id="toc-day-5-rna-seq---from-fastq-to-count-matrix">Day 5: RNA-seq - from FASTQ to count matrix</a>
<ul>
<li><a href="#connecting-to-the-virtual-machine" id="toc-connecting-to-the-virtual-machine">Connecting to the virtual machine</a></li>
<li><a href="#quality-control" id="toc-quality-control">Quality control</a></li>
<li><a href="#alignment" id="toc-alignment">Alignment</a></li>
<li><a href="#quantification" id="toc-quantification">Quantification</a></li>
<li><a href="#visualization-2" id="toc-visualization-2">Visualization</a></li>
<li><a href="#excercises" id="toc-excercises">Excercises</a></li>
</ul></li>
</ul></li>
<li><a href="#project-cancer-hallmark-and-metabolic-pathways-in-cancer" id="toc-project-cancer-hallmark-and-metabolic-pathways-in-cancer">Project: Cancer Hallmark and Metabolic Pathways in Cancer</a>
<ul>
<li><a href="#introduction-1" id="toc-introduction-1">Introduction</a></li>
<li><a href="#objectives-and-workplan" id="toc-objectives-and-workplan">Objectives and workplan</a></li>
<li><a href="#description-of-datasets" id="toc-description-of-datasets">Description of datasets</a></li>
<li><a href="#literature-review" id="toc-literature-review">Literature review</a></li>
<li><a href="#how-to-structure-your-project" id="toc-how-to-structure-your-project">How to structure your project</a>
<ul>
<li><a href="#project" id="toc-project">Project</a></li>
</ul></li>
</ul></li>
<li><a href="#project-report-guide" id="toc-project-report-guide">Project report guide</a>
<ul>
<li><a href="#knitting" id="toc-knitting">Knitting</a></li>
<li><a href="#formatting" id="toc-formatting">Formatting</a>
<ul>
<li><a href="#new-pages-and-paragraphs" id="toc-new-pages-and-paragraphs">New pages and paragraphs</a></li>
</ul></li>
<li><a href="#chunk-options" id="toc-chunk-options">Chunk options</a></li>
<li><a href="#external-images" id="toc-external-images">External images</a></li>
<li><a href="#references" id="toc-references">References</a>
<ul>
<li><a href="#citations" id="toc-citations">Citations</a></li>
<li><a href="#footnotes" id="toc-footnotes">Footnotes</a></li>
</ul></li>
</ul></li>
</ul>
<!--bookdown:toc2:end-->
      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Bioinformatics Introductory Course - Master Molecular Biotechnology</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<!--bookdown:toc:end-->
<!--bookdown:body:start-->
<p><img src="logos_IPMB_final_202309.jpg" /></p>
<div id="introduction" class="section level1 unnumbered hasAnchor">
<h1 class="unnumbered hasAnchor">Introduction<a href="#introduction" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>R is a powerful programming language for the analysis of data.
R is very versatile, it is free, and very easy to understand.
That is because, unlike other programming languages, R is quite “human-friendly” and dynamic, thus it does not require compiling to be “machine-readable”.
All you have to do is type R commands in the console or script.</p>
<p>Through this course, you will learn:</p>
<blockquote>
<ul>
<li>Basic R programming for data analysis (<em>before we start</em>, go though the sections present in this Introduction)</li>
<li>How to use R to work on different datasets and apply statistics (<em>Week One</em>)</li>
<li>How to annotate, analyse, obtain, pre-process, and visualize genomic data (<em>Week Two</em>)</li>
<li>How to perform RNA-seq analysis (<em>Week Three</em>)</li>
</ul>
</blockquote>
<div id="where-to-start" class="section level2 unnumbered hasAnchor">
<h2 class="unnumbered hasAnchor">Where to start<a href="#where-to-start" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="log-into-the-server" class="section level3 unnumbered hasAnchor">
<h3 class="unnumbered hasAnchor">Log into the server<a href="#log-into-the-server" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We will use an <strong>RStudio Server</strong> running on a virtual machine.
Here are the instructions to connect:</p>
<p>You can find the list of students <a href="https://docs.google.com/spreadsheets/d/1eMCr4b6UcOPIG8tJkfoEkMoZzJycWJPMYN5nMZWiU48/edit?usp=sharing">here</a>.
Check your name, your user name on which you will connect.</p>
<div id="if-you-are-running-machine-with-linux-or-macos" class="section level4 unnumbered hasAnchor">
<h4 class="unnumbered hasAnchor">if you are running machine with Linux or MacOS<a href="#if-you-are-running-machine-with-linux-or-macos" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<ul>
<li>open a terminal and type the following commang:
<ul>
<li>GK1: <code>ssh &lt;USER&gt;@194.94.113.18 -p 30210 -L 8787:127.0.0.1:8787</code>; replace <code>&lt;USER&gt;</code> with your user name from the Google Sheet</li>
<li>GK2: <code>ssh &lt;USER&gt;@194.94.113.18 -p 30108 -L 8787:127.0.0.1:8787</code>; replace <code>&lt;USER&gt;</code> with your user name from the Google Sheet</li>
</ul></li>
<li>in a browser, go to <a href="http://127.0.0.1:8787">this url</a></li>
<li>log in with your user name and the password that will be provided to you.</li>
</ul>
<p>You can now start working in RStudio!</p>
</div>
<div id="if-you-are-running-machine-with-windows" class="section level4 unnumbered hasAnchor">
<h4 class="unnumbered hasAnchor">if you are running machine with Windows<a href="#if-you-are-running-machine-with-windows" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Here, the user should install <a href="https://www.putty.org/">Putty</a>.
In the GUI, the following settings are needed:</p>
<ul>
<li>In the main panel:
<ul>
<li>Host name (or IP address): <USER>@194.94.113.18</li>
<li>Port: 30210 (GK1) or 30108 (GK2)</li>
<li>Connection type: SSH</li>
</ul></li>
<li>In the panel “Connection / SSH / Tunnel”:
<ul>
<li>Source port: 8787</li>
<li>Destination port: 127.0.0.1:8787</li>
<li>Check options “Local” and “Auto”</li>
</ul></li>
<li>Then click “Add”</li>
<li>Now click “Open”</li>
<li>in a browser, go to <a href="http://127.0.0.1:8787">this url</a></li>
<li>log in with your user name and the password that will be provided to you.</li>
</ul>
<p>You can now start working in RStudio!</p>
</div>
</div>
<div id="installing-and-configuring-cyberduck" class="section level3 unnumbered hasAnchor">
<h3 class="unnumbered hasAnchor">Installing and configuring Cyberduck<a href="#installing-and-configuring-cyberduck" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>In order to access remote files (for example opening a pdf file), you will need to install an additional tool, called cyberduck (<a href="https://cyberduck.io/" class="uri">https://cyberduck.io/</a>).</p>
<ul>
<li>Install Cyberduck according to your operating system</li>
<li>Once installed, open Cyberduck</li>
<li>On the top left, click on Neue Verbindung(or new connection)</li>
</ul>
<p>In the new window, make the following changes:</p>
<ul>
<li>select sftp in the dropdown menu</li>
<li>in the server box, type 194.94.113.18</li>
<li>in the port box, type 30210 (GK1) or 30108 (GK2)</li>
<li>in Benutzername (or Username), type your username (like user1)</li>
<li>password box: (ask for it!)</li>
</ul>
<p>Click on Verbinden (or Connect)</p>
</div>
<div id="installing-r-and-rstudio" class="section level3 unnumbered hasAnchor">
<h3 class="unnumbered hasAnchor">Installing R and RStudio<a href="#installing-r-and-rstudio" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Alternatively, you can also choose to install both R and RStudio.
Rstudio is the most commonly used IDE (or Integrated Development Environment) for R, and it will <strong>use the R version installed on your computer</strong>.
RStudio is meant to make R programming quite visual and way easier for you as a beginner.</p>
<p>Start by installing <a href="https://cran.r-project.org/">R</a> and then <a href="https://posit.co/download/rstudio-desktop/">RStudio</a>.</p>
<p>RStudio is composed of 4 main windows/panels.<br />
The <strong>Editor</strong>, the <strong>Environment</strong>, the <strong>Console</strong>, and the <strong>Files</strong> window/panel.
Order can vary.</p>
<div class="float">
<img src="figures/RstudioLayout.png" alt="In Editor (1), you can find the area where you usually write most code, like a Rscript or a Rmarkdown. Environment (2) shows you your variables, history, among others. Here, you can see any variable you define in your session. You can use this to do a very basic inspection of objects you create or import. Console (3) is where all your code gets processed when you run it. The console remembers (which is why when you click arrow up or down, it shows you commands you ran before), but it does not keep! Therefore, ALWAYS write the meaningful code lines in a script/markdown (1). Lastly, 4 can be used for browsing documents (Files), see the plots you produce (Plots), searching/managing packages (Packages), or find help on functions (Help)." />
<div class="figcaption">In <strong>Editor (1)</strong>, you can find the area where you usually write most code, like a Rscript or a Rmarkdown. <strong>Environment (2)</strong> shows you your variables, history, among others. Here, you can see any variable you define in your session. You can use this to do a very basic inspection of objects you create or import. <strong>Console (3)</strong> is where all your code gets processed when you run it. The console remembers (which is why when you click arrow up or down, it shows you commands you ran before), but it does not keep! Therefore, ALWAYS write the meaningful code lines in a script/markdown (1). Lastly, <strong>4</strong> can be used for browsing documents (<em>Files</em>), see the plots you produce (<em>Plots</em>), searching/managing packages (<em>Packages</em>), or find help on functions (<em>Help</em>).</div>
</div>
</div>
<div id="courses" class="section level3 unnumbered hasAnchor">
<h3 class="unnumbered hasAnchor">Courses<a href="#courses" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>It is recommended that you take an introductory class for R to get to know this language before you start coding here.
Start by going through <a href="https://moderndive.netlify.app/1-getting-started">this Chapter on Getting Started with Data in R</a>.
We will cover other basics on the next sections.</p>
<p>You can find more resources here:</p>
<blockquote>
<ul>
<li>R Tutorial for Beginners at <a href="https://www.guru99.com/r-tutorial.html">guru99</a></li>
<li>R courses at <a href="https://www.bioinformatics.babraham.ac.uk/training.html#advancedrtidy">Babraham Bioinformatics</a></li>
<li>R for <a href="https://r4ds.had.co.nz">Data Science</a></li>
<li>R markdown at <a href="https://rmarkdown.rstudio.com/lesson-1.html">RStudio</a> and in this <a href="https://www.markdownguide.org/cheat-sheet/">cheatsheet</a></li>
<li><a href="https://www.youtube.com/watch?v=h29g21z0a68">Youtube tutorial</a> on ggplot2 with one of its developers</li>
</ul>
</blockquote>
</div>
</div>
<div id="good-practice" class="section level2 unnumbered hasAnchor">
<h2 class="unnumbered hasAnchor">Good practice<a href="#good-practice" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Two notes of advice about good practice now that your journey is about to start:</p>
<ul>
<li><strong>Commenting</strong>: Comment your code as you go with <code>#</code>! This way you avoid forgetting the meaning of a given line. This is also important if you want to share your code with others. Trust me, you will not remember what you are writing today in this class one week from now, so it will be best to have a reminder.<br />
</li>
<li><strong>Clean and tidy code</strong>: See that weird line in the middle of your script/markdown? This is meant to be a guideline about code length. If your code is so long that it crosses the line, please find a way to make it shorter. In addition, avoid making your chunks too long and make separate markdowns or scripts for each class or topic, it will keep everything more organised and easy to track.</li>
</ul>
</div>
<div id="r-markdown" class="section level2 unnumbered hasAnchor">
<h2 class="unnumbered hasAnchor">R Markdown<a href="#r-markdown" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>R Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents.
You will be working with R Markdown during this course.
For more details on using R Markdown see <a href="http://rmarkdown.rstudio.com" class="uri">http://rmarkdown.rstudio.com</a> or <a href="https://www.markdownguide.org/basic-syntax/" class="uri">https://www.markdownguide.org/basic-syntax/</a>.</p>
<p>R instructions in a markdown file are written in a “chunk”, the one below.<br />
Chunks can be added using the <code>+C</code> icon at the top right corner of this panel/editor.<br />
There is also a keyboard shortcut for it.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="co"># Sentences written after &quot;#&quot; are comments. Comments are ignored during the execution of the code.</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="co"># print &quot;Hello world!&quot; on the screen.</span></span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="fu">print</span>(<span class="st">&quot;Hello world!&quot;</span>)</span></code></pre></div>
<pre><code>## [1] &quot;Hello world!&quot;</code></pre>
<blockquote>
<p>What happens in the editor if you remove the second parenthesis?
Do you see the red cross appearing?
You can click on it and read the comment.<br />
And what happens if you press “enter” while the second parenthesis is missing?<br />
</p>
</blockquote>
<p>R is very smart and it will usually let know know when there is a simple syntax mistake like this one on your code.
R will also try to guess variable and function names from your input, all you have to do is <em>click TAB for auto-complete</em>.</p>
</div>
<div id="r-basics" class="section level2 unnumbered hasAnchor">
<h2 class="unnumbered hasAnchor">R Basics<a href="#r-basics" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="variable-assignment-in-r" class="section level3 unnumbered hasAnchor">
<h3 class="unnumbered hasAnchor">Variable assignment in R<a href="#variable-assignment-in-r" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>You can start by testing one of the most basic tasks - variable assignment.
Variable assignments in R can be done using either the <code>&lt;-</code> symbol, or the <code>=</code> symbol.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a>a <span class="ot">&lt;-</span> <span class="st">&quot;Hello world!&quot;</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="co"># or</span></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>hello <span class="ot">=</span> <span class="st">&quot;Hello world!&quot;</span></span></code></pre></div>
<p>You can now run <code>print(a)</code> and <code>print(hello)</code>.
Note that after assignment both variables (a and hello) are listed in the “Environment” window of RStudio.</p>
<blockquote>
<p>R is case sensitive.
<code>hello</code> is a known variable, but <code>Hello</code> is not!</p>
</blockquote>
<p>You can remove variables too…</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">rm</span>(a)</span></code></pre></div>
</div>
<div id="classes" class="section level3 unnumbered hasAnchor">
<h3 class="unnumbered hasAnchor">Classes<a href="#classes" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>In the same way that a orange is a fruit or Malta is a country, any R variable corresponds to a class and holds a given structure.
Classes Let us say you generate a new variable (<code>name</code>).
How can you know more about it?</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a>name <span class="ot">=</span> <span class="st">&quot;Marie&quot;</span></span></code></pre></div>
<p>The function <code>str()</code> gives you the <strong>structure</strong> of the variable.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="fu">str</span>(name)</span></code></pre></div>
<pre><code>##  chr &quot;Marie&quot;</code></pre>
<p>where <em>chr</em> stands for character, the class of this variable.</p>
<p>The function <code>class()</code> gives you the <strong>class</strong> of the variable.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="fu">class</span>(name)</span></code></pre></div>
<pre><code>## [1] &quot;character&quot;</code></pre>
</div>
<div id="logical-operations" class="section level3 unnumbered hasAnchor">
<h3 class="unnumbered hasAnchor">Logical operations<a href="#logical-operations" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We can perform tests on simple variables using the <code>==</code>, <code>&gt;</code>,<code>&lt;</code> operators:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a>name <span class="sc">==</span> <span class="st">&quot;Harry&quot;</span></span></code></pre></div>
<pre><code>## [1] FALSE</code></pre>
<p>The output of this test will be either FALSE/TRUE.
You can test these using numbers too:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="dv">33</span> <span class="sc">&gt;</span> <span class="dv">23</span></span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
</div>
<div id="vectors" class="section level3 unnumbered hasAnchor">
<h3 class="unnumbered hasAnchor">Vectors<a href="#vectors" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We can make any vector using <code>c()</code>.
Try it with some numbers.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a>var1 <span class="ot">=</span> <span class="fu">c</span>(<span class="dv">3</span>,<span class="dv">7</span>,<span class="dv">12</span>,<span class="dv">2</span>)</span></code></pre></div>
<p>If you want to identify the specific number of this vector by position, you can use square brackets.
Like this:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a>var1[<span class="dv">1</span>]     <span class="co"># outputs the first element</span></span>
<span id="cb15-2"><a href="#cb15-2" tabindex="-1"></a>var1[<span class="dv">2</span><span class="sc">:</span><span class="dv">3</span>]   <span class="co"># outputs the 2nd to 3rd elements</span></span>
<span id="cb15-3"><a href="#cb15-3" tabindex="-1"></a>var1[<span class="sc">-</span><span class="dv">3</span>]    <span class="co"># outputs all except the 3rd element</span></span></code></pre></div>
</div>
<div id="matrixes" class="section level3 unnumbered hasAnchor">
<h3 class="unnumbered hasAnchor">Matrixes<a href="#matrixes" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>A matrix is a two-dimensional <strong>array</strong> of numbers.
They are defined using the <code>matrix()</code> function as follows: Take note of the difference between <code>byrow=FALSE</code> and <code>byrow=TRUE</code>.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">4</span>, <span class="dv">5</span>, <span class="dv">6</span>)</span>
<span id="cb16-2"><a href="#cb16-2" tabindex="-1"></a></span>
<span id="cb16-3"><a href="#cb16-3" tabindex="-1"></a>X <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="at">data =</span> x, <span class="at">nrow =</span> <span class="dv">2</span>, <span class="at">ncol =</span> <span class="dv">3</span>, <span class="at">byrow =</span> <span class="cn">TRUE</span>)</span>
<span id="cb16-4"><a href="#cb16-4" tabindex="-1"></a>X <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="at">data =</span> x, <span class="at">nrow =</span> <span class="dv">2</span>, <span class="at">ncol =</span> <span class="dv">3</span>, <span class="at">byrow =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<blockquote>
<p>Do you understand all parameters (“nrow”, “ncol”, “byrow”) of the <code>matrix()</code> function?
What happens if you do not specify “byrow” as TRUE or FALSE?</p>
</blockquote>
<p>You can ask for the <strong>dimensions</strong> of a matrix (and see later also for data frames) using the function <code>dim()</code>.
Do it for X.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a><span class="fu">dim</span>(X)</span></code></pre></div>
<p>To access the elements of a matrix, it is similar to the vector but with 2 indexes:</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a>X[<span class="dv">1</span>, <span class="dv">2</span>]                     <span class="co"># [1] 3  -&gt; element in the first row and second column</span></span>
<span id="cb18-2"><a href="#cb18-2" tabindex="-1"></a>X[<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>, <span class="dv">1</span>]                     <span class="co"># [1] 1 2 </span></span>
<span id="cb18-3"><a href="#cb18-3" tabindex="-1"></a>X[<span class="dv">2</span>, ]                      <span class="co"># [1] 2 4 6  -&gt; all elements of the second row</span></span>
<span id="cb18-4"><a href="#cb18-4" tabindex="-1"></a><span class="fu">length</span>(X[<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>, <span class="dv">1</span>])             <span class="co"># [1] 2  -&gt; X[1:2, 1] is a vector of length 2</span></span></code></pre></div>
</div>
<div id="dataframes" class="section level3 unnumbered hasAnchor">
<h3 class="unnumbered hasAnchor">Dataframes<a href="#dataframes" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Data frames are like matrices, but they can contain <strong>multiple types</strong> of values mixed.<br />
We can create a data frame using the <code>data.frame()</code> function.
We can also convert a matrix into a data frame with the <code>as.data.frame()</code> function.</p>
<p>We can prepare three vectors of the same length (for example 4 elements) and create a data frame:</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a>Name <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;Leah&quot;</span>, <span class="st">&quot;Alice&quot;</span>, <span class="st">&quot;Jonas&quot;</span>, <span class="st">&quot;Paula&quot;</span>)</span>
<span id="cb19-2"><a href="#cb19-2" tabindex="-1"></a>Age <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">21</span>, <span class="dv">22</span>, <span class="dv">20</span>, <span class="dv">22</span>)</span>
<span id="cb19-3"><a href="#cb19-3" tabindex="-1"></a>Course <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;Mathematics&quot;</span>, <span class="st">&quot;Physics&quot;</span>, <span class="st">&quot;Medicine&quot;</span>, <span class="st">&quot;Biology&quot;</span>)</span>
<span id="cb19-4"><a href="#cb19-4" tabindex="-1"></a>Place_of_birth <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;USA&quot;</span>, <span class="st">&quot;Germany&quot;</span>, <span class="st">&quot;Germany&quot;</span>, <span class="st">&quot;France&quot;</span>)</span>
<span id="cb19-5"><a href="#cb19-5" tabindex="-1"></a></span>
<span id="cb19-6"><a href="#cb19-6" tabindex="-1"></a>Students <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">First_Name =</span> Name, </span>
<span id="cb19-7"><a href="#cb19-7" tabindex="-1"></a>                       <span class="at">Age =</span> Age, </span>
<span id="cb19-8"><a href="#cb19-8" tabindex="-1"></a>                       <span class="at">Course =</span> Course,</span>
<span id="cb19-9"><a href="#cb19-9" tabindex="-1"></a>                       <span class="at">Place_of_birth =</span> Place_of_birth)</span>
<span id="cb19-10"><a href="#cb19-10" tabindex="-1"></a>Students</span></code></pre></div>
<pre><code>##   First_Name Age      Course Place_of_birth
## 1       Leah  21 Mathematics            USA
## 2      Alice  22     Physics        Germany
## 3      Jonas  20    Medicine        Germany
## 4      Paula  22     Biology         France</code></pre>
<p>We can again access specific elements in a similar manner to the matrix before.
Try some of these examples:</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a><span class="co"># Columns</span></span>
<span id="cb21-2"><a href="#cb21-2" tabindex="-1"></a>Students<span class="sc">$</span>Age </span>
<span id="cb21-3"><a href="#cb21-3" tabindex="-1"></a>Students[[<span class="st">&quot;Age&quot;</span>]]</span>
<span id="cb21-4"><a href="#cb21-4" tabindex="-1"></a></span>
<span id="cb21-5"><a href="#cb21-5" tabindex="-1"></a><span class="co"># Rows or columns</span></span>
<span id="cb21-6"><a href="#cb21-6" tabindex="-1"></a>Students[, <span class="dv">3</span>]</span>
<span id="cb21-7"><a href="#cb21-7" tabindex="-1"></a>Students[<span class="dv">3</span>,]</span>
<span id="cb21-8"><a href="#cb21-8" tabindex="-1"></a></span>
<span id="cb21-9"><a href="#cb21-9" tabindex="-1"></a><span class="co"># Elements</span></span>
<span id="cb21-10"><a href="#cb21-10" tabindex="-1"></a>Students<span class="sc">$</span>Course[<span class="dv">3</span>]    <span class="co"># access element 3 of the column &quot;Course&quot;</span></span>
<span id="cb21-11"><a href="#cb21-11" tabindex="-1"></a>Students[<span class="dv">2</span><span class="sc">:</span><span class="dv">3</span>, <span class="dv">3</span>]      <span class="co"># select elements 2 and 3 of column 3 </span></span></code></pre></div>
</div>
<div id="apply-sapply-lapply" class="section level3 unnumbered hasAnchor">
<h3 class="unnumbered hasAnchor">Apply, sapply, lapply<a href="#apply-sapply-lapply" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><code>Apply()</code> is used to repeat the same operation over all columns/rows of a <strong>matrix</strong> or <strong>data frame</strong>.
Therefore, we need to specify three parameters:<br />
1.
The matrix on which the operation should be performed<br />
2.
Whether to repeat the operation over rows, which is defined by a “1”, or columns (use a “2” instead)<br />
3.
The operation that should be performed.<br />
</p>
<p>Let’s have a look at an example to make this more clear: We want to determine the minimum for each row of the matrix “Mat”.
Therefore, we can use the function <code>min()</code>.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" tabindex="-1"></a><span class="co">#First, we build the matrix </span></span>
<span id="cb22-2"><a href="#cb22-2" tabindex="-1"></a>m <span class="ot">&lt;-</span> <span class="fu">rnorm</span>(<span class="dv">30</span>)   <span class="co"># generate 30 random numbers for the matrix</span></span>
<span id="cb22-3"><a href="#cb22-3" tabindex="-1"></a>Mat <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="at">data =</span> m, <span class="at">nrow =</span> <span class="dv">6</span>)</span>
<span id="cb22-4"><a href="#cb22-4" tabindex="-1"></a>Mat</span></code></pre></div>
<pre><code>##             [,1]        [,2]       [,3]        [,4]       [,5]
## [1,]  0.81075342  1.63010444  0.1009744 -0.51431302  0.3188881
## [2,] -1.02544562  0.45038368  0.5050776  0.19423683  1.5527958
## [3,]  1.06936183  0.06153936 -0.1269797 -0.02154715 -0.3657364
## [4,]  0.88974304 -0.43830831  0.1023935  0.75687749 -1.0931076
## [5,] -0.07781049  0.49469362  0.3494489 -0.77850067 -0.1861264
## [6,]  0.02723114 -1.68036554 -0.9208189 -1.10706835  1.0189460</code></pre>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" tabindex="-1"></a><span class="co">#Next, we want to determine the minimum value of each row: </span></span>
<span id="cb24-2"><a href="#cb24-2" tabindex="-1"></a><span class="fu">apply</span>(Mat, <span class="dv">1</span>, min)</span></code></pre></div>
<pre><code>## [1] -0.5143130 -1.0254456 -0.3657364 -1.0931076 -0.7785007 -1.6803655</code></pre>
<blockquote>
<p>What happens if you change the “1” in the apply function to “2”?
Try it!</p>
</blockquote>
<blockquote>
<p>Remember that you can use <code>help(apply)</code> or <code>?apply</code> to get help on this function!</p>
</blockquote>
<p>We can perform a loop over all elements of a <strong>vector</strong> or <strong>list</strong> using the <code>sapply()</code> function.<br />
<code>sapply()</code> needs two information:<br />
1.
which vector do we consider?<br />
2.
which function do we want to apply to each element of this vector?<br />
<code>sapply()</code> will return a <strong>vector</strong> containing the results.</p>
<p>For instance, we can calculate the square root of every element of a vector using the function <code>sqrt()</code>:</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>)</span>
<span id="cb26-2"><a href="#cb26-2" tabindex="-1"></a><span class="fu">sapply</span>(x, sqrt)</span></code></pre></div>
<pre><code>## [1] 1.000000 1.414214 1.732051 2.000000 2.236068</code></pre>
<p>Instead of using built-in functions, we can also write our own little function and combine it with <code>sapply()</code>, <code>apply()</code> or <code>lapply()</code> (see below):</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" tabindex="-1"></a><span class="co"># sapply()</span></span>
<span id="cb28-2"><a href="#cb28-2" tabindex="-1"></a>z <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>)</span>
<span id="cb28-3"><a href="#cb28-3" tabindex="-1"></a><span class="fu">sapply</span>(z, <span class="cf">function</span>(x) {x<span class="sc">*</span><span class="dv">2</span>}) </span></code></pre></div>
<pre><code>## [1]  2  4  6  8 10</code></pre>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" tabindex="-1"></a><span class="co"># apply()</span></span>
<span id="cb30-2"><a href="#cb30-2" tabindex="-1"></a>Mat </span></code></pre></div>
<pre><code>##             [,1]        [,2]       [,3]        [,4]       [,5]
## [1,]  0.81075342  1.63010444  0.1009744 -0.51431302  0.3188881
## [2,] -1.02544562  0.45038368  0.5050776  0.19423683  1.5527958
## [3,]  1.06936183  0.06153936 -0.1269797 -0.02154715 -0.3657364
## [4,]  0.88974304 -0.43830831  0.1023935  0.75687749 -1.0931076
## [5,] -0.07781049  0.49469362  0.3494489 -0.77850067 -0.1861264
## [6,]  0.02723114 -1.68036554 -0.9208189 -1.10706835  1.0189460</code></pre>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" tabindex="-1"></a><span class="fu">apply</span>(Mat, <span class="dv">1</span>, <span class="cf">function</span>(x) {<span class="fu">sum</span>(x<span class="sc">*</span><span class="dv">2</span>)}) <span class="co">#the operation is performed on every row of &quot;Mat&quot;; every element is multiplied by two and the sum of the row is calculated</span></span></code></pre></div>
<pre><code>## [1]  4.6928147  3.3540966  1.2332761  0.4351964 -0.3965901 -5.3241513</code></pre>
<p>Besides, we can use <code>sapply()</code> for a <strong>list</strong>, for example to calculate the length of every element using <code>length()</code>:</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" tabindex="-1"></a>List1 <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">color =</span> <span class="fu">c</span>(<span class="st">&quot;blue&quot;</span>, <span class="st">&quot;red&quot;</span>), <span class="at">size =</span> <span class="dv">5</span>, <span class="at">state =</span> <span class="fu">c</span>(<span class="cn">TRUE</span>, <span class="cn">FALSE</span>, <span class="cn">TRUE</span>, <span class="cn">TRUE</span>))</span>
<span id="cb34-2"><a href="#cb34-2" tabindex="-1"></a></span>
<span id="cb34-3"><a href="#cb34-3" tabindex="-1"></a><span class="fu">sapply</span>(List1, length)</span></code></pre></div>
<pre><code>## color  size state 
##     2     1     4</code></pre>
<p>As you can see, this returns a vector with the length of every list element.</p>
<p>However, sometimes it can be useful to keep the results stored in a list.
Therefore, we can use <code>lapply()</code> instead.
Let’s have a look at the difference between <code>sapply()</code> and <code>lapply()</code> using the example from above:</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" tabindex="-1"></a><span class="fu">lapply</span>(List1, length)</span></code></pre></div>
<pre><code>## $color
## [1] 2
## 
## $size
## [1] 1
## 
## $state
## [1] 4</code></pre>
<blockquote>
<p>Do you understand the difference?</p>
</blockquote>
</div>
<div id="for-loops" class="section level3 unnumbered hasAnchor">
<h3 class="unnumbered hasAnchor">For Loops<a href="#for-loops" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Besides the apply-family, we can use <strong>for loops</strong> for iterating over a sequence:</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;a&quot;</span>, <span class="st">&quot;b&quot;</span>, <span class="st">&quot;c&quot;</span>)</span>
<span id="cb38-2"><a href="#cb38-2" tabindex="-1"></a></span>
<span id="cb38-3"><a href="#cb38-3" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> x) {</span>
<span id="cb38-4"><a href="#cb38-4" tabindex="-1"></a>  <span class="fu">print</span>(i)</span>
<span id="cb38-5"><a href="#cb38-5" tabindex="-1"></a>}</span></code></pre></div>
<pre><code>## [1] &quot;a&quot;
## [1] &quot;b&quot;
## [1] &quot;c&quot;</code></pre>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" tabindex="-1"></a><span class="co">#or </span></span>
<span id="cb40-2"><a href="#cb40-2" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>) {</span>
<span id="cb40-3"><a href="#cb40-3" tabindex="-1"></a>  <span class="fu">print</span>(i<span class="sc">*</span><span class="dv">3</span>)</span>
<span id="cb40-4"><a href="#cb40-4" tabindex="-1"></a>}</span></code></pre></div>
<pre><code>## [1] 3
## [1] 6
## [1] 9
## [1] 12</code></pre>
</div>
<div id="if-statements" class="section level3 unnumbered hasAnchor">
<h3 class="unnumbered hasAnchor">If Statements<a href="#if-statements" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>You can use an <strong>if statement</strong> to execute a block of code only, if a condition is TRUE.</p>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" tabindex="-1"></a>a <span class="ot">&lt;-</span> <span class="dv">5</span></span>
<span id="cb42-2"><a href="#cb42-2" tabindex="-1"></a>b <span class="ot">&lt;-</span> <span class="dv">10</span></span>
<span id="cb42-3"><a href="#cb42-3" tabindex="-1"></a></span>
<span id="cb42-4"><a href="#cb42-4" tabindex="-1"></a><span class="cf">if</span> (a <span class="sc">&lt;</span> b) {</span>
<span id="cb42-5"><a href="#cb42-5" tabindex="-1"></a>  <span class="fu">print</span>(<span class="st">&quot;a is smaller than b&quot;</span>)</span>
<span id="cb42-6"><a href="#cb42-6" tabindex="-1"></a>}</span></code></pre></div>
<pre><code>## [1] &quot;a is smaller than b&quot;</code></pre>
<blockquote>
<p>Modify “a” or “b” and see how the output changes!</p>
</blockquote>
<p>We can also add <strong>else if</strong> to this statement.
In that case, if the <strong>if</strong> condition is FALSE, the <strong>else if</strong> condition will be tried:</p>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" tabindex="-1"></a>a <span class="ot">&lt;-</span> <span class="dv">10</span></span>
<span id="cb44-2"><a href="#cb44-2" tabindex="-1"></a>b <span class="ot">&lt;-</span> <span class="dv">10</span></span>
<span id="cb44-3"><a href="#cb44-3" tabindex="-1"></a></span>
<span id="cb44-4"><a href="#cb44-4" tabindex="-1"></a><span class="cf">if</span> (a <span class="sc">&lt;</span> b) {</span>
<span id="cb44-5"><a href="#cb44-5" tabindex="-1"></a>  <span class="fu">print</span>(<span class="st">&quot;a is smaller than b&quot;</span>)</span>
<span id="cb44-6"><a href="#cb44-6" tabindex="-1"></a>} <span class="cf">else</span> <span class="cf">if</span> (a <span class="sc">==</span> b) {</span>
<span id="cb44-7"><a href="#cb44-7" tabindex="-1"></a>  <span class="fu">print</span>(<span class="st">&quot;a is equal to b&quot;</span>)</span>
<span id="cb44-8"><a href="#cb44-8" tabindex="-1"></a>}</span></code></pre></div>
<pre><code>## [1] &quot;a is equal to b&quot;</code></pre>
<p>Besides, we can add an <strong>else</strong> statement to be executed when all previous conditions are not TRUE.</p>
<div class="sourceCode" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" tabindex="-1"></a>a <span class="ot">&lt;-</span> <span class="dv">10</span></span>
<span id="cb46-2"><a href="#cb46-2" tabindex="-1"></a>b <span class="ot">&lt;-</span> <span class="dv">5</span></span>
<span id="cb46-3"><a href="#cb46-3" tabindex="-1"></a></span>
<span id="cb46-4"><a href="#cb46-4" tabindex="-1"></a><span class="cf">if</span> (a <span class="sc">&lt;</span> b) {</span>
<span id="cb46-5"><a href="#cb46-5" tabindex="-1"></a>  <span class="fu">print</span>(<span class="st">&quot;a is smaller than b&quot;</span>)</span>
<span id="cb46-6"><a href="#cb46-6" tabindex="-1"></a>} <span class="cf">else</span> <span class="cf">if</span> (a <span class="sc">==</span> b) {</span>
<span id="cb46-7"><a href="#cb46-7" tabindex="-1"></a>  <span class="fu">print</span>(<span class="st">&quot;a is equal to b&quot;</span>)</span>
<span id="cb46-8"><a href="#cb46-8" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb46-9"><a href="#cb46-9" tabindex="-1"></a>  <span class="fu">print</span>(<span class="st">&quot;a is greater than b&quot;</span>)</span>
<span id="cb46-10"><a href="#cb46-10" tabindex="-1"></a>}</span></code></pre></div>
<pre><code>## [1] &quot;a is greater than b&quot;</code></pre>
<blockquote>
<p>Do you understand how this works?
What would happen if a = 10 and b = 10?
Try it!</p>
</blockquote>
</div>
</div>
<div id="data-wrangling-with-tidyverse" class="section level2 unnumbered hasAnchor">
<h2 class="unnumbered hasAnchor">Data wrangling with tidyverse<a href="#data-wrangling-with-tidyverse" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="introducing-tidyverse" class="section level3 unnumbered hasAnchor">
<h3 class="unnumbered hasAnchor">Introducing tidyverse<a href="#introducing-tidyverse" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><strong>Tidyverse</strong> is a set of packages often used to manipulate data.
You can think of tidyverse as a variation of the original R, which is meant to be a grammar specific to data manipulation.</p>
<p>You can install these through:</p>
<div class="sourceCode" id="cb48"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;tidyverse&quot;</span>)</span></code></pre></div>
<p>And load it using</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code></pre></div>
<pre><code>## ── Attaching core tidyverse packages ─────────── tidyverse 2.0.0 ──
## ✔ dplyr     1.1.4     ✔ readr     2.1.5
## ✔ forcats   1.0.0     ✔ stringr   1.5.1
## ✔ ggplot2   3.5.1     ✔ tibble    3.2.1
## ✔ lubridate 1.9.3     ✔ tidyr     1.3.1
## ✔ purrr     1.0.2     
## ── Conflicts ───────────────────────────── tidyverse_conflicts() ──
## ✖ dplyr::filter() masks stats::filter()
## ✖ dplyr::lag()    masks stats::lag()
## ℹ Use the conflicted package (&lt;http://conflicted.r-lib.org/&gt;) to force all conflicts to become errors</code></pre>
<p>You can find R packages in CRAN.
CRAN is a large repository containing almost 20 thousand packages!
For this course, we will just use CRAN packages.
To install them, you will use <code>install.packages()</code>.</p>
<p>We will learn a couple of functions from this framework which will make the next tasks a lot easier.
Tidyverse works with the <em>tibble</em> class, which is very similar to a the dataframe class, only faster and tidier.</p>
<p>Start by loading a default tibble including sleeping patterns from different mammals.
It also includes information on the animal order, genus, diet, or body weight.
To do so, run the following:</p>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" tabindex="-1"></a><span class="fu">data</span>(msleep)</span></code></pre></div>
<p>We can start by inspecting <code>msleep</code>.</p>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" tabindex="-1"></a><span class="fu">glimpse</span>(msleep)</span></code></pre></div>
<p>We can print the first rows of a dataframe using <code>head()</code>.
Try it out.</p>
<div class="sourceCode" id="cb53"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" tabindex="-1"></a><span class="fu">head</span>(msleep)</span></code></pre></div>
<pre><code>## # A tibble: 6 × 11
##   name    genus vore  order conservation sleep_total sleep_rem sleep_cycle awake
##   &lt;chr&gt;   &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;              &lt;dbl&gt;     &lt;dbl&gt;       &lt;dbl&gt; &lt;dbl&gt;
## 1 Cheetah Acin… carni Carn… lc                  12.1      NA        NA      11.9
## 2 Owl mo… Aotus omni  Prim… &lt;NA&gt;                17         1.8      NA       7  
## 3 Mounta… Aplo… herbi Rode… nt                  14.4       2.4      NA       9.6
## 4 Greate… Blar… omni  Sori… lc                  14.9       2.3       0.133   9.1
## 5 Cow     Bos   herbi Arti… domesticated         4         0.7       0.667  20  
## 6 Three-… Brad… herbi Pilo… &lt;NA&gt;                14.4       2.2       0.767   9.6
## # ℹ 2 more variables: brainwt &lt;dbl&gt;, bodywt &lt;dbl&gt;</code></pre>
</div>
<div id="filter-based-on-conditions" class="section level3 unnumbered hasAnchor">
<h3 class="unnumbered hasAnchor">Filter based on conditions<a href="#filter-based-on-conditions" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Now that we know how this dataset looks like, we can start doing more with it.
Let us see the information for the domestic pig.<br />
We can do it using <code>filter()</code> on the <strong>genus</strong> column:</p>
<div class="sourceCode" id="cb55"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" tabindex="-1"></a>msleep <span class="sc">%&gt;%</span></span>
<span id="cb55-2"><a href="#cb55-2" tabindex="-1"></a>  <span class="fu">filter</span>(genus <span class="sc">==</span> <span class="st">&quot;Sus&quot;</span>)</span></code></pre></div>
<pre><code>## # A tibble: 1 × 11
##   name  genus vore  order   conservation sleep_total sleep_rem sleep_cycle awake
##   &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;   &lt;chr&gt;              &lt;dbl&gt;     &lt;dbl&gt;       &lt;dbl&gt; &lt;dbl&gt;
## 1 Pig   Sus   omni  Artiod… domesticated         9.1       2.4         0.5  14.9
## # ℹ 2 more variables: brainwt &lt;dbl&gt;, bodywt &lt;dbl&gt;</code></pre>
<p>We also use the <code>%&gt;%</code> operator here.
This is a <strong>pipe operator</strong> and you will use it very often.</p>
<p>To understand piping, let’s imagine you are making a three layer cake in a factory.
Instead of manually carrying the cake from one station to another to add layers and frosting, you have a conveyor belt transporting the cake through the stations automatically to make the process much more efficient.
Piping avoids that you have to create multiple intermediate objects to achieve your final result when performing multiple functions.</p>
<blockquote>
<p>Try filtering other columns based on numeric conditions.
Use <code>filter()</code> to see which 7 animals in the datasets have a body weight superior to 200 (kg).</p>
</blockquote>
</div>
<div id="select-specific-columns" class="section level3 unnumbered hasAnchor">
<h3 class="unnumbered hasAnchor">Select specific columns<a href="#select-specific-columns" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Now, let’s say you want to <em>select</em> only specific columns.
Use the <code>select()</code> function to select the columns <code>name</code>, <code>order</code>, and <code>bodywt</code> (body weight).</p>
<div class="sourceCode" id="cb57"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" tabindex="-1"></a>body_weights <span class="ot">=</span> msleep <span class="sc">%&gt;%</span></span>
<span id="cb57-2"><a href="#cb57-2" tabindex="-1"></a>  <span class="fu">select</span>(name, order, bodywt)</span>
<span id="cb57-3"><a href="#cb57-3" tabindex="-1"></a></span>
<span id="cb57-4"><a href="#cb57-4" tabindex="-1"></a><span class="fu">head</span>(body_weights)</span></code></pre></div>
<pre><code>## # A tibble: 6 × 3
##   name                       order         bodywt
##   &lt;chr&gt;                      &lt;chr&gt;          &lt;dbl&gt;
## 1 Cheetah                    Carnivora     50    
## 2 Owl monkey                 Primates       0.48 
## 3 Mountain beaver            Rodentia       1.35 
## 4 Greater short-tailed shrew Soricomorpha   0.019
## 5 Cow                        Artiodactyla 600    
## 6 Three-toed sloth           Pilosa         3.85</code></pre>
</div>
<div id="operations-on-groups" class="section level3 unnumbered hasAnchor">
<h3 class="unnumbered hasAnchor">Operations on groups<a href="#operations-on-groups" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We can assess the average (<code>mean()</code>) body weight by animal order present.
To do so, we will use <code>summarise()</code> and <code>group_by()</code> functions on the <em>body_weights</em> we generated previously.</p>
<ul>
<li><code>group_by()</code>: Assigns a group to a given set of rows (observations).<br />
</li>
<li><code>summarise()</code>: Makes calculations for a given group based on the row values assigned to it.</li>
</ul>
<div class="sourceCode" id="cb59"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" tabindex="-1"></a>body_weights <span class="sc">%&gt;%</span> </span>
<span id="cb59-2"><a href="#cb59-2" tabindex="-1"></a>  <span class="fu">group_by</span>(order) <span class="sc">%&gt;%</span></span>
<span id="cb59-3"><a href="#cb59-3" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">mean_bweight =</span> <span class="fu">mean</span>(bodywt, <span class="at">na.rm =</span> T)) <span class="co"># na.rm = T will exclude missing values (NA)</span></span></code></pre></div>
<pre><code>## # A tibble: 19 × 2
##    order           mean_bweight
##    &lt;chr&gt;                  &lt;dbl&gt;
##  1 Afrosoricida          0.9   
##  2 Artiodactyla        282.    
##  3 Carnivora            57.7   
##  4 Cetacea             342.    
##  5 Chiroptera            0.0165
##  6 Cingulata            31.8   
##  7 Didelphimorphia       1.03  
##  8 Diprotodontia         1.36  
##  9 Erinaceomorpha        0.66  
## 10 Hyracoidea            3.06  
## 11 Lagomorpha            2.5   
## 12 Monotremata           4.5   
## 13 Perissodactyla      305.    
## 14 Pilosa                3.85  
## 15 Primates             13.9   
## 16 Proboscidea        4600.    
## 17 Rodentia              0.288 
## 18 Scandentia            0.104 
## 19 Soricomorpha          0.0414</code></pre>
</div>
<div id="reshaping-data" class="section level3 unnumbered hasAnchor">
<h3 class="unnumbered hasAnchor">Reshaping data<a href="#reshaping-data" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>On tidyverse, he objects you are working on are sometimes not in a tidy format.
There are three rules which make a dataset tidy:</p>
<ul>
<li>Each variable has its own column.</li>
<li>Each observation has its own row.</li>
<li>Each value has its own cell.</li>
</ul>
<p><img src="figures/tidyexp.png" alt="Here is the comparison between object types. Tidy objects (right) require an extra step of processing." /> <a href="https://r4ds.hadley.nz/data-tidy.html">Image source</a></p>
<p>How can we convert them then?
The mammalian sleep dataset we used earlier is already in a tidy format, so we will generate an example.</p>
<p>You are testing a new diet on 7 cats, and thus registering their weight before and after 1 month on the diet.</p>
<div class="sourceCode" id="cb61"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" tabindex="-1"></a>cat.weights <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">CatName=</span><span class="fu">c</span>(<span class="st">&#39;Muffin&#39;</span>, <span class="st">&#39;Lily&#39;</span>, <span class="st">&#39;Mittens&#39;</span>, <span class="st">&#39;Oreo&#39;</span>, <span class="st">&#39;Loki&#39;</span>, <span class="st">&#39;Fluffy&#39;</span>, <span class="st">&#39;Honey&#39;</span>),</span>
<span id="cb61-2"><a href="#cb61-2" tabindex="-1"></a>                          <span class="at">month0=</span><span class="fu">c</span>(<span class="fl">4.3</span>, <span class="fl">4.5</span>, <span class="dv">5</span>, <span class="fl">4.4</span>, <span class="fl">3.8</span>, <span class="fl">5.5</span>, <span class="fl">4.5</span>),</span>
<span id="cb61-3"><a href="#cb61-3" tabindex="-1"></a>                          <span class="at">month1=</span><span class="fu">c</span>(<span class="fl">4.4</span>, <span class="fl">4.5</span>, <span class="fl">4.9</span>, <span class="fl">4.3</span>, <span class="fl">3.9</span>, <span class="fl">5.4</span>, <span class="fl">4.6</span>))</span>
<span id="cb61-4"><a href="#cb61-4" tabindex="-1"></a></span>
<span id="cb61-5"><a href="#cb61-5" tabindex="-1"></a>cat.weights</span></code></pre></div>
<pre><code>##   CatName month0 month1
## 1  Muffin    4.3    4.4
## 2    Lily    4.5    4.5
## 3 Mittens    5.0    4.9
## 4    Oreo    4.4    4.3
## 5    Loki    3.8    3.9
## 6  Fluffy    5.5    5.4
## 7   Honey    4.5    4.6</code></pre>
<p>In this example, you can see that we have 2 different variables for the same unit (weight), these are our <em>gatherable</em> columns.
We can then make this dataframe tidy using the <code>gather()</code> from the <code>tidyr</code> package:</p>
<div class="sourceCode" id="cb63"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" tabindex="-1"></a><span class="fu">library</span>(tidyr)</span>
<span id="cb63-2"><a href="#cb63-2" tabindex="-1"></a></span>
<span id="cb63-3"><a href="#cb63-3" tabindex="-1"></a>cat.weights <span class="sc">%&gt;%</span> </span>
<span id="cb63-4"><a href="#cb63-4" tabindex="-1"></a>  <span class="fu">gather</span>(<span class="at">key =</span> Time, <span class="co"># column name given to the gathered column names</span></span>
<span id="cb63-5"><a href="#cb63-5" tabindex="-1"></a>         <span class="at">value =</span> Weight, <span class="co"># column name will be given to the values</span></span>
<span id="cb63-6"><a href="#cb63-6" tabindex="-1"></a>         month0<span class="sc">:</span>month1) <span class="co"># columns to gather</span></span></code></pre></div>
<pre><code>##    CatName   Time Weight
## 1   Muffin month0    4.3
## 2     Lily month0    4.5
## 3  Mittens month0    5.0
## 4     Oreo month0    4.4
## 5     Loki month0    3.8
## 6   Fluffy month0    5.5
## 7    Honey month0    4.5
## 8   Muffin month1    4.4
## 9     Lily month1    4.5
## 10 Mittens month1    4.9
## 11    Oreo month1    4.3
## 12    Loki month1    3.9
## 13  Fluffy month1    5.4
## 14   Honey month1    4.6</code></pre>
<p>Do you understand the differences?
Note that now each name shows up twice.</p>
</div>
</div>
<div id="plotting-using-ggplot2" class="section level2 unnumbered hasAnchor">
<h2 class="unnumbered hasAnchor">Plotting using ggplot2<a href="#plotting-using-ggplot2" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>One of the many advantages of tidyverse is how compatible it is with <code>ggplot2</code>.
Thanks to the wonders of the <code>%&gt;%</code> operator, you can generate a basic plot instantly from a gathered/tidy tibble without creating intermediate objects.</p>
<p><strong>The structure of any <code>ggplot()</code> is always based on two essential elements: the aesthetics (<code>aes()</code>) and the plot layers (named as <code>geom_something()</code>. <code>aes()</code> can found within the <code>ggplot()</code> function or on the <code>geom_something()</code>.</strong> There are multiple options for <code>geom_X()</code> depending on the information you are plotting.
You can find all possibilities <a href="https://ggplot2.tidyverse.org/reference/">here</a>.</p>
<p>The <code>aes()</code> function can harbor information on axis (x, y) and other aesthetics.
Let us see an example within the scatter plot next.</p>
<p>In this course we will work with <code>ggplot()</code>, but plots can also be generated by another framework.
The alternative to <code>ggplot()</code> is <code>plot()</code>.
Beware when looking for solutions for your data visualization problems.</p>
<div id="scatter-plot" class="section level3 unnumbered hasAnchor">
<h3 class="unnumbered hasAnchor">Scatter plot<a href="#scatter-plot" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>For example, if we want to see the relationship between the sleep time (total, in hours) and the body weights, we use this formula:</p>
<div class="sourceCode" id="cb65"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" tabindex="-1"></a><span class="fu">ggplot</span>(msleep,                    <span class="co"># start by picking the dataframe to plot</span></span>
<span id="cb65-2"><a href="#cb65-2" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> bodywt,            <span class="co"># column to use for x-axis</span></span>
<span id="cb65-3"><a href="#cb65-3" tabindex="-1"></a>           <span class="at">y =</span> sleep_total)) <span class="sc">+</span>    <span class="co"># column to use for y-axis</span></span>
<span id="cb65-4"><a href="#cb65-4" tabindex="-1"></a>  <span class="fu">geom_point</span>()                    <span class="co"># plot layer for points</span></span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-19-1.png" width="384" /></p>
<blockquote>
<p>The x-axis looks a little squished, so it will be hard to view many of the trends.
We can then log-transform this scale using the <code>log()</code> function.
Try it out!</p>
</blockquote>
<p>We can try to play with the other aesthetics of the plot.
For example, we can use the colour easthetic to colour the dots based on the <em>vore</em>/diet.</p>
<div class="sourceCode" id="cb66"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" tabindex="-1"></a><span class="fu">ggplot</span>(msleep,                    </span>
<span id="cb66-2"><a href="#cb66-2" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">log</span>(bodywt),            <span class="co"># column to use for x-axis</span></span>
<span id="cb66-3"><a href="#cb66-3" tabindex="-1"></a>           <span class="at">y =</span> sleep_total,            <span class="co"># column to use for y-axis</span></span>
<span id="cb66-4"><a href="#cb66-4" tabindex="-1"></a>           <span class="at">colour =</span> vore)) <span class="sc">+</span>           <span class="co"># column used for colouring dots</span></span>
<span id="cb66-5"><a href="#cb66-5" tabindex="-1"></a>  <span class="fu">geom_point</span>() </span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-20-1.png" width="480" /></p>
</div>
<div id="histogram" class="section level3 unnumbered hasAnchor">
<h3 class="unnumbered hasAnchor">Histogram<a href="#histogram" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Let us start by making an histogram, as we will work with distributions over this course.
If we want to see the global distribution of sleep time (total, in hours) for all animals in the dataset, we will use <code>geom_histogram()</code>.
Like this:</p>
<div class="sourceCode" id="cb67"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" tabindex="-1"></a><span class="fu">ggplot</span>(msleep,                    <span class="co"># start by picking the dataframe to plot</span></span>
<span id="cb67-2"><a href="#cb67-2" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> sleep_total)) <span class="sc">+</span>    <span class="co"># column to use for y-axis</span></span>
<span id="cb67-3"><a href="#cb67-3" tabindex="-1"></a>  <span class="fu">geom_histogram</span>()                <span class="co"># plot layer for histogram</span></span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-21-1.png" width="480" /></p>
<p>In some cases, you will not have the data on a tidy tibble and rather on a vector for example.
For such cases, the structure would be slightly different.
Check how you can plot the distribution of a random set of numbers.</p>
<div class="sourceCode" id="cb68"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" tabindex="-1"></a>rand.numbers <span class="ot">=</span> <span class="fu">rnorm</span>(<span class="dv">100</span>) <span class="co"># we will learn this function later on</span></span>
<span id="cb68-2"><a href="#cb68-2" tabindex="-1"></a></span>
<span id="cb68-3"><a href="#cb68-3" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span>   </span>
<span id="cb68-4"><a href="#cb68-4" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="fu">aes</span>(<span class="at">x =</span> rand.numbers),  <span class="co"># aes goes in the plot layer for histogram</span></span>
<span id="cb68-5"><a href="#cb68-5" tabindex="-1"></a>                 <span class="at">bins =</span> <span class="dv">50</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-22-1.png" width="672" /></p>
<blockquote>
<p>Try changing the option <code>bins</code>.
What do you see?</p>
</blockquote>
</div>
<div id="boxplot" class="section level3 unnumbered hasAnchor">
<h3 class="unnumbered hasAnchor">Boxplot<a href="#boxplot" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>In this next example, we are creating a boxplot with the sleep time (total, in hours) for each dietary group (<code>vore</code>).</p>
<div class="sourceCode" id="cb69"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" tabindex="-1"></a><span class="fu">ggplot</span>(msleep, <span class="co"># we use na.omit here to hide NAs from the plot</span></span>
<span id="cb69-2"><a href="#cb69-2" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> vore,</span>
<span id="cb69-3"><a href="#cb69-3" tabindex="-1"></a>           <span class="at">y =</span> sleep_total)) <span class="sc">+</span></span>
<span id="cb69-4"><a href="#cb69-4" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>()</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-23-1.png" width="384" /></p>
<p>See the NA on the x-axis?
In the course (week one) we will learn how to clean the data to avoid missing values like these ones.</p>
<blockquote>
<p>Try some of the other geom_ layers available.
As an example, <code>geom_violin()</code> can be used in a similar way to <code>geom_boxplot()</code>.</p>
</blockquote>
</div>
</div>
<div id="common-errors" class="section level2 unnumbered hasAnchor">
<h2 class="unnumbered hasAnchor">Common errors<a href="#common-errors" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p><code>Error in library(viridis) : there is no package called ‘viridis’</code>:<br />
<strong>Solution:</strong> install the required package using <code>install.packages("package")</code> and load it using <code>library(package)</code>.</p>
<p><code>Error in gather : could not find function "gather"</code>:<br />
<strong>Solution:</strong> load the package the function belongs to.
You can see it by running <code>??gather</code></p>
<p><code>Error in fortify(): ! data must be a &lt;data.frame&gt;, or an object coercible by fortify(), not a &lt;uneval&gt; object.</code>:<br />
<strong>Solution:</strong> <code>aes()</code> is within <code>ggplot()</code> when it should be on <code>geom_...()</code>.</p>
</div>
<div id="introductory-exercises" class="section level2 unnumbered hasAnchor">
<h2 class="unnumbered hasAnchor">Introductory exercises<a href="#introductory-exercises" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="exercise-1-vectors" class="section level3 unnumbered hasAnchor">
<h3 class="unnumbered hasAnchor">Exercise 1: Vectors<a href="#exercise-1-vectors" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ol style="list-style-type: decimal">
<li><p>Create a vector named “test_scores” containing the test scores 75, 90, 65, 68 and 83 of students.</p></li>
<li><p>Create a vector “passing” that contains TRUE for test scores above 70 and FALSE for scores equal or below 70.
<em>Hint</em>: Remember that you can perform tests on vector elements.
Use <code>sum()</code> to check how many students passed the test.</p></li>
<li><p><em>(expert)</em> Can you create a vector “high_scores” that contains only scores above 70, using the results obtained in 1.
and 2.?</p></li>
</ol>
</div>
<div id="exercise-2-matrices" class="section level3 unnumbered hasAnchor">
<h3 class="unnumbered hasAnchor">Exercise 2: Matrices<a href="#exercise-2-matrices" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ol style="list-style-type: decimal">
<li>Create a matrix that looks like this:</li>
</ol>
<div class="sourceCode" id="cb70"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" tabindex="-1"></a><span class="co">#      [,1] [,2] [,3]</span></span>
<span id="cb70-2"><a href="#cb70-2" tabindex="-1"></a><span class="co"># [1,]    2    4    6</span></span>
<span id="cb70-3"><a href="#cb70-3" tabindex="-1"></a><span class="co"># [2,]    8    10   12</span></span>
<span id="cb70-4"><a href="#cb70-4" tabindex="-1"></a><span class="co"># [3,]    3    5    7</span></span>
<span id="cb70-5"><a href="#cb70-5" tabindex="-1"></a><span class="co"># [4,]    2    9    11</span></span></code></pre></div>
<ol start="2" style="list-style-type: decimal">
<li><p>Select different parts of the matrix, for example:</p>
<ul>
<li>the element of the second row and first column</li>
<li>the second and third element of the second column</li>
<li>all elements of the last column</li>
<li>Try something else!</li>
</ul></li>
<li><p>Add a new column to the matrix using <code>cbind()</code>.</p></li>
<li><p>Can you name the columns of the matrix as “A”, “B”, “C” and “D” using <code>colnames()</code>?</p></li>
</ol>
</div>
<div id="exercise-3-data-frames" class="section level3 unnumbered hasAnchor">
<h3 class="unnumbered hasAnchor">Exercise 3: Data Frames<a href="#exercise-3-data-frames" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ol style="list-style-type: decimal">
<li>Create a data frame from these vectors, with three columns and five rows:</li>
</ol>
<div class="sourceCode" id="cb71"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" tabindex="-1"></a>Product_name <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;orange&quot;</span>, <span class="st">&quot;strawberry&quot;</span>, <span class="st">&quot;broccoli&quot;</span>, <span class="st">&quot;blueberry&quot;</span>, <span class="st">&quot;cucumber&quot;</span>)</span>
<span id="cb71-2"><a href="#cb71-2" tabindex="-1"></a>Fruit <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="cn">TRUE</span>, <span class="cn">TRUE</span>, <span class="cn">FALSE</span>, <span class="cn">TRUE</span>, <span class="cn">FALSE</span>)</span>
<span id="cb71-3"><a href="#cb71-3" tabindex="-1"></a>Color <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;orange&quot;</span>, <span class="st">&quot;red&quot;</span>, <span class="st">&quot;green&quot;</span>, <span class="st">&quot;blue&quot;</span>, <span class="st">&quot;green&quot;</span>)</span></code></pre></div>
<ol start="2" style="list-style-type: decimal">
<li><p>Select the column “Fruit”.
How many Fruits are there?
<em>Hint</em>: Remember the function <code>sum()</code>.</p></li>
<li><p>Add a column named “Berry” containing TRUE (for products that are berries) or FALSE to the data frame.</p></li>
</ol>
<!--chapter:end:index.Rmd-->
</div>
</div>
</div>
<div id="lecture-files" class="section level1 hasAnchor">
<h1 class="hasAnchor">Lecture files<a href="#lecture-files" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<hr />
<div id="week-1---data-analysis-statistics" class="section level2 hasAnchor">
<h2 class="hasAnchor">Week 1 - data analysis &amp; statistics<a href="#week-1---data-analysis-statistics" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="day-1---data-typesgraphscorrelation" class="section level4 hasAnchor">
<h4 class="hasAnchor">Day 1 - data types/graphs/correlation<a href="#day-1---data-typesgraphscorrelation" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<ul>
<li><a href="./slides/GKBioinfo_week1_day1.pdf">[slides]</a></li>
<li><a href="./slides/solution_day1.Rmd">[solutions to exercises]</a></li>
</ul>
</div>
<div id="day-2---clustering-pca" class="section level4 hasAnchor">
<h4 class="hasAnchor">Day 2 - clustering &amp; PCA<a href="#day-2---clustering-pca" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<ul>
<li><a href="./slides/GKBioinfo_week1_day2.pdf">[slides]</a></li>
<li><a href="./slides/solution_day2.Rmd">[solutions to exercises]</a></li>
</ul>
</div>
<div id="day-3---distributions-statistical-inference" class="section level4 hasAnchor">
<h4 class="hasAnchor">Day 3 - distributions &amp; statistical inference<a href="#day-3---distributions-statistical-inference" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<ul>
<li><a href="./slides/GKBioinfo_week1_day3.pdf">[slides]</a></li>
<li><a href="./slides/solution_day3.Rmd">[solutions to exercises]</a></li>
</ul>
</div>
<div id="day-4---hypothesis-testing" class="section level4 hasAnchor">
<h4 class="hasAnchor">Day 4 - hypothesis testing<a href="#day-4---hypothesis-testing" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<ul>
<li><a href="./slides/GKBioinfo_week1_day4.pdf">[slides]</a></li>
<li><a href="./slides/solution_day4.Rmd">[solutions to exercises]</a></li>
</ul>
</div>
<div id="day-5---regression" class="section level4 hasAnchor">
<h4 class="hasAnchor">Day 5 - regression<a href="#day-5---regression" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<ul>
<li><a href="./slides/GKBioinfo_week1_day5.pdf">[slides]</a></li>
</ul>
<hr />
</div>
</div>
<div id="week-2---sequence-analysis" class="section level2 hasAnchor">
<h2 class="hasAnchor">Week 2 - sequence analysis<a href="#week-2---sequence-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="day-1---intro-to-sequence-analysis" class="section level4 hasAnchor">
<h4 class="hasAnchor">Day 1 - intro to sequence analysis<a href="#day-1---intro-to-sequence-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<ul>
<li><a href="./slides/GKBioinfo_week2.pdf">[slides]</a></li>
</ul>
</div>
<div id="day-2---blast-alignments-etc" class="section level4 hasAnchor">
<h4 class="hasAnchor">Day 2 - Blast, alignments, etc…<a href="#day-2---blast-alignments-etc" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<ul>
<li><a href="./slides/GKBioinfo_week2_2.pdf">[slides]</a></li>
</ul>
</div>
<div id="day-45---next-generation-sequencing-rna-seq" class="section level4 hasAnchor">
<h4 class="hasAnchor">Day 4/5 - Next generation sequencing, RNA-seq<a href="#day-45---next-generation-sequencing-rna-seq" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<ul>
<li><a href="./slides/2024_RNAseq_Part1.pdf">[slides RNA-seq - part 1]</a></li>
<li><a href="./slides/Oscar_Introduction%20RNAseq%201-2%202024.pdf">[slides RNA-seq - part 1.2]</a></li>
</ul>
<!--chapter:end:00-slides.Rmd-->
</div>
</div>
</div>
<div id="week-1-data-analysis" class="section level1 unnumbered hasAnchor">
<h1 class="unnumbered hasAnchor">Week 1 : data analysis<a href="#week-1-data-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<div id="day-1-descriptive-statistics-and-data-types" class="section level2 unnumbered hasAnchor">
<h2 class="unnumbered hasAnchor">Day 1: Descriptive statistics and data types<a href="#day-1-descriptive-statistics-and-data-types" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Today, you will learn how to perform basic tasks on a dataframe/tibble, descriptive statistics, perform data cleaning, and plotting.</p>
<div id="data-features-and-where-to-find-them" class="section level3 hasAnchor">
<h3 class="hasAnchor">Data features and where to find them<a href="#data-features-and-where-to-find-them" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="load-the-data" class="section level4 hasAnchor">
<h4 class="hasAnchor">Load the data<a href="#load-the-data" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>The diabetes dataset, which we will be using in this practical class will be downloaded from an online repository.
We will load that into R and have a sneak peek into how it looks like with the console.
In the following you will see several functions that give us information about our dataset.</p>
<div class="sourceCode" id="cb72"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" tabindex="-1"></a>dat <span class="ot">=</span> <span class="fu">as_tibble</span>(<span class="fu">read.delim</span>(<span class="st">&#39;https://tinyurl.com/y4fark9g&#39;</span>)) <span class="co"># Load the dataset</span></span>
<span id="cb72-2"><a href="#cb72-2" tabindex="-1"></a><span class="fu">head</span>(dat, <span class="dv">10</span>) <span class="co"># Look at the first 10 lines of the table</span></span></code></pre></div>
<pre><code>## # A tibble: 10 × 19
##       id  chol stab.glu   hdl ratio glyhb location     age gender height weight
##    &lt;int&gt; &lt;int&gt;    &lt;int&gt; &lt;int&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;      &lt;int&gt; &lt;chr&gt;   &lt;int&gt;  &lt;int&gt;
##  1  1000   203       82    56  3.60  4.31 Buckingham    46 female     62    121
##  2  1001   165       97    24  6.90  4.44 Buckingham    29 female     64    218
##  3  1002   228       92    37  6.20  4.64 Buckingham    58 female     61    256
##  4  1003    78       93    12  6.5   4.63 Buckingham    67 male       67    119
##  5  1005   249       90    28  8.90  7.72 Buckingham    64 male       68    183
##  6  1008   248       94    69  3.60  4.81 Buckingham    34 male       71    190
##  7  1011   195       92    41  4.80  4.84 Buckingham    30 male       69    191
##  8  1015   227       75    44  5.20  3.94 Buckingham    37 male       59    170
##  9  1016   177       87    49  3.60  4.84 Buckingham    45 male       69    166
## 10  1022   263       89    40  6.60  5.78 Buckingham    55 female     63    202
## # ℹ 8 more variables: frame &lt;chr&gt;, bp.1s &lt;int&gt;, bp.1d &lt;int&gt;, bp.2s &lt;int&gt;,
## #   bp.2d &lt;int&gt;, waist &lt;int&gt;, hip &lt;int&gt;, time.ppn &lt;int&gt;</code></pre>
</div>
<div id="dimensions-and-naming" class="section level4 hasAnchor">
<h4 class="hasAnchor">Dimensions and naming<a href="#dimensions-and-naming" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p><strong>1.</strong> What is the dimension of our dataset (i.e. how many rows/columns are there in our data)</p>
<div class="sourceCode" id="cb74"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" tabindex="-1"></a><span class="co"># Dimension</span></span>
<span id="cb74-2"><a href="#cb74-2" tabindex="-1"></a><span class="fu">dim</span>(dat)</span></code></pre></div>
<pre><code>## [1] 403  19</code></pre>
<div class="sourceCode" id="cb76"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" tabindex="-1"></a><span class="co"># Number of columns</span></span>
<span id="cb76-2"><a href="#cb76-2" tabindex="-1"></a><span class="fu">ncol</span>(dat)</span>
<span id="cb76-3"><a href="#cb76-3" tabindex="-1"></a><span class="co"># Number of rows</span></span>
<span id="cb76-4"><a href="#cb76-4" tabindex="-1"></a><span class="fu">nrow</span>(dat)</span></code></pre></div>
<p><strong>2.</strong> What are the column names of our dataset</p>
<div class="sourceCode" id="cb77"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" tabindex="-1"></a><span class="fu">colnames</span>(dat) <span class="co"># Similarly rownames() for rows</span></span></code></pre></div>
<pre><code>##  [1] &quot;id&quot;       &quot;chol&quot;     &quot;stab.glu&quot; &quot;hdl&quot;      &quot;ratio&quot;    &quot;glyhb&quot;   
##  [7] &quot;location&quot; &quot;age&quot;      &quot;gender&quot;   &quot;height&quot;   &quot;weight&quot;   &quot;frame&quot;   
## [13] &quot;bp.1s&quot;    &quot;bp.1d&quot;    &quot;bp.2s&quot;    &quot;bp.2d&quot;    &quot;waist&quot;    &quot;hip&quot;     
## [19] &quot;time.ppn&quot;</code></pre>
<p>Probably you are confused about what these column names mean.
For more description on these values <a href="https://biostat.app.vumc.org/wiki/pub/Main/DataSets/Cdiabetes.html">look here</a></p>
</div>
<div id="numerical-features" class="section level4 hasAnchor">
<h4 class="hasAnchor">Numerical features<a href="#numerical-features" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p><strong>3.</strong> How do we extract the minimum and maximum age of patients in our dataset?</p>
<div class="sourceCode" id="cb79"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" tabindex="-1"></a><span class="fu">min</span>(dat<span class="sc">$</span>age)</span>
<span id="cb79-2"><a href="#cb79-2" tabindex="-1"></a><span class="fu">max</span>(dat<span class="sc">$</span>age)</span>
<span id="cb79-3"><a href="#cb79-3" tabindex="-1"></a><span class="fu">range</span>(dat<span class="sc">$</span>age)</span></code></pre></div>
<blockquote>
<p>Can you find out the same for height and weight?</p>
</blockquote>
<p><strong>4.</strong> How does the overall summary of our entire dataset look like?</p>
<div class="sourceCode" id="cb80"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" tabindex="-1"></a><span class="fu">summary</span>(dat)</span></code></pre></div>
<blockquote>
<p>Can you explain what you see after you run the <code>summary()</code> function?</p>
</blockquote>
<p>Feel free to play around with this syntax until you feel comfortable with it.
You can open a window with <code>View(dat)</code> to compare your results.</p>
<hr />
</div>
</div>
<div id="data-cleaning" class="section level3 hasAnchor">
<h3 class="hasAnchor">Data cleaning<a href="#data-cleaning" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Very often the first thing one needs to do before any data science project is to clean up the raw data and transform it into a format that is readily understood and easy to use for all downstream analysis.
This process usually involves: –</p>
<ul>
<li>Removing empty value rows/columns</li>
<li>Removing unused or unnecessary rows/columns</li>
<li>Reordering the data matrix</li>
<li>Keeping columns uniformly numeric (age, weight etc) or string (names, places etc) or logical (TRUE/FALSE, 1/0)</li>
<li>Handling strange caveats which are data specific like replacing <code>,</code> or <code>.</code>, or <code>;</code> from numbers etc</li>
</ul>
<p>Lets do some clean up of our own diabetes data</p>
<ol style="list-style-type: decimal">
<li>We will make the <code>id</code> column the row names for the dataset;</li>
<li>We will remove the <code>bp.2s</code> and <code>bp.2d</code> columns as it has mostly missing values (see summary above);</li>
<li>We will also remove the column <code>time.ppn</code> which will not be required in our analysis;</li>
<li>We will reorder the columns of the data such that all the qualitative and quantitative values are separated.</li>
</ol>
<p>To perform this cleanup, we need a couple of important functions, that we will first discuss:</p>
<ul>
<li><code>filter</code></li>
<li><code>is.na</code></li>
<li><code>mutate</code></li>
<li><code>across</code></li>
<li><code>%in%</code></li>
</ul>
<div id="filter" class="section level4 hasAnchor">
<h4 class="hasAnchor">filter()<a href="#filter" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p><code>filter()</code> is used on a dataset to filter rows satisfying a condition you specify like we saw previously (Introduction).
Let’s look at an example. We are only filtering for senior individuals in our dataset.</p>
<div class="sourceCode" id="cb81"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1" tabindex="-1"></a>dat_seniors <span class="ot">=</span> dat <span class="sc">%&gt;%</span></span>
<span id="cb81-2"><a href="#cb81-2" tabindex="-1"></a>  <span class="fu">filter</span>(age <span class="sc">&lt;=</span> <span class="dv">65</span>)</span></code></pre></div>
<p>We can also filter based on other conditions, like location, sex, among others.
In some cases, we can also use <code>which()</code> to filter values. The syntax is different…</p>
<div class="sourceCode" id="cb82"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" tabindex="-1"></a>dat[dat<span class="sc">$</span>age <span class="sc">&lt;=</span> <span class="dv">65</span>,]</span></code></pre></div>
<p>… but it works for vectors and other classes. Let’s see the next example.</p>
<div class="sourceCode" id="cb83"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb83-1"><a href="#cb83-1" tabindex="-1"></a><span class="co"># number of animals you have</span></span>
<span id="cb83-2"><a href="#cb83-2" tabindex="-1"></a>number <span class="ot">=</span> <span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">4</span>,<span class="dv">5</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">5</span>)</span>
<span id="cb83-3"><a href="#cb83-3" tabindex="-1"></a><span class="co"># Let&#39;s create a different vector (of the same length)</span></span>
<span id="cb83-4"><a href="#cb83-4" tabindex="-1"></a>animals <span class="ot">=</span> <span class="fu">c</span>(<span class="st">&quot;cat&quot;</span>, <span class="st">&quot;dog&quot;</span>, <span class="st">&quot;cow&quot;</span>, <span class="st">&quot;parrot&quot;</span>, <span class="st">&quot;zebra&quot;</span>, <span class="st">&quot;sparrow&quot;</span>, <span class="st">&quot;lizard&quot;</span>)</span>
<span id="cb83-5"><a href="#cb83-5" tabindex="-1"></a><span class="co"># Let&#39;s use the &quot;which()&quot; function now</span></span>
<span id="cb83-6"><a href="#cb83-6" tabindex="-1"></a>animals[<span class="fu">which</span>(number <span class="sc">&gt;</span> <span class="dv">2</span>)]</span></code></pre></div>
<pre><code>## [1] &quot;dog&quot;    &quot;cow&quot;    &quot;parrot&quot; &quot;lizard&quot;</code></pre>
<p>We selected all animals from the “animals” vector that correspond to more than three individuals in the “number” vector.</p>
</div>
<div id="is.na" class="section level4 hasAnchor">
<h4 class="hasAnchor">is.na()<a href="#is.na" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p><code>is.na()</code> is used to determine if NA values are present in a given object. We can try a simple example with one variable being assigned as NA.</p>
<div class="sourceCode" id="cb85"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb85-1"><a href="#cb85-1" tabindex="-1"></a>x <span class="ot">=</span> <span class="dv">2</span></span>
<span id="cb85-2"><a href="#cb85-2" tabindex="-1"></a><span class="fu">is.na</span>(x)</span></code></pre></div>
<pre><code>## [1] FALSE</code></pre>
<div class="sourceCode" id="cb87"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb87-1"><a href="#cb87-1" tabindex="-1"></a>y <span class="ot">=</span> <span class="cn">NA</span></span>
<span id="cb87-2"><a href="#cb87-2" tabindex="-1"></a><span class="fu">is.na</span>(y)</span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<p>We can do this with vectors obtained from <code>dat</code>. What class is the output in?</p>
<div class="sourceCode" id="cb89"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb89-1"><a href="#cb89-1" tabindex="-1"></a><span class="fu">is.na</span>(dat<span class="sc">$</span>glyhb)</span></code></pre></div>
</div>
<div id="mutate" class="section level4 hasAnchor">
<h4 class="hasAnchor">mutate()<a href="#mutate" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p><code>mutate()</code> is often used to create a new column based on another column of the dataframe. Let us use this function to <em>mutate</em> two new columns including the weight in kilograms and the height in centimeters. <strong>The conversion from pounds to kilograms can be done by multiplying weight in pounds by 0.454. To covert height to centimeters we only need to multiply height (inches) by 2.54</strong>.</p>
<div class="sourceCode" id="cb90"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb90-1"><a href="#cb90-1" tabindex="-1"></a>dat <span class="sc">%&gt;%</span></span>
<span id="cb90-2"><a href="#cb90-2" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">weight.kg =</span> weight <span class="sc">*</span> <span class="fl">0.454</span>,        <span class="co"># you can generate both columns using the same mutate!</span></span>
<span id="cb90-3"><a href="#cb90-3" tabindex="-1"></a>         <span class="at">height.cm =</span> height <span class="sc">*</span> <span class="fl">2.54</span>) <span class="sc">%&gt;%</span>     <span class="co"># we do not need to save this output</span></span>
<span id="cb90-4"><a href="#cb90-4" tabindex="-1"></a>  <span class="fu">select</span>(id, weight.kg, height.cm)</span></code></pre></div>
<pre><code>## # A tibble: 403 × 3
##       id weight.kg height.cm
##    &lt;int&gt;     &lt;dbl&gt;     &lt;dbl&gt;
##  1  1000      54.9      157.
##  2  1001      99.0      163.
##  3  1002     116.       155.
##  4  1003      54.0      170.
##  5  1005      83.1      173.
##  6  1008      86.3      180.
##  7  1011      86.7      175.
##  8  1015      77.2      150.
##  9  1016      75.4      175.
## 10  1022      91.7      160.
## # ℹ 393 more rows</code></pre>
</div>
<div id="across" class="section level4 hasAnchor">
<h4 class="hasAnchor">across()<a href="#across" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p><code>across()</code> is very often used together with <code>mutate()</code> and another helper function, like <code>everywhere()</code>, <code>starts_with()</code>, <code>ends_with()</code>, or <code>contains()</code>.
Later, we will use <code>across()</code> together with the other functions we learned previously to remove NAs like this:</p>
<div class="sourceCode" id="cb92"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb92-1"><a href="#cb92-1" tabindex="-1"></a>dat <span class="sc">%&gt;%</span></span>
<span id="cb92-2"><a href="#cb92-2" tabindex="-1"></a>  <span class="fu">rowwise</span>() <span class="sc">%&gt;%</span></span>
<span id="cb92-3"><a href="#cb92-3" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">na_count =</span> <span class="fu">sum</span>(<span class="fu">is.na</span>(<span class="fu">across</span>(<span class="fu">everything</span>()))))</span></code></pre></div>
<p>There is much to unpack here:</p>
<ul>
<li><code>rowwise()</code> ensures that the next operations are applied by row.<br />
</li>
<li><code>mutate()</code> adds a new column called na_count to the dataframe.<br />
</li>
<li><code>across(everything())</code> selects all columns in the current row.<br />
</li>
<li><code>sum(is.na(...))</code> calculates the sum of missing values for each row.</li>
</ul>
<blockquote>
<p>Try to run the previous example without <code>rowwise()</code>. What does it look like?</p>
</blockquote>
</div>
<div id="in" class="section level4 hasAnchor">
<h4 class="hasAnchor">%in%<a href="#in" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>This is an operator to check which elements of a first vector are inside a second vector.</p>
<div class="sourceCode" id="cb93"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb93-1"><a href="#cb93-1" tabindex="-1"></a><span class="fu">c</span>(<span class="st">&#39;Frodo&#39;</span>,<span class="st">&#39;Sam&#39;</span>) <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">&#39;Pippin&#39;</span>,<span class="st">&#39;Sam&#39;</span>,<span class="st">&#39;Frodo&#39;</span>,<span class="st">&#39;Merry&#39;</span>)</span></code></pre></div>
<pre><code>## [1] TRUE TRUE</code></pre>
</div>
<div id="ready-for-the-cleaning" class="section level4 hasAnchor">
<h4 class="hasAnchor">Ready for the cleaning!<a href="#ready-for-the-cleaning" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>The first column of the dataframe is the column with the name “id”.
The rows are just numbered, without names.
We are going to rename the rows using the column “id”. The function <code>column_to_rownames()</code> allows us to do this efficiently.</p>
<div class="sourceCode" id="cb95"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb95-1"><a href="#cb95-1" tabindex="-1"></a><span class="co"># set the row names using the column id</span></span>
<span id="cb95-2"><a href="#cb95-2" tabindex="-1"></a>dat <span class="ot">=</span> dat <span class="sc">%&gt;%</span></span>
<span id="cb95-3"><a href="#cb95-3" tabindex="-1"></a>  <span class="fu">column_to_rownames</span>(<span class="at">var =</span> <span class="st">&#39;id&#39;</span>)</span></code></pre></div>
<blockquote>
<p>Keep in mind that rownames must be unique!</p>
</blockquote>
<p>The <em>na_count</em> column will then include the number of NAs per row. Do you understand how it works?
We finally apply <code>filter</code> again to keep only rows with less than or 2 NAs.</p>
<div class="sourceCode" id="cb96"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb96-1"><a href="#cb96-1" tabindex="-1"></a>dat <span class="ot">=</span> dat <span class="sc">%&gt;%</span></span>
<span id="cb96-2"><a href="#cb96-2" tabindex="-1"></a>  <span class="fu">rowwise</span>() <span class="sc">%&gt;%</span></span>
<span id="cb96-3"><a href="#cb96-3" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">na_count =</span> <span class="fu">sum</span>(<span class="fu">is.na</span>(<span class="fu">across</span>(<span class="fu">everything</span>())))) <span class="sc">%&gt;%</span></span>
<span id="cb96-4"><a href="#cb96-4" tabindex="-1"></a>  <span class="fu">filter</span>(na_count <span class="sc">&lt;=</span> <span class="dv">2</span>)</span></code></pre></div>
<p>We will also remove the <em>na_count</em> and some problematic columns (<em>bp.2s</em>, <em>bp.2d</em> and <em>time.ppn</em>) by <strong>selecting the ones which are not these</strong>. We can do this using <code>!</code>, as this character can be used to invert results. Let us try it with <code>select()</code>.</p>
<div class="sourceCode" id="cb97"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb97-1"><a href="#cb97-1" tabindex="-1"></a>dat <span class="ot">=</span> dat <span class="sc">%&gt;%</span></span>
<span id="cb97-2"><a href="#cb97-2" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">!</span><span class="fu">c</span>(na_count, time.ppn, bp<span class="fl">.2</span>d, bp<span class="fl">.2</span>s))</span></code></pre></div>
<p>Next, we can re-order the remaining columns, in order to put the categorical columns first, and numerical columns after. We can use <code>select</code> to order columns too, but we need to combine it with <code>where()</code> and functions which verify the class of the columns, like <code>is.character()</code> or <code>is.numeric()</code>.</p>
<p>Here is a simple example:</p>
<div class="sourceCode" id="cb98"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb98-1"><a href="#cb98-1" tabindex="-1"></a><span class="co"># Create a character and numeric </span></span>
<span id="cb98-2"><a href="#cb98-2" tabindex="-1"></a>name <span class="ot">=</span> <span class="fu">c</span>(<span class="st">&quot;Antonia&quot;</span>)</span>
<span id="cb98-3"><a href="#cb98-3" tabindex="-1"></a>age <span class="ot">=</span> <span class="fu">c</span>(<span class="dv">23</span>)</span>
<span id="cb98-4"><a href="#cb98-4" tabindex="-1"></a></span>
<span id="cb98-5"><a href="#cb98-5" tabindex="-1"></a><span class="co"># Verify if the previous object are from the character/numeric classes</span></span>
<span id="cb98-6"><a href="#cb98-6" tabindex="-1"></a><span class="fu">is.character</span>(name)</span>
<span id="cb98-7"><a href="#cb98-7" tabindex="-1"></a><span class="fu">is.character</span>(age)</span>
<span id="cb98-8"><a href="#cb98-8" tabindex="-1"></a><span class="fu">is.numeric</span>(age)</span></code></pre></div>
<p>And here we can apply the same principle to the re-ordering:</p>
<div class="sourceCode" id="cb99"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb99-1"><a href="#cb99-1" tabindex="-1"></a>dat <span class="ot">&lt;-</span> dat <span class="sc">%&gt;%</span></span>
<span id="cb99-2"><a href="#cb99-2" tabindex="-1"></a>  <span class="fu">select</span>(</span>
<span id="cb99-3"><a href="#cb99-3" tabindex="-1"></a>    <span class="co"># Select categorical columns</span></span>
<span id="cb99-4"><a href="#cb99-4" tabindex="-1"></a>    <span class="fu">where</span>(is.character), </span>
<span id="cb99-5"><a href="#cb99-5" tabindex="-1"></a>    <span class="co"># Select numerical columns</span></span>
<span id="cb99-6"><a href="#cb99-6" tabindex="-1"></a>    <span class="fu">where</span>(is.numeric)</span>
<span id="cb99-7"><a href="#cb99-7" tabindex="-1"></a>  )</span>
<span id="cb99-8"><a href="#cb99-8" tabindex="-1"></a></span>
<span id="cb99-9"><a href="#cb99-9" tabindex="-1"></a><span class="co"># OR you can use the indexes too, but if you more than 10-20 columns, that is not ideal</span></span>
<span id="cb99-10"><a href="#cb99-10" tabindex="-1"></a><span class="co"># dat = dat[,c(8,6,11,9,10,14,15,2,5,1,3,4,12,13)]</span></span></code></pre></div>
<p>Now lets look at our cleaned data:</p>
<div class="sourceCode" id="cb100"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb100-1"><a href="#cb100-1" tabindex="-1"></a><span class="fu">summary</span>(dat)</span></code></pre></div>
<pre><code>##    location            gender             frame                chol      
##  Length:377         Length:377         Length:377         Min.   : 78.0  
##  Class :character   Class :character   Class :character   1st Qu.:179.0  
##  Mode  :character   Mode  :character   Mode  :character   Median :204.0  
##                                                           Mean   :208.2  
##                                                           3rd Qu.:230.0  
##                                                           Max.   :443.0  
##                                                                          
##     stab.glu          hdl             ratio            glyhb       
##  Min.   : 48.0   Min.   : 12.00   Min.   : 1.500   Min.   : 2.680  
##  1st Qu.: 81.0   1st Qu.: 38.00   1st Qu.: 3.200   1st Qu.: 4.390  
##  Median : 90.0   Median : 46.00   Median : 4.200   Median : 4.860  
##  Mean   :107.3   Mean   : 50.36   Mean   : 4.538   Mean   : 5.594  
##  3rd Qu.:108.0   3rd Qu.: 59.00   3rd Qu.: 5.400   3rd Qu.: 5.622  
##  Max.   :385.0   Max.   :120.00   Max.   :19.300   Max.   :16.110  
##                                                    NA&#39;s   :3       
##       age           height          weight          bp.1s      
##  Min.   :19.0   Min.   :52.00   Min.   : 99.0   Min.   : 90.0  
##  1st Qu.:34.0   1st Qu.:63.00   1st Qu.:151.0   1st Qu.:122.0  
##  Median :45.0   Median :66.00   Median :174.0   Median :136.0  
##  Mean   :46.9   Mean   :66.02   Mean   :178.1   Mean   :137.4  
##  3rd Qu.:60.0   3rd Qu.:69.00   3rd Qu.:200.0   3rd Qu.:148.0  
##  Max.   :92.0   Max.   :76.00   Max.   :325.0   Max.   :250.0  
##                                                                
##      bp.1d            waist            hip       
##  Min.   : 48.00   Min.   :26.00   Min.   :30.00  
##  1st Qu.: 75.00   1st Qu.:33.00   1st Qu.:39.00  
##  Median : 82.00   Median :37.00   Median :42.00  
##  Mean   : 83.69   Mean   :37.95   Mean   :43.08  
##  3rd Qu.: 92.00   3rd Qu.:41.25   3rd Qu.:46.00  
##  Max.   :124.00   Max.   :56.00   Max.   :64.00  
##                   NA&#39;s   :1       NA&#39;s   :1</code></pre>
<p>Hold up, the ordering and selection of columns looks right, but it seems that there are certain rows that have missing values still (like <code>glyhb</code> column has 3 <code>NA</code> values still).
Lets remove all rows with any missing value using <code>na.omit()</code>.
Remember, 1 row = 1 patient.</p>
<div class="sourceCode" id="cb102"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb102-1"><a href="#cb102-1" tabindex="-1"></a>dat <span class="ot">=</span> dat <span class="sc">%&gt;%</span></span>
<span id="cb102-2"><a href="#cb102-2" tabindex="-1"></a>  <span class="fu">na.omit</span>()</span></code></pre></div>
<blockquote>
<p>How many patients were removed because they were associated with missing values?</p>
</blockquote>
<p>Now our cleaned data has no missing values, columns are cleanly ordered and each column is in the right format</p>
<div class="sourceCode" id="cb103"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb103-1"><a href="#cb103-1" tabindex="-1"></a><span class="fu">summary</span>(dat)</span></code></pre></div>
<pre><code>##    location            gender             frame                chol      
##  Length:367         Length:367         Length:367         Min.   : 78.0  
##  Class :character   Class :character   Class :character   1st Qu.:179.0  
##  Mode  :character   Mode  :character   Mode  :character   Median :204.0  
##                                                           Mean   :207.5  
##                                                           3rd Qu.:229.0  
##                                                           Max.   :443.0  
##     stab.glu          hdl             ratio            glyhb       
##  Min.   : 48.0   Min.   : 12.00   Min.   : 1.500   Min.   : 2.680  
##  1st Qu.: 81.0   1st Qu.: 38.00   1st Qu.: 3.200   1st Qu.: 4.390  
##  Median : 90.0   Median : 46.00   Median : 4.200   Median : 4.860  
##  Mean   :107.3   Mean   : 50.28   Mean   : 4.536   Mean   : 5.602  
##  3rd Qu.:108.0   3rd Qu.: 59.00   3rd Qu.: 5.400   3rd Qu.: 5.630  
##  Max.   :385.0   Max.   :120.00   Max.   :19.300   Max.   :16.110  
##       age            height          weight          bp.1s      
##  Min.   :19.00   Min.   :52.00   Min.   : 99.0   Min.   : 90.0  
##  1st Qu.:34.00   1st Qu.:63.00   1st Qu.:151.0   1st Qu.:121.5  
##  Median :45.00   Median :66.00   Median :174.0   Median :136.0  
##  Mean   :46.68   Mean   :66.05   Mean   :178.1   Mean   :137.1  
##  3rd Qu.:60.00   3rd Qu.:69.00   3rd Qu.:200.0   3rd Qu.:148.0  
##  Max.   :92.00   Max.   :76.00   Max.   :325.0   Max.   :250.0  
##      bp.1d           waist            hip       
##  Min.   : 48.0   Min.   :26.00   Min.   :30.00  
##  1st Qu.: 75.0   1st Qu.:33.00   1st Qu.:39.00  
##  Median : 82.0   Median :37.00   Median :42.00  
##  Mean   : 83.4   Mean   :37.93   Mean   :43.04  
##  3rd Qu.: 92.0   3rd Qu.:41.50   3rd Qu.:46.00  
##  Max.   :124.0   Max.   :56.00   Max.   :64.00</code></pre>
<blockquote>
<p>Can you identify which types of data (continuous, discrete etc) each column above represents and why?</p>
</blockquote>
<hr />
</div>
</div>
<div id="visualizing-data-distribution" class="section level3 hasAnchor">
<h3 class="hasAnchor">Visualizing data distribution<a href="#visualizing-data-distribution" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>In this section you will also learn the essential functions to plot data in an intuitive and useful way using the <code>ggplot2</code> package, just like in the introductory section to tidyverse.</p>
<div id="histograms" class="section level4 hasAnchor">
<h4 class="hasAnchor">Histograms<a href="#histograms" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>We can plot the column “stab.glu” as a histogram using the <code>hist()</code> function:</p>
<div class="sourceCode" id="cb105"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb105-1"><a href="#cb105-1" tabindex="-1"></a><span class="fu">ggplot</span>(dat,</span>
<span id="cb105-2"><a href="#cb105-2" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> stab.glu)) <span class="sc">+</span></span>
<span id="cb105-3"><a href="#cb105-3" tabindex="-1"></a>  <span class="fu">geom_histogram</span>() <span class="sc">+</span></span>
<span id="cb105-4"><a href="#cb105-4" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&quot;Stabilized Glucose concentration in blood&quot;</span>,  <span class="co"># add labels to the x-axis</span></span>
<span id="cb105-5"><a href="#cb105-5" tabindex="-1"></a>       <span class="at">title =</span> <span class="st">&quot;Glucose concentration&quot;</span>)                  <span class="co"># add title</span></span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-47-1.png" width="672" /></p>
<blockquote>
<p>Add the parameter <code>bins = 50</code> in the above lines of code (inside <code>geom_histogram</code>) and see what happens.
Try different values for <code>bins</code> like <code>10, 20, 75, 100</code>. Can you interpret the differences?
Is this a good or bad thing about histograms?</p>
</blockquote>
</div>
<div id="density-plots" class="section level4 hasAnchor">
<h4 class="hasAnchor">Density plots<a href="#density-plots" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>For density plots, we use the <code>geom_density()</code> function to estimate the probability density function for a given variable.</p>
<div class="sourceCode" id="cb106"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb106-1"><a href="#cb106-1" tabindex="-1"></a><span class="fu">ggplot</span>(dat,</span>
<span id="cb106-2"><a href="#cb106-2" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> stab.glu)) <span class="sc">+</span></span>
<span id="cb106-3"><a href="#cb106-3" tabindex="-1"></a>  <span class="fu">geom_density</span>() <span class="sc">+</span></span>
<span id="cb106-4"><a href="#cb106-4" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&quot;Stabilized Glucose concentration in blood&quot;</span>,  <span class="co"># add labels to the x-axis</span></span>
<span id="cb106-5"><a href="#cb106-5" tabindex="-1"></a>       <span class="at">title =</span> <span class="st">&quot;Glucose concentration&quot;</span>)                  <span class="co"># add title</span></span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-48-1.png" width="672" /></p>
</div>
<div id="boxplots" class="section level4 hasAnchor">
<h4 class="hasAnchor">Boxplots<a href="#boxplots" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>The <code>boxplot()</code> function produces a boxplot for a given variable:</p>
<div class="sourceCode" id="cb107"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb107-1"><a href="#cb107-1" tabindex="-1"></a><span class="fu">ggplot</span>(dat,</span>
<span id="cb107-2"><a href="#cb107-2" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> stab.glu)) <span class="sc">+</span></span>
<span id="cb107-3"><a href="#cb107-3" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb107-4"><a href="#cb107-4" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span><span class="st">&quot;Stabilized Glucose concentration in blood&quot;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-49-1.png" width="384" /></p>
<blockquote>
<p>Can you explain all features of this graph, such as upper/lower whisker, 25% quantile, …?</p>
</blockquote>
</div>
<div id="qq-plots" class="section level4 hasAnchor">
<h4 class="hasAnchor">QQ-plots<a href="#qq-plots" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>We can use <strong>QQ-plots</strong> to either (1) compare two distributions, or (2) compare a distribution with a theoretical distribution (typically the normal distribution).</p>
<p>We can for example compare the distribution of the blood pressure values to check if they are normally distributed</p>
<div class="sourceCode" id="cb108"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb108-1"><a href="#cb108-1" tabindex="-1"></a><span class="do">## Let&#39;s first make a histogram</span></span>
<span id="cb108-2"><a href="#cb108-2" tabindex="-1"></a><span class="fu">ggplot</span>(dat,</span>
<span id="cb108-3"><a href="#cb108-3" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> bp<span class="fl">.1</span>s)) <span class="sc">+</span></span>
<span id="cb108-4"><a href="#cb108-4" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins =</span> <span class="dv">50</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-50-1.png" width="672" /></p>
<p>Now we can use the function <code>geom_qq()</code> to generate the <strong>QQ-plot</strong> of this distribution against the standard normal distribution:</p>
<div class="sourceCode" id="cb109"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb109-1"><a href="#cb109-1" tabindex="-1"></a><span class="fu">ggplot</span>(dat,</span>
<span id="cb109-2"><a href="#cb109-2" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">sample =</span> bp<span class="fl">.1</span>s)) <span class="sc">+</span>     <span class="co"># we use sample= inside aes for the QQ-plot</span></span>
<span id="cb109-3"><a href="#cb109-3" tabindex="-1"></a>  <span class="fu">geom_qq</span>()                       <span class="co"># creates the QQ-plot</span></span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-51-1.png" width="288" /></p>
<p>Using the additional command <code>geom_qq_line()</code>, we can add a straight line that goes through the first and third quartile:</p>
<div class="sourceCode" id="cb110"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb110-1"><a href="#cb110-1" tabindex="-1"></a><span class="fu">ggplot</span>(dat,</span>
<span id="cb110-2"><a href="#cb110-2" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">sample =</span> bp<span class="fl">.1</span>s)) <span class="sc">+</span>     <span class="co"># we use sample= inside aes for the QQ-plot</span></span>
<span id="cb110-3"><a href="#cb110-3" tabindex="-1"></a>  <span class="fu">geom_qq</span>() <span class="sc">+</span></span>
<span id="cb110-4"><a href="#cb110-4" tabindex="-1"></a>  <span class="fu">geom_qq_line</span>(<span class="at">colour =</span> <span class="st">&#39;red&#39;</span>)                  <span class="co"># adds in the QQ-line on top</span></span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-52-1.png" width="288" /></p>
<blockquote>
<p>So, is the distribution normal??</p>
</blockquote>
<p>Now let’s compare the quantiles of the cholesterol values by biological sex.
<strong>Notes on <code>ggplot()</code> here:</strong> Rather than <code>ggplot(dataset, aes(...))</code> we use <code>ggplot() + geom_xx(aes(...))</code> for situations where the data we wish to plot is not in a dataframe.</p>
<div class="sourceCode" id="cb111"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb111-1"><a href="#cb111-1" tabindex="-1"></a><span class="co"># We can use &quot;filter()&quot; to filter the cholesterol values for men and women</span></span>
<span id="cb111-2"><a href="#cb111-2" tabindex="-1"></a>dat.male <span class="ot">=</span> dat <span class="sc">%&gt;%</span></span>
<span id="cb111-3"><a href="#cb111-3" tabindex="-1"></a>  <span class="fu">filter</span>(gender <span class="sc">==</span> <span class="st">&#39;male&#39;</span>)</span>
<span id="cb111-4"><a href="#cb111-4" tabindex="-1"></a></span>
<span id="cb111-5"><a href="#cb111-5" tabindex="-1"></a>dat.female <span class="ot">=</span> dat <span class="sc">%&gt;%</span></span>
<span id="cb111-6"><a href="#cb111-6" tabindex="-1"></a>  <span class="fu">filter</span>(gender <span class="sc">==</span> <span class="st">&#39;female&#39;</span>)</span>
<span id="cb111-7"><a href="#cb111-7" tabindex="-1"></a></span>
<span id="cb111-8"><a href="#cb111-8" tabindex="-1"></a><span class="co"># Compute the quantiles (note the &quot;na.rm&quot; option to ignore missing NA values!)</span></span>
<span id="cb111-9"><a href="#cb111-9" tabindex="-1"></a>q.male <span class="ot">=</span> <span class="fu">quantile</span>(dat.male<span class="sc">$</span>bp<span class="fl">.1</span>s, </span>
<span id="cb111-10"><a href="#cb111-10" tabindex="-1"></a>                  <span class="at">probs=</span><span class="fu">seq</span>(<span class="dv">0</span>,<span class="dv">1</span>,<span class="at">by=</span><span class="fl">0.05</span>), </span>
<span id="cb111-11"><a href="#cb111-11" tabindex="-1"></a>                  <span class="at">na.rm=</span><span class="cn">TRUE</span>)</span>
<span id="cb111-12"><a href="#cb111-12" tabindex="-1"></a>q.female <span class="ot">=</span> <span class="fu">quantile</span>(dat.female<span class="sc">$</span>bp<span class="fl">.1</span>s, </span>
<span id="cb111-13"><a href="#cb111-13" tabindex="-1"></a>                    <span class="at">probs=</span><span class="fu">seq</span>(<span class="dv">0</span>,<span class="dv">1</span>,<span class="at">by=</span><span class="fl">0.05</span>),</span>
<span id="cb111-14"><a href="#cb111-14" tabindex="-1"></a>                    <span class="at">na.rm=</span><span class="cn">TRUE</span>)</span>
<span id="cb111-15"><a href="#cb111-15" tabindex="-1"></a></span>
<span id="cb111-16"><a href="#cb111-16" tabindex="-1"></a><span class="co"># Now plot against each other!</span></span>
<span id="cb111-17"><a href="#cb111-17" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb111-18"><a href="#cb111-18" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> q.male, <span class="at">y =</span> q.female)) <span class="sc">+</span></span>
<span id="cb111-19"><a href="#cb111-19" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">&quot;Quantiles&quot;</span>, <span class="at">x =</span> <span class="st">&quot;Male quantiles&quot;</span>, <span class="at">y =</span> <span class="st">&quot;Female quantiles&quot;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-53-1.png" width="288" /></p>
<hr />
</div>
</div>
<div id="correlation" class="section level3 hasAnchor">
<h3 class="hasAnchor">Correlation<a href="#correlation" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="measuring-the-centrality-in-data" class="section level4 hasAnchor">
<h4 class="hasAnchor">Measuring the centrality in data<a href="#measuring-the-centrality-in-data" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Before you begin, think back to the mean, median and quantiles we saw on the boxplot. Do you remember what these terms mean? How does an asymmetrical distribution influence mean and median?
We have already seen that the <code>summary()</code> and <code>quantile()</code> functions in R can compute the mean, median and quantiles of any given data.</p>
<div class="sourceCode" id="cb112"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb112-1"><a href="#cb112-1" tabindex="-1"></a><span class="fu">mean</span>(dat<span class="sc">$</span>stab.glu) </span>
<span id="cb112-2"><a href="#cb112-2" tabindex="-1"></a><span class="fu">median</span>(dat<span class="sc">$</span>stab.glu) </span>
<span id="cb112-3"><a href="#cb112-3" tabindex="-1"></a><span class="fu">quantile</span>(dat<span class="sc">$</span>stab.glu) </span></code></pre></div>
<blockquote>
<p>Calculate the mean and median of other continuous numeric data in the diabetes dataset and measure the difference between them. (a) Why is there a difference between the mean and median? (b) Why do you think there are larger differences for some and almost no difference for others?</p>
</blockquote>
</div>
<div id="association-between-variables" class="section level4 hasAnchor">
<h4 class="hasAnchor">Association between variables<a href="#association-between-variables" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Often a common step during any data analysis project is to find associations between variables present in the dataset. Such associations helps us to decipher the underlying structure in the data.</p>
<p>For instance, in our diabetes dataset we would expect a high correlation between free blood glucose levels and glycosylated blood levels or between waist and hip sizes. One of the most common ways of measuring associations is <em>correlations</em>.</p>
<p>Let us start by producing a <strong>scatter plot</strong> between a pair of variables:</p>
<div class="sourceCode" id="cb113"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb113-1"><a href="#cb113-1" tabindex="-1"></a><span class="fu">ggplot</span>(dat,</span>
<span id="cb113-2"><a href="#cb113-2" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> stab.glu, <span class="at">y =</span> glyhb)) <span class="sc">+</span></span>
<span id="cb113-3"><a href="#cb113-3" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb113-4"><a href="#cb113-4" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x=</span><span class="st">&#39;Stabilized glucose&#39;</span>, <span class="at">y=</span><span class="st">&#39;Glycosylated hemoglobin&#39;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-55-1.png" width="672" /></p>
<blockquote>
<p>Do you suspect that the two variables have a relationship? Do the scatter plot for other pairs of numerical variables!</p>
</blockquote>
<p>We now can compute the correlation of the two variables. We can compute the <strong>Pearson correlation</strong> or the <strong>Spearman correlation</strong>:</p>
<div class="sourceCode" id="cb114"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb114-1"><a href="#cb114-1" tabindex="-1"></a><span class="do">## compute the Pearson correlation</span></span>
<span id="cb114-2"><a href="#cb114-2" tabindex="-1"></a><span class="fu">cor</span>(dat<span class="sc">$</span>stab.glu, dat<span class="sc">$</span>glyhb, <span class="at">method=</span><span class="st">&#39;pearson&#39;</span>)</span></code></pre></div>
<pre><code>## [1] 0.7411355</code></pre>
<div class="sourceCode" id="cb116"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb116-1"><a href="#cb116-1" tabindex="-1"></a><span class="do">## compute the Spearman correlation</span></span>
<span id="cb116-2"><a href="#cb116-2" tabindex="-1"></a><span class="fu">cor</span>(dat<span class="sc">$</span>stab.glu, dat<span class="sc">$</span>glyhb, <span class="at">method=</span><span class="st">&#39;spearman&#39;</span>)</span></code></pre></div>
<pre><code>## [1] 0.5214866</code></pre>
<p>The Spearman correlation seems much lower, right? To understand why, we can do a scatter plot between the <strong>ranks</strong> of the two variables:</p>
<div class="sourceCode" id="cb118"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb118-1"><a href="#cb118-1" tabindex="-1"></a><span class="fu">ggplot</span>(dat,</span>
<span id="cb118-2"><a href="#cb118-2" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">rank</span>(stab.glu), <span class="at">y =</span> <span class="fu">rank</span>(glyhb))) <span class="sc">+</span></span>
<span id="cb118-3"><a href="#cb118-3" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb118-4"><a href="#cb118-4" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x=</span><span class="st">&#39;Rank - Stabilized glucose&#39;</span>, <span class="at">y=</span><span class="st">&#39;Rank - Glycosylated hemoglobin&#39;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-57-1.png" width="672" /></p>
<p>Do you understand the usage of ranks here? Run `rank()`` on a vector like <em>c(3,5,10,1,23)</em> to see how the output looks like.</p>
<blockquote>
<p>Associations are among the simplest forms of structure in the data! It is important to remember that <em>Association does not imply correlation</em> and <em>Correlation does not imply causation</em>. Take a look at this page to view few common logical fallacies. <a href="https://en.wikipedia.org/wiki/Fallacy">see here</a></p>
</blockquote>
<hr />
</div>
</div>
<div id="exercises" class="section level3 hasAnchor">
<h3 class="hasAnchor">EXERCISES<a href="#exercises" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="exercise-1-data-features" class="section level4 hasAnchor">
<h4 class="hasAnchor">Exercise 1: Data features<a href="#exercise-1-data-features" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<ol style="list-style-type: decimal">
<li><p>Try to obtain the same result of the <code>head()</code> function by using slicing on the dataset “dat” using row indexes.</p></li>
<li><p>Print out the last element in the last column of “dat” using the <code>dim()</code> function instead of using numerals.</p></li>
</ol>
</div>
<div id="exercise-2-visualization-and-correlation" class="section level4 hasAnchor">
<h4 class="hasAnchor">Exercise 2: Visualization and correlation<a href="#exercise-2-visualization-and-correlation" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<ol style="list-style-type: decimal">
<li><p>Visualize the cholesterol levels of all patients with a histogram using the <code>geom_histogram()</code> function.</p></li>
<li><p>Visualize the cholesterol levels of all <strong>male</strong> patients with a histogram using <code>geom_histogram()</code>.
<em>(expert):</em> Mark the median, first and third quartile with vertical lines using <code>geom_vline()</code>. The values are defined using <code>xintercept</code> (as in which value of x should the vertical line intercept).
Then mark median, first and third quantile for <strong>female</strong> patients in the same graph with a different color.
What can you tell from the differences in these values?</p></li>
<li><p>Is there an association between “hip” and “waist” on the data frame “dat”? Use the <code>geom_point()</code> function to do a scatter plot of the <strong>values</strong> and of the <strong>ranks</strong> (as determined by the <code>rank()</code> function). Compute both the <strong>Pearson</strong> and <strong>Spearman</strong> correlation values.</p></li>
</ol>
</div>
<div id="going-further-expert" class="section level4 hasAnchor">
<h4 class="hasAnchor">Going further <em>(expert)</em><a href="#going-further-expert" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<ol style="list-style-type: decimal">
<li><p>Select only the numerical columns (using <code>select(where(is.numeric))</code>), and apply the function <code>cor()</code> <strong>directly</strong>: <code>%&gt;% cor()</code> . What happens? What is the output? Store the result of this command in a new variable named “all.cor”. Plot a heatmap of these results using the <code>pheatmap()</code> function from the “pheatmap” package. Remember that you first have to install and then activate this package using <code>library("pheatmap")</code>.</p></li>
<li><p>Find the <strong>highest</strong> and <strong>lowest</strong> Pearson correlation value of the result from exercise 2.2. To which pair of variables do they correspond? Plot the corresponding scatter plots using <code>geom_point()</code>! <em>Hint</em>: Before finding the highest Pearson correlation value, use the <code>diag()</code> function to set all diagonal values (=1) of “all.cor” to NA.</p></li>
</ol>
</div>
</div>
</div>
<div id="day-2-dimensionality-reduction-and-unsupervised-learning" class="section level2 hasAnchor">
<h2 class="hasAnchor">Day 2: Dimensionality reduction and unsupervised learning<a href="#day-2-dimensionality-reduction-and-unsupervised-learning" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="preparing-the-data" class="section level3 hasAnchor">
<h3 class="hasAnchor">Preparing the data<a href="#preparing-the-data" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><strong>Unsupervised clustering</strong> is one of the most basic data analysis techniques. It allows to identify groups (or clusters) or observations (here: patients) or variables. <em>Unsupervised</em> means that you are not using any prior knowledge about groups of variables or associations. <strong>K-means clustering</strong> is a good example of unsupervised learning because the method categorizes sample based uniquely on the data.</p>
<p>In this part, we will use a dataset of gene expression data from the TCGA (The Cancer Genome Atlas) project. This project has sequenced several thousand samples from cancer patients of more than 30 cancer types. We will use a subset of this data, containing 200 samples (=patients, as columns) , for which the expression of 300 genes (= rows) has been measured.</p>
<div id="load-data" class="section level4 hasAnchor">
<h4 class="hasAnchor">Load data<a href="#load-data" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>We will start by reading the gene expression data. The columns are the samples and the rows are the genes. This is matrix, which allows some numerical operations to be conducted directly.</p>
<div class="sourceCode" id="cb119"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb119-1"><a href="#cb119-1" tabindex="-1"></a>brca.exp <span class="ot">=</span> <span class="fu">readRDS</span>(<span class="fu">url</span>(<span class="st">&#39;https://www.dropbox.com/s/qububmfvtv443mq/brca.exp.rds?dl=1&#39;</span>))</span>
<span id="cb119-2"><a href="#cb119-2" tabindex="-1"></a><span class="fu">dim</span>(brca.exp)</span></code></pre></div>
<pre><code>## [1] 100 200</code></pre>
<p><strong>WARNING</strong>: If you have problem loading the data, please download <a href="https://www.dropbox.com/s/qububmfvtv443mq/brca.exp.rds?dl=1">this file</a>, store it on your disk, and open it with the following command:</p>
<div class="sourceCode" id="cb121"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb121-1"><a href="#cb121-1" tabindex="-1"></a><span class="co">#brca.exp = readRDS(&quot;xxxx&quot;) # xxxx should be replaced with the path to the downloaded file in your device</span></span></code></pre></div>
<p>Next we will load the clinical annotation file for this gene expression data and explore it</p>
<div class="sourceCode" id="cb122"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb122-1"><a href="#cb122-1" tabindex="-1"></a>brca.anno <span class="ot">=</span> <span class="fu">readRDS</span>(<span class="fu">url</span>(<span class="st">&#39;https://www.dropbox.com/s/9xlivejqkj77llc/brca.anno.rds?dl=1&#39;</span>))</span>
<span id="cb122-2"><a href="#cb122-2" tabindex="-1"></a><span class="fu">head</span>(brca.anno)</span></code></pre></div>
<pre><code>##                 Age ER_status HER2_status Classification
## TCGA-BH-A1EO-01  68  Positive    Negative      Luminal A
## TCGA-E2-A14N-01  37  Negative    Negative     Basal-like
## TCGA-AN-A0FF-01  32  Positive    Negative      Luminal B
## TCGA-A2-A04V-01  39  Positive    Negative      Luminal A
## TCGA-AN-A0XP-01  69  Positive    Negative      Luminal A
## TCGA-C8-A12U-01  46  Positive    Negative      Luminal B</code></pre>
<p>Same here: if you have issues running the previous <code>readRDS</code> command, download <a href="https://www.dropbox.com/s/z6bzwzgzdhky1qz/brca.anno.rds?dl=1">this file</a>, save it on your disk and load it with</p>
<div class="sourceCode" id="cb124"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb124-1"><a href="#cb124-1" tabindex="-1"></a><span class="do">### brca.anno = readRDS(xxx)</span></span></code></pre></div>
<p>You can check the number of samples for each tumor type using the <code>table()</code> function, applied to a specific column (here, there is only one column…)</p>
<div class="sourceCode" id="cb125"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb125-1"><a href="#cb125-1" tabindex="-1"></a><span class="fu">table</span>(brca.anno<span class="sc">$</span>HER2_status)</span></code></pre></div>
<pre><code>## 
## Equivocal  Negative  Positive 
##         4       174        22</code></pre>
</div>
<div id="data-transformation" class="section level4 hasAnchor">
<h4 class="hasAnchor">Data transformation<a href="#data-transformation" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>You will see that the distribution of the data is extremely squeezed due to <strong>outliers</strong> with very high or low values. We will need to make the data more homogeneous, so that our downstream analysis is not affected by these very large magnitude numbers.</p>
<p>We will carry out the following data processing steps. Some of these steps use rather arbitrary values, which come from visually inspecting the data!</p>
<ol style="list-style-type: decimal">
<li><p><strong>Thresholding:</strong> cap the values to the 95th percentile</p></li>
<li><p><strong>Homogenization:</strong> base-2 logarithmic transformation of the entire dataset</p></li>
<li><p><strong>Scaling:</strong> standardize the data so that across genes the mean = 0 and variance = 1.</p></li>
</ol>
<p><strong>Before we start modifying the data, we will store the original data frame into a variable, so that in case of problems we can revert back to the initial data!!</strong></p>
<div class="sourceCode" id="cb127"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb127-1"><a href="#cb127-1" tabindex="-1"></a>brca.exp.original <span class="ot">=</span> brca.exp <span class="co"># keeps the original as matrix</span></span></code></pre></div>
<p><strong>Thresholding</strong></p>
<div class="sourceCode" id="cb128"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb128-1"><a href="#cb128-1" tabindex="-1"></a><span class="do">## what is the value of the 95th percent percentile?</span></span>
<span id="cb128-2"><a href="#cb128-2" tabindex="-1"></a>q95 <span class="ot">=</span> <span class="fu">quantile</span>(brca.exp,<span class="at">probs=</span><span class="fl">0.95</span>)</span>
<span id="cb128-3"><a href="#cb128-3" tabindex="-1"></a></span>
<span id="cb128-4"><a href="#cb128-4" tabindex="-1"></a><span class="do">## set all values above to this value</span></span>
<span id="cb128-5"><a href="#cb128-5" tabindex="-1"></a>brca.exp[brca.exp<span class="sc">&gt;</span>q95] <span class="ot">=</span> q95</span></code></pre></div>
<p><strong>Homogenization and Scaling</strong></p>
<p>We will perform this step by log-transforming the data. We are able to use this operation because the data is still in a matrix.</p>
<div class="sourceCode" id="cb129"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb129-1"><a href="#cb129-1" tabindex="-1"></a>brca.exp <span class="ot">=</span> <span class="fu">log2</span>(brca.exp<span class="sc">+</span><span class="dv">1</span>)</span></code></pre></div>
<blockquote>
<p>Why do we add +1 ?</p>
</blockquote>
<p>Next, we will scale the data and plot its distribution. To do this efficient, we need to convert the data to tibble first, make it tidy, and then plot.</p>
<p>Conversion to tibble can be done using <code>as_tibble(brca.exp, rownames = NA)</code>, where <code>rownames = NA</code> is meant to keep the original rownames (in this case, gene names) in the new tibble, although they are invisible.</p>
<p>Check this out:</p>
<div class="sourceCode" id="cb130"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb130-1"><a href="#cb130-1" tabindex="-1"></a><span class="fu">rownames</span>(<span class="fu">as_tibble</span>(brca.exp, <span class="at">rownames =</span> <span class="cn">NA</span>))[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code></pre></div>
<pre><code>## [1] &quot;TFF1&quot;    &quot;C4orf7&quot;  &quot;AGR3&quot;    &quot;GABRP&quot;   &quot;C1orf64&quot;</code></pre>
<p>In addition, <code>gather(key = "sample", value = "expression")</code> converts the tibble to a long format, where “sample” represents the original column names, and “expression” represents the values present in the initial matrix.</p>
<div class="sourceCode" id="cb132"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb132-1"><a href="#cb132-1" tabindex="-1"></a><span class="do">## scaling</span></span>
<span id="cb132-2"><a href="#cb132-2" tabindex="-1"></a>brca.exp <span class="ot">=</span> <span class="fu">scale</span>(brca.exp)</span>
<span id="cb132-3"><a href="#cb132-3" tabindex="-1"></a></span>
<span id="cb132-4"><a href="#cb132-4" tabindex="-1"></a><span class="do">## plotting the density</span></span>
<span id="cb132-5"><a href="#cb132-5" tabindex="-1"></a><span class="fu">as_tibble</span>(brca.exp, <span class="at">rownames =</span> <span class="cn">NA</span>) <span class="sc">%&gt;%</span></span>
<span id="cb132-6"><a href="#cb132-6" tabindex="-1"></a>  <span class="fu">gather</span>(<span class="at">key =</span> <span class="st">&quot;sample&quot;</span>, </span>
<span id="cb132-7"><a href="#cb132-7" tabindex="-1"></a>         <span class="at">value =</span> <span class="st">&quot;expression&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb132-8"><a href="#cb132-8" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> expression)) <span class="sc">+</span></span>
<span id="cb132-9"><a href="#cb132-9" tabindex="-1"></a>  <span class="fu">geom_histogram</span>() <span class="sc">+</span></span>
<span id="cb132-10"><a href="#cb132-10" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">&quot;Transformed data&quot;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-67-1.png" width="672" /></p>
<p>Compare to the density plot before these pre-processing steps using the same strategy.</p>
<div class="sourceCode" id="cb133"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb133-1"><a href="#cb133-1" tabindex="-1"></a><span class="fu">as_tibble</span>(brca.exp.original, <span class="at">rownames =</span> <span class="cn">NA</span>) <span class="sc">%&gt;%</span></span>
<span id="cb133-2"><a href="#cb133-2" tabindex="-1"></a>  <span class="fu">gather</span>(<span class="at">key =</span> <span class="st">&quot;sample&quot;</span>, <span class="at">value =</span> <span class="st">&quot;expression&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb133-3"><a href="#cb133-3" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> expression)) <span class="sc">+</span></span>
<span id="cb133-4"><a href="#cb133-4" tabindex="-1"></a>  <span class="fu">geom_histogram</span>() <span class="sc">+</span></span>
<span id="cb133-5"><a href="#cb133-5" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">&quot;Untransformed data&quot;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-68-1.png" width="672" /></p>
<hr />
</div>
</div>
<div id="k-means-clustering" class="section level3 hasAnchor">
<h3 class="hasAnchor">k-means clustering<a href="#k-means-clustering" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Another widely used method for grouping observations is the k-means clustering. Now we will cluster our dataset using k-means and explore the underlying structure of the data. In this dataset, different clusters could represent different batches, different tumour subtypes, among other features.</p>
<div id="performing-k-means" class="section level4 hasAnchor">
<h4 class="hasAnchor">Performing k-means<a href="#performing-k-means" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>We use the function <code>kmeans()</code> in R on our matrix. You can check the options and usage in the help panel on the right. The parameter <code>centers</code> indicates how many clusters are requested.</p>
<div class="sourceCode" id="cb134"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb134-1"><a href="#cb134-1" tabindex="-1"></a>km <span class="ot">=</span> <span class="fu">kmeans</span>(<span class="at">x=</span><span class="fu">t</span>(brca.exp), </span>
<span id="cb134-2"><a href="#cb134-2" tabindex="-1"></a>            <span class="at">centers =</span> <span class="dv">2</span>, </span>
<span id="cb134-3"><a href="#cb134-3" tabindex="-1"></a>            <span class="at">nstart =</span> <span class="dv">10</span>)</span></code></pre></div>
<blockquote>
<p>What does the <code>t</code> mean in the command?</p>
</blockquote>
<blockquote>
<p>Just type <code>km</code> in your console and check all results generated. Play around with the <code>centers</code> parameter. See cluster assignments by typing <code>table(km$cluster)</code></p>
</blockquote>
</div>
<div id="quality-of-the-clustering" class="section level4 hasAnchor">
<h4 class="hasAnchor">Quality of the clustering<a href="#quality-of-the-clustering" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>We can judge the quality of the clustering by computing the <strong>intra</strong>-cluster distances, i.e. the sum (squared) of all distances between pairs of objects belonging to the same cluster. This is called the <strong>within sum of squares (WSS)</strong>. The better the clustering, the smaller WSS should be. However, it also automatically decreases with increasing k.</p>
<blockquote>
<p>What would be WSS if we request a number of clusters equal to the number of data points? You can check what the WSS is for a particular clustering by typing</p>
</blockquote>
<div class="sourceCode" id="cb135"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb135-1"><a href="#cb135-1" tabindex="-1"></a>km<span class="sc">$</span>tot.withinss</span></code></pre></div>
<pre><code>## [1] 4123.459</code></pre>
<blockquote>
<p>run k-means for k=2 to k=7 clusters, and for each k check the WSS value. How does WSS evolve with increasing k?</p>
</blockquote>
<p>We can also run a little loop to test different k. Loops are very important structures in any programming language.
We can test a simple scenario before. Check how the output of this simple loop looks like.</p>
<div class="sourceCode" id="cb137"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb137-1"><a href="#cb137-1" tabindex="-1"></a>k_to_test <span class="ot">=</span> <span class="fu">c</span>(<span class="dv">2</span><span class="sc">:</span><span class="dv">7</span>)</span>
<span id="cb137-2"><a href="#cb137-2" tabindex="-1"></a></span>
<span id="cb137-3"><a href="#cb137-3" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(k_to_test)) {</span>
<span id="cb137-4"><a href="#cb137-4" tabindex="-1"></a>  <span class="fu">print</span>(i) <span class="co"># We can print the indexes</span></span>
<span id="cb137-5"><a href="#cb137-5" tabindex="-1"></a>}</span></code></pre></div>
<pre><code>## [1] 1
## [1] 2
## [1] 3
## [1] 4
## [1] 5
## [1] 6</code></pre>
<div class="sourceCode" id="cb139"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb139-1"><a href="#cb139-1" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(k_to_test)) {</span>
<span id="cb139-2"><a href="#cb139-2" tabindex="-1"></a>  <span class="fu">print</span>(k_to_test[i]) <span class="co"># or the actual elements</span></span>
<span id="cb139-3"><a href="#cb139-3" tabindex="-1"></a>}</span></code></pre></div>
<pre><code>## [1] 2
## [1] 3
## [1] 4
## [1] 5
## [1] 6
## [1] 7</code></pre>
<blockquote>
<p>Do you understand the difference between the 2 previous for loops? Try to make your own for loop.</p>
</blockquote>
<p>Now we can make one to test k from 2 to 7:</p>
<div class="sourceCode" id="cb141"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb141-1"><a href="#cb141-1" tabindex="-1"></a>km_wws <span class="ot">=</span> <span class="fu">numeric</span>() <span class="co"># we start by creating an empty vector</span></span>
<span id="cb141-2"><a href="#cb141-2" tabindex="-1"></a></span>
<span id="cb141-3"><a href="#cb141-3" tabindex="-1"></a><span class="co"># To write in the position 1, we use i</span></span>
<span id="cb141-4"><a href="#cb141-4" tabindex="-1"></a><span class="co"># to find the 1st element, we use k_to_test[i]</span></span>
<span id="cb141-5"><a href="#cb141-5" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(k_to_test)) {</span>
<span id="cb141-6"><a href="#cb141-6" tabindex="-1"></a>  km_wws[i] <span class="ot">=</span> <span class="fu">kmeans</span>(<span class="at">x=</span><span class="fu">t</span>(brca.exp), </span>
<span id="cb141-7"><a href="#cb141-7" tabindex="-1"></a>                     <span class="at">centers =</span> k_to_test[i])<span class="sc">$</span>tot.withinss</span>
<span id="cb141-8"><a href="#cb141-8" tabindex="-1"></a>}</span>
<span id="cb141-9"><a href="#cb141-9" tabindex="-1"></a></span>
<span id="cb141-10"><a href="#cb141-10" tabindex="-1"></a><span class="co"># We can plot the k against WSS using geom_line</span></span>
<span id="cb141-11"><a href="#cb141-11" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb141-12"><a href="#cb141-12" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">x =</span> k_to_test, <span class="at">y =</span> km_wws)) <span class="sc">+</span></span>
<span id="cb141-13"><a href="#cb141-13" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x=</span><span class="st">&quot;Number of clusters K&quot;</span>,</span>
<span id="cb141-14"><a href="#cb141-14" tabindex="-1"></a>       <span class="at">y=</span><span class="st">&quot;Total within-clusters sum of squares&quot;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-72-1.png" width="672" /></p>
<blockquote>
<p>Do you see an obvious “elbow” or “kink” in the curve?? Another criteria for the quality of the clustering is the <strong>silhouette</strong> method.</p>
</blockquote>
<p>To run the silhouette method, we need to compute the pairwise distances between all objects (i.e. patients) in the data matrix. This is done with the <code>dist</code> function, which can take different metrics (euclidean, …)</p>
<div class="sourceCode" id="cb142"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb142-1"><a href="#cb142-1" tabindex="-1"></a><span class="do">## compute the patient-patient distance matrix (this is why we transpose using the `t()` function)</span></span>
<span id="cb142-2"><a href="#cb142-2" tabindex="-1"></a>D <span class="ot">=</span> <span class="fu">dist</span>(<span class="fu">t</span>(brca.exp))</span></code></pre></div>
<p>We now compute the silhouette for a specific k-means clustering:</p>
<div class="sourceCode" id="cb143"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb143-1"><a href="#cb143-1" tabindex="-1"></a><span class="fu">library</span>(cluster)</span></code></pre></div>
<pre><code>## 
## Attaching package: &#39;cluster&#39;</code></pre>
<pre><code>## The following object is masked _by_ &#39;.GlobalEnv&#39;:
## 
##     animals</code></pre>
<div class="sourceCode" id="cb146"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb146-1"><a href="#cb146-1" tabindex="-1"></a>km <span class="ot">=</span> <span class="fu">kmeans</span>(<span class="at">x=</span><span class="fu">t</span>(brca.exp), <span class="at">centers =</span> <span class="dv">3</span>, <span class="at">nstart =</span> <span class="dv">10</span>)</span>
<span id="cb146-2"><a href="#cb146-2" tabindex="-1"></a>s <span class="ot">=</span> <span class="fu">silhouette</span>(km<span class="sc">$</span>cluster,D)</span>
<span id="cb146-3"><a href="#cb146-3" tabindex="-1"></a></span>
<span id="cb146-4"><a href="#cb146-4" tabindex="-1"></a><span class="co"># Let us use the basic R function plot() to see the results</span></span>
<span id="cb146-5"><a href="#cb146-5" tabindex="-1"></a><span class="fu">plot</span>(s)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-74-1.png" width="672" /></p>
<blockquote>
<p>Run this for various values of k and compare the silhouette scores
Compare to the Elbow method!</p>
</blockquote>
<hr />
</div>
</div>
<div id="hierarchical-clustering" class="section level3 hasAnchor">
<h3 class="hasAnchor">Hierarchical clustering<a href="#hierarchical-clustering" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Clustering is a method by which we group together similar observations while separating out the dissimilar ones. We will cluster our samples from the cancer dataset to see which samples cluster together or separately. Hierarchical clustering does not generate discrete clusters of datapoints, but rather creates a dendrogram that indicates the magnitude of similitude between samples. Once again is up to the Data Scientist to decide the right amount of clusters.</p>
<div id="determine-the-most-variable-genes" class="section level4 hasAnchor">
<h4 class="hasAnchor">Determine the most variable genes<a href="#determine-the-most-variable-genes" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>When performing clustering, we usually reduce the number of genes used, as some of them are not informative. For example, genes that show a mostly constant expression across all samples will not be useful to distinguish the samples, right? One simple method is to select genes that show a <strong>high variance</strong> across all samples.</p>
<div class="sourceCode" id="cb147"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb147-1"><a href="#cb147-1" tabindex="-1"></a>brca.exp.tibble <span class="ot">=</span> <span class="fu">as_tibble</span>(brca.exp, <span class="at">rownames=</span><span class="cn">NA</span>) <span class="sc">%&gt;%</span></span>
<span id="cb147-2"><a href="#cb147-2" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>(<span class="st">&quot;gene&quot;</span>)</span>
<span id="cb147-3"><a href="#cb147-3" tabindex="-1"></a></span>
<span id="cb147-4"><a href="#cb147-4" tabindex="-1"></a><span class="do">## create a new column with the variance for all genes across all samples</span></span>
<span id="cb147-5"><a href="#cb147-5" tabindex="-1"></a>brca.exp.var <span class="ot">=</span> brca.exp.tibble <span class="sc">%&gt;%</span></span>
<span id="cb147-6"><a href="#cb147-6" tabindex="-1"></a>  <span class="fu">rowwise</span>() <span class="sc">%&gt;%</span></span>
<span id="cb147-7"><a href="#cb147-7" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">variance =</span> <span class="fu">var</span>(<span class="fu">c_across</span>(<span class="fu">starts_with</span>(<span class="st">&quot;TCGA&quot;</span>)))) </span>
<span id="cb147-8"><a href="#cb147-8" tabindex="-1"></a><span class="co"># only includes the columns starting with TCGA</span></span></code></pre></div>
<p>We now want to find the top 25% with the highest variance</p>
<div class="sourceCode" id="cb148"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb148-1"><a href="#cb148-1" tabindex="-1"></a><span class="do">## what is the 75% quantile of the variance?</span></span>
<span id="cb148-2"><a href="#cb148-2" tabindex="-1"></a>q75 <span class="ot">=</span> <span class="fu">quantile</span>(brca.exp.var<span class="sc">$</span>variance, <span class="at">probs =</span> <span class="fl">0.75</span>)</span>
<span id="cb148-3"><a href="#cb148-3" tabindex="-1"></a>q75</span></code></pre></div>
<pre><code>##       75% 
## 0.4934196</code></pre>
<p>So let us select all rows (genes) with a variance higher than or equal to <code>q75</code>:</p>
<div class="sourceCode" id="cb150"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb150-1"><a href="#cb150-1" tabindex="-1"></a><span class="do">## only select the genes with a variance in the top 25%</span></span>
<span id="cb150-2"><a href="#cb150-2" tabindex="-1"></a>topVariantGenes <span class="ot">&lt;-</span> brca.exp.var <span class="sc">%&gt;%</span></span>
<span id="cb150-3"><a href="#cb150-3" tabindex="-1"></a>  <span class="fu">filter</span>(variance <span class="sc">&gt;=</span> q75)</span>
<span id="cb150-4"><a href="#cb150-4" tabindex="-1"></a></span>
<span id="cb150-5"><a href="#cb150-5" tabindex="-1"></a><span class="fu">print</span>(topVariantGenes<span class="sc">$</span>gene)</span></code></pre></div>
<pre><code>##  [1] &quot;TFF1&quot;      &quot;C4orf7&quot;    &quot;AGR3&quot;      &quot;GABRP&quot;     &quot;C1orf64&quot;   &quot;TFF3&quot;     
##  [7] &quot;ABCC11&quot;    &quot;PGR&quot;       &quot;FABP7&quot;     &quot;SERPINA11&quot; &quot;VGLL1&quot;     &quot;A2ML1&quot;    
## [13] &quot;ELF5&quot;      &quot;PROM1&quot;     &quot;CYP4Z2P&quot;   &quot;SLC6A14&quot;   &quot;CAPN8&quot;     &quot;ABCC8&quot;    
## [19] &quot;SYTL5&quot;     &quot;SFRP1&quot;     &quot;ART3&quot;      &quot;GABBR2&quot;    &quot;PPP1R14C&quot;  &quot;HORMAD1&quot;  
## [25] &quot;LOC84740&quot;</code></pre>
</div>
<div id="computing-the-correlation-between-all-patients" class="section level4 hasAnchor">
<h4 class="hasAnchor">Computing the correlation between all patients<a href="#computing-the-correlation-between-all-patients" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Let us start by filtering for only the highly variable genes. Then we can directly calculate Spearman correlation.</p>
<div class="sourceCode" id="cb152"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb152-1"><a href="#cb152-1" tabindex="-1"></a>brca.exp.highvar.cor <span class="ot">=</span> brca.exp.tibble <span class="sc">%&gt;%</span></span>
<span id="cb152-2"><a href="#cb152-2" tabindex="-1"></a>  <span class="fu">filter</span>(gene <span class="sc">%in%</span> topVariantGenes<span class="sc">$</span>gene) <span class="sc">%&gt;%</span>        <span class="co"># from the whole list, select only high variable</span></span>
<span id="cb152-3"><a href="#cb152-3" tabindex="-1"></a>  <span class="fu">select</span>(<span class="fu">where</span>(is.numeric)) <span class="sc">%&gt;%</span>                     <span class="co"># get only numerical columns</span></span>
<span id="cb152-4"><a href="#cb152-4" tabindex="-1"></a>  <span class="fu">cor</span>(<span class="at">method=</span><span class="st">&quot;spearman&quot;</span>)                            <span class="co"># create correlation-based distance matrix</span></span></code></pre></div>
<p>If we want to display the correlation matrix as a heatmap, we can use the <code>pheatmap</code> function as before:</p>
<div class="sourceCode" id="cb153"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb153-1"><a href="#cb153-1" tabindex="-1"></a><span class="fu">library</span>(pheatmap)</span>
<span id="cb153-2"><a href="#cb153-2" tabindex="-1"></a><span class="fu">pheatmap</span>(brca.exp.highvar.cor, </span>
<span id="cb153-3"><a href="#cb153-3" tabindex="-1"></a>         <span class="at">show_rownames =</span> <span class="cn">FALSE</span>, </span>
<span id="cb153-4"><a href="#cb153-4" tabindex="-1"></a>         <span class="at">show_colnames =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-79-1.png" width="672" /></p>
<p>Each cell of this heatmap represents the correlation value between the sample in the row and the sample in the column. The correlation of a sample to itself is always 1 (red diagonal).</p>
<p>The function automatically determines the clustering trees of the rows and columns (which are identical, since the correlation matrix is symmetrical!)</p>
</div>
<div id="including-clinical-annotations-in-the-heatmap" class="section level4 hasAnchor">
<h4 class="hasAnchor">Including clinical annotations in the heatmap<a href="#including-clinical-annotations-in-the-heatmap" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>This is a nice representation, but in order to interpret this clustering, we need to add some additional (clinical) information to interpret the clustering structure. To do this, we use an annotation data frame containing as columns a number of clinical features.</p>
<p>The clinical annotation is stored in the <code>brca.anno</code> data frame.
We can now plot again the heatmap, using the annotation dataframe to add additional information</p>
<div class="sourceCode" id="cb154"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb154-1"><a href="#cb154-1" tabindex="-1"></a><span class="fu">pheatmap</span>(brca.exp.highvar.cor,</span>
<span id="cb154-2"><a href="#cb154-2" tabindex="-1"></a>         <span class="at">annotation_row =</span> brca.anno,</span>
<span id="cb154-3"><a href="#cb154-3" tabindex="-1"></a>         <span class="at">show_rownames =</span> <span class="cn">FALSE</span>, </span>
<span id="cb154-4"><a href="#cb154-4" tabindex="-1"></a>         <span class="at">show_colnames =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-80-1.png" width="768" /></p>
<blockquote>
<p>How would you interpret this dendrogram? Do the clusters you observe make any sense? What are the parameters by which the samples cluster together? How many meaningful clusters can you observe? Do you see any relation between the distribution of the data and your clusters ?</p>
</blockquote>
<blockquote>
<p>The function <code>pheatmap</code> accepts a parameter clustering_method to indicate alternative linkage methods; try other linkage methods (check which are available with the pheatmap help page, which can be accessed by typing <code>?pheatmap</code> in the console!)</p>
</blockquote>
<hr />
</div>
</div>
<div id="principal-component-analysis" class="section level3 hasAnchor">
<h3 class="hasAnchor">Principal component analysis<a href="#principal-component-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We will now use principal component analysis to explore the same dataset, and identify directions (i.e. principal components) with maximal variance. Principal components analysis finds n-dimensional vectors (Principal Components) in the direction of the largest variance, thereby allowing you to describe an n-dimensional dataset with just a few dimensions.</p>
<div class="sourceCode" id="cb155"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb155-1"><a href="#cb155-1" tabindex="-1"></a>pca <span class="ot">=</span> topVariantGenes <span class="sc">%&gt;%</span></span>
<span id="cb155-2"><a href="#cb155-2" tabindex="-1"></a>  <span class="fu">select</span>(<span class="fu">where</span>(is.numeric)) <span class="sc">%&gt;%</span></span>
<span id="cb155-3"><a href="#cb155-3" tabindex="-1"></a>  <span class="fu">t</span>() <span class="sc">%&gt;%</span> <span class="co"># do not forget to transpose the data!</span></span>
<span id="cb155-4"><a href="#cb155-4" tabindex="-1"></a>  <span class="fu">prcomp</span>(<span class="at">center =</span> <span class="cn">FALSE</span>, <span class="at">scale =</span> <span class="cn">FALSE</span>) <span class="co"># We set these as false as we have already scaled our data</span></span>
<span id="cb155-5"><a href="#cb155-5" tabindex="-1"></a><span class="fu">summary</span>(pca)</span></code></pre></div>
<pre><code>## Importance of components:
##                          PC1    PC2     PC3     PC4     PC5     PC6     PC7
## Standard deviation     3.719 2.1886 0.89796 0.86195 0.69326 0.62793 0.61645
## Proportion of Variance 0.559 0.1936 0.03259 0.03003 0.01942 0.01593 0.01536
## Cumulative Proportion  0.559 0.7526 0.78514 0.81516 0.83458 0.85052 0.86588
##                            PC8     PC9    PC10    PC11    PC12   PC13    PC14
## Standard deviation     0.59717 0.55556 0.53665 0.52007 0.50060 0.4900 0.45974
## Proportion of Variance 0.01441 0.01247 0.01164 0.01093 0.01013 0.0097 0.00854
## Cumulative Proportion  0.88029 0.89276 0.90440 0.91533 0.92546 0.9352 0.94370
##                           PC15    PC16   PC17    PC18    PC19    PC20    PC21
## Standard deviation     0.44696 0.43284 0.4009 0.39542 0.38101 0.35781 0.34271
## Proportion of Variance 0.00807 0.00757 0.0065 0.00632 0.00587 0.00517 0.00475
## Cumulative Proportion  0.95178 0.95935 0.9658 0.97216 0.97803 0.98320 0.98795
##                           PC22    PC23    PC24    PC25
## Standard deviation     0.30347 0.28452 0.26505 0.23424
## Proportion of Variance 0.00372 0.00327 0.00284 0.00222
## Cumulative Proportion  0.99167 0.99494 0.99778 1.00000</code></pre>
<blockquote>
<p>How many principal components do you obtain? Compare this to the dimension of the matrix using the <code>dim()</code> function!</p>
</blockquote>
<blockquote>
<p>What would happen if you would not transpose the matrix with t(…) in the prcomp function?</p>
</blockquote>
<p>Principal components are ranked by the amount of variance that they explain. This can be visualized using a <strong>scree plot</strong>, indicating how much variance each PC explains: the <strong>standard deviation</strong> explained by each principal component is contained in the <code>pca$sdev</code> vector:</p>
<div class="sourceCode" id="cb157"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb157-1"><a href="#cb157-1" tabindex="-1"></a>pca<span class="sc">$</span>sdev</span></code></pre></div>
<pre><code>##  [1] 3.7190077 2.1886234 0.8979609 0.8619466 0.6932587 0.6279258 0.6164472
##  [8] 0.5971745 0.5555603 0.5366514 0.5200662 0.5006037 0.4899632 0.4597383
## [15] 0.4469622 0.4328397 0.4008966 0.3954205 0.3810141 0.3578084 0.3427089
## [22] 0.3034732 0.2845180 0.2650462 0.2342352</code></pre>
<p>We see that the standard deviation is indeed going down! Let us now plot the proportion of total variance explained by each PC</p>
<div class="sourceCode" id="cb159"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb159-1"><a href="#cb159-1" tabindex="-1"></a>variance <span class="ot">=</span> (pca<span class="sc">$</span>sdev)<span class="sc">^</span><span class="dv">2</span></span>
<span id="cb159-2"><a href="#cb159-2" tabindex="-1"></a>prop.variance <span class="ot">=</span> variance<span class="sc">/</span><span class="fu">sum</span>(variance)</span>
<span id="cb159-3"><a href="#cb159-3" tabindex="-1"></a><span class="fu">names</span>(prop.variance) <span class="ot">=</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(prop.variance)</span>
<span id="cb159-4"><a href="#cb159-4" tabindex="-1"></a></span>
<span id="cb159-5"><a href="#cb159-5" tabindex="-1"></a><span class="co"># We make a data.frame from the prop.variance and the PC it corresponds to </span></span>
<span id="cb159-6"><a href="#cb159-6" tabindex="-1"></a><span class="co"># we can obtain the PCs using names()</span></span>
<span id="cb159-7"><a href="#cb159-7" tabindex="-1"></a><span class="fu">data.frame</span>(<span class="at">proportion =</span> prop.variance, </span>
<span id="cb159-8"><a href="#cb159-8" tabindex="-1"></a>           <span class="at">PCs =</span> <span class="fu">as.numeric</span>(<span class="fu">names</span>(prop.variance))) <span class="sc">%&gt;%</span></span>
<span id="cb159-9"><a href="#cb159-9" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> PCs, <span class="at">y =</span> proportion)) <span class="sc">+</span></span>
<span id="cb159-10"><a href="#cb159-10" tabindex="-1"></a>  <span class="fu">geom_col</span>() <span class="sc">+</span>                               <span class="co"># to make the barplot</span></span>
<span id="cb159-11"><a href="#cb159-11" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">y=</span><span class="st">&#39;Proportion of variance&#39;</span>)           <span class="co"># we only plot the first 20 PCs</span></span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-83-1.png" width="672" /></p>
<p>Principal component analysis represents each data point (here: patient) in a new space in which the coordinates are principal components. Check the following output:</p>
<div class="sourceCode" id="cb160"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb160-1"><a href="#cb160-1" tabindex="-1"></a><span class="fu">head</span>(pca<span class="sc">$</span>x)</span></code></pre></div>
<p>We can now display the data points (i.e. patients) in the first two principal components. In addition, we can color the dots according to certain clinical parameters:</p>
<div class="sourceCode" id="cb161"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb161-1"><a href="#cb161-1" tabindex="-1"></a><span class="co"># We start by creating a dataframe and combining it with the annotation</span></span>
<span id="cb161-2"><a href="#cb161-2" tabindex="-1"></a>pca_with_annot <span class="ot">=</span> <span class="fu">as.data.frame</span>(pca<span class="sc">$</span>x) <span class="sc">%&gt;%</span></span>
<span id="cb161-3"><a href="#cb161-3" tabindex="-1"></a>  <span class="fu">merge</span>(brca.anno, <span class="at">by =</span> <span class="dv">0</span>) <span class="co"># by = 0 makes use of the rownames as common information</span></span>
<span id="cb161-4"><a href="#cb161-4" tabindex="-1"></a></span>
<span id="cb161-5"><a href="#cb161-5" tabindex="-1"></a><span class="do">## Now the object is in a ggplot2 friendly format</span></span>
<span id="cb161-6"><a href="#cb161-6" tabindex="-1"></a><span class="fu">ggplot</span>(pca_with_annot,</span>
<span id="cb161-7"><a href="#cb161-7" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> PC1, <span class="at">y =</span> PC2, <span class="at">colour =</span> ER_status)) <span class="sc">+</span></span>
<span id="cb161-8"><a href="#cb161-8" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb161-9"><a href="#cb161-9" tabindex="-1"></a>  <span class="fu">scale_colour_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">&quot;grey&quot;</span>, <span class="st">&quot;red&quot;</span>, <span class="st">&quot;navy&quot;</span>)) <span class="co"># scale_colour_manual can be used to change colours</span></span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-85-1.png" width="576" /></p>
<blockquote>
<p>Choose different PCs for this plot. Can you still observe the two clusters corresponding to the ER_status of the patients?</p>
</blockquote>
<hr />
</div>
<div id="exercises-1" class="section level3 hasAnchor">
<h3 class="hasAnchor">EXERCISES<a href="#exercises-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="exercise-1-variance" class="section level4 hasAnchor">
<h4 class="hasAnchor">Exercise 1: Variance<a href="#exercise-1-variance" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p><em>For this exercise set it is given that all the data cleanup steps have been taken, you don’t need to put them in the results.</em></p>
<ol style="list-style-type: decimal">
<li><p>Make a heatmap of the reduced matrix “topVariantGenes” using the <code>pheatmap()</code> function of the <code>pheatmap</code> library (do not forget to select only for numerical columns). Check for parameters that might change the style of the heatmap (column names, row names, etc..). How is this heatmap different from the heatmap in section 2?</p></li>
<li><p>Repeat the selection of top variable genes (apply the same quantile used to generate “topVariantGenes”), but using the median absolute deviation (or MAD) using the <code>mad()</code> function instead of the <code>sd()</code> function, and store into as <code>brca.topMAD</code></p></li>
<li><p>Extract the gene names of <code>topVariantGenes</code> and <code>brca.topMAD</code> and check how many overlap using the <code>intersect()</code> function.</p></li>
</ol>
</div>
<div id="exercise-2-hierarchical-clustering" class="section level4 hasAnchor">
<h4 class="hasAnchor">Exercise 2: Hierarchical clustering<a href="#exercise-2-hierarchical-clustering" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>In section 2, we have computed a correlation matrix, and used this matrix to build a clustering tree.</p>
<p>Try different linkage methods using the <code>clustering_method</code> parameter to see if the topology of the dendrogram changes!</p>
<ol start="2" style="list-style-type: decimal">
<li>Try building a distance matrix which would lead to different topologies of the dendrogram, depending on which linkage method is used! Show the dendrograms built with different linkage methods!</li>
</ol>
</div>
<div id="exercise-3-pca" class="section level4 hasAnchor">
<h4 class="hasAnchor">Exercise 3: PCA<a href="#exercise-3-pca" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<ol style="list-style-type: decimal">
<li><p>Display the patients in the first two principal components (available in <code>pca_with_annot</code>) using <code>geom_point()</code>. Color the patients in the PCA plot according to <code>HER2_status</code>.</p></li>
<li><p>(optional) Color the patients in the PCA plot according to <code>Classification</code>; you will probably need to define some more colors… You can check available colors <a href="http://www.stat.columbia.edu/~tzheng/files/Rcolor.pdf">here</a></p></li>
</ol>
</div>
<div id="going-further-expert-1" class="section level4 hasAnchor">
<h4 class="hasAnchor">Going further <em>(expert)</em><a href="#going-further-expert-1" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Instead of performing the k-means on the whole gene expression matrix, we can run k-means on the space in which each patient is represented by the first k principal components.</p>
<ol style="list-style-type: decimal">
<li><p>Run k-means with different numbers of clusters (1-10) on the patients using the first 2, 4, 6,… principal components (i.e. the first columns of <code>pca_with_annot</code>). Use the elbow method to evaluate how the within sum of squares (WSS) evolves. What is the optimal number of clusters?</p></li>
<li><p>Represent the patients in the PCA plot as previously, but color them according to the cluster they belong to! Run kmeans with two clusters for this and merge the k-means results.</p></li>
</ol>
</div>
</div>
</div>
<div id="day-3-probability-distributions" class="section level2 hasAnchor">
<h2 class="hasAnchor">Day 3: Probability distributions<a href="#day-3-probability-distributions" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="probability-distributions" class="section level3 hasAnchor">
<h3 class="hasAnchor">Probability distributions<a href="#probability-distributions" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>In the previous Exercise Sheet we have learnt more about unsupervised learning, like hierarchical clustering and especially PCA. These are among the fundamental methods of Data Analysis. Today we will learn more about another milestone of statistics: <strong>probability distributions</strong>.</p>
<div class="float">
<img src="figures/distrib.png" alt="Various kinds of distributions and the relations among them." />
<div class="figcaption">Various kinds of distributions and the relations among them.</div>
</div>
<p><a href="https://medium.com/@srowen/common-probability-distributions-347e6b945ce4">Figure source</a></p>
<p>Distributions represented by sparse lines represent outcomes which will be discrete (for example the <em>roll of dice</em> will always have discrete integer values from 1 to 6). Distributions represented by dense lines represent outcomes which can be continuous i.e real numbers (for example the height of people living in Heidelberg).</p>
<p>R has in-built functions for almost all probability distributions</p>
<div class="sourceCode" id="cb162"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb162-1"><a href="#cb162-1" tabindex="-1"></a><span class="co"># Get the list of all probability distribution related functions</span></span>
<span id="cb162-2"><a href="#cb162-2" tabindex="-1"></a><span class="fu">help</span>(distributions)</span></code></pre></div>
<p>All of these functions for probability distributions follow the same common scheme, the <strong>root name</strong> of the function prefixed by either of <strong>p</strong>, <strong>d</strong>, <strong>q</strong> and <strong>r</strong>. For example for the Normal distribution we have the four variants of function available - <strong>pnorm</strong>, <strong>dnorm</strong>, <strong>qnorm</strong> and <strong>rnorm</strong>.</p>
<p>Here, you can find some specific help on these functions:</p>
<div class="sourceCode" id="cb163"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb163-1"><a href="#cb163-1" tabindex="-1"></a>?rnorm</span>
<span id="cb163-2"><a href="#cb163-2" tabindex="-1"></a><span class="co"># or</span></span>
<span id="cb163-3"><a href="#cb163-3" tabindex="-1"></a>?rpois</span>
<span id="cb163-4"><a href="#cb163-4" tabindex="-1"></a><span class="co"># etc ...</span></span></code></pre></div>
<p>In the respective help documentation, you will find details on each of the functions.<br />
Probably, the most difficult to distinguish or to remember are <strong>pnorm()</strong> and <strong>qnorm()</strong> (respectively <strong>ppois()</strong> and <strong>qpois()</strong>, etc …).<br />
We are going to look at them more deeply in the following.</p>
<hr />
<div id="getting-to-know-the-various-functions" class="section level4 hasAnchor">
<h4 class="hasAnchor">Getting to know the various functions<a href="#getting-to-know-the-various-functions" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Let us get a grasp of what these functions actually do. You should be familiar with the cumulative distribution function, let’s take a look at that and its inverse first. We will work with a Normal distribution. We calculate the <strong>cumulative probability</strong> for the values 1,2,3,4 in three different distributions, using one of the functions described previously.<br />
Short hint: <strong>p</strong> like <strong>cumulative P-robability</strong>. Which function are you going to use?</p>
<div class="sourceCode" id="cb164"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb164-1"><a href="#cb164-1" tabindex="-1"></a><span class="co"># Distribution 1: Mean = 2, Standard Deviation = 1</span></span>
<span id="cb164-2"><a href="#cb164-2" tabindex="-1"></a><span class="fu">pnorm</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>,<span class="at">mean=</span><span class="dv">2</span>,<span class="at">sd=</span><span class="dv">1</span>)</span></code></pre></div>
<pre><code>## [1] 0.1586553 0.5000000 0.8413447 0.9772499</code></pre>
<div class="sourceCode" id="cb166"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb166-1"><a href="#cb166-1" tabindex="-1"></a><span class="co"># Distribution 2: Mean = 2, Standard Deviation = 2</span></span>
<span id="cb166-2"><a href="#cb166-2" tabindex="-1"></a><span class="fu">pnorm</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>,<span class="at">mean=</span><span class="dv">2</span>,<span class="at">sd=</span><span class="dv">2</span>)</span></code></pre></div>
<pre><code>## [1] 0.3085375 0.5000000 0.6914625 0.8413447</code></pre>
<div class="sourceCode" id="cb168"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb168-1"><a href="#cb168-1" tabindex="-1"></a><span class="co"># Distribution 2: Mean = 4, Standard Deviation = 1</span></span>
<span id="cb168-2"><a href="#cb168-2" tabindex="-1"></a><span class="fu">pnorm</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>,<span class="at">mean=</span><span class="dv">4</span>,<span class="at">sd=</span><span class="dv">1</span>)</span></code></pre></div>
<pre><code>## [1] 0.001349898 0.022750132 0.158655254 0.500000000</code></pre>
<blockquote>
<p>Do you understand why the cumulative distribution functions change the way they do?</p>
</blockquote>
<p>Now, on the same distributions, we calculate the <strong>inverse cdf</strong> (inverse cumulative distribution function) for the cumulative probabilities of 25%, 50% and 75%. We use the <code>qnorm()</code> function for this (<strong>q- like quantile</strong>):</p>
<div class="sourceCode" id="cb170"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb170-1"><a href="#cb170-1" tabindex="-1"></a><span class="co"># Distribution 1: Mean = 2, Standard Deviation = 1</span></span>
<span id="cb170-2"><a href="#cb170-2" tabindex="-1"></a><span class="fu">qnorm</span>(<span class="fu">c</span>(<span class="fl">0.25</span>,<span class="fl">0.5</span>,<span class="fl">0.75</span>),<span class="at">mean=</span><span class="dv">2</span>,<span class="at">sd=</span><span class="dv">1</span>)</span></code></pre></div>
<pre><code>## [1] 1.32551 2.00000 2.67449</code></pre>
<div class="sourceCode" id="cb172"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb172-1"><a href="#cb172-1" tabindex="-1"></a><span class="co"># Distribution 2: Mean = 2, Standard Deviation = 2</span></span>
<span id="cb172-2"><a href="#cb172-2" tabindex="-1"></a><span class="fu">qnorm</span>(<span class="fu">c</span>(<span class="fl">0.25</span>,<span class="fl">0.5</span>,<span class="fl">0.75</span>),<span class="at">mean=</span><span class="dv">2</span>,<span class="at">sd=</span><span class="dv">2</span>)</span></code></pre></div>
<pre><code>## [1] 0.6510205 2.0000000 3.3489795</code></pre>
<div class="sourceCode" id="cb174"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb174-1"><a href="#cb174-1" tabindex="-1"></a><span class="co"># Distribution 3: Mean = 4, Standard Deviation = 1</span></span>
<span id="cb174-2"><a href="#cb174-2" tabindex="-1"></a><span class="fu">qnorm</span>(<span class="fu">c</span>(<span class="fl">0.25</span>,<span class="fl">0.5</span>,<span class="fl">0.75</span>),<span class="at">mean=</span><span class="dv">4</span>,<span class="at">sd=</span><span class="dv">1</span>)</span></code></pre></div>
<pre><code>## [1] 3.32551 4.00000 4.67449</code></pre>
<blockquote>
<p>Try with 100% on any of the distributions. Can you explain this result? Do you expect the result to be different in the other ones?</p>
</blockquote>
<p>Now that you know the output of the p- and q- functions, let’s look at <strong>d- like density probability</strong> functions. For any continuous distribution, the value of the function at any specific value is 0. This is why this probability function is used in discrete distribution such as the binomial distribution (function <strong>dbinom()</strong>). We first calculate the probability of getting 5 events out of 5 in a binomial distribution with a probability of 0.5. Then, we calculate the odds of <strong>not</strong> getting 5 out of 5.</p>
<div class="sourceCode" id="cb176"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb176-1"><a href="#cb176-1" tabindex="-1"></a><span class="co"># probability of 5 out of 5</span></span>
<span id="cb176-2"><a href="#cb176-2" tabindex="-1"></a><span class="fu">dbinom</span>(<span class="dv">5</span>, <span class="at">size=</span><span class="dv">5</span>, <span class="at">prob=</span><span class="fl">0.5</span>) <span class="co"># size = number of trials; prob = probability of success on each trial</span></span></code></pre></div>
<pre><code>## [1] 0.03125</code></pre>
<div class="sourceCode" id="cb178"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb178-1"><a href="#cb178-1" tabindex="-1"></a><span class="co"># probability of NOT getting 5 out of 5</span></span>
<span id="cb178-2"><a href="#cb178-2" tabindex="-1"></a><span class="dv">1</span><span class="sc">-</span><span class="fu">dbinom</span>(<span class="dv">5</span>, <span class="at">size=</span><span class="dv">5</span>, <span class="at">prob=</span><span class="fl">0.5</span>) </span></code></pre></div>
<pre><code>## [1] 0.96875</code></pre>
<div class="sourceCode" id="cb180"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb180-1"><a href="#cb180-1" tabindex="-1"></a><span class="co"># or</span></span>
<span id="cb180-2"><a href="#cb180-2" tabindex="-1"></a><span class="fu">pbinom</span>(<span class="dv">4</span>, <span class="at">size=</span><span class="dv">5</span>, <span class="at">prob=</span><span class="fl">0.5</span>) <span class="co"># Remember, pbinom() returns the &quot;cumulative probability&quot; of ...</span></span></code></pre></div>
<pre><code>## [1] 0.96875</code></pre>
<blockquote>
<p>What is the probability of getting 5 out of 10? And NOT getting 5 out of 10?</p>
</blockquote>
<p>Suppose that the distribution of body height is described by a <strong>normal distribution</strong> with <strong>mean = 1.75 m</strong> and standard deviation <strong>sd = 0.15</strong>. What is the probability that someone is taller than 1.9 m? We can use the parameter <strong>lower.tail</strong> of the function <code>pnorm()</code> for this.<br />
What is the probability that someone is smaller than 1.60m?</p>
<div class="sourceCode" id="cb182"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb182-1"><a href="#cb182-1" tabindex="-1"></a><span class="do">## taller than 1.9</span></span>
<span id="cb182-2"><a href="#cb182-2" tabindex="-1"></a><span class="fu">pnorm</span>(<span class="fl">1.9</span>, <span class="at">mean=</span><span class="fl">1.75</span>, <span class="at">sd=</span><span class="fl">0.15</span>, <span class="at">lower.tail=</span><span class="cn">FALSE</span>)</span></code></pre></div>
<pre><code>## [1] 0.1586553</code></pre>
<div class="sourceCode" id="cb184"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb184-1"><a href="#cb184-1" tabindex="-1"></a><span class="co"># or</span></span>
<span id="cb184-2"><a href="#cb184-2" tabindex="-1"></a><span class="dv">1</span><span class="sc">-</span><span class="fu">pnorm</span>(<span class="fl">1.9</span>, <span class="at">mean=</span><span class="fl">1.75</span>, <span class="at">sd=</span><span class="fl">0.15</span>)</span></code></pre></div>
<pre><code>## [1] 0.1586553</code></pre>
<div class="sourceCode" id="cb186"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb186-1"><a href="#cb186-1" tabindex="-1"></a><span class="do">## smaller than 1.6</span></span>
<span id="cb186-2"><a href="#cb186-2" tabindex="-1"></a><span class="fu">pnorm</span>(<span class="fl">1.6</span>, <span class="at">mean=</span><span class="fl">1.75</span>, <span class="at">sd=</span><span class="fl">0.15</span>, <span class="at">lower.tail=</span><span class="cn">TRUE</span>)</span></code></pre></div>
<pre><code>## [1] 0.1586553</code></pre>
<div class="sourceCode" id="cb188"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb188-1"><a href="#cb188-1" tabindex="-1"></a><span class="co"># equivalent to </span></span>
<span id="cb188-2"><a href="#cb188-2" tabindex="-1"></a><span class="fu">pnorm</span>(<span class="fl">1.6</span>, <span class="at">mean=</span><span class="fl">1.75</span>, <span class="at">sd=</span><span class="fl">0.15</span>) <span class="co"># lower.tail is set to TRUE by default (check the help page using `?pnorm`)</span></span></code></pre></div>
<pre><code>## [1] 0.1586553</code></pre>
<p>Let’s have a look at this distribution using the <strong>dnorm()</strong> function:</p>
<div class="sourceCode" id="cb190"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb190-1"><a href="#cb190-1" tabindex="-1"></a>x <span class="ot">=</span> <span class="fu">seq</span>(<span class="dv">1</span>, <span class="fl">2.5</span>, <span class="at">by=</span><span class="fl">0.01</span>)</span>
<span id="cb190-2"><a href="#cb190-2" tabindex="-1"></a>y <span class="ot">=</span> <span class="fu">dnorm</span>(x, <span class="at">mean=</span><span class="fl">1.75</span>, <span class="at">sd=</span><span class="fl">0.15</span>)</span>
<span id="cb190-3"><a href="#cb190-3" tabindex="-1"></a></span>
<span id="cb190-4"><a href="#cb190-4" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb190-5"><a href="#cb190-5" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">x =</span> x, <span class="at">y =</span> y),</span>
<span id="cb190-6"><a href="#cb190-6" tabindex="-1"></a>            <span class="at">colour =</span> <span class="st">&quot;red&quot;</span>, <span class="at">linewidth =</span> <span class="fl">0.8</span>) <span class="sc">+</span></span>
<span id="cb190-7"><a href="#cb190-7" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">c</span>(<span class="fl">1.6</span>, <span class="fl">1.9</span>),</span>
<span id="cb190-8"><a href="#cb190-8" tabindex="-1"></a>             <span class="at">linetype =</span> <span class="st">&#39;dashed&#39;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-92-1.png" width="672" /></p>
<p>Let’s finally look at <strong>r- like random</strong> functions. These, unlike the others, don’t return single probabilities or values but rather generate a random distribution of values. We can use this to generate a normal distribution with mean = 10, sd = 5.</p>
<div class="sourceCode" id="cb191"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb191-1"><a href="#cb191-1" tabindex="-1"></a><span class="do">## normal distribution</span></span>
<span id="cb191-2"><a href="#cb191-2" tabindex="-1"></a>x <span class="ot">=</span> <span class="fu">rnorm</span>(<span class="at">n=</span><span class="dv">1000</span>,<span class="at">mean=</span><span class="dv">10</span>,<span class="at">sd=</span><span class="dv">5</span>)</span>
<span id="cb191-3"><a href="#cb191-3" tabindex="-1"></a></span>
<span id="cb191-4"><a href="#cb191-4" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb191-5"><a href="#cb191-5" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="fu">aes</span>(<span class="at">x=</span>x), <span class="at">bins =</span> <span class="dv">20</span>,</span>
<span id="cb191-6"><a href="#cb191-6" tabindex="-1"></a>                 <span class="at">fill =</span> <span class="st">&quot;orange&quot;</span>) <span class="sc">+</span></span>
<span id="cb191-7"><a href="#cb191-7" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">y =</span> <span class="st">&quot;Frequency&quot;</span>) </span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-93-1.png" width="576" /></p>
<blockquote>
<p>Can you generate a Poisson distribution and a binomial distribution?</p>
</blockquote>
<hr />
</div>
<div id="normalgaussian-distribution" class="section level4 hasAnchor">
<h4 class="hasAnchor">Normal/Gaussian distribution<a href="#normalgaussian-distribution" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>The normal or the Gaussian distribution is given as:</p>
<p><span class="math display">\[P(x) = \frac{1}{{\sigma \sqrt {2\pi } }} \cdot e ^ \frac{-(x- \mu)^2}{{2\sigma ^2 }} \]</span> where <span class="math inline">\(\mu\)</span> is the mean of the distribution and <span class="math inline">\(\sigma\)</span> is the standard deviation.</p>
<p>The <strong>standard normal distribution</strong> is a special case of <strong>normal distribution</strong> where the values for <span class="math inline">\(\mu = 0\)</span> and <span class="math inline">\(\sigma = 1\)</span>. Thus, the above equation for the Normal distribution simplifies to:</p>
<p><span class="math display">\[P(x) = \frac{1}{{\sqrt {2\pi } }} \cdot e ^ \frac{-x^2}{2} \]</span> Now for any <span class="math inline">\(x\)</span> we can easily solve this equation since <span class="math inline">\(\pi\)</span> and <span class="math inline">\(e\)</span> are known constants.</p>
<hr />
</div>
<div id="visualization" class="section level4 hasAnchor">
<h4 class="hasAnchor">Visualization<a href="#visualization" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Let’s generate three random normal distributions with different means and standard deviations and visualize them together</p>
<div class="sourceCode" id="cb192"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb192-1"><a href="#cb192-1" tabindex="-1"></a><span class="do">## Use the function `dnorm()` to plot the density distribution</span></span>
<span id="cb192-2"><a href="#cb192-2" tabindex="-1"></a>x <span class="ot">=</span> <span class="fu">seq</span>(<span class="sc">-</span><span class="dv">10</span>, <span class="dv">30</span>, <span class="at">by=</span>.<span class="dv">1</span>)</span>
<span id="cb192-3"><a href="#cb192-3" tabindex="-1"></a>d1 <span class="ot">=</span> <span class="fu">dnorm</span>(x, <span class="at">mean=</span><span class="dv">0</span>, <span class="at">sd=</span><span class="dv">1</span>)</span>
<span id="cb192-4"><a href="#cb192-4" tabindex="-1"></a>d2 <span class="ot">=</span> <span class="fu">dnorm</span>(x, <span class="at">mean=</span><span class="dv">10</span>, <span class="at">sd=</span><span class="dv">1</span>)</span>
<span id="cb192-5"><a href="#cb192-5" tabindex="-1"></a>d3 <span class="ot">=</span> <span class="fu">dnorm</span>(x, <span class="at">mean=</span><span class="dv">20</span>, <span class="at">sd=</span><span class="dv">1</span>)</span>
<span id="cb192-6"><a href="#cb192-6" tabindex="-1"></a></span>
<span id="cb192-7"><a href="#cb192-7" tabindex="-1"></a><span class="co"># Compare with the histogram build from 1000 random number drawn from the standard normal distribution</span></span>
<span id="cb192-8"><a href="#cb192-8" tabindex="-1"></a>r1 <span class="ot">=</span> <span class="fu">rnorm</span>(<span class="at">n=</span><span class="dv">1000</span>, <span class="at">mean=</span><span class="dv">0</span>, <span class="at">sd =</span><span class="dv">1</span>)  <span class="co"># random distributions of values</span></span>
<span id="cb192-9"><a href="#cb192-9" tabindex="-1"></a>r2 <span class="ot">=</span> <span class="fu">rnorm</span>(<span class="at">n=</span><span class="dv">1000</span>, <span class="at">mean=</span><span class="dv">10</span>, <span class="at">sd=</span><span class="dv">1</span>)</span>
<span id="cb192-10"><a href="#cb192-10" tabindex="-1"></a>r3 <span class="ot">=</span> <span class="fu">rnorm</span>(<span class="at">n=</span><span class="dv">1000</span>, <span class="at">mean=</span><span class="dv">20</span>, <span class="at">sd=</span><span class="dv">1</span>)</span>
<span id="cb192-11"><a href="#cb192-11" tabindex="-1"></a></span>
<span id="cb192-12"><a href="#cb192-12" tabindex="-1"></a><span class="co"># Histogram visualization</span></span>
<span id="cb192-13"><a href="#cb192-13" tabindex="-1"></a><span class="co"># We will start by converting the rnorm outputs into one single df</span></span>
<span id="cb192-14"><a href="#cb192-14" tabindex="-1"></a><span class="fu">data.frame</span>(<span class="at">vals =</span> <span class="fu">c</span>(r1, r2, r3), <span class="co"># column vals is actual numbers</span></span>
<span id="cb192-15"><a href="#cb192-15" tabindex="-1"></a>           <span class="at">Specs =</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">&quot;mean=0, sd=1&quot;</span>, <span class="dv">1000</span>), <span class="co"># column Specs is mean/sd </span></span>
<span id="cb192-16"><a href="#cb192-16" tabindex="-1"></a>                     <span class="fu">rep</span>(<span class="st">&quot;mean=10, sd=1&quot;</span>, <span class="dv">1000</span>),</span>
<span id="cb192-17"><a href="#cb192-17" tabindex="-1"></a>                     <span class="fu">rep</span>(<span class="st">&quot;mean=20, sd=1&quot;</span>, <span class="dv">1000</span>))</span>
<span id="cb192-18"><a href="#cb192-18" tabindex="-1"></a>             ) <span class="sc">%&gt;%</span></span>
<span id="cb192-19"><a href="#cb192-19" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> vals, </span>
<span id="cb192-20"><a href="#cb192-20" tabindex="-1"></a>             <span class="at">colour =</span> Specs)) <span class="sc">+</span></span>
<span id="cb192-21"><a href="#cb192-21" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins =</span> <span class="dv">50</span>) <span class="sc">+</span> </span>
<span id="cb192-22"><a href="#cb192-22" tabindex="-1"></a>  <span class="fu">geom_freqpoly</span>(<span class="at">bins =</span> <span class="dv">50</span>) <span class="co"># geom_freqpoly adds in the line on top of the histogram</span></span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-94-1.png" width="672" /></p>
<blockquote>
<p>Play with the mean and sd parameters and visualize the distributions (plain lines) as well as the corresponding histograms.</p>
</blockquote>
<hr />
</div>
<div id="application-on-a-real-dataset" class="section level4 hasAnchor">
<h4 class="hasAnchor">Application on a real dataset<a href="#application-on-a-real-dataset" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Now we will use the Normal distribution to make predictions about gene expression of TP53 in lung cancer. TP53 is the most commonly mutated gene across multiple cancer types especially in lung cancers. We will read a table (import) containing measurements of TP53 expression levels in 586 patients.</p>
<div class="sourceCode" id="cb193"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb193-1"><a href="#cb193-1" tabindex="-1"></a>tp53.exp <span class="ot">=</span> <span class="fu">read.table</span>(<span class="st">&quot;https://www.dropbox.com/s/rwopdr8ycmdg8bd/TP53_expression_LungAdeno.txt?dl=1&quot;</span>, </span>
<span id="cb193-2"><a href="#cb193-2" tabindex="-1"></a>                      <span class="at">header=</span>T, <span class="at">sep=</span><span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>)[,<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]</span>
<span id="cb193-3"><a href="#cb193-3" tabindex="-1"></a><span class="fu">summary</span>(tp53.exp)</span></code></pre></div>
<pre><code>##    Samples          TP53_expression 
##  Length:586         Min.   :  92.6  
##  Class :character   1st Qu.: 911.8  
##  Mode  :character   Median :1313.3  
##                     Mean   :1380.8  
##                     3rd Qu.:1778.1  
##                     Max.   :4703.9  
##                     NA&#39;s   :69</code></pre>
<div id="data-cleaning-and-central-values" class="section level5 hasAnchor">
<h5 class="hasAnchor">Data cleaning and central values<a href="#data-cleaning-and-central-values" class="anchor-section" aria-label="Anchor link to header"></a></h5>
<p>We will remove all the missing values and calculate the mean and standard deviation for the TP53 gene expression.</p>
<div class="sourceCode" id="cb195"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb195-1"><a href="#cb195-1" tabindex="-1"></a>tp53.exp <span class="ot">=</span> tp53.exp <span class="sc">%&gt;%</span></span>
<span id="cb195-2"><a href="#cb195-2" tabindex="-1"></a>  <span class="fu">na.omit</span>() </span>
<span id="cb195-3"><a href="#cb195-3" tabindex="-1"></a></span>
<span id="cb195-4"><a href="#cb195-4" tabindex="-1"></a>m.tp53 <span class="ot">=</span> <span class="fu">mean</span>(tp53.exp<span class="sc">$</span>TP53_expression) <span class="co"># mean</span></span>
<span id="cb195-5"><a href="#cb195-5" tabindex="-1"></a>m.tp53</span></code></pre></div>
<pre><code>## [1] 1380.822</code></pre>
<div class="sourceCode" id="cb197"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb197-1"><a href="#cb197-1" tabindex="-1"></a>s.tp53 <span class="ot">=</span> <span class="fu">sd</span>(tp53.exp<span class="sc">$</span>TP53_expression) <span class="co"># standard deviation</span></span>
<span id="cb197-2"><a href="#cb197-2" tabindex="-1"></a>s.tp53</span></code></pre></div>
<pre><code>## [1] 719.5934</code></pre>
</div>
<div id="modeling-using-a-normal-distribution" class="section level5 hasAnchor">
<h5 class="hasAnchor">Modeling using a normal distribution<a href="#modeling-using-a-normal-distribution" class="anchor-section" aria-label="Anchor link to header"></a></h5>
<p>Let’s see how well a normal distribution with <span class="math inline">\(\mu = 1380.822\)</span> (m.tp53) and <span class="math inline">\(\sigma = 719.5934\)</span> (s.tp53) can approximate the <strong>real</strong> distribution of TP53 expression.<br />
We assume that the population mean and standard deviation is similar as calculated above since we cannot measure the expression of TP53 in each and every lung cancer patient in the world.</p>
<div class="sourceCode" id="cb199"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb199-1"><a href="#cb199-1" tabindex="-1"></a><span class="co"># distribution of the measured data</span></span>
<span id="cb199-2"><a href="#cb199-2" tabindex="-1"></a><span class="fu">ggplot</span>(tp53.exp,</span>
<span id="cb199-3"><a href="#cb199-3" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> TP53_expression)) <span class="sc">+</span></span>
<span id="cb199-4"><a href="#cb199-4" tabindex="-1"></a>  <span class="fu">geom_density</span>() <span class="sc">+</span> </span>
<span id="cb199-5"><a href="#cb199-5" tabindex="-1"></a>  <span class="fu">xlim</span>(<span class="sc">-</span><span class="dv">1500</span>, <span class="dv">6000</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-97-1.png" width="672" /></p>
<p>Make a normal distribution with the above parameters</p>
<div class="sourceCode" id="cb200"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb200-1"><a href="#cb200-1" tabindex="-1"></a>x <span class="ot">=</span> <span class="fu">seq</span>(<span class="dv">0</span>,<span class="dv">5000</span>,<span class="at">by=</span><span class="dv">5</span>)</span>
<span id="cb200-2"><a href="#cb200-2" tabindex="-1"></a>d.pred <span class="ot">=</span> <span class="fu">dnorm</span>(x, <span class="at">mean =</span> m.tp53, <span class="at">sd =</span> s.tp53)</span>
<span id="cb200-3"><a href="#cb200-3" tabindex="-1"></a></span>
<span id="cb200-4"><a href="#cb200-4" tabindex="-1"></a><span class="co"># Now plot both, predicted and measured data</span></span>
<span id="cb200-5"><a href="#cb200-5" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb200-6"><a href="#cb200-6" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="fu">aes</span>(<span class="at">x =</span> tp53.exp<span class="sc">$</span>TP53_expression)) <span class="sc">+</span></span>
<span id="cb200-7"><a href="#cb200-7" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">x =</span> x, </span>
<span id="cb200-8"><a href="#cb200-8" tabindex="-1"></a>                <span class="at">y =</span> d.pred),</span>
<span id="cb200-9"><a href="#cb200-9" tabindex="-1"></a>                <span class="at">colour =</span> <span class="st">&quot;red&quot;</span>) <span class="sc">+</span></span>
<span id="cb200-10"><a href="#cb200-10" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">c</span>(<span class="fu">quantile</span>(tp53.exp<span class="sc">$</span>TP53_expression, <span class="at">probs =</span> <span class="fu">c</span>(<span class="fl">0.1</span>, <span class="fl">0.9</span>)))) <span class="sc">+</span></span>
<span id="cb200-11"><a href="#cb200-11" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">c</span>(<span class="fu">qnorm</span>(<span class="at">p =</span> <span class="fl">0.1</span>, <span class="at">mean =</span> m.tp53, <span class="at">sd =</span> s.tp53), </span>
<span id="cb200-12"><a href="#cb200-12" tabindex="-1"></a>                            <span class="fu">qnorm</span>(<span class="at">p =</span> <span class="fl">0.9</span>, <span class="at">mean =</span> m.tp53, <span class="at">sd =</span> s.tp53)),</span>
<span id="cb200-13"><a href="#cb200-13" tabindex="-1"></a>             <span class="at">colour =</span> <span class="st">&#39;red&#39;</span>, <span class="at">linetype =</span> <span class="st">&#39;dotted&#39;</span>) </span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-98-1.png" width="672" /></p>
</div>
<div id="data-prediction-using-the-normal-distribution-model" class="section level5 hasAnchor">
<h5 class="hasAnchor">Data prediction using the normal distribution model<a href="#data-prediction-using-the-normal-distribution-model" class="anchor-section" aria-label="Anchor link to header"></a></h5>
<p>Using a normal distribution with <span class="math inline">\(\mu = 1380.822\)</span> (m.tp53) and <span class="math inline">\(\sigma = 719.5934\)</span> (s.tp53), we will ask the following questions -</p>
<p>– <strong>(Q1)</strong> What is the probability of observing the expression of TP53 to be <strong>less</strong> than 1000?</p>
<div class="sourceCode" id="cb201"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb201-1"><a href="#cb201-1" tabindex="-1"></a><span class="fu">pnorm</span>(<span class="at">q=</span><span class="dv">1000</span>, <span class="at">mean =</span>m.tp53, <span class="at">sd =</span> s.tp53) <span class="co"># returns the cumulative probability</span></span></code></pre></div>
<pre><code>## [1] 0.2983271</code></pre>
<p>– <strong>(Q2)</strong> What is the probability of observing the expression of TP53 to be <strong>greater</strong> than 1000?</p>
<div class="sourceCode" id="cb203"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb203-1"><a href="#cb203-1" tabindex="-1"></a><span class="dv">1</span> <span class="sc">-</span> <span class="fu">pnorm</span>(<span class="at">q=</span><span class="dv">1000</span>, <span class="at">mean =</span>m.tp53, <span class="at">sd =</span> s.tp53) </span></code></pre></div>
<pre><code>## [1] 0.7016729</code></pre>
<div class="sourceCode" id="cb205"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb205-1"><a href="#cb205-1" tabindex="-1"></a><span class="co"># is same as</span></span>
<span id="cb205-2"><a href="#cb205-2" tabindex="-1"></a><span class="fu">pnorm</span>(<span class="at">q=</span><span class="dv">1000</span>, <span class="at">mean =</span>m.tp53, <span class="at">sd =</span> s.tp53, <span class="at">lower.tail =</span> <span class="cn">FALSE</span>) <span class="co"># or pnorm(q=1000, mean =m.tp53, sd = s.tp53, lower.tail = F)</span></span></code></pre></div>
<pre><code>## [1] 0.7016729</code></pre>
</div>
<div id="evaluating-the-quality-of-the-predictions" class="section level5 hasAnchor">
<h5 class="hasAnchor">Evaluating the quality of the predictions<a href="#evaluating-the-quality-of-the-predictions" class="anchor-section" aria-label="Anchor link to header"></a></h5>
<p>Let’s check how good these predictions are compared to <strong>real</strong> data.
– <strong>(Q1)</strong> What is the probability of observing the expression of TP53 to be <strong>less</strong> than 1000?</p>
<div class="sourceCode" id="cb207"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb207-1"><a href="#cb207-1" tabindex="-1"></a><span class="fu">sum</span>(tp53.exp<span class="sc">$</span>TP53_expression <span class="sc">&lt;</span> <span class="dv">1000</span>)<span class="sc">/</span><span class="fu">nrow</span>(tp53.exp)</span></code></pre></div>
<pre><code>## [1] 0.2978723</code></pre>
<p>– <strong>(Q2)</strong> What is the probability of observing the expression of TP53 to be <strong>greater</strong> than 1000?</p>
<div class="sourceCode" id="cb209"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb209-1"><a href="#cb209-1" tabindex="-1"></a><span class="fu">sum</span>(tp53.exp<span class="sc">$</span>TP53_expression <span class="sc">&gt;</span> <span class="dv">1000</span>)<span class="sc">/</span><span class="fu">nrow</span>(tp53.exp)</span></code></pre></div>
<pre><code>## [1] 0.7021277</code></pre>
<blockquote>
<p>I would say those predictions are pretty good !! Now, let’s try to break this model. Re-execute the code above with different <span class="math inline">\(q\)</span> values <code>q=100, q=500, q=4000, q=4500</code> etc. At what values do you think the model would not perform well. HINT: Look at the tails of the distribution!</p>
</blockquote>
<div class="sourceCode" id="cb211"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb211-1"><a href="#cb211-1" tabindex="-1"></a><span class="co"># What is the probability of observing the expression of TP53 to be less than q?</span></span>
<span id="cb211-2"><a href="#cb211-2" tabindex="-1"></a>q <span class="ot">=</span> <span class="fu">c</span>(<span class="dv">100</span>,<span class="dv">500</span>,<span class="dv">1000</span>,<span class="dv">4000</span>,<span class="dv">4500</span>)</span>
<span id="cb211-3"><a href="#cb211-3" tabindex="-1"></a></span>
<span id="cb211-4"><a href="#cb211-4" tabindex="-1"></a><span class="co"># for loop used to calculate and store the predicted (a) and real (b) values</span></span>
<span id="cb211-5"><a href="#cb211-5" tabindex="-1"></a>pred <span class="ot">=</span> <span class="fu">numeric</span>()</span>
<span id="cb211-6"><a href="#cb211-6" tabindex="-1"></a>meas <span class="ot">=</span> <span class="fu">numeric</span>()</span>
<span id="cb211-7"><a href="#cb211-7" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(q)){</span>
<span id="cb211-8"><a href="#cb211-8" tabindex="-1"></a>  pred[i] <span class="ot">=</span> <span class="fu">pnorm</span>(<span class="at">q=</span>q[i], <span class="at">mean =</span> m.tp53, <span class="at">sd =</span> s.tp53) </span>
<span id="cb211-9"><a href="#cb211-9" tabindex="-1"></a>  meas[i] <span class="ot">=</span> <span class="fu">sum</span>(tp53.exp<span class="sc">$</span>TP53_expression <span class="sc">&lt;</span> q[i])<span class="sc">/</span><span class="fu">nrow</span>(tp53.exp)</span>
<span id="cb211-10"><a href="#cb211-10" tabindex="-1"></a>}</span>
<span id="cb211-11"><a href="#cb211-11" tabindex="-1"></a></span>
<span id="cb211-12"><a href="#cb211-12" tabindex="-1"></a><span class="co"># Change into a dataframe </span></span>
<span id="cb211-13"><a href="#cb211-13" tabindex="-1"></a>model_mat <span class="ot">=</span> <span class="fu">data.frame</span>(pred, meas, q)</span>
<span id="cb211-14"><a href="#cb211-14" tabindex="-1"></a>model_mat</span></code></pre></div>
<pre><code>##         pred        meas    q
## 1 0.03754417 0.001934236  100
## 2 0.11046576 0.104448743  500
## 3 0.29832708 0.297872340 1000
## 4 0.99986358 0.992263056 4000
## 5 0.99999270 0.996131528 4500</code></pre>
<p>Again, using a normal distribution with <span class="math inline">\(\mu = 1380.822\)</span> and <span class="math inline">\(\sigma = 719.5934\)</span>, what if we ask what is the value of TP53 expression at the 10% and 90% quantiles:</p>
<div class="sourceCode" id="cb213"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb213-1"><a href="#cb213-1" tabindex="-1"></a><span class="fu">qnorm</span>(<span class="at">p =</span> <span class="fl">0.1</span>, <span class="at">mean =</span> m.tp53, <span class="at">sd =</span> s.tp53)</span></code></pre></div>
<pre><code>## [1] 458.6258</code></pre>
<div class="sourceCode" id="cb215"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb215-1"><a href="#cb215-1" tabindex="-1"></a><span class="fu">qnorm</span>(<span class="at">p =</span> <span class="fl">0.9</span>, <span class="at">mean =</span> m.tp53, <span class="at">sd =</span> s.tp53)</span></code></pre></div>
<pre><code>## [1] 2303.018</code></pre>
<p>Let’s check how good these predictions are compared to our real data.</p>
<div class="sourceCode" id="cb217"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb217-1"><a href="#cb217-1" tabindex="-1"></a><span class="fu">quantile</span>(tp53.exp<span class="sc">$</span>TP53_expression, </span>
<span id="cb217-2"><a href="#cb217-2" tabindex="-1"></a>         <span class="at">probs =</span> <span class="fu">c</span>(<span class="fl">0.1</span>, <span class="fl">0.9</span>))</span></code></pre></div>
<pre><code>##       10%       90% 
##  478.2844 2240.1913</code></pre>
<p>Again the predictions are pretty good!</p>
</div>
<div id="visualization-1" class="section level5 hasAnchor">
<h5 class="hasAnchor">Visualization<a href="#visualization-1" class="anchor-section" aria-label="Anchor link to header"></a></h5>
<p>We can also visualize all of this on a simple graph:</p>
<div class="sourceCode" id="cb219"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb219-1"><a href="#cb219-1" tabindex="-1"></a><span class="co"># Model prediction:</span></span>
<span id="cb219-2"><a href="#cb219-2" tabindex="-1"></a>x <span class="ot">=</span> <span class="fu">seq</span>(<span class="dv">0</span>,<span class="dv">5000</span>,<span class="at">by=</span><span class="dv">5</span>)</span>
<span id="cb219-3"><a href="#cb219-3" tabindex="-1"></a>d.pred <span class="ot">=</span> <span class="fu">dnorm</span>(x,<span class="at">mean =</span> m.tp53, <span class="at">sd =</span> s.tp53)</span>
<span id="cb219-4"><a href="#cb219-4" tabindex="-1"></a></span>
<span id="cb219-5"><a href="#cb219-5" tabindex="-1"></a><span class="co"># Model and measured data and predicted versus measured 0.1 and 0.9 quantiles:</span></span>
<span id="cb219-6"><a href="#cb219-6" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb219-7"><a href="#cb219-7" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">x =</span> x, </span>
<span id="cb219-8"><a href="#cb219-8" tabindex="-1"></a>                <span class="at">y =</span> d.pred),</span>
<span id="cb219-9"><a href="#cb219-9" tabindex="-1"></a>            <span class="at">colour =</span> <span class="st">&quot;red&quot;</span>) <span class="sc">+</span></span>
<span id="cb219-10"><a href="#cb219-10" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="fu">aes</span>(<span class="at">x =</span> tp53.exp<span class="sc">$</span>TP53_expression)) </span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-106-1.png" width="672" /></p>
<p>Compare the black and red vertical lines (real vs predicted).</p>
<blockquote>
<p>Re-execute the code above with <code>p=0.25, p=0.5, p=0.75 etc</code> and check how good the predictions are.</p>
</blockquote>
</div>
<div id="visualization-using-a-q-q-plot" class="section level5 hasAnchor">
<h5 class="hasAnchor">Visualization using a Q-Q plot<a href="#visualization-using-a-q-q-plot" class="anchor-section" aria-label="Anchor link to header"></a></h5>
<p>Now, let’s plot the sample quantiles against theoretical quantiles to check the similarity between the two. This is called a quantile - quantile plot or a Q-Q plot, which you are familiar with (see Exercises Sheet 1).</p>
<div class="sourceCode" id="cb220"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb220-1"><a href="#cb220-1" tabindex="-1"></a>q <span class="ot">=</span> <span class="fu">seq</span>(<span class="dv">0</span>,<span class="dv">1</span>,<span class="fl">0.01</span>) <span class="co"># Creating a vector of quantiles</span></span>
<span id="cb220-2"><a href="#cb220-2" tabindex="-1"></a></span>
<span id="cb220-3"><a href="#cb220-3" tabindex="-1"></a><span class="co"># Find values corresponding to these quantiles in the real data</span></span>
<span id="cb220-4"><a href="#cb220-4" tabindex="-1"></a>q.observed <span class="ot">=</span> <span class="fu">quantile</span>(tp53.exp<span class="sc">$</span>TP53_expression, <span class="at">probs =</span> q)</span>
<span id="cb220-5"><a href="#cb220-5" tabindex="-1"></a></span>
<span id="cb220-6"><a href="#cb220-6" tabindex="-1"></a><span class="co"># Find values corresponding to these quantiles in the theoretical normal distribution</span></span>
<span id="cb220-7"><a href="#cb220-7" tabindex="-1"></a>q.theoretical <span class="ot">=</span> <span class="fu">qnorm</span>(<span class="at">p =</span> q, <span class="at">mean =</span> m.tp53, <span class="at">sd =</span> s.tp53)</span>
<span id="cb220-8"><a href="#cb220-8" tabindex="-1"></a></span>
<span id="cb220-9"><a href="#cb220-9" tabindex="-1"></a><span class="co"># # Correlate the above two values</span></span>
<span id="cb220-10"><a href="#cb220-10" tabindex="-1"></a><span class="fu">ggplot</span>(tp53.exp,</span>
<span id="cb220-11"><a href="#cb220-11" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">sample =</span> TP53_expression)) <span class="sc">+</span></span>
<span id="cb220-12"><a href="#cb220-12" tabindex="-1"></a>  <span class="fu">geom_qq</span>() <span class="sc">+</span></span>
<span id="cb220-13"><a href="#cb220-13" tabindex="-1"></a>  <span class="fu">geom_qq_line</span>(<span class="at">colour =</span> <span class="st">&#39;red&#39;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-107-1.png" width="480" /></p>
<div class="sourceCode" id="cb221"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb221-1"><a href="#cb221-1" tabindex="-1"></a><span class="do">## Would be the same as:</span></span>
<span id="cb221-2"><a href="#cb221-2" tabindex="-1"></a><span class="co"># ggplot() +</span></span>
<span id="cb221-3"><a href="#cb221-3" tabindex="-1"></a><span class="co">#   geom_point(aes(x = q.theoretical, </span></span>
<span id="cb221-4"><a href="#cb221-4" tabindex="-1"></a><span class="co">#                  y = q.observed), size = 1) +</span></span>
<span id="cb221-5"><a href="#cb221-5" tabindex="-1"></a><span class="co">#   geom_abline(intercept = 0, slope = 1, </span></span>
<span id="cb221-6"><a href="#cb221-6" tabindex="-1"></a><span class="co">#               size = 1, color = &quot;red&quot;) </span></span></code></pre></div>
<hr />
</div>
</div>
<div id="binomial-distribution" class="section level4 hasAnchor">
<h4 class="hasAnchor">Binomial distribution<a href="#binomial-distribution" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>A binomial distribution can be defined as -</p>
<p><span class="math display">\[P(x) = \frac{n!}{x!(n-x)!}\cdot p^x \cdot (1-p)^{n-x}\]</span></p>
<p>Where <span class="math inline">\(x\)</span> is the number of successes out of <span class="math inline">\(n\)</span> experiments and <span class="math inline">\(p\)</span> is the probability of success.</p>
<ul>
<li><span class="math inline">\(mean = n \cdot p\)</span></li>
<li><span class="math inline">\(variance = np \cdot (1 - p)\)</span></li>
<li><span class="math inline">\(sd = \sqrt{np \cdot (1 - p)}\)</span></li>
</ul>
<p>The design of the experiment is as follows -</p>
<ul>
<li>The experiment is repeated and are independent of one another</li>
<li>Each experiment has just two outcomes</li>
<li>The probability of success is constant and does not change with experiments</li>
</ul>
<p>We can for example compute the probability of having 7 heads in a series of 10 throws of a coin:</p>
<div class="sourceCode" id="cb222"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb222-1"><a href="#cb222-1" tabindex="-1"></a><span class="co"># x = number of successes; size = number of trials; prob = probability of success on each trial</span></span>
<span id="cb222-2"><a href="#cb222-2" tabindex="-1"></a><span class="fu">dbinom</span>(<span class="at">x=</span><span class="dv">7</span>, <span class="at">size=</span><span class="dv">10</span>, <span class="at">prob =</span> <span class="fl">0.5</span>) </span></code></pre></div>
<pre><code>## [1] 0.1171875</code></pre>
<p>Or we can compute what the probability is to get 7 or more heads using the function <code>pbinom()</code>. Remember that the parameter “lower.tail” is used to specify whether to calculate the probability of observing x or fewer successes (if <code>lower.tail = TRUE</code>) or the probability of observing more than x successes (<code>lower.tail = FALSE</code>):</p>
<div class="sourceCode" id="cb224"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb224-1"><a href="#cb224-1" tabindex="-1"></a><span class="fu">pbinom</span>(<span class="dv">6</span>, <span class="at">size=</span><span class="dv">10</span>,</span>
<span id="cb224-2"><a href="#cb224-2" tabindex="-1"></a>       <span class="at">prob=</span><span class="fl">0.5</span>,</span>
<span id="cb224-3"><a href="#cb224-3" tabindex="-1"></a>       <span class="at">lower.tail=</span><span class="cn">FALSE</span>) </span></code></pre></div>
<pre><code>## [1] 0.171875</code></pre>
<p>Beware that this syntax means <strong>strictly more than 6</strong>, i.e. 7 or more!!</p>
<blockquote>
<p>How would you compute the probability to get less than 5? What would <code>qbinom(0.3,size=10,prob=0.5,lower.tail=FALSE)</code> represent?</p>
</blockquote>
<hr />
</div>
</div>
<div id="confidence-interval" class="section level3 hasAnchor">
<h3 class="hasAnchor">Confidence interval<a href="#confidence-interval" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The <strong>confidence interval</strong> describes the interval containing the (unknown) expectation value of a distribution with 95% confidence. This means that out of 100 random realizations of this random variable, the true expectation value <span class="math inline">\(\mu\)</span> will indeed be in this interval.</p>
<p>Let us try a simulation: we consider a random variable distributed according to a Poisson distribution <span class="math display">\[P(x) = \frac{{e^{ - \lambda } \lambda ^x }}{{x!}}\]</span> Here, <em>we know the true value of the expectation value</em>. We want to get an estimate for <span class="math inline">\(\lambda\)</span>, and check if the confidence interval contains the true expectation value.</p>
<p>For example, a farmer expects to collect 75 eggs from his hens per hour.</p>
<div class="sourceCode" id="cb226"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb226-1"><a href="#cb226-1" tabindex="-1"></a>lambda <span class="ot">=</span> <span class="dv">75</span></span></code></pre></div>
<p>He now collects during 100 days the eggs <span class="math inline">\(N=8\)</span> times a day (each time during one hour). We want to compute the mean <span class="math inline">\(m_N\)</span> over these <span class="math inline">\(N=8\)</span> realizations and determine the 95% confidence interval, and check, how often the expectation value <span class="math inline">\(\mu\)</span> is inside the confidence interval.</p>
<p>Remember that the 95% CI is given by <span class="math display">\[[m_N-t_{95,N-1}\frac{\sigma}{\sqrt{N}},m_N+t_{95,N-1}\frac{\sigma}{\sqrt{N}}]\]</span> where <span class="math inline">\(t_{95,N-1}\)</span> is the critical value for the <span class="math inline">\(t\)</span>-distribution with <span class="math inline">\(n-1\)</span> degrees of freedom.</p>
<p>Let’s start by creating our samples:</p>
<div class="sourceCode" id="cb227"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb227-1"><a href="#cb227-1" tabindex="-1"></a><span class="co"># size of the sample</span></span>
<span id="cb227-2"><a href="#cb227-2" tabindex="-1"></a>N <span class="ot">=</span> <span class="dv">8</span></span>
<span id="cb227-3"><a href="#cb227-3" tabindex="-1"></a></span>
<span id="cb227-4"><a href="#cb227-4" tabindex="-1"></a><span class="co"># we now draw 100 times samples of size N=8</span></span>
<span id="cb227-5"><a href="#cb227-5" tabindex="-1"></a><span class="do">## rpois is the function used to generate the Poisson distribution</span></span>
<span id="cb227-6"><a href="#cb227-6" tabindex="-1"></a>X <span class="ot">=</span> <span class="fu">lapply</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">100</span>, <span class="cf">function</span>(i) {</span>
<span id="cb227-7"><a href="#cb227-7" tabindex="-1"></a>  <span class="fu">rpois</span>(N, <span class="at">lambda =</span> lambda)</span>
<span id="cb227-8"><a href="#cb227-8" tabindex="-1"></a>  })</span></code></pre></div>
<p><code>lapply()</code> is a function in R that stands for “list apply”. It is used to <em>apply</em> a function to each element of a list or vector and produces a list with the same length as an output.</p>
<p>In this previous example, the input is a vector (<code>1:100</code>).
Then, we <code>lapply(v, function(i))</code>:
+ v: The list or vector you want to apply the function to.
+ function: The function you want to apply where <code>i</code> is each element in X. In this case, we are applying <code>rpois</code> 100 times.</p>
<p>As an output, <code>lapply</code> returns a list where each element has had the specified function applied to it.</p>
<blockquote>
<p>Run View(X) to see how this object looks like. Try using lapply to obtain the mean of all elements in each X using <code>lapply(X, function(i){mean(i)})</code> or <code>lapply(X, mean)</code>.</p>
</blockquote>
<p>Now, we calculate the mean and the standard deviation of the respective samples:</p>
<div class="sourceCode" id="cb228"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb228-1"><a href="#cb228-1" tabindex="-1"></a><span class="co"># we compute the sample means</span></span>
<span id="cb228-2"><a href="#cb228-2" tabindex="-1"></a>Xm <span class="ot">=</span> <span class="fu">sapply</span>(X,mean)</span>
<span id="cb228-3"><a href="#cb228-3" tabindex="-1"></a><span class="co"># and the sample standard deviations</span></span>
<span id="cb228-4"><a href="#cb228-4" tabindex="-1"></a>Xsd <span class="ot">=</span> <span class="fu">sapply</span>(X,sd) </span></code></pre></div>
<p>Next, we determine the upper and lower bounds of the 95% CI. Remember that the confidence interval is based on a <span class="math inline">\(t\)</span>-distribution. The degrees of freedom of this distribution is the sample size -1 (<span class="math inline">\(N\)</span>-1=7 in this case)</p>
<div class="sourceCode" id="cb229"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb229-1"><a href="#cb229-1" tabindex="-1"></a>df <span class="ot">=</span> N<span class="dv">-1</span></span>
<span id="cb229-2"><a href="#cb229-2" tabindex="-1"></a>tc <span class="ot">=</span> <span class="fu">qt</span>(<span class="fu">c</span>(<span class="fl">0.975</span>),df) <span class="co"># this is the critical value for the t-distribution for df = N-1 degrees of freedom and 95% CI</span></span>
<span id="cb229-3"><a href="#cb229-3" tabindex="-1"></a></span>
<span id="cb229-4"><a href="#cb229-4" tabindex="-1"></a>Xl <span class="ot">=</span> Xm<span class="sc">-</span>tc<span class="sc">*</span>Xsd<span class="sc">/</span><span class="fu">sqrt</span>(N) <span class="co"># upper bound of the 95% CI</span></span>
<span id="cb229-5"><a href="#cb229-5" tabindex="-1"></a>Xh <span class="ot">=</span> Xm<span class="sc">+</span>tc<span class="sc">*</span>Xsd<span class="sc">/</span><span class="fu">sqrt</span>(N) <span class="co"># lower bound of the 95% CI</span></span></code></pre></div>
<p>Finally, we determine whether each sample mean is found within the 95% CI or not:</p>
<div class="sourceCode" id="cb230"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb230-1"><a href="#cb230-1" tabindex="-1"></a><span class="do">## vector of TRUE/FALSE if the real expectation value lambda is inside the interval</span></span>
<span id="cb230-2"><a href="#cb230-2" tabindex="-1"></a>i.ok <span class="ot">=</span>  <span class="fu">as.factor</span>(Xl <span class="sc">&lt;</span> lambda <span class="sc">&amp;</span> Xh <span class="sc">&gt;</span> lambda)</span>
<span id="cb230-3"><a href="#cb230-3" tabindex="-1"></a></span>
<span id="cb230-4"><a href="#cb230-4" tabindex="-1"></a>plot_data <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb230-5"><a href="#cb230-5" tabindex="-1"></a>  <span class="at">n =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">100</span>,</span>
<span id="cb230-6"><a href="#cb230-6" tabindex="-1"></a>  <span class="at">Xm =</span> Xm,</span>
<span id="cb230-7"><a href="#cb230-7" tabindex="-1"></a>  <span class="at">Xl =</span> Xl,</span>
<span id="cb230-8"><a href="#cb230-8" tabindex="-1"></a>  <span class="at">Xh =</span> Xh,</span>
<span id="cb230-9"><a href="#cb230-9" tabindex="-1"></a>  <span class="at">i.ok =</span> i.ok</span>
<span id="cb230-10"><a href="#cb230-10" tabindex="-1"></a>) </span>
<span id="cb230-11"><a href="#cb230-11" tabindex="-1"></a></span>
<span id="cb230-12"><a href="#cb230-12" tabindex="-1"></a><span class="co"># Plot using ggplot2</span></span>
<span id="cb230-13"><a href="#cb230-13" tabindex="-1"></a><span class="fu">ggplot</span>(plot_data, </span>
<span id="cb230-14"><a href="#cb230-14" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> n, <span class="at">y =</span> Xm, </span>
<span id="cb230-15"><a href="#cb230-15" tabindex="-1"></a>           <span class="at">color =</span> <span class="st">`</span><span class="at">i.ok</span><span class="st">`</span>)) <span class="sc">+</span></span>
<span id="cb230-16"><a href="#cb230-16" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb230-17"><a href="#cb230-17" tabindex="-1"></a>  <span class="fu">geom_errorbar</span>(<span class="fu">aes</span>(<span class="at">ymin =</span> Xl, <span class="at">ymax =</span> Xh, <span class="at">color =</span> i.ok), <span class="at">width =</span> <span class="fl">0.2</span>) <span class="sc">+</span></span>
<span id="cb230-18"><a href="#cb230-18" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> lambda, <span class="at">linetype =</span> <span class="st">&quot;dashed&quot;</span>, <span class="at">color =</span> <span class="st">&quot;black&quot;</span>) <span class="sc">+</span></span>
<span id="cb230-19"><a href="#cb230-19" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">&quot;grey&quot;</span>, <span class="st">&quot;orange&quot;</span>)) <span class="sc">+</span></span>
<span id="cb230-20"><a href="#cb230-20" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">y =</span> <span class="st">&quot;Mean values&quot;</span>,</span>
<span id="cb230-21"><a href="#cb230-21" tabindex="-1"></a>       <span class="at">title =</span> <span class="fu">paste</span>(<span class="st">&quot;Mean values and confidence intervals, N=&quot;</span>,N),</span>
<span id="cb230-22"><a href="#cb230-22" tabindex="-1"></a>       <span class="at">color =</span> <span class="st">&quot;95% CI Status&quot;</span>) </span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-114-1.png" width="768" /></p>
<p>Here, the orange/grey bars represent the confidence interval, the dot is the mean of the sample values, and the dotted line at <code>\lambda</code> represents the true expectation value. Whenever the true expectation value is within the CI, the bar is orange, if not, the bar is grey. How often is the true expectation value outside the CI? Count the grey bars!</p>
<p>It happens 5 times, which fits pretty well with the expected 5%.</p>
<blockquote>
<p>Repeat this simulation, but now with samples of <span class="math inline">\(N=24\)</span> (again 100 times)<br />
What do you observe?<br />
How often is the true expectation value outside the CI? Change to 90% CI and check if that works!</p>
</blockquote>
<hr />
</div>
<div id="exercises-2" class="section level3 hasAnchor">
<h3 class="hasAnchor">EXERCISES<a href="#exercises-2" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="exercise-1-probability-distributions" class="section level4 hasAnchor">
<h4 class="hasAnchor">Exercise 1: Probability distributions<a href="#exercise-1-probability-distributions" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<ol style="list-style-type: decimal">
<li><p>What is the expression of TP53 observed at 10th percentile? What is the expression of TP53 observed at 90th percentile?</p></li>
<li><p>(optional) Other distribution types can become very similar to the normal distribution under certain conditions. Plot the histogram of 1000 random numbers drawn from the Poisson distribution with lambda = 1, 10, 100, 1000. What do you observe?</p></li>
</ol>
</div>
<div id="exercise-2-confidence-intervals" class="section level4 hasAnchor">
<h4 class="hasAnchor">Exercise 2: Confidence intervals<a href="#exercise-2-confidence-intervals" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>You are buying 10 packs of gummy bears. You particularly like the red ones and the green ones. A pack contains 6 different colors and you expect them to be equally distributed. There are 84 pieces per 200g pack.</p>
<ol style="list-style-type: decimal">
<li><p>What is the expected amount of red or green gummy bears?</p></li>
<li><p>You selected your 10 packs according to the colors you could see in the pack. At home, you counted the following bears per pack:</p></li>
</ol>
<ul>
<li>for the red ones: 12 16 17 12 16 13 11 18 13 19<br />
</li>
<li>for the green ones: 11 10 15 16 12 14 13 10 13 17<br />
Was your selection procedure a success? In other words, is the expected value bellow (congrats!), within or above (bad luck!) the 95% CI?</li>
</ul>
</div>
</div>
</div>
<div id="day-4-hypothesis-testing" class="section level2 hasAnchor">
<h2 class="hasAnchor">Day 4: Hypothesis testing<a href="#day-4-hypothesis-testing" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>On this section, we will go through hypothesis testing. You will start to see how to <em>formulate hypotheses</em> and how to <em>test</em> them.
In addition, we want to learn how to use and interpret hypothesis tests.</p>
<p>We will work again with the diabetes dataset that we used previously.</p>
<div class="sourceCode" id="cb231"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb231-1"><a href="#cb231-1" tabindex="-1"></a>dat <span class="ot">=</span> <span class="fu">read.delim</span>(<span class="st">&#39;https://tinyurl.com/y4fark9g&#39;</span>)</span>
<span id="cb231-2"><a href="#cb231-2" tabindex="-1"></a></span>
<span id="cb231-3"><a href="#cb231-3" tabindex="-1"></a><span class="co"># set the row names using the column id</span></span>
<span id="cb231-4"><a href="#cb231-4" tabindex="-1"></a><span class="fu">rownames</span>(dat) <span class="ot">=</span> dat<span class="sc">$</span>id</span></code></pre></div>
<p>Load the required packages</p>
<div class="sourceCode" id="cb232"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb232-1"><a href="#cb232-1" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb232-2"><a href="#cb232-2" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb232-3"><a href="#cb232-3" tabindex="-1"></a><span class="fu">library</span>(tibble)</span></code></pre></div>
<p>Check out the content of the dataset using the summary function</p>
<div class="sourceCode" id="cb233"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb233-1"><a href="#cb233-1" tabindex="-1"></a><span class="fu">summary</span>(dat)</span></code></pre></div>
<pre><code>##        id             chol          stab.glu          hdl        
##  Min.   : 1000   Min.   : 78.0   Min.   : 48.0   Min.   : 12.00  
##  1st Qu.: 4792   1st Qu.:179.0   1st Qu.: 81.0   1st Qu.: 38.00  
##  Median :15766   Median :204.0   Median : 89.0   Median : 46.00  
##  Mean   :15978   Mean   :207.8   Mean   :106.7   Mean   : 50.45  
##  3rd Qu.:20336   3rd Qu.:230.0   3rd Qu.:106.0   3rd Qu.: 59.00  
##  Max.   :41756   Max.   :443.0   Max.   :385.0   Max.   :120.00  
##                  NA&#39;s   :1                       NA&#39;s   :1       
##      ratio            glyhb         location              age       
##  Min.   : 1.500   Min.   : 2.68   Length:403         Min.   :19.00  
##  1st Qu.: 3.200   1st Qu.: 4.38   Class :character   1st Qu.:34.00  
##  Median : 4.200   Median : 4.84   Mode  :character   Median :45.00  
##  Mean   : 4.522   Mean   : 5.59                      Mean   :46.85  
##  3rd Qu.: 5.400   3rd Qu.: 5.60                      3rd Qu.:60.00  
##  Max.   :19.300   Max.   :16.11                      Max.   :92.00  
##  NA&#39;s   :1        NA&#39;s   :13                                        
##     gender              height          weight         frame          
##  Length:403         Min.   :52.00   Min.   : 99.0   Length:403        
##  Class :character   1st Qu.:63.00   1st Qu.:151.0   Class :character  
##  Mode  :character   Median :66.00   Median :172.5   Mode  :character  
##                     Mean   :66.02   Mean   :177.6                     
##                     3rd Qu.:69.00   3rd Qu.:200.0                     
##                     Max.   :76.00   Max.   :325.0                     
##                     NA&#39;s   :5       NA&#39;s   :1                         
##      bp.1s           bp.1d            bp.2s           bp.2d       
##  Min.   : 90.0   Min.   : 48.00   Min.   :110.0   Min.   : 60.00  
##  1st Qu.:121.2   1st Qu.: 75.00   1st Qu.:138.0   1st Qu.: 84.00  
##  Median :136.0   Median : 82.00   Median :149.0   Median : 92.00  
##  Mean   :136.9   Mean   : 83.32   Mean   :152.4   Mean   : 92.52  
##  3rd Qu.:146.8   3rd Qu.: 90.00   3rd Qu.:161.0   3rd Qu.:100.00  
##  Max.   :250.0   Max.   :124.00   Max.   :238.0   Max.   :124.00  
##  NA&#39;s   :5       NA&#39;s   :5        NA&#39;s   :262     NA&#39;s   :262     
##      waist           hip           time.ppn     
##  Min.   :26.0   Min.   :30.00   Min.   :   5.0  
##  1st Qu.:33.0   1st Qu.:39.00   1st Qu.:  90.0  
##  Median :37.0   Median :42.00   Median : 240.0  
##  Mean   :37.9   Mean   :43.04   Mean   : 341.2  
##  3rd Qu.:41.0   3rd Qu.:46.00   3rd Qu.: 517.5  
##  Max.   :56.0   Max.   :64.00   Max.   :1560.0  
##  NA&#39;s   :2      NA&#39;s   :2       NA&#39;s   :3</code></pre>
<p>How can we inspect the differences between the weight of men and women? We can start by ploting two histograms or density plots representing the weight by biological sex.</p>
<div class="sourceCode" id="cb235"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb235-1"><a href="#cb235-1" tabindex="-1"></a><span class="fu">ggplot</span>(dat,</span>
<span id="cb235-2"><a href="#cb235-2" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> weight, </span>
<span id="cb235-3"><a href="#cb235-3" tabindex="-1"></a>           <span class="at">fill =</span> gender)) <span class="sc">+</span></span>
<span id="cb235-4"><a href="#cb235-4" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="at">alpha =</span> <span class="fl">0.5</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-118-1.png" width="672" /></p>
<p>The distributions look different in shape, right? Where do you think the mean would be located in the plot?</p>
<div class="sourceCode" id="cb236"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb236-1"><a href="#cb236-1" tabindex="-1"></a><span class="co"># We can use &quot;filter()&quot; to filter the cholesterol values for men and women</span></span>
<span id="cb236-2"><a href="#cb236-2" tabindex="-1"></a>dat.male <span class="ot">=</span> dat <span class="sc">%&gt;%</span></span>
<span id="cb236-3"><a href="#cb236-3" tabindex="-1"></a>  <span class="fu">filter</span>(gender <span class="sc">==</span> <span class="st">&#39;male&#39;</span>)</span>
<span id="cb236-4"><a href="#cb236-4" tabindex="-1"></a></span>
<span id="cb236-5"><a href="#cb236-5" tabindex="-1"></a>dat.female <span class="ot">=</span> dat <span class="sc">%&gt;%</span></span>
<span id="cb236-6"><a href="#cb236-6" tabindex="-1"></a>  <span class="fu">filter</span>(gender <span class="sc">==</span> <span class="st">&#39;female&#39;</span>)</span>
<span id="cb236-7"><a href="#cb236-7" tabindex="-1"></a></span>
<span id="cb236-8"><a href="#cb236-8" tabindex="-1"></a><span class="co"># we will calculate the mean weight by sex.</span></span>
<span id="cb236-9"><a href="#cb236-9" tabindex="-1"></a>mean.men <span class="ot">&lt;-</span> <span class="fu">mean</span>(dat.male<span class="sc">$</span>weight, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb236-10"><a href="#cb236-10" tabindex="-1"></a>mean.women <span class="ot">&lt;-</span> <span class="fu">mean</span>(dat.female<span class="sc">$</span>weight, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p>And we can then add in the means to the plot as vertical lines.</p>
<div class="sourceCode" id="cb237"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb237-1"><a href="#cb237-1" tabindex="-1"></a><span class="fu">ggplot</span>(dat,</span>
<span id="cb237-2"><a href="#cb237-2" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> weight, </span>
<span id="cb237-3"><a href="#cb237-3" tabindex="-1"></a>           <span class="at">fill =</span> gender)) <span class="sc">+</span></span>
<span id="cb237-4"><a href="#cb237-4" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb237-5"><a href="#cb237-5" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> mean.men, <span class="at">colour =</span> <span class="st">&quot;cyan&quot;</span>) <span class="sc">+</span></span>
<span id="cb237-6"><a href="#cb237-6" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> mean.women, <span class="at">colour =</span> <span class="st">&quot;salmon&quot;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-120-1.png" width="672" /></p>
<blockquote>
<p>Do you think that the mean of the male weight is lower than 180?
Do you think that the mean of the female weight is <strong>really</strong> different from the mean of the male weights?
Do you think that the mean of the male weight is higher than that of the females?</p>
</blockquote>
<hr />
<div id="tests-of-mean" class="section level3 hasAnchor">
<h3 class="hasAnchor">Tests of mean<a href="#tests-of-mean" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Now, we want to use a statistical test to check if,</p>
<ul>
<li>males have a mean weight that is <strong>significantly</strong> lower than a specific value (one-sample, one-tailed test)</li>
<li>there is a <strong>significant</strong> difference in the mean of the weights between the two groups (two-sample, two-sided test)</li>
<li>male have a <strong>significantly</strong> higher mean weight than females (two-sample, one-sided test)</li>
</ul>
<p>Note the use of the word <strong>significant</strong> in the previous statements!</p>
<p>This is exactly what <strong>mean tests</strong> such as the t-test (or the Wilcoxon test) are designed for!
We will perform here a <strong>t-test</strong>.</p>
<hr />
<div id="one-sample-t-test" class="section level4 hasAnchor">
<h4 class="hasAnchor">One sample t-test<a href="#one-sample-t-test" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Using a one-sample t-test, we can check if values differ significantly from a target value.
For example, you could sample 10 chocolate bars, and test if they significantly differ from the expected weight of 100 g:</p>
<blockquote>
<p>Should we perform a one- or two-sided test?</p>
</blockquote>
<div class="sourceCode" id="cb238"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb238-1"><a href="#cb238-1" tabindex="-1"></a>bars <span class="ot">=</span> <span class="fu">c</span>(<span class="dv">103</span>,<span class="dv">103</span>,<span class="dv">97</span>,<span class="fl">102.5</span>,<span class="fl">100.5</span>,<span class="dv">103</span>,<span class="fl">101.3</span>,<span class="fl">99.5</span>,<span class="dv">101</span>,<span class="dv">104</span>) <span class="co"># weights of 10 chocolate bars</span></span>
<span id="cb238-2"><a href="#cb238-2" tabindex="-1"></a>chocbar.mean <span class="ot">=</span> <span class="dv">100</span> <span class="co"># expected weight</span></span></code></pre></div>
<p>The function <code>t.test()</code> offers three <strong>alternative</strong> options: <em>two.sided</em>, <em>less</em> and <em>greater</em>.
Here, if we want to test whether the mean weight of the 10 chocolate bars is <strong>different</strong> from the expected weight of 100 g, we want to perform a <em>two sided</em> test and use the alternative <em>two.sided</em>.</p>
<p>It is essential to <strong>clearly formulate the H0 and H1 hypothesis</strong>.
There are two alternative but equivalent ways to do so.
Either:</p>
<ul>
<li>H0: the expectation value of the random variable “Weight of a chocolate bar” is <strong>equal</strong> to 100 g.<br />
</li>
<li>H1: the expectation value of the random variable “Weight of a chocolate bar” is <strong>different</strong> from 100 g.</li>
</ul>
<p>or</p>
<ul>
<li>H0: the mean weight of a chocolate bar is not significantly different from 100 g.<br />
</li>
<li>H1: the mean weight of a chocolate bar is significantly different from 100 g.</li>
</ul>
<p>Note the difference between these two formulations, and ask for help if you have questions about this!</p>
<div class="sourceCode" id="cb239"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb239-1"><a href="#cb239-1" tabindex="-1"></a><span class="fu">t.test</span>(<span class="at">x =</span> bars, <span class="at">mu =</span> chocbar.mean, <span class="at">alternative =</span> <span class="st">&quot;two.sided&quot;</span>)</span></code></pre></div>
<pre><code>## 
##  One Sample t-test
## 
## data:  bars
## t = 2.233, df = 9, p-value = 0.05244
## alternative hypothesis: true mean is not equal to 100
## 95 percent confidence interval:
##   99.98067 102.97933
## sample estimates:
## mean of x 
##    101.48</code></pre>
<blockquote>
<p>How would you interpret this result?
Can you reject the H0 hypothesis?</p>
</blockquote>
<p>Using <span class="math inline">\(\alpha=0.05\)</span>, the H0 hypothesis can not be rejected as the p-value is 0.05244 (p-value &gt;= 0.05).
With <span class="math inline">\(\alpha=0.05\)</span>, the mean weight of the chocolate bars is not significantly different from 100 g.</p>
<p>Using <span class="math inline">\(\alpha=0.1\)</span>, the H0 hypothesis can be rejected (p-value &lt; 0.1).
With <span class="math inline">\(\alpha=0.1\)</span>, the mean weight of the chocolate bars is significantly different from 100 g.</p>
<p><strong>BUT</strong> … It does not mean that <span class="math inline">\(\alpha\)</span> should be chosen with respect to the results of the t.test!!!</p>
<p>Before running a t.test, <strong>formulate the null hypothesis H0 and the alternative hypothesis H1</strong> and <strong>decide about the alpha</strong> value.
Remember, <strong>alpha</strong> represent the <strong>false positive rate</strong>: under the H0 hypothesis (test of two identical distributions), this is the proportion of tests that will detect a difference between the two groups (p-value &lt; <span class="math inline">\(\alpha\)</span>).</p>
<p><strong>Beware not to get confused between one-/two-sample tests, and one-/two-sided tests!</strong></p>
<p>Regarding the mean weight of the males, we would like to check whether males have a mean weight that is significantly <strong>lower</strong> than 180.
Here as well, we will perform a <strong>one-sample test</strong>, with mu = 180.
However, we will perform a <strong>one-sided</strong> test using the alternative option <strong>less</strong>.</p>
<p>The hypotheses can be formulated as:</p>
<ul>
<li><p>H0: the expectation value of the random variable “Weight of male patients” is <strong>equal or greater</strong> 180.<br />
</p></li>
<li><p>H1: the expectation value of the random variable “Weight of male patients” is <strong>less</strong> than 180.</p></li>
</ul>
<div class="sourceCode" id="cb241"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb241-1"><a href="#cb241-1" tabindex="-1"></a><span class="fu">t.test</span>(dat.male<span class="sc">$</span>weight, <span class="at">mu =</span> <span class="dv">180</span>, <span class="at">alternative =</span> <span class="st">&quot;less&quot;</span>)</span></code></pre></div>
<pre><code>## 
##  One Sample t-test
## 
## data:  dat.male$weight
## t = 0.64094, df = 167, p-value = 0.7388
## alternative hypothesis: true mean is less than 180
## 95 percent confidence interval:
##      -Inf 186.8629
## sample estimates:
## mean of x 
##  181.9167</code></pre>
<blockquote>
<p>How would you interpret the result with <span class="math inline">\(\alpha=0.05\)</span>?
Can you reject the H0 hypothesis?</p>
</blockquote>
<hr />
</div>
<div id="two-sample-t-test-2-sided" class="section level4 hasAnchor">
<h4 class="hasAnchor">Two-sample t-test (2-sided)<a href="#two-sample-t-test-2-sided" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Now, we will compare the mean of the weights between males and females.<br />
</p>
<p>According to the previous histogram, females have a different mean weight as males.
This can be tested using a <em>two-sample and two.tailed</em> t.test.</p>
<div class="sourceCode" id="cb243"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb243-1"><a href="#cb243-1" tabindex="-1"></a><span class="fu">ggplot</span>(dat,</span>
<span id="cb243-2"><a href="#cb243-2" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> weight, </span>
<span id="cb243-3"><a href="#cb243-3" tabindex="-1"></a>           <span class="at">fill =</span> gender)) <span class="sc">+</span></span>
<span id="cb243-4"><a href="#cb243-4" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb243-5"><a href="#cb243-5" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> mean.men, <span class="at">colour =</span> <span class="st">&quot;cyan&quot;</span>) <span class="sc">+</span></span>
<span id="cb243-6"><a href="#cb243-6" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> mean.women, <span class="at">colour =</span> <span class="st">&quot;salmon&quot;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-124-1.png" width="672" /></p>
<p>The hypotheses can be formulated as:</p>
<ul>
<li><p>H0: the expectation value of the random variable “Weight of male patients” is <strong>equal</strong> to the expectation value of the random variable “Weight of female patients”.<br />
</p></li>
<li><p>H1: the expectation value of the random variable “Weight of male patients” is <strong>different</strong> from the expectation value of the random variable “Weight of female patients”.</p></li>
</ul>
<div class="sourceCode" id="cb244"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb244-1"><a href="#cb244-1" tabindex="-1"></a><span class="fu">t.test</span>(dat.male<span class="sc">$</span>weight, </span>
<span id="cb244-2"><a href="#cb244-2" tabindex="-1"></a>       dat.female<span class="sc">$</span>weight) <span class="co"># remember, alternative = &quot;two.sided&quot; per default.</span></span></code></pre></div>
<pre><code>## 
##  Welch Two Sample t-test
## 
## data:  dat.male$weight and dat.female$weight
## t = 1.8453, df = 372.45, p-value = 0.06579
## alternative hypothesis: true difference in means is not equal to 0
## 95 percent confidence interval:
##  -0.4875828 15.3465572
## sample estimates:
## mean of x mean of y 
##  181.9167  174.4872</code></pre>
<blockquote>
<p>How would you interpret the result with <span class="math inline">\(\alpha=0.05\)</span>?
Can you reject the H0 hypothesis?</p>
</blockquote>
<hr />
</div>
<div id="two-sample-t-test-1-sided" class="section level4 hasAnchor">
<h4 class="hasAnchor">Two-sample t-test (1-sided)<a href="#two-sample-t-test-1-sided" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Looking at the histogram, other observers could in principle see a difference between the mean values of the weights and formulate the following hypotheses:</p>
<ul>
<li><p>H0: the expectation value of the random variable “Weight of male patients” is <strong>equal or lower</strong> to the expectation value of the random variable “Weight of female patients”.<br />
</p></li>
<li><p>H1: the expectation value of the random variable “Weight of male patients” is <strong>higher</strong> than the expectation value of the random variable “Weight of female patients”.</p></li>
</ul>
<div class="sourceCode" id="cb246"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb246-1"><a href="#cb246-1" tabindex="-1"></a><span class="fu">t.test</span>(dat.male<span class="sc">$</span>weight, </span>
<span id="cb246-2"><a href="#cb246-2" tabindex="-1"></a>       dat.female<span class="sc">$</span>weight, <span class="at">alternative =</span> <span class="st">&quot;greater&quot;</span>)</span></code></pre></div>
<pre><code>## 
##  Welch Two Sample t-test
## 
## data:  dat.male$weight and dat.female$weight
## t = 1.8453, df = 372.45, p-value = 0.0329
## alternative hypothesis: true difference in means is greater than 0
## 95 percent confidence interval:
##  0.7903498       Inf
## sample estimates:
## mean of x mean of y 
##  181.9167  174.4872</code></pre>
<blockquote>
<p>How would you interpret the result with <span class="math inline">\(\alpha=0.05\)</span>?</p>
</blockquote>
<blockquote>
<p>Can you explain why the p-value of the one-tailed t.test is lower than the p-value of the two-tailed t.test?
What is the relation between these two values?</p>
</blockquote>
<div class="sourceCode" id="cb248"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb248-1"><a href="#cb248-1" tabindex="-1"></a><span class="co"># p-value of the two-sided t-test versus p-value of the one-sided t-test: </span></span>
<span id="cb248-2"><a href="#cb248-2" tabindex="-1"></a><span class="fu">t.test</span>(dat.male<span class="sc">$</span>weight, </span>
<span id="cb248-3"><a href="#cb248-3" tabindex="-1"></a>       dat.female<span class="sc">$</span>weight, <span class="at">alternative =</span> <span class="st">&quot;two.sided&quot;</span>)<span class="sc">$</span>p.value <span class="co"># two-tailed</span></span></code></pre></div>
<pre><code>## [1] 0.06579432</code></pre>
<div class="sourceCode" id="cb250"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb250-1"><a href="#cb250-1" tabindex="-1"></a><span class="fu">t.test</span>(dat.male<span class="sc">$</span>weight, </span>
<span id="cb250-2"><a href="#cb250-2" tabindex="-1"></a>       dat.female<span class="sc">$</span>weight, <span class="at">alternative =</span> <span class="st">&quot;greater&quot;</span>)<span class="sc">$</span>p.value <span class="co"># one-tailed</span></span></code></pre></div>
<pre><code>## [1] 0.03289716</code></pre>
<p>This can be visualized using the t-distribution.
Here (see above result of the t.test), <em>t = 1.8453</em> and <em>df = 372.45</em>.</p>
<p>In the <strong>one-tailed t.test</strong>, the p-value is the area under the curve for t &gt; 1.8453 (alternative greater) <strong>OR</strong> for t &lt; -1.8453 (alternative less).</p>
<p><img src="_main_files/figure-html/unnamed-chunk-128-1.png" width="480" /></p>
<p>In the <strong>two-tailed t.test</strong>, the p-value is the area under the curve for t &gt; 1.8453 (alternative greater) <strong>AND</strong> for t &lt; -1.8453 (alternative less).
It is two times the p-value of the one-sided t.test!</p>
<p><img src="_main_files/figure-html/unnamed-chunk-129-1.png" width="480" /></p>
<p>IMPORTANTLY, a t-test can only be performed if the data is <strong>normally distributed</strong>!
If the data is not normally distributed, you will need to use a non-parametric test, like Wilcoxon test.</p>
<hr />
</div>
<div id="wilcoxon-test" class="section level4 hasAnchor">
<h4 class="hasAnchor">Wilcoxon test<a href="#wilcoxon-test" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>What if the data is not normally distributed?
In that case, we are not supposed to use the t-test for testing differences between mean values!
Let us us see an example.</p>
<p>Let us start by loading the data.</p>
<div class="sourceCode" id="cb252"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb252-1"><a href="#cb252-1" tabindex="-1"></a>all.aml <span class="ot">=</span> <span class="fu">read.delim</span>(<span class="st">&#39;http://bioinfo.ipmb.uni-heidelberg.de/crg/datascience3fs/practicals/data/all.aml.cleaned.csv&#39;</span>,</span>
<span id="cb252-2"><a href="#cb252-2" tabindex="-1"></a>                     <span class="at">header=</span><span class="cn">TRUE</span>)</span>
<span id="cb252-3"><a href="#cb252-3" tabindex="-1"></a>all.aml.anno <span class="ot">=</span> <span class="fu">read.delim</span>(<span class="st">&quot;http://bioinfo.ipmb.uni-heidelberg.de/crg/datascience3fs/practicals/data/all.aml.anno.cleaned.csv&quot;</span>,</span>
<span id="cb252-4"><a href="#cb252-4" tabindex="-1"></a>                          <span class="at">header=</span><span class="cn">TRUE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb252-5"><a href="#cb252-5" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">id =</span> <span class="fu">paste0</span>(<span class="st">&quot;pat&quot;</span>, Samples))</span></code></pre></div>
<p>We check whether the distribution of the expression values for the gene <em>SOX4</em> corresponds to a normal distribution:</p>
<div class="sourceCode" id="cb253"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb253-1"><a href="#cb253-1" tabindex="-1"></a>expression <span class="ot">=</span> all.aml <span class="sc">%&gt;%</span></span>
<span id="cb253-2"><a href="#cb253-2" tabindex="-1"></a>  <span class="fu">t</span>() <span class="sc">%&gt;%</span></span>
<span id="cb253-3"><a href="#cb253-3" tabindex="-1"></a>  <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb253-4"><a href="#cb253-4" tabindex="-1"></a>  <span class="fu">na.omit</span>()</span>
<span id="cb253-5"><a href="#cb253-5" tabindex="-1"></a></span>
<span id="cb253-6"><a href="#cb253-6" tabindex="-1"></a><span class="co"># Plot histogram </span></span>
<span id="cb253-7"><a href="#cb253-7" tabindex="-1"></a><span class="fu">ggplot</span>(expression,</span>
<span id="cb253-8"><a href="#cb253-8" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> SOX4)) <span class="sc">+</span></span>
<span id="cb253-9"><a href="#cb253-9" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins =</span> <span class="dv">20</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-131-1.png" width="528" /></p>
<div class="sourceCode" id="cb254"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb254-1"><a href="#cb254-1" tabindex="-1"></a><span class="co"># Check if it&#39;s normally distributed using a QQ-plot </span></span>
<span id="cb254-2"><a href="#cb254-2" tabindex="-1"></a><span class="fu">ggplot</span>(expression,</span>
<span id="cb254-3"><a href="#cb254-3" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">sample =</span> SOX4)) <span class="sc">+</span></span>
<span id="cb254-4"><a href="#cb254-4" tabindex="-1"></a>  <span class="fu">geom_qq</span>() <span class="sc">+</span></span>
<span id="cb254-5"><a href="#cb254-5" tabindex="-1"></a>  <span class="fu">geom_qq_line</span>(<span class="at">colour =</span> <span class="st">&#39;red&#39;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-131-2.png" width="528" /></p>
<p>This looks everything but normal!
In that case, we cannot apply the t.test, but need to apply a <strong>non-parametric test</strong> called the <em>Wilcoxon test</em>.
This test is performed not on the <em>values</em> (like the t-test) but on the <em>ranks</em> of these values (remember the difference between the Pearson’s and the Spearman’s correlations!)</p>
<div class="sourceCode" id="cb255"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb255-1"><a href="#cb255-1" tabindex="-1"></a><span class="co"># divide the gene expression data in two groups according to ALL or AML patients:</span></span>
<span id="cb255-2"><a href="#cb255-2" tabindex="-1"></a><span class="co"># obtain the AML and rest</span></span>
<span id="cb255-3"><a href="#cb255-3" tabindex="-1"></a>aml.patient.id <span class="ot">=</span> all.aml.anno <span class="sc">%&gt;%</span></span>
<span id="cb255-4"><a href="#cb255-4" tabindex="-1"></a>  <span class="fu">filter</span>(ALL.AML <span class="sc">==</span> <span class="st">&quot;AML&quot;</span>)</span>
<span id="cb255-5"><a href="#cb255-5" tabindex="-1"></a>other.id <span class="ot">=</span> all.aml.anno <span class="sc">%&gt;%</span></span>
<span id="cb255-6"><a href="#cb255-6" tabindex="-1"></a>  <span class="fu">filter</span>(ALL.AML <span class="sc">!=</span> <span class="st">&quot;AML&quot;</span>) <span class="co"># filters for those which are not labeled AML</span></span>
<span id="cb255-7"><a href="#cb255-7" tabindex="-1"></a></span>
<span id="cb255-8"><a href="#cb255-8" tabindex="-1"></a>gene.all <span class="ot">=</span> expression <span class="sc">%&gt;%</span></span>
<span id="cb255-9"><a href="#cb255-9" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>(<span class="st">&quot;id&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb255-10"><a href="#cb255-10" tabindex="-1"></a>  <span class="fu">filter</span>(id <span class="sc">%in%</span> other.id<span class="sc">$</span>id)</span>
<span id="cb255-11"><a href="#cb255-11" tabindex="-1"></a></span>
<span id="cb255-12"><a href="#cb255-12" tabindex="-1"></a>gene.aml <span class="ot">=</span> expression <span class="sc">%&gt;%</span></span>
<span id="cb255-13"><a href="#cb255-13" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>(<span class="st">&quot;id&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb255-14"><a href="#cb255-14" tabindex="-1"></a>  <span class="fu">filter</span>(id <span class="sc">%in%</span> aml.patient.id<span class="sc">$</span>id)</span>
<span id="cb255-15"><a href="#cb255-15" tabindex="-1"></a>  </span>
<span id="cb255-16"><a href="#cb255-16" tabindex="-1"></a><span class="co"># test for a difference in the mean expression values using the Wilcoxon test:</span></span>
<span id="cb255-17"><a href="#cb255-17" tabindex="-1"></a><span class="fu">wilcox.test</span>(gene.aml<span class="sc">$</span>SOX4, gene.all<span class="sc">$</span>SOX4)</span></code></pre></div>
<pre><code>## 
##  Wilcoxon rank sum test with continuity correction
## 
## data:  gene.aml$SOX4 and gene.all$SOX4
## W = 260.5, p-value = 0.0001125
## alternative hypothesis: true location shift is not equal to 0</code></pre>
<p>Compare the obtained p-value with the p-value obtained if we would have used the t-test:</p>
<div class="sourceCode" id="cb257"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb257-1"><a href="#cb257-1" tabindex="-1"></a><span class="fu">t.test</span>(gene.aml<span class="sc">$</span>SOX4, gene.all<span class="sc">$</span>SOX4)</span></code></pre></div>
<pre><code>## 
##  Welch Two Sample t-test
## 
## data:  gene.aml$SOX4 and gene.all$SOX4
## t = -3.6216, df = 54.611, p-value = 0.000642
## alternative hypothesis: true difference in means is not equal to 0
## 95 percent confidence interval:
##  -2605.3197  -748.9356
## sample estimates:
## mean of x mean of y 
##  1653.000  3330.128</code></pre>
<p>The p-values are very different!!
So is the difference of expression between ALL and AML patients for this gene significant or not taking <span class="math inline">\(\alpha=0.05\)</span>?</p>
<p>Here, we <strong>cannot</strong> trust the t-test due to the non-normality of the data!
Hence, the correct p-value is the one from the Wilcoxon test.</p>
<hr />
</div>
</div>
<div id="proportion-tests" class="section level3 hasAnchor">
<h3 class="hasAnchor">Proportion tests<a href="#proportion-tests" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The t-test and Wilcoxon tests are <strong>tests of the mean</strong>, meaning that we are <strong>comparing the means of two samples and looking for significant differences</strong>.</p>
<p>But there are other hypothesis that one might want to test, related to the relationship between two <strong>categorical</strong> variables:</p>
<ul>
<li>is the proportion of men <strong>significantly</strong> higher in the patients from Louisa compared to the ones from Buckingham?</li>
<li>is the proportion of smokers under 18 in Germany <strong>significantly</strong> higher than in other European countries?</li>
</ul>
<p>The proportion test (<strong>Fisher Exact Test</strong> or <strong>chi-squared test</strong>) are used to investigate the relationship between 2 categorical variables, starting from a <strong>contingency table</strong>.
We will use a dataset with clinical information about breast cancer patients.</p>
<div class="sourceCode" id="cb259"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb259-1"><a href="#cb259-1" tabindex="-1"></a>dat.brca <span class="ot">=</span> <span class="fu">read.delim</span>(<span class="st">&#39;http://bioinfo.ipmb.uni-heidelberg.de/crg/datascience3fs/practicals/data/gbsg_ba_ca.dat&#39;</span>, </span>
<span id="cb259-2"><a href="#cb259-2" tabindex="-1"></a>                      <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<blockquote>
<p>Check which variables in this dataset are categorical/ordinal/numerical.</p>
</blockquote>
<p>We can now check if there is a significant relationship between some variables.
For example, we can verify if the choice of treatment with tamoxifen (variable <code>hormon</code>) is related to the pre-/post-menopausal status (variable <code>meno</code>)</p>
<p>First, we can build the contingency table for these 2 variables:</p>
<div class="sourceCode" id="cb260"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb260-1"><a href="#cb260-1" tabindex="-1"></a><span class="do">## build contingency table</span></span>
<span id="cb260-2"><a href="#cb260-2" tabindex="-1"></a><span class="fu">table</span>(dat.brca<span class="sc">$</span>meno, dat.brca<span class="sc">$</span>hormon)</span></code></pre></div>
<pre><code>##                 
##                  had tamoxifen no tamoxifen
##   Postmenopausal           187          209
##   premenopausal             59          231</code></pre>
<p>We can compute the <strong>odds-ratio (OR)</strong> for these two variables.
This metric measures the strength of the association between two events, which in this case would be treatment with tamoxifen and menopausal status.</p>
<p>Here’s how to interpret the odds ratio (OR) value:</p>
<ul>
<li><strong>OR = 1</strong>: events are independent<br />
</li>
<li><strong>OR &gt; 1</strong>: events are positively correlated<br />
</li>
<li><strong>OR &lt; 1</strong>: events are negatively correlated</li>
</ul>
<div class="sourceCode" id="cb262"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb262-1"><a href="#cb262-1" tabindex="-1"></a>CT <span class="ot">=</span> <span class="fu">table</span>(dat.brca<span class="sc">$</span>meno,dat.brca<span class="sc">$</span>hormon)</span>
<span id="cb262-2"><a href="#cb262-2" tabindex="-1"></a></span>
<span id="cb262-3"><a href="#cb262-3" tabindex="-1"></a>OR <span class="ot">=</span> (CT[<span class="dv">1</span>,<span class="dv">1</span>]<span class="sc">/</span>CT[<span class="dv">1</span>,<span class="dv">2</span>])<span class="sc">/</span>(CT[<span class="dv">2</span>,<span class="dv">1</span>]<span class="sc">/</span>CT[<span class="dv">2</span>,<span class="dv">2</span>])</span>
<span id="cb262-4"><a href="#cb262-4" tabindex="-1"></a>OR</span></code></pre></div>
<pre><code>## [1] 3.503122</code></pre>
<blockquote>
<p>How would the odds-ratio look like if you would transpose the matrix?</p>
</blockquote>
<p>Now we can run the one-sided <strong>Fisher Exact Test</strong> (FET).
The H0/H1 hypothesis are:</p>
<ul>
<li>H0: the odds-ratio is not significantly larger than one</li>
<li>H1: the odds-ratio is significantly larger than one</li>
</ul>
<div class="sourceCode" id="cb264"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb264-1"><a href="#cb264-1" tabindex="-1"></a><span class="do">## build contingency table</span></span>
<span id="cb264-2"><a href="#cb264-2" tabindex="-1"></a>tab <span class="ot">=</span> <span class="fu">table</span>(dat.brca<span class="sc">$</span>meno,dat.brca<span class="sc">$</span>hormon)</span>
<span id="cb264-3"><a href="#cb264-3" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb264-4"><a href="#cb264-4" tabindex="-1"></a><span class="do">## run the FET</span></span>
<span id="cb264-5"><a href="#cb264-5" tabindex="-1"></a><span class="fu">fisher.test</span>(tab, <span class="at">alternative =</span> <span class="st">&#39;greater&#39;</span>)</span></code></pre></div>
<pre><code>## 
##  Fisher&#39;s Exact Test for Count Data
## 
## data:  tab
## p-value = 1.388e-13
## alternative hypothesis: true odds ratio is greater than 1
## 95 percent confidence interval:
##  2.580638      Inf
## sample estimates:
## odds ratio 
##   3.496639</code></pre>
<p>Check if your computation of the odds-ratio is right!
Compute also the <strong>two-sided</strong> test:</p>
<blockquote>
<p>Formulate the H0/H1 hypothesis!</p>
</blockquote>
<div class="sourceCode" id="cb266"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb266-1"><a href="#cb266-1" tabindex="-1"></a><span class="fu">fisher.test</span>(tab)</span></code></pre></div>
<pre><code>## 
##  Fisher&#39;s Exact Test for Count Data
## 
## data:  tab
## p-value = 2.475e-13
## alternative hypothesis: true odds ratio is not equal to 1
## 95 percent confidence interval:
##  2.444410 5.049548
## sample estimates:
## odds ratio 
##   3.496639</code></pre>
<p>We can also use the <strong>chi-square test</strong> to answer the same question.
The Chi-square test compares the <strong>observed</strong> number of occurrences in the contingency table to the <strong>expected</strong> number of occurrences if there was no relationship between the variables.</p>
<ul>
<li>H0: the observed and expected occurrences are not significantly different</li>
<li>H1: the observed and expected occurrences are significantly different</li>
</ul>
<div class="sourceCode" id="cb268"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb268-1"><a href="#cb268-1" tabindex="-1"></a><span class="fu">chisq.test</span>(tab)</span></code></pre></div>
<pre><code>## 
##  Pearson&#39;s Chi-squared test with Yates&#39; continuity correction
## 
## data:  tab
## X-squared = 51.416, df = 1, p-value = 7.473e-13</code></pre>
<p>Now we want to verify the impact of age on the grade of the tumor.
We categorize the patients in under and over 40 year groups, and perform a chi-squared test:</p>
<div class="sourceCode" id="cb270"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb270-1"><a href="#cb270-1" tabindex="-1"></a><span class="do">## contingency table</span></span>
<span id="cb270-2"><a href="#cb270-2" tabindex="-1"></a>tab <span class="ot">=</span> <span class="fu">table</span>(dat.brca<span class="sc">$</span>age<span class="sc">&gt;</span><span class="dv">40</span>,</span>
<span id="cb270-3"><a href="#cb270-3" tabindex="-1"></a>            dat.brca<span class="sc">$</span>grade)</span>
<span id="cb270-4"><a href="#cb270-4" tabindex="-1"></a>tab <span class="co"># tumor grades 1,2,3; age under 40 (FALSE) and over 40 (TRUE)</span></span></code></pre></div>
<pre><code>##        
##           1   2   3
##   FALSE   6  41  26
##   TRUE   75 403 135</code></pre>
<div class="sourceCode" id="cb272"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb272-1"><a href="#cb272-1" tabindex="-1"></a><span class="do">##</span></span>
<span id="cb272-2"><a href="#cb272-2" tabindex="-1"></a><span class="fu">chisq.test</span>(tab)</span></code></pre></div>
<pre><code>## 
##  Pearson&#39;s Chi-squared test
## 
## data:  tab
## X-squared = 6.9515, df = 2, p-value = 0.03094</code></pre>
<p>We can determine the table of expected counts using the <code>apply()</code> function.
This works similarly to the <code>lapply</code>, but it can be used to <em>apply</em> a function by rows (if 1), columns (if 2), or both:</p>
<div class="sourceCode" id="cb274"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb274-1"><a href="#cb274-1" tabindex="-1"></a>tot <span class="ot">=</span> <span class="fu">apply</span>(tab,<span class="dv">2</span>,sum) <span class="co"># this is the total number of occurrences in the 3 categories of &quot;grade&quot;, independently of &quot;age&quot;</span></span>
<span id="cb274-2"><a href="#cb274-2" tabindex="-1"></a>tot</span></code></pre></div>
<pre><code>##   1   2   3 
##  81 444 161</code></pre>
<div class="sourceCode" id="cb276"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb276-1"><a href="#cb276-1" tabindex="-1"></a>age <span class="ot">=</span> <span class="fu">apply</span>(tab,<span class="dv">1</span>,sum) <span class="co"># this is the total number of persons above/below 40, independently of &quot;grade&quot;</span></span>
<span id="cb276-2"><a href="#cb276-2" tabindex="-1"></a>age</span></code></pre></div>
<pre><code>## FALSE  TRUE 
##    73   613</code></pre>
<p>On the other hand, <code>sapply</code> is commonly used for dataframe or other tabular formats:</p>
<div class="sourceCode" id="cb278"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb278-1"><a href="#cb278-1" tabindex="-1"></a>tot.proportions <span class="ot">=</span> tot<span class="sc">/</span><span class="fu">sum</span>(tot) <span class="co"># HO proportions (the proportions of occurrences in the three categories of grade, independently of age)</span></span>
<span id="cb278-2"><a href="#cb278-2" tabindex="-1"></a></span>
<span id="cb278-3"><a href="#cb278-3" tabindex="-1"></a>tab.exp <span class="ot">=</span> <span class="fu">sapply</span>(tot.proportions,<span class="cf">function</span>(x) {x<span class="sc">*</span>age}) <span class="co"># expected counts under H0</span></span>
<span id="cb278-4"><a href="#cb278-4" tabindex="-1"></a>tab.exp</span></code></pre></div>
<pre><code>##               1         2         3
## FALSE  8.619534  47.24781  17.13265
## TRUE  72.380466 396.75219 143.86735</code></pre>
<blockquote>
<p>How would you compute the chi-square test statistic using the <code>tab</code> and <code>tab.exp</code> tables?</p>
</blockquote>
<hr />
</div>
<div id="exercises-3" class="section level3 hasAnchor">
<h3 class="hasAnchor">EXERCISES<a href="#exercises-3" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="exercise-1-one-sided-t-test" class="section level4 hasAnchor">
<h4 class="hasAnchor">Exercise 1: One-sided t-test<a href="#exercise-1-one-sided-t-test" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<ol style="list-style-type: decimal">
<li>Consider the following graph, formulate the hypotheses H0 and H1 and perform a (one-sided) t-test.
Interpret the result using <span class="math inline">\(\alpha=0.05\)</span>.</li>
</ol>
<p><img src="_main_files/figure-html/unnamed-chunk-143-1.png" width="672" /></p>
<ol start="2" style="list-style-type: decimal">
<li>Calculate the mean age of the men.</li>
<li>Compare it to age = 50. Formulate the hypotheses H0 and H1 and perform a (one-sided) t-test. Interpret the result using <span class="math inline">\(\alpha=0.05\)</span>.</li>
</ol>
</div>
<div id="exercice-2-two-sided-t-test" class="section level4 hasAnchor">
<h4 class="hasAnchor">Exercice 2: Two-sided t-test<a href="#exercice-2-two-sided-t-test" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Can you find interesting differences in the mean values of parameters of the dataset <strong>dat</strong> for the two groups defined by the <strong>location</strong>?
Follow these steps:</p>
<ol style="list-style-type: decimal">
<li><p>Select the two groups according to the location.
To do so, check the result of <code>distinct(location)</code> and create two tibbles corresponding to each of possible the locations.</p></li>
<li><p>Calculate the mean values of the numerical parameters for each group (ex: age, height, weight, waist, hip, …).
<em>Hint</em>: create first a dataframe with numeric columns only (use <code>select(which(is.numerical))</code>).
Select the rows corresponding to the two groups and use an <code>summarise()</code> and <code>group_by</code> loop to calculate the mean values (grouped by location).</p></li>
<li><p>Select one of these, formulate the H0 and H1 hypotheses and perform a (two-sided) t-test.
Interpret the result (<span class="math inline">\(\alpha=0.05\)</span>).</p></li>
</ol>
</div>
<div id="exercise-3-mean-and-proportion-testing" class="section level4 hasAnchor">
<h4 class="hasAnchor">Exercise 3: Mean and proportion testing<a href="#exercise-3-mean-and-proportion-testing" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>What test would you use for the following questions?</p>
<ul>
<li>A lotion company has to figure out whether their last product is more likely to give acne to men rather than women.</li>
<li>The department of education wants to find out whether social science students have higher grades than science students.</li>
<li>A biologist needs to find out whether a specific gene is more likely to be silenced in lactose intolerant people.</li>
</ul>
</div>
<div id="going-further-checking-the-normality-of-the-distribution" class="section level4 hasAnchor">
<h4 class="hasAnchor">Going further: Checking the normality of the distribution<a href="#going-further-checking-the-normality-of-the-distribution" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>In principle, t-tests require that the data is approximately normally distributed.
If not, we can use <strong>non-parametric</strong> tests (see next lecture).</p>
<p>In order to check whether the data is normally distributed or not, it is possible to perform a <strong>Shapiro-Wilk</strong> normality test (see lecture).
This statistical test is implemented in R in the function <code>shapiro.test()</code>. Try it out with any of the datasets we used before.</p>
<p>A p-value inferior to the chosen <span class="math inline">\(\alpha\)</span> level here (we will use 0.05), means that the null hypothesis can be rejected. Therefore, the data distribution tested is not normal and you need to use a non-parametric test.</p>
</div>
</div>
</div>
<div id="day-5-multiple-testing-and-regression" class="section level2 hasAnchor">
<h2 class="hasAnchor">Day 5: Multiple testing and regression<a href="#day-5-multiple-testing-and-regression" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="multiple-testing" class="section level3 hasAnchor">
<h3 class="hasAnchor">Multiple testing<a href="#multiple-testing" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>In the previous examples we were only testing for one single hypothesis. However, it is common in biology that we have more than one hypothesis being tested.<br />
The table below is a reminder of which errors can be made when testing hypotheses:</p>
<table>
<thead>
<tr class="header">
<th align="left"></th>
<th align="left">H0 is true</th>
<th align="left">H0 is false</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">Decision: Reject H0</td>
<td align="left">Type I Error</td>
<td align="left">Correct Decision</td>
</tr>
<tr class="even">
<td align="left">Decision: Do not reject H0</td>
<td align="left">Correct Decision</td>
<td align="left">Type II Error</td>
</tr>
</tbody>
</table>
<p>Decision is associated with <span class="math inline">\(\alpha\)</span>, so if the p-value is under 0.05, we reject the null hypothesis. However, when testing multiple hypothesis, we are prone to erroneously rejecting the 1st, 2nd… or Nth null hypothesis. I.e., if we test 1000 hypotheses simultaneously, we expect to erroneously reject 50 (5%) just by chance!</p>
<p>The <strong>family wise error rate (FWER)</strong> is the probability of making at least one Type I error and we can control this probability using a p-value correction.</p>
<div id="example" class="section level4 hasAnchor">
<h4 class="hasAnchor">Example<a href="#example" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Imagine that you are trying to test the effect of a treatment A on a cell line in the lab. You run a gene expression experiment to compare treatment with control (untreated) for a panel of 20 genes you pre-selected.</p>
<p>The comparison between treatment and control for the panel gives you a set p-values of the 20 genes</p>
<div class="sourceCode" id="cb280"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb280-1"><a href="#cb280-1" tabindex="-1"></a>p.vals <span class="ot">=</span> <span class="fu">c</span>(<span class="fl">0.1</span>, <span class="fl">0.0001</span>, <span class="fl">0.04</span>, <span class="fl">1.2e-20</span>, <span class="fl">0.002</span>, <span class="fl">0.01</span>, <span class="fl">0.5</span>, <span class="dv">1</span>, <span class="fl">0.02</span>, <span class="fl">1.9e-5</span>,</span>
<span id="cb280-2"><a href="#cb280-2" tabindex="-1"></a>           <span class="fl">1.1e-6</span>, <span class="fl">0.03</span>, <span class="fl">0.7</span>, <span class="fl">0.06</span>, <span class="fl">0.01</span>, <span class="fl">0.3</span>, <span class="fl">0.05</span>, <span class="fl">1e-13</span>, <span class="fl">0.032</span>, <span class="fl">0.004</span>)</span></code></pre></div>
<p>But since you ran 20 different tests to generate these, there is a need to correct the p-values. You can do this on the p-values directly using the <code>p.adjust()</code> function. Run <code>?p.adjust()</code> to see which methods can be used.</p>
<p>We will use the false discovery rate (FDR) correction on this example.</p>
<div class="sourceCode" id="cb281"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb281-1"><a href="#cb281-1" tabindex="-1"></a>p.vals_FDR <span class="ot">=</span> <span class="fu">p.adjust</span>(p.vals, <span class="at">method=</span><span class="st">&quot;fdr&quot;</span>)</span>
<span id="cb281-2"><a href="#cb281-2" tabindex="-1"></a>p.vals_FDR</span></code></pre></div>
<pre><code>##  [1] 1.250000e-01 4.000000e-04 6.153846e-02 2.400000e-19 6.666667e-03
##  [6] 2.222222e-02 5.555556e-01 1.000000e+00 4.000000e-02 9.500000e-05
## [11] 7.333333e-06 5.333333e-02 7.368421e-01 8.000000e-02 2.222222e-02
## [16] 3.529412e-01 7.142857e-02 1.000000e-12 5.333333e-02 1.142857e-02</code></pre>
<p>Check how the p-values change!</p>
<p><img src="_main_files/figure-html/unnamed-chunk-147-1.png" width="50%" /><img src="_main_files/figure-html/unnamed-chunk-147-2.png" width="50%" /></p>
<blockquote>
<p>How many genes change from being significant to non-significant? How does this impact your experiment?</p>
</blockquote>
</div>
</div>
<div id="regression" class="section level3 hasAnchor">
<h3 class="hasAnchor">Regression<a href="#regression" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>In this last part, we want to learn how to build a <strong>regression model</strong> in order to make predictions on certain quantitative variables using other quantitative variables. The important steps here are:</p>
<ul>
<li><strong>learning</strong> the model</li>
<li><strong>testing</strong> the model to evaluate its performances, and also check that the assumptions of the linearity are given</li>
<li><strong>predict</strong> values based on new data points</li>
</ul>
<p>We will use again the diabetes data set, and build a simple linear regression models with a single explanatory variable.
We will focus on predicting the <strong>cholesterol level</strong>.</p>
<p>Load the data and perform some basic inspection on the data:</p>
<div class="sourceCode" id="cb283"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb283-1"><a href="#cb283-1" tabindex="-1"></a>tmp <span class="ot">=</span> <span class="fu">read.table</span>(<span class="st">&#39;https://www.dropbox.com/s/zviurze7c85quyw/diabetes_full.csv?dl=1&#39;</span>,<span class="at">header=</span><span class="cn">TRUE</span>,<span class="at">sep=</span><span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>) </span></code></pre></div>
<p>We will limit the dataset to the numerical variables</p>
<div class="sourceCode" id="cb284"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb284-1"><a href="#cb284-1" tabindex="-1"></a>dat <span class="ot">=</span> tmp <span class="sc">%&gt;%</span></span>
<span id="cb284-2"><a href="#cb284-2" tabindex="-1"></a>  <span class="fu">select</span>(</span>
<span id="cb284-3"><a href="#cb284-3" tabindex="-1"></a>    <span class="fu">where</span>(is.numeric)</span>
<span id="cb284-4"><a href="#cb284-4" tabindex="-1"></a>  )</span>
<span id="cb284-5"><a href="#cb284-5" tabindex="-1"></a><span class="fu">head</span>(dat)</span></code></pre></div>
<pre><code>##     id chol stab.glu hdl ratio glyhb age height weight bp.1s bp.1d bp.2s bp.2d
## 1 1000  203       82  56   3.6  4.31  46     62    121   118    59    NA    NA
## 2 1001  165       97  24   6.9  4.44  29     64    218   112    68    NA    NA
## 3 1002  228       92  37   6.2  4.64  58     61    256   190    92   185    92
## 4 1003   78       93  12   6.5  4.63  67     67    119   110    50    NA    NA
## 5 1005  249       90  28   8.9  7.72  64     68    183   138    80    NA    NA
## 6 1008  248       94  69   3.6  4.81  34     71    190   132    86    NA    NA
##   waist hip time.ppn
## 1    29  38      720
## 2    46  48      360
## 3    49  57      180
## 4    33  38      480
## 5    44  41      300
## 6    36  42      195</code></pre>
<p>First, we have to do some data cleaning and remove all patients with at least 1 “NA”. Use <code>na.omit()</code>.</p>
<div class="sourceCode" id="cb286"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb286-1"><a href="#cb286-1" tabindex="-1"></a><span class="co"># Select the patients without NAs</span></span>
<span id="cb286-2"><a href="#cb286-2" tabindex="-1"></a>dat <span class="ot">=</span> dat <span class="sc">%&gt;%</span></span>
<span id="cb286-3"><a href="#cb286-3" tabindex="-1"></a>  <span class="fu">na.omit</span>()</span></code></pre></div>
<p>Then, we can make a heatmap and visualize the correlation values between each of the variables in the dataset.</p>
<div class="sourceCode" id="cb287"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb287-1"><a href="#cb287-1" tabindex="-1"></a><span class="fu">library</span>(pheatmap)</span>
<span id="cb287-2"><a href="#cb287-2" tabindex="-1"></a>cor.vals <span class="ot">=</span> <span class="fu">cor</span>(dat, <span class="at">method =</span> <span class="st">&quot;spearman&quot;</span>)</span>
<span id="cb287-3"><a href="#cb287-3" tabindex="-1"></a></span>
<span id="cb287-4"><a href="#cb287-4" tabindex="-1"></a><span class="fu">pheatmap</span>(cor.vals,</span>
<span id="cb287-5"><a href="#cb287-5" tabindex="-1"></a>         <span class="at">cluster_cols =</span> <span class="cn">FALSE</span>,</span>
<span id="cb287-6"><a href="#cb287-6" tabindex="-1"></a>         <span class="at">cluster_rows =</span> <span class="cn">FALSE</span>,</span>
<span id="cb287-7"><a href="#cb287-7" tabindex="-1"></a>         <span class="at">display_numbers =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-151-1.png" width="672" /></p>
<blockquote>
<p>What are the strongest correlations? Do they make sense?</p>
</blockquote>
<p>Apart from displaying the raw correlation values, we can test if any of those are <strong>statistically significant</strong>, i.e. if they are significantly positive (one-sided test), negative (one-sided test), or non zero (two-sided test). To do so, we will use <code>cor.test</code> function.</p>
<p>For example, there seems to be a positive correlation between <em>stabilized glucose</em> (stab.glu) and <em>hip circumference</em> (hip).</p>
<div class="sourceCode" id="cb288"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb288-1"><a href="#cb288-1" tabindex="-1"></a><span class="do">## compute correlation</span></span>
<span id="cb288-2"><a href="#cb288-2" tabindex="-1"></a><span class="fu">cor</span>(dat<span class="sc">$</span>stab.glu,dat<span class="sc">$</span>hip)</span></code></pre></div>
<pre><code>## [1] 0.03528725</code></pre>
<div class="sourceCode" id="cb290"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb290-1"><a href="#cb290-1" tabindex="-1"></a><span class="do">##</span></span>
<span id="cb290-2"><a href="#cb290-2" tabindex="-1"></a><span class="do">## test for significance</span></span>
<span id="cb290-3"><a href="#cb290-3" tabindex="-1"></a><span class="fu">cor.test</span>(dat<span class="sc">$</span>stab.glu, dat<span class="sc">$</span>hip)</span></code></pre></div>
<pre><code>## 
##  Pearson&#39;s product-moment correlation
## 
## data:  dat$stab.glu and dat$hip
## t = 0.40873, df = 134, p-value = 0.6834
## alternative hypothesis: true correlation is not equal to 0
## 95 percent confidence interval:
##  -0.1338406  0.2024178
## sample estimates:
##        cor 
## 0.03528725</code></pre>
<blockquote>
<p>Read carefully this output, and make sure you understand it. Check other pairs!</p>
</blockquote>
<div id="univariate-linear-regression" class="section level4 hasAnchor">
<h4 class="hasAnchor">Univariate linear regression<a href="#univariate-linear-regression" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Next, we will assess what is the most promising variable to predict cholesterol level. Go back to the correlation heatmap to see if there are variables highly correlated with it.</p>
<p>We will use glycosilated hemoglobin (<code>glyhb</code>) as a predictor of the cholesterol level and the function <code>lm()</code>.</p>
<div class="sourceCode" id="cb292"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb292-1"><a href="#cb292-1" tabindex="-1"></a>l.g <span class="ot">=</span> <span class="fu">lm</span>(chol <span class="sc">~</span> glyhb, <span class="at">data=</span>dat)</span>
<span id="cb292-2"><a href="#cb292-2" tabindex="-1"></a><span class="fu">summary</span>(l.g)</span></code></pre></div>
<pre><code>## 
## Call:
## lm(formula = chol ~ glyhb, data = dat)
## 
## Residuals:
##     Min      1Q  Median      3Q     Max 
## -84.458 -32.011  -7.436  20.584 156.407 
## 
## Coefficients:
##             Estimate Std. Error t value Pr(&gt;|t|)    
## (Intercept)  169.514      9.774  17.343  &lt; 2e-16 ***
## glyhb          8.182      1.517   5.393 3.04e-07 ***
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## Residual standard error: 44.02 on 134 degrees of freedom
## Multiple R-squared:  0.1783, Adjusted R-squared:  0.1722 
## F-statistic: 29.08 on 1 and 134 DF,  p-value: 3.045e-07</code></pre>
<p>For a simple linear regression (with only one explanatory variable), the p-value of the t-test for the slope is identical with the p-value of the F-test for the global model. <em>This will no longer be the case when including several explanatory variables!</em></p>
<p>By the way, have we checked that a linear regression makes sense in this case? Remember that we have to check that:</p>
<ul>
<li>the residuals are <strong>normally distributed</strong></li>
<li>there is <strong>no correlation</strong> between the residuals and the explanatory variable</li>
</ul>
<div class="sourceCode" id="cb294"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb294-1"><a href="#cb294-1" tabindex="-1"></a><span class="co"># normal distribution of residuals?</span></span>
<span id="cb294-2"><a href="#cb294-2" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb294-3"><a href="#cb294-3" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="fu">aes</span>(<span class="at">x =</span> l.g<span class="sc">$</span>residuals)) </span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-154-1.png" width="384" /></p>
<div class="sourceCode" id="cb295"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb295-1"><a href="#cb295-1" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb295-2"><a href="#cb295-2" tabindex="-1"></a>  <span class="fu">geom_qq</span>(<span class="fu">aes</span>(<span class="at">sample =</span> l.g<span class="sc">$</span>residuals)) <span class="sc">+</span></span>
<span id="cb295-3"><a href="#cb295-3" tabindex="-1"></a>  <span class="fu">geom_qq_line</span>(<span class="fu">aes</span>(<span class="at">sample =</span> l.g<span class="sc">$</span>residuals),</span>
<span id="cb295-4"><a href="#cb295-4" tabindex="-1"></a>               <span class="at">colour =</span> <span class="st">&#39;red&#39;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-154-2.png" width="384" /></p>
<div class="sourceCode" id="cb296"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb296-1"><a href="#cb296-1" tabindex="-1"></a><span class="do">## correlation residuals x-values?</span></span>
<span id="cb296-2"><a href="#cb296-2" tabindex="-1"></a>corr.gly_resi <span class="ot">=</span> <span class="fu">cor</span>(dat<span class="sc">$</span>glyhb,l.g<span class="sc">$</span>residuals)</span>
<span id="cb296-3"><a href="#cb296-3" tabindex="-1"></a></span>
<span id="cb296-4"><a href="#cb296-4" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb296-5"><a href="#cb296-5" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> dat<span class="sc">$</span>glyhb, <span class="at">y =</span> l.g<span class="sc">$</span>residuals))</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-154-3.png" width="384" /></p>
<blockquote>
<p>What is your overall opinion about the validity of the regression model here?</p>
</blockquote>
<p>We can now use the model to predict the cholesterol values, and compare them to the real cholesterol values, since we have the information in this dataset.</p>
<div class="sourceCode" id="cb297"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb297-1"><a href="#cb297-1" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb297-2"><a href="#cb297-2" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> dat<span class="sc">$</span>chol, </span>
<span id="cb297-3"><a href="#cb297-3" tabindex="-1"></a>                 <span class="at">y =</span> l.g<span class="sc">$</span>fitted.values),</span>
<span id="cb297-4"><a href="#cb297-4" tabindex="-1"></a>             <span class="at">colour =</span> <span class="st">&quot;navy&quot;</span>) <span class="sc">+</span></span>
<span id="cb297-5"><a href="#cb297-5" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&#39;Real values&#39;</span>,</span>
<span id="cb297-6"><a href="#cb297-6" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">&#39;Predicted values&#39;</span>) <span class="sc">+</span></span>
<span id="cb297-7"><a href="#cb297-7" tabindex="-1"></a>  <span class="fu">geom_abline</span>(<span class="at">intercept =</span> <span class="dv">0</span>, <span class="at">slope =</span> <span class="dv">1</span>, </span>
<span id="cb297-8"><a href="#cb297-8" tabindex="-1"></a>              <span class="at">color =</span> <span class="st">&#39;red&#39;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-155-1.png" width="384" /></p>
<p>Not really super convincing, right? Let’s put more information into the model!</p>
</div>
<div id="multiple-regression-model" class="section level4 hasAnchor">
<h4 class="hasAnchor">Multiple regression model<a href="#multiple-regression-model" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Let us include all the information to try to predict cholesterol level:</p>
<div class="sourceCode" id="cb298"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb298-1"><a href="#cb298-1" tabindex="-1"></a>l.all <span class="ot">=</span> <span class="fu">lm</span>(chol <span class="sc">~</span> .,<span class="at">data=</span>dat)</span>
<span id="cb298-2"><a href="#cb298-2" tabindex="-1"></a><span class="fu">summary</span>(l.all)</span></code></pre></div>
<pre><code>## 
## Call:
## lm(formula = chol ~ ., data = dat)
## 
## Residuals:
##    Min     1Q Median     3Q    Max 
## -77.81 -10.15  -0.88  10.96  55.28 
## 
## Coefficients:
##               Estimate Std. Error t value Pr(&gt;|t|)    
## (Intercept)  2.662e+01  5.724e+01   0.465   0.6428    
## id          -2.722e-05  1.410e-04  -0.193   0.8472    
## stab.glu     6.908e-02  5.848e-02   1.181   0.2398    
## hdl          2.249e+00  1.500e-01  14.988   &lt;2e-16 ***
## ratio        2.520e+01  1.354e+00  18.611   &lt;2e-16 ***
## glyhb        4.451e-01  1.491e+00   0.299   0.7658    
## age          1.566e-01  1.654e-01   0.947   0.3457    
## height      -1.294e+00  6.645e-01  -1.947   0.0539 .  
## weight      -6.610e-02  1.284e-01  -0.515   0.6077    
## bp.1s        9.021e-02  2.053e-01   0.440   0.6611    
## bp.1d       -1.778e-01  2.945e-01  -0.604   0.5472    
## bp.2s       -1.940e-01  2.004e-01  -0.968   0.3350    
## bp.2d        5.934e-01  2.983e-01   1.989   0.0490 *  
## waist        1.013e+00  7.203e-01   1.406   0.1623    
## hip         -5.150e-01  8.398e-01  -0.613   0.5409    
## time.ppn    -9.561e-03  7.239e-03  -1.321   0.1891    
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## Residual standard error: 21.78 on 120 degrees of freedom
## Multiple R-squared:  0.8199, Adjusted R-squared:  0.7974 
## F-statistic: 36.42 on 15 and 120 DF,  p-value: &lt; 2.2e-16</code></pre>
<blockquote>
<p>Do you note something unexpected in this report?</p>
</blockquote>
<p>We can see that the inclusion of several explanatory variables improves the regression. Check the <span class="math inline">\(R^2\)</span> values for example!</p>
<p>We can see that the variables <code>weight</code>, <code>waist</code> and <code>hip</code> do not reach the significance level. Two explanations are possible</p>
<ol style="list-style-type: decimal">
<li>either these variables are indeed non-informative regarding the prediction of the cholesterol level</li>
<li>or the mutual correlation between these 3 variables interferes with the model.</li>
</ol>
<p>We can remove waist and hip for example, and redo the regression</p>
<div class="sourceCode" id="cb300"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb300-1"><a href="#cb300-1" tabindex="-1"></a>l.less <span class="ot">=</span> <span class="fu">lm</span>(chol <span class="sc">~</span> stab.glu <span class="sc">+</span> hdl <span class="sc">+</span> glyhb <span class="sc">+</span> age <span class="sc">+</span> height <span class="sc">+</span> weight  <span class="sc">+</span> bp<span class="fl">.1</span>s <span class="sc">+</span> bp<span class="fl">.1</span>d,<span class="at">data=</span>dat)</span>
<span id="cb300-2"><a href="#cb300-2" tabindex="-1"></a><span class="fu">summary</span>(l.less)</span></code></pre></div>
<pre><code>## 
## Call:
## lm(formula = chol ~ stab.glu + hdl + glyhb + age + height + weight + 
##     bp.1s + bp.1d, data = dat)
## 
## Residuals:
##     Min      1Q  Median      3Q     Max 
## -71.763 -29.004  -7.782  26.894 166.868 
## 
## Coefficients:
##               Estimate Std. Error t value Pr(&gt;|t|)    
## (Intercept) 137.126166  75.534099   1.815 0.071819 .  
## stab.glu     -0.055663   0.109905  -0.506 0.613406    
## hdl           0.607084   0.227293   2.671 0.008555 ** 
## glyhb         9.914922   2.634210   3.764 0.000254 ***
## age           0.242536   0.312447   0.776 0.439046    
## height       -0.632074   0.985494  -0.641 0.522433    
## weight       -0.058539   0.100419  -0.583 0.560962    
## bp.1s         0.007349   0.215115   0.034 0.972801    
## bp.1d         0.387031   0.373868   1.035 0.302539    
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## Residual standard error: 42.44 on 127 degrees of freedom
## Multiple R-squared:  0.2759, Adjusted R-squared:  0.2303 
## F-statistic: 6.049 on 8 and 127 DF,  p-value: 1.39e-06</code></pre>
<p>We can see how the <code>weight</code> variable seems indeed to contribute to the prediction of the cholesterol. The removal of the strongly correlated variables has increased its significance. It now almost reaches significance at the 5% level!</p>
<blockquote>
<p>Check the result of the F-test to compare both models!</p>
</blockquote>
<p>We can now check if the prediction are better that with the univariate model</p>
<div class="sourceCode" id="cb302"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb302-1"><a href="#cb302-1" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb302-2"><a href="#cb302-2" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> dat<span class="sc">$</span>chol, </span>
<span id="cb302-3"><a href="#cb302-3" tabindex="-1"></a>                 <span class="at">y =</span> l.less<span class="sc">$</span>fitted.values),</span>
<span id="cb302-4"><a href="#cb302-4" tabindex="-1"></a>             <span class="at">colour =</span> <span class="st">&quot;navy&quot;</span>) <span class="sc">+</span></span>
<span id="cb302-5"><a href="#cb302-5" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&#39;Real values&#39;</span>,</span>
<span id="cb302-6"><a href="#cb302-6" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">&#39;Predicted values&#39;</span>) <span class="sc">+</span></span>
<span id="cb302-7"><a href="#cb302-7" tabindex="-1"></a>  <span class="fu">geom_abline</span>(<span class="at">intercept =</span> <span class="dv">0</span>, <span class="at">slope =</span> <span class="dv">1</span>, </span>
<span id="cb302-8"><a href="#cb302-8" tabindex="-1"></a>              <span class="at">color =</span> <span class="st">&#39;red&#39;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-158-1.png" width="384" /></p>
<p>Better? I would say so… To determine the accuracy, we can compute the so called <strong>root mean squared error (RMSE)</strong>:</p>
<p><span class="math display">\[
RMSE = \sqrt{\frac{1}{n}\sum_{i=1}^n (x_i-\hat{x_i})^2}
\]</span></p>
<div class="sourceCode" id="cb303"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb303-1"><a href="#cb303-1" tabindex="-1"></a>n <span class="ot">=</span> <span class="fu">nrow</span>(dat)</span>
<span id="cb303-2"><a href="#cb303-2" tabindex="-1"></a>rmse <span class="ot">=</span> <span class="fu">sqrt</span>(<span class="dv">1</span><span class="sc">/</span>n<span class="sc">*</span><span class="fu">sum</span>(l.less<span class="sc">$</span>residuals<span class="sc">^</span><span class="dv">2</span>))</span>
<span id="cb303-3"><a href="#cb303-3" tabindex="-1"></a>rmse</span></code></pre></div>
<pre><code>## [1] 41.01479</code></pre>
<p>Of course, this is cheating, right? We are predicting the values on <strong>exactly</strong> the same data we used to learn the model. In real machine learning, we need to perform <strong>cross-validation</strong>, i.e. learn the model on one part of the data (<em>training set</em>) and validate it on another set (<em>test set</em>).</p>
<p>Let us split the dataset in a test and training set randomly:</p>
<div class="sourceCode" id="cb305"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb305-1"><a href="#cb305-1" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1234</span>)</span>
<span id="cb305-2"><a href="#cb305-2" tabindex="-1"></a><span class="do">## take 200 random patients to form the training set</span></span>
<span id="cb305-3"><a href="#cb305-3" tabindex="-1"></a>i.train <span class="ot">=</span> <span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(dat),<span class="dv">100</span>)</span>
<span id="cb305-4"><a href="#cb305-4" tabindex="-1"></a><span class="do">##</span></span>
<span id="cb305-5"><a href="#cb305-5" tabindex="-1"></a>dat.train <span class="ot">=</span> dat[i.train,]</span>
<span id="cb305-6"><a href="#cb305-6" tabindex="-1"></a>dat.test <span class="ot">=</span> dat[<span class="sc">-</span>i.train,]</span></code></pre></div>
<p>We now learn a new model on the train dataset:</p>
<div class="sourceCode" id="cb306"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb306-1"><a href="#cb306-1" tabindex="-1"></a>l.train <span class="ot">=</span> <span class="fu">lm</span>(chol <span class="sc">~</span> stab.glu <span class="sc">+</span> hdl <span class="sc">+</span> glyhb <span class="sc">+</span> age <span class="sc">+</span> height <span class="sc">+</span> weight  <span class="sc">+</span> bp<span class="fl">.1</span>s <span class="sc">+</span> bp<span class="fl">.1</span>d,<span class="at">data=</span>dat.train)</span>
<span id="cb306-2"><a href="#cb306-2" tabindex="-1"></a><span class="fu">summary</span>(l.train)</span></code></pre></div>
<pre><code>## 
## Call:
## lm(formula = chol ~ stab.glu + hdl + glyhb + age + height + weight + 
##     bp.1s + bp.1d, data = dat.train)
## 
## Residuals:
##     Min      1Q  Median      3Q     Max 
## -76.806 -28.259  -6.705  22.934 158.094 
## 
## Coefficients:
##              Estimate Std. Error t value Pr(&gt;|t|)   
## (Intercept) 214.29240   95.18482   2.251  0.02677 * 
## stab.glu     -0.04772    0.13938  -0.342  0.73288   
## hdl           0.24900    0.31841   0.782  0.43624   
## glyhb        10.14338    3.22686   3.143  0.00225 **
## age           0.41957    0.39691   1.057  0.29327   
## height       -1.11670    1.26775  -0.881  0.38072   
## weight       -0.11006    0.12823  -0.858  0.39298   
## bp.1s        -0.12595    0.28635  -0.440  0.66110   
## bp.1d         0.31732    0.50740   0.625  0.53328   
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## Residual standard error: 46 on 91 degrees of freedom
## Multiple R-squared:  0.2749, Adjusted R-squared:  0.2111 
## F-statistic: 4.312 on 8 and 91 DF,  p-value: 0.0001965</code></pre>
<p>We can compute the RMSE for the training dataset</p>
<div class="sourceCode" id="cb308"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb308-1"><a href="#cb308-1" tabindex="-1"></a>n.train <span class="ot">=</span> <span class="fu">nrow</span>(dat.train)</span>
<span id="cb308-2"><a href="#cb308-2" tabindex="-1"></a>rmse.train <span class="ot">=</span> <span class="fu">sqrt</span>(<span class="dv">1</span><span class="sc">/</span>n.train<span class="sc">*</span><span class="fu">sum</span>(l.train<span class="sc">$</span>residuals<span class="sc">^</span><span class="dv">2</span>))</span>
<span id="cb308-3"><a href="#cb308-3" tabindex="-1"></a>rmse.train</span></code></pre></div>
<pre><code>## [1] 43.88137</code></pre>
<p>Let use use that model to predict cholesterol values for the left out test set</p>
<div class="sourceCode" id="cb310"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb310-1"><a href="#cb310-1" tabindex="-1"></a>pred <span class="ot">=</span> <span class="fu">predict</span>(l.train, <span class="at">newdata =</span> dat.test)</span></code></pre></div>
<p>and compute the rmse:</p>
<div class="sourceCode" id="cb311"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb311-1"><a href="#cb311-1" tabindex="-1"></a>n.test <span class="ot">=</span> <span class="fu">nrow</span>(dat.test)</span>
<span id="cb311-2"><a href="#cb311-2" tabindex="-1"></a>residuals <span class="ot">=</span> dat.test<span class="sc">$</span>chol <span class="sc">-</span> pred</span>
<span id="cb311-3"><a href="#cb311-3" tabindex="-1"></a>rmse.test <span class="ot">=</span> <span class="fu">sqrt</span>(<span class="dv">1</span><span class="sc">/</span>n.test<span class="sc">*</span><span class="fu">sum</span>(residuals<span class="sc">^</span><span class="dv">2</span>))</span>
<span id="cb311-4"><a href="#cb311-4" tabindex="-1"></a>rmse.test</span></code></pre></div>
<pre><code>## [1] 34.75285</code></pre>
<p>Of course, the RMSE is higher on the test dataset, since this does not include the data used for the establishment of the regression model; however, this is a more realistic estimation of the validity of the model, as it indicates how well the model could be extended to novel, independent data!</p>
<p>An important topic here is <strong>feature selection</strong>, i.e. finding the optimal and minimal set of explanatory variables that allow to predict well the output variable.</p>
<p>We now repeat the train/test split 10 times with each time a different random split; plot the 10 <code>rmse.train</code> and <code>rmse.test</code> values!</p>
<div class="sourceCode" id="cb313"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb313-1"><a href="#cb313-1" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">345</span>)</span>
<span id="cb313-2"><a href="#cb313-2" tabindex="-1"></a>RMSE <span class="ot">&lt;-</span> <span class="fu">sapply</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="cf">function</span>(x) {</span>
<span id="cb313-3"><a href="#cb313-3" tabindex="-1"></a>  i.train <span class="ot">=</span> <span class="fu">sample</span> (<span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(dat),<span class="dv">100</span>)</span>
<span id="cb313-4"><a href="#cb313-4" tabindex="-1"></a>  <span class="do">##</span></span>
<span id="cb313-5"><a href="#cb313-5" tabindex="-1"></a>  dat.train <span class="ot">=</span> dat[i.train,]</span>
<span id="cb313-6"><a href="#cb313-6" tabindex="-1"></a>  dat.test <span class="ot">=</span> dat[<span class="sc">-</span>i.train,]</span>
<span id="cb313-7"><a href="#cb313-7" tabindex="-1"></a>  <span class="do">##</span></span>
<span id="cb313-8"><a href="#cb313-8" tabindex="-1"></a>  l.train <span class="ot">=</span> <span class="fu">lm</span>(chol <span class="sc">~</span> stab.glu <span class="sc">+</span> hdl <span class="sc">+</span> glyhb <span class="sc">+</span> age <span class="sc">+</span> height <span class="sc">+</span> weight  <span class="sc">+</span> bp<span class="fl">.1</span>s <span class="sc">+</span> bp<span class="fl">.1</span>d,<span class="at">data=</span>dat.train)</span>
<span id="cb313-9"><a href="#cb313-9" tabindex="-1"></a>  <span class="do">##</span></span>
<span id="cb313-10"><a href="#cb313-10" tabindex="-1"></a>  n.train <span class="ot">=</span> <span class="fu">nrow</span>(dat.train)</span>
<span id="cb313-11"><a href="#cb313-11" tabindex="-1"></a>  rmse.train <span class="ot">=</span> <span class="fu">sqrt</span>(<span class="dv">1</span><span class="sc">/</span>n.train<span class="sc">*</span><span class="fu">sum</span>(l.train<span class="sc">$</span>residuals<span class="sc">^</span><span class="dv">2</span>))</span>
<span id="cb313-12"><a href="#cb313-12" tabindex="-1"></a>  <span class="do">##</span></span>
<span id="cb313-13"><a href="#cb313-13" tabindex="-1"></a>  pred <span class="ot">=</span> <span class="fu">predict</span>(l.train,<span class="at">newdata =</span> dat.test)</span>
<span id="cb313-14"><a href="#cb313-14" tabindex="-1"></a>  <span class="do">##</span></span>
<span id="cb313-15"><a href="#cb313-15" tabindex="-1"></a>  n.test <span class="ot">=</span> <span class="fu">nrow</span>(dat.test)</span>
<span id="cb313-16"><a href="#cb313-16" tabindex="-1"></a>  residuals <span class="ot">=</span> dat.test<span class="sc">$</span>chol <span class="sc">-</span> pred</span>
<span id="cb313-17"><a href="#cb313-17" tabindex="-1"></a>  rmse.test <span class="ot">=</span> <span class="fu">sqrt</span>(<span class="dv">1</span><span class="sc">/</span>n.test<span class="sc">*</span><span class="fu">sum</span>(residuals<span class="sc">^</span><span class="dv">2</span>))</span>
<span id="cb313-18"><a href="#cb313-18" tabindex="-1"></a>  RMSE <span class="ot">&lt;-</span> <span class="fu">c</span>(rmse.train,rmse.test)</span>
<span id="cb313-19"><a href="#cb313-19" tabindex="-1"></a>  RMSE</span>
<span id="cb313-20"><a href="#cb313-20" tabindex="-1"></a>})</span>
<span id="cb313-21"><a href="#cb313-21" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb313-22"><a href="#cb313-22" tabindex="-1"></a></span>
<span id="cb313-23"><a href="#cb313-23" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb313-24"><a href="#cb313-24" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">y =</span> RMSE[<span class="dv">1</span>,],</span>
<span id="cb313-25"><a href="#cb313-25" tabindex="-1"></a>                 <span class="at">x =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="at">colour =</span> <span class="st">&quot;gold&quot;</span>)) <span class="sc">+</span></span>
<span id="cb313-26"><a href="#cb313-26" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">y =</span> RMSE[<span class="dv">2</span>,],</span>
<span id="cb313-27"><a href="#cb313-27" tabindex="-1"></a>                 <span class="at">x =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="at">colour =</span> <span class="st">&quot;steelblue&quot;</span>)) <span class="sc">+</span></span>
<span id="cb313-28"><a href="#cb313-28" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">breaks=</span><span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">4</span>,<span class="dv">6</span>,<span class="dv">8</span>,<span class="dv">10</span>)) <span class="sc">+</span></span>
<span id="cb313-29"><a href="#cb313-29" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="fu">mean</span>(RMSE[<span class="dv">1</span>,]),</span>
<span id="cb313-30"><a href="#cb313-30" tabindex="-1"></a>             <span class="at">colour =</span> <span class="st">&quot;gold&quot;</span>) <span class="sc">+</span></span>
<span id="cb313-31"><a href="#cb313-31" tabindex="-1"></a>  <span class="fu">ylim</span>(<span class="fu">min</span>(RMSE), <span class="fu">max</span>(RMSE <span class="sc">+</span><span class="dv">2</span>)) <span class="sc">+</span></span>
<span id="cb313-32"><a href="#cb313-32" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="fu">mean</span>(RMSE[<span class="dv">2</span>,]),</span>
<span id="cb313-33"><a href="#cb313-33" tabindex="-1"></a>             <span class="at">colour =</span> <span class="st">&quot;steelblue&quot;</span>) <span class="sc">+</span></span>
<span id="cb313-34"><a href="#cb313-34" tabindex="-1"></a>  <span class="fu">scale_colour_manual</span>(<span class="at">name =</span> <span class="st">&quot;RMSE&quot;</span>, <span class="at">guide =</span> <span class="st">&#39;legend&#39;</span>,</span>
<span id="cb313-35"><a href="#cb313-35" tabindex="-1"></a>                      <span class="at">values =</span><span class="fu">c</span>(<span class="st">&#39;gold&#39;</span><span class="ot">=</span><span class="st">&#39;gold&#39;</span>,<span class="st">&#39;steelblue&#39;</span><span class="ot">=</span><span class="st">&#39;steelblue&#39;</span>), </span>
<span id="cb313-36"><a href="#cb313-36" tabindex="-1"></a>                      <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">&#39;train&#39;</span>,<span class="st">&#39;test&#39;</span>)) <span class="sc">+</span></span>
<span id="cb313-37"><a href="#cb313-37" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&quot;Iteration&quot;</span>, <span class="at">y =</span> <span class="st">&quot;RMSE values&quot;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-165-1.png" width="672" /></p>
<!--chapter:end:01-week.Rmd-->
</div>
</div>
</div>
</div>
<div id="week-2-sequence-analysis" class="section level1 unnumbered hasAnchor">
<h1 class="unnumbered hasAnchor">Week 2 : sequence analysis<a href="#week-2-sequence-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<div id="day-1-annotathon-identifying-orfs" class="section level2 hasAnchor">
<h2 class="hasAnchor">Day 1: Annotathon! Identifying ORFs<a href="#day-1-annotathon-identifying-orfs" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The Annotathon project aims to give the students the opportunity to learn how to annotate unidentified sequences while helping the scientific community. In this course, the results you will produce will help progress on the discovery of new species or advance knowledge on known ones.</p>
<p>Each group will be affected an unknown sequence of DNA. Your mission, if you accept it, is to analyse this sequence and to gather the maximum of information you can on it.</p>
<p>Which means that there is no already known answer for the questions you will have. Each of you will follow the same bioinformatic process but the results will be uniq to the sequence.</p>
<div id="getting-ready" class="section level3 hasAnchor">
<h3 class="hasAnchor">Getting ready<a href="#getting-ready" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="creating-an-account" class="section level4 hasAnchor">
<h4 class="hasAnchor">Creating an account<a href="#creating-an-account" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>To create you account on <a href="https://annotathon.org/" class="uri">https://annotathon.org/</a>, clic on the “<em>New account</em>” tab
<img src="figures/Annotathon/Annot_create_account.png" />
Follow the instructions to open a new account; make sure you select the appropriate Team</p>
<p><img src="figures/Annotathon/Annot_select_affiliation.png" />
And the correct “<strong>Team code</strong>”</p>
<p><strong>GKBIOINFO2024</strong></p>
<p>You are required to enter at least one firstname/lastname pair, and one email address in order to receive Annotathon specific notifications. Your email address is secure and will under no circumstance be made public or passed on to any third party. Only low traffic messages specific to your course duration will be mailed to this address; no further messages will be sent after the course is completed.</p>
<p>Finally a clic on “<strong>Open account</strong>” should be followed by the message “<strong>Account ‘XYZ’ has been created</strong>”. Use your ‘username’ and ‘password’ and clic “<strong>Connect</strong>” in the form at top of page to open an Annotathon session. You will be reminded that your email address is not validated until you have followed the special link included in an email automatically sent to you at account creation.</p>
</div>
<div id="reading-the-doc" class="section level4 hasAnchor">
<h4 class="hasAnchor">Reading the doc<a href="#reading-the-doc" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>To guide you for the Annotathon, you have the document you are reading right now that contains the main informations and guidelines. To go further in details, you can access a complete “<strong>Rule Book</strong>” directly on the website.</p>
<p><img src="figures/Annotathon/Annot_access_rulebook.png" /></p>
</div>
</div>
<div id="getting-started" class="section level3 hasAnchor">
<h3 class="hasAnchor">Getting started<a href="#getting-started" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="getting-the-sequence" class="section level4 hasAnchor">
<h4 class="hasAnchor">Getting the sequence<a href="#getting-the-sequence" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>The first step is to get your sequence. To do that, go to “<strong>Cart</strong>”, select the sample and clic “<strong>Add a new sequence to your cart</strong>”.</p>
<p><img src="figures/Annotathon/Annot_add_sequence_to_cart.png" /></p>
<p>You now have your sequence and should be on a page with a lot of informations about your sequence. (This is the new look of your “<strong>Cart</strong>” page)</p>
<p><img src="figures/Annotathon/Annot_your_sequence.png" /></p>
<ul>
<li>In this example, GOS_11742010.1 is the name of the sequence i am analysing.</li>
<li>In the “<strong>Actions</strong>” section, the eye is to view the state of your annotation, the notebook is to change the annotation</li>
<li>the “<strong>version</strong>” is to go back to a previous save of your annotation</li>
<li>the “<strong>Message</strong>” section is to send messages to me in case you have questions</li>
<li>“<strong>Genomic Sequence</strong>” is to get your actual sequence in a FASTA format</li>
<li>…</li>
</ul>
</div>
<div id="and-now-what" class="section level4 hasAnchor">
<h4 class="hasAnchor">And now what ??<a href="#and-now-what" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Well, now it is time to dive into the world of sequence annotation. You will follow a protocol described in the next chapter. For each step you will have to fill a report directly on the website. To start editing the report, clic on the notebook.</p>
<p><img src="figures/Annotathon/Annot_edit_report.png" /></p>
<p>Every step of the analysis will be written following the same pattern :</p>
<p>Protocol :
Describe what tool is used, how it is used …</p>
<p>Results Analysis:
your observations</p>
<p>Raw results:
output of the tool used</p>
</div>
</div>
<div id="finding-orfs" class="section level3 hasAnchor">
<h3 class="hasAnchor">Finding ORFs<a href="#finding-orfs" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Steps to find the Open Reading Frame (ORF)</p>
<div id="copy-your-fasta-sequence" class="section level4 hasAnchor">
<h4 class="hasAnchor">copy your fasta sequence<a href="#copy-your-fasta-sequence" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p><img src="figures/Annotathon/Annot_copy_sequence.png" />
go to : <a href="https://www.ncbi.nlm.nih.gov/orffinder/" class="uri">https://www.ncbi.nlm.nih.gov/orffinder/</a></p>
<p>Paste your sequence and choose the parameters (do not forget to note the selected parameters to your report).</p>
<p>For this study, you will only consider ORFs that verify the following criteria:</p>
<ol style="list-style-type: decimal">
<li>do not contain any <strong>STOP codons</strong> (basic ORF definition…)</li>
<li>contains at least <strong>60 codons</strong></li>
<li>can be on either direct or reverse <strong>strands</strong></li>
<li>can be in <strong>frames 1, 2 or 3</strong> on each strand</li>
<li>can be <strong>complete or incomplete</strong> at the 5’ or 3’ ends, or both!</li>
</ol>
<p><img src="figures/Annotathon/Annot_ORF_input.png" />
Then clic on submit and you will have your first output for your sequence, congratulations !!</p>
<p><img src="figures/Annotathon/Annot_ORF_output.png" />
Now you can gather all of your ORFs sequences into your report (FASTA format) on the Annotathon website (only the ones that are more than 60 aa).</p>
<p>This should be your first entry to the annotathon report page, so it is time to talk about one really important thing here to avoid frustration : <strong>REMEMBER TO SAVE YOUR WORK !!!!!!</strong>
To do that, you clic on the “<strong>save your annotations</strong>” button on the top of the page.</p>
<p>Gather informations about all the ORFs in a table like this one :</p>
<p><img src="figures/Annotathon/Annot_table1_ORF.png" /></p>
<p>then, make a representation of the ORFs on a schematic figure :</p>
<p><img src="figures/Annotathon/Annot_figure1_ORF.png" /></p>
<p>If you do have multiple ORFs, just arbitrary select the biggest one. If you want more details on this part, you can refer to the rule book, but here we will skip the classification of each ORF part.</p>
<p>You now have one ORF that you will annotate deeply, do remember to keep the sequence of the ORF on your report and to <strong>save your work</strong>.</p>
</div>
</div>
<div id="conserved-protein-domains" class="section level3 hasAnchor">
<h3 class="hasAnchor">Conserved protein domains<a href="#conserved-protein-domains" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Your new mission is now to find if your ORF contains any known domains (comparison to databases)
go to <a href="https://www.ebi.ac.uk/jdispatcher/pfa/iprscan5" class="uri">https://www.ebi.ac.uk/jdispatcher/pfa/iprscan5</a></p>
<p><img src="figures/Annotathon/Annot_protein_domain.png" />
Copy paste your sequence, keep all the databases selected, put a name for the run and clic on submit. After 2 - 5 minutes you will obtain your results. (you need to use the protein sequence here)</p>
<p>Then copy the raw output to your report and make a table with the informations as described here:</p>
<p><img src="figures/Annotathon/Annot_table2_prot_domain.png" /></p>
<p>then analyse your results following the next guidelines :</p>
<p><strong>1. Selected domains</strong> (if present)</p>
<pre><code>- Which domains predictions do you select to annotate the ORF? Specifiy their sizes, E-value! Justify your selection!
- Refer clearly to Table 2 for your detailled analysis.</code></pre>
<p><strong>2. Rejected domains</strong> (if present)</p>
<pre><code>- Why are some functional domains rejected? (high E-value?  No IPR domains?)</code></pre>
<p>Here you have to make choices based on the output of interproscan and other ressources you could find on the internet. Do remember to cite any exterior tool and also to <strong>save your work</strong>.</p>
</div>
</div>
<div id="day-2-annotathon-alignments" class="section level2 hasAnchor">
<h2 class="hasAnchor">Day 2: Annotathon! Alignments<a href="#day-2-annotathon-alignments" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="blast-homolog-search" class="section level3 hasAnchor">
<h3 class="hasAnchor">BLAST homolog search<a href="#blast-homolog-search" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>BLAST is a well known tool to identify homologs of sequences. To use it, first go here :
<a href="https://blast.ncbi.nlm.nih.gov/Blast.cgi?PROGRAM=blastp&amp;PAGE_TYPE=BlastSearch&amp;LINK_LOC=blasthome" class="uri">https://blast.ncbi.nlm.nih.gov/Blast.cgi?PROGRAM=blastp&amp;PAGE_TYPE=BlastSearch&amp;LINK_LOC=blasthome</a>
Here, since we already have the protein sequence we will use the BLASTp (protein BLAST).</p>
<p>The BLAST will be done twice, once against the NR database</p>
<p><img src="figures/Annotathon/Annot_BLASTp_NR.png" />
and once against the swissprot database (just change the database in the yellow part). Remember to copy the raw results for both and to note the informations to create this table :</p>
<p><img src="figures/Annotathon/Annot_table3_BLASTp.png" />
Then, using the raw results and this tool : <a href="https://annotathon.org/outils/blast_definition_list.php" class="uri">https://annotathon.org/outils/blast_definition_list.php</a>
create a definition list.</p>
<p><img src="figures/Annotathon/Annot_BLAST_hit_list.png" />
look at your list and merge the similar categories into a table in your results with this format :</p>
<p><img src="figures/Annotathon/Annot_table4_BLAST.png" />
Then, take time to discuss your analysis following those guidelines:</p>
<p><strong>RESULTS ANALYSIS</strong>
Do not continue the analysis of the fragment if:</p>
<ul>
<li>No homologs (or very small number of homologs in NR database, ie &lt;100)</li>
<li>Your gene is already present in the NR biological database (nucleotide BLASTn with <strong>ID &gt; 95%</strong>)</li>
</ul>
<p><em><strong>Proposed structure of your analysis section:</strong></em></p>
<p><strong>1. Overview of the alignments</strong></p>
<ul>
<li>Synthetic description of your alignment results (number, known functions, quality of the alignments, …)</li>
<li>Give details on E-value range, % identity/similarity range, what about indels, alignment coverage, ….)</li>
</ul>
<p><strong>2. Identification of protein homologs</strong></p>
<ul>
<li>Justify E-value thresholds (NR &amp; SP) (Evalue cutoffs, changes in putative homolog functions), refer to tables 3 &amp; 4.</li>
</ul>
<p><strong>3. Function of homologs from SWISSPROT analysis</strong></p>
<ul>
<li>From SP entries, give details on closest homolog functions, specific role of important amino acids involved in catalytic function, cross your analysis with protein domain analysis. In all cases, cite your sources with web links.</li>
</ul>
</div>
<div id="blast-taxonomic-report" class="section level3 hasAnchor">
<h3 class="hasAnchor">BLAST Taxonomic report<a href="#blast-taxonomic-report" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Using the output of your BLAST analysis you will now be able to construct a taxonomic report. To do that, go to this website <a href="https://annotathon.org/outils/blast_tax_report2.php" class="uri">https://annotathon.org/outils/blast_tax_report2.php</a> , and copy your BLAST hits.</p>
<p><img src="figures/Annotathon/Annot_Taxonomy_from_BLAST.png" />
Hit the <strong>Taxonomy Report</strong> button. Copy the output to your report. Then, use this second tool : <a href="https://annotathon.org/outils/blast_taxonomy_list.php" class="uri">https://annotathon.org/outils/blast_taxonomy_list.php</a> to produce a summary table.</p>
<p><img src="figures/Annotathon/Annot_taxonomy_summary_tool.png" />
hit the <strong>Taxa table</strong> button and you will get a summary table like this :</p>
<p><img src="figures/Annotathon/Annot_taxonomy_summary.png" /></p>
<p>you can change the classification level parameter to have a more precise prediction</p>
</div>
<div id="selecting-sequences-for-multiple-alignment" class="section level3 hasAnchor">
<h3 class="hasAnchor">Selecting sequences for multiple alignment<a href="#selecting-sequences-for-multiple-alignment" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Now, using the previous results, you will create two groups of sequences :</p>
<ul>
<li><p>The study group (20 - 30 sequences) is composed with sequences of the same taxonomic group than your ORF and with a good E-value (the lowest possible).</p></li>
<li><p>The external group (10 - 15 sequences) with the closest homologs outside of the taxonomic group of your ORF (Lowest e-value here too)</p></li>
</ul>
<p>keep the amino acide sequences in a FASTA format.</p>
</div>
</div>
<div id="day-3-annotathon-phylogenetic-trees-and-conclusion" class="section level2 hasAnchor">
<h2 class="hasAnchor">Day 3: Annotathon! Phylogenetic trees and conclusion<a href="#day-3-annotathon-phylogenetic-trees-and-conclusion" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="multiple-alignments" class="section level3 hasAnchor">
<h3 class="hasAnchor">Multiple alignments<a href="#multiple-alignments" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Now that you have your list of sequences you can go to this website to do a multiple sequence alignment : <a href="http://www.phylogeny.fr/simple_phylogeny.cgi?workflow_id=b02e40313c3ca8c0e3d4dbc97db10078&amp;tab_index=2" class="uri">http://www.phylogeny.fr/simple_phylogeny.cgi?workflow_id=b02e40313c3ca8c0e3d4dbc97db10078&amp;tab_index=2</a></p>
<p>Copy paste your sequences here :
<img src="figures/Annotathon/Anno_Multiple_sequence_alignment.png" />
and clic submit.</p>
<p>The limitation in the number of sequences to align is simply due to computation time of multiple alignment programs, as well as subsequent phylogenetic tree reconstruction. Computation time is reasonable up to around thirty or fifty sequences of a few hundred residues.</p>
<p>Copy &amp; paste the <strong>“ClustalW” formated</strong> multiple alignment in the ‘<strong>Multiple Alignement</strong>’ Annotathon field.</p>
<p><img src="figures/Annotathon/Annot_Clustal_format.png" />
<strong>RESULTS ANALYSIS:</strong>
<strong>1. Quality of the multiple alignment</strong></p>
<ul>
<li>Can you confirm that that sequences are really homologs? Similar lengths? How many identical positions? How many conservative substitutions positions? Number of indels? can you find that the conservation of sequences within alignment reflects the subgroups (In and out groups)?</li>
<li>After curation with GBLOCKS, what is the number of conserved homolog positions (informative sites) for phylogenetic reconstruction? It is enough?</li>
</ul>
<p><strong>2. Identification of conserved blocks</strong></p>
<ul>
<li>You can annotate well conserved blocks in your alignment with codes (such as A, B, C etc.) and refer to them in your analysis.</li>
<li>Are there any conserved amino acids that are known as actives sites for this protein family? If yes, position in alignment, function, activity?</li>
</ul>
<p><strong>3. N and C-termini of the studied ORF</strong></p>
<ul>
<li>Analysis of the N-ter/C-term of the alignment (complete? start codon? potentially missing number of amino acids in N and C-termini?)</li>
</ul>
</div>
<div id="phylogeny" class="section level3 hasAnchor">
<h3 class="hasAnchor">Phylogeny<a href="#phylogeny" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>To construct the phylogenic tree :
Use the previous multiple alignment to infer a phylogenetic tree using two distinct tree reconstruction approaches:</p>
<ul>
<li>‘distance’ method (e.g. ‘<strong>neighbor-joining (NJ)</strong>’, ‘<strong>BioNJ</strong>’ or ‘<strong>Phylip protdist/neighbor</strong>’)</li>
<li>‘maximun likelyhood’ method (e.g. ‘<strong>PhyML)</strong>’)</li>
</ul>
<p>Go to this place : <a href="http://www.phylogeny.fr/alacarte.cgi" class="uri">http://www.phylogeny.fr/alacarte.cgi</a></p>
<p><img src="figures/Annotathon/Annot_phylogeny_alacarte.png" />
Once you have the two versions for the report, you can play with the different parameters to see how the results are impacted.</p>
<p>Then use this tool : <a href="https://annotathon.org/outils/nw_utils.php" class="uri">https://annotathon.org/outils/nw_utils.php</a>
to create a tree to the format expected for your report. (copy the NEWICK format of your tree and paste it in the box)</p>
<p><img src="figures/Annotathon/Annot_newick_to_annot_tree.png" /></p>
<p>Your result should look like this :
<img src="figures/Annotathon/Annot_newick_example.png" /></p>
<p>Copy &amp; paste the <strong>textual</strong> tree representation in the ‘<strong>Tree</strong>’ Annotathon field. Remember to include a protocol line in the ‘<strong>Tree</strong>’ field that includes the program name and run parameters (ex ‘Phylip / Protdist+neighbor / Randomized input - Random number seed = 11 / rooted on: Coccidioides immitis (ascomycetes)’).</p>
<p>**Suggested plan for the analysis section:</p>
<p><strong>1. Tree topologies</strong></p>
<pre><code>- Describe the topology of each tree. What are the monophyletic groups?
- Do the two independant trees describe the same evolutionary history? the same topology? Similar or different clades? 
- Identify the commonalities as well as the potential incoherencies.</code></pre>
<p><strong>2. Coherence with reference trees</strong></p>
<pre><code>- Are the in- and out-groups correctly separated?
- Are your gene trees coherent with the reference species trees (&quot;tree of life&quot;)? 
- Identify each discrepancy with the reference species tree, and suggest some explanations (HGTs, gene duplications...).</code></pre>
<p><strong>3. Predict the most likely taxonomic origin of the metagenomic ORF</strong></p>
<pre><code>- In which monophyletic clade does the the metagenomic sequence seem to emerge?
- Propose a hypothetical taxonomic classification for the metagenomic ORF!
- Provide detailed justification of your hypothesis, do not under/over interpret the infered phylogenetic trees!</code></pre>
<p>After you have analysed the phylogenetic tree produced, specify the most likely taxonomic group (e.g. “Alphaproteobacteria”) to which belongs the organism carrying your DNA fragment. To specify this group in the ‘<strong>Taxonomy</strong>’ Annotathon field you have two options:</p>
<ul>
<li>specify in the ‘<strong>NCBI numerical identifier</strong>’ box the taxonomic group code (for instance 204455 for <em>Rhodobacterales</em>, these codes are found in GENBANK records in the feature table, such as /db_xref=“taxon:204455” or can be found by querying the NCBI taxonomy database using the link in the Help tab)</li>
<li>specify the exact scientific name for this group in the ‘<strong>Scientific name</strong>’ box (e.g. <em>Rhodobacterales</em>)</li>
</ul>
</div>
<div id="conclusion" class="section level3 hasAnchor">
<h3 class="hasAnchor">Conclusion<a href="#conclusion" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Write up your interpretations and hypotheses based on the observations you have made in the preceeding “RESULTS ANALYSES”. Imagine you are trying to convince a very sceptical colleague: use rigorous argumentation, cite precise evidence and numerical values when ever possible, highlight important findings, cross information from independent sources. Remember that in silico analyses generally do not constitute final proof, only suggestions. Terms such as “putative”, “suggests” or “probably” can show understanding of the limitations of computational biology results.
Make sure you have at least covered:</p>
<ul>
<li>arguments supporting your coding versus non-coding hypothesis; make sure you discuss the start position of your ORF (refer to the FAQ for all the subtlities and pitfalls)!</li>
<li>functionnal predictions for the protein, both at the biochemical level (e.g. “ubiquitin conjugation enzyme”), and the biological role at the organism level (e.g. “implicated in the control of the cell cycle”). Make careful use of annotations available for sequence homologs or conserved domains</li>
<li>your taxonomic classification hypothesis for the organism carrying this DNA fragment.</li>
</ul>
<p>Some common pitfalls to avoid at all cost:</p>
<ul>
<li>explain the theoretical aims and methodology of the tools used (please consider the readers knowledgable in these matters)</li>
<li>describe on which button you clicked (please consider that readers are highly familiar with running BLASTs)</li>
<li>write in telegraphic style</li>
<li>dilute, inflate, digress in the hope that evaluation is proportional to word volume</li>
<li>copy raw results in extenso in the discussion when the reader has direct access to them in the appropriate fields</li>
<li>write linearly without any structure, or purely chronologically</li>
<li>insulate the analyses from each other (you can, and certainly should, make reference to the multiple alignment while discussing the ORF boundaries)</li>
<li>conclude without references to the results and observations</li>
<li>propose hypotheses without providing the supporting evidence</li>
<li>make approximate statements, such as citing BLAST homologs or conserved domains without citing their respective E-values</li>
</ul>
<p>Concentrate on producing a scientific, structured, synthetic and rigorous argumentation that will hold up to peer scrutiny!</p>
</div>
</div>
<div id="day-4-data-formats-and-where-to-find-them" class="section level2 hasAnchor">
<h2 class="hasAnchor">Day 4: Data formats and where to find them<a href="#day-4-data-formats-and-where-to-find-them" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Today we will go through commonly used tools and databases applied in genomic research. You will learn how to:</p>
<ul>
<li>Identify common data formats and know how they are produced;<br />
</li>
<li>Search and obtain genomic data from commonly used databases - GEO;<br />
</li>
<li>Handle different data formats used for transcriptomics and epigenomics;<br />
</li>
<li>Visualize genomic data in IGV.</li>
</ul>
<div id="meet-igv" class="section level3 hasAnchor">
<h3 class="hasAnchor">Meet IGV<a href="#meet-igv" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Alongside the next sections, you will find some examples of files in IGV Viewer and you will use it in the exercises. Access it <a href="https://igv.org/app/">here</a>. We will work only with the Web app, but IGV provides a local software that is quite versatile too.</p>
<div class="float">
<img src="figures/IGVintro.png" alt="Menu overview in IGV Viewer (web app). Along this course, you will load local files and ENCODE tracks into IGV Viewer. This can be done in the Tracks menu (pink arrow). Genome assembly needs to be adjusted according to the local file you are importing (orange arrow). We will mostly work with hg38 and mm10, the most commonly applied assemblies for human and mouse respectively. In addition, we will import a session from a local file too (green arrow)." />
<div class="figcaption">Menu overview in IGV Viewer (web app). Along this course, you will load local files and ENCODE tracks into IGV Viewer. This can be done in the <em>Tracks</em> menu (pink arrow). Genome assembly needs to be adjusted according to the local file you are importing (orange arrow). We will mostly work with hg38 and mm10, the most commonly applied assemblies for human and mouse respectively. In addition, we will import a session from a local file too (green arrow).</div>
</div>
<p>Below this menu, you can select genome locations (in the format <em>chrN:start-end</em>) and search for genes.</p>
<div id="example-1" class="section level4 hasAnchor">
<h4 class="hasAnchor">Example 1<a href="#example-1" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Try IGV out using <a href="IGV_sessions/heart_Histones.json">this session</a>. Download the file, and upload it to IGV using Session &gt; local file.</p>
<p>Do you recognize the differences between the different histone modifications? Blue represents active histone modifications (linked to active transcription). H3K27ac is a classical marker of active gene promoters and other open chromatin regions. H3K36me3 is linked with gene bodies.</p>
<p>Conversely, red represents repressive histone modifications. H3K27me3 and H3K9me3 are both associated with repression/heterochromatin. However, H3K9me3 is meant to be much more permanent, while H3K27me3 is often associated with regions that can be activated in development.</p>
<blockquote>
<p>What genome assembly is being used?</p>
</blockquote>
<blockquote>
<p>Inspect the HOXD gene cluster (Find it in <em>chr2:175,992,177-176,263,537</em>). Based on these signal tracks, do you think this cluster of genes is being expressed in heart?</p>
</blockquote>
<blockquote>
<p>Check different genes (eg. BDP1, NPLOC4, PIK3C3) and observed the shape of the H3K27ac (activation) signal at their promoters.</p>
</blockquote>
<p>Using <code>Tracks &gt; ENCODE Signals - other</code> look for the corresponding RNA-seq file (type RNA-seq heart in the search box, and look for the female patient with the right age…).
Check the expression of the previous genes.</p>
<blockquote>
<p>Do you understand the RNA-seq profile along the genes?</p>
</blockquote>
</div>
</div>
<div id="identifying-formats" class="section level3 hasAnchor">
<h3 class="hasAnchor">Identifying formats<a href="#identifying-formats" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Genomic data will generally fit into one of four classes:</p>
<ul>
<li>Nucleotide sequences (like FASTQ);</li>
<li>Read alignments (like BAM files);</li>
<li>Annotations (as in GTF/GFF files);</li>
<li>Quantitative outputs (like count tables, peaks, among others).</li>
</ul>
<p>Different formats and/or classes will be processed and visualised in different ways because they have distinct features.</p>
<p>Let us start with an example of a <strong>FASTQ file</strong>:</p>
<p>FASTQ files are text-based (as many other sequence classes) and are the <strong>most common format for storage of sequences and quality scores from high-throughput sequencing technologies</strong>. These are large files, so they are generally compressed (extension <code>.fastq.gz</code>). When opening a FASTQ file (using <code>zless</code> or <code>less</code> in the terminal), this what an entry will generally look like:</p>
<pre><code>@A00665:126:HTNC2DRXX:1:2101:1136:1047 1:N:0:AGGCAGAA+NTAAGGAG
GNCCTTACTAGACCAATGGGACTTAAACCCACAAACACTTAGTTAACAGCT
+
F#:FFFFFFFFF:FFFFFFFFFFFFFFFFFFFFF:FFFFFFFFFFFFF:FF</code></pre>
<p>One FASTQ file contains millions of these entries. Each entry (starting with <code>@</code>) is composed of a sequence identifier (in this case, <code>@A00665:126:HTNC2DRXX ...</code>), the sequence itself, a separator (in this case, <code>+</code>), and the quality scores (Phred +33 encoded <a href="https://people.duke.edu/~ccc14/duke-hts-2018/bioinformatics/quality_scores.html">check this page for further explanations</a>) for each base called. Sequence identifiers include information which allows us to know the name of the instrument which produced it, run identifiers, flowcell identifiers, pair number, index sequence, among others.</p>
<p>This example is from paired-end sequencing, meaning that there are two FASTQ files per sample and entries match between them. I.e., <code>@A00665:126:HTNC2DRXX</code> will also exist in FASTQ file 2. When the sequencing is single-end, only one FASTQ file is produced per sample.</p>
<p>Upon alignment to the reference genome, the FASTQ generates a BAM file. BAM files are compressed formats of SAM files (stands for <strong>S</strong>equence <strong>A</strong>lignment/<strong>M</strong>ap format), and they represent read alignments to a genomic location. SAM files are tabular-based and include header (which start with <code>@</code>) and alignment lines.</p>
<div class="float">
<img src="https://www.samformat.info/images/sam_format_annotated_example.5108a0cd.jpg" alt="SAM format and the meaning of elements in the alignment lines. Source: https://www.samformat.info/sam-format-flag" />
<div class="figcaption">SAM format and the meaning of elements in the alignment lines. Source: <a href="https://www.samformat.info/sam-format-flag" class="uri">https://www.samformat.info/sam-format-flag</a></div>
</div>
<p>Alignment lines, corresponding to reads, will include different elements as shown above. One of the most important is the <strong>flag</strong>, which includes information on the alignment of a given read. The flag (in red) is just a number but it represents different combinations of read properties as you can see here <a href="https://broadinstitute.github.io/picard/explain-flags.html">here</a>. Flags can be used to remove unmapped read pairs or failing reads from further downstream processing.</p>
<blockquote>
<p>Check in the link above what would the flag 99, 512, and 13 mean in a SAM file.</p>
</blockquote>
<p>Because these are large files, it is much more common to see read alignments in BAM format. However, BAM files are not human-readable as FASTQ is. You will visualize a BAM file later on in IGV. This is what it looks like:</p>
<div class="float">
<img src="figures/bamexample.png" alt="Example of BAM file for a genomic region within the NPLOC4 gene as present in IGV example session. Grey rectangules represent reads." />
<div class="figcaption">Example of BAM file for a genomic region within the NPLOC4 gene as present in IGV example session. Grey rectangules represent reads.</div>
</div>
<p>While FASTQ and read alignment formats are generally common between different kinds of assays, the output after alignments depends on the assay. Let us explore different formats and to which assay they are associated to.</p>
<div id="where-do-different-formats-come-from" class="section level4 hasAnchor">
<h4 class="hasAnchor">Where do different formats come from?<a href="#where-do-different-formats-come-from" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<div class="float">
<img src="figures/fileformatorigin.png" alt="File outputs after alignment for common experiments." />
<div class="figcaption">File outputs after alignment for common experiments.</div>
</div>
<p>Assays meant to assess chromatin accessibility, transcription factors, or histone modifications are typically based in peaks, which are saved in a BED-like format. A BED file is structured as:</p>
<pre><code>chr1    9922    10469   
chr1    564407  565441  
chr1    565595  567010  
chr1    2583259 2588018 
chr1    2627935 2629661 </code></pre>
<p>In this example, each line is a detected peak at a given location in the genome (indicated as chromosome start end). Additional fields (like p-value, log fold change, length, among others) are often included but depend on the software that generated them.</p>
<p>Chromatin accessibility, transcription factors, or histone modifications can also be represented in BigWig/bedGraph format. This formats, which look like summits, represent <em>signal</em>. Some assays require a control to distinguish this signal from noise (in ChIP-seq, this is usually refered to as Input).</p>
<p>Let us see an example of BigWigs and peaks.</p>
</div>
<div id="example-2" class="section level4 hasAnchor">
<h4 class="hasAnchor">Example 2<a href="#example-2" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Load the session <a href="IGV_sessions/ExampleATAC_ChIP_CTCF_hg38.json">here</a> to see an example of how ChIP-seq and ATAC-seq correlate with each other.</p>
<p>Observe how peaks (rectangles below the coverage tracks) are not always identified for all the “bumps” in the BigWig.</p>
<blockquote>
<p>Search for the ENCODE track for this cell line representing the H3K27ac signal (bigWig format!), and compare to ATAC-seq.</p>
</blockquote>
</div>
</div>
<div id="a-practical-guide-to-geo" class="section level3 hasAnchor">
<h3 class="hasAnchor">A practical guide to GEO<a href="#a-practical-guide-to-geo" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><a href="https://www.ncbi.nlm.nih.gov/gds/">GEO</a>, standing for Gene Expression Omnibus is one of the most used repositories for microarray, next-generation sequencing, and other genomic datasets. When publishing a manuscript, authors must upload the sequencing data they produced into a public database to make it accessible to others, and GEO is often the chosen one. For this reason, GEO includes thousands of original submitter-supplied records that can be accessed from any place by anyone.</p>
<p>We will learn how to search in GEO and how to read a GEO record to obtain specific information.
Datasets in GEO can be found directly through a valid GEO accession number.</p>
<div id="using-filters-and-querying" class="section level4 hasAnchor">
<h4 class="hasAnchor">Using filters and querying<a href="#using-filters-and-querying" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Searching for datasets is not always easy because although some aspects of GEO are common, others are not. Submitters have freedom to adapt descriptions fitting to their specific submission and experiment, but that can make things harder when filtering for datasets.</p>
<p>You can see the global aspect of GEO search here:</p>
<div class="float">
<img src="figures/GEOfilters.png" alt="GEO search: On the top, you can write your own query. Left menus indicate filters that are then added to your query. Try it out with different organisms and study types. Attribute types can also be applied. You can see suggested organisms on the right (yellow). Additional filters can also be seen (eg. Supplementary file to see only records including BED, WIG, …)." />
<div class="figcaption">GEO search: On the top, you can write your own query. Left menus indicate filters that are then added to your query. Try it out with different organisms and study types. Attribute types can also be applied. You can see suggested organisms on the right (yellow). Additional filters can also be seen (eg. <em>Supplementary file</em> to see only records including BED, WIG, …).</div>
</div>
<p>The study section is one of the most important, as it indicates what kind of assay was performed and for which goal. For example, you would apply <em>Expression profiling …</em> when looking for transcription data and <em>Genome binding/occupancy profiling … </em> when looking for chromatin accessibility, histone modifications, or transcription factor data. In addition, you can also select the type of assay (<em>high throughput sequencing</em>, <em>array</em>, <em>Mass Spec</em>, …).</p>
<p>After this, you are ready to dive into a GEO record. Feel free to open one to see information inside.</p>
</div>
<div id="geo-records" class="section level4 hasAnchor">
<h4 class="hasAnchor">GEO records<a href="#geo-records" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<div class="float">
<img src="figures/GEOrecord.png" alt="This is a GEO record. Appearance and amount of information can widely vary, particularly in summary and overall design (red). Inside each record, you will find multiple samples (green). These include information on any experimental or phenotypic differences (knock-out, wild type, treatments, among others). You will also find information on the processing of the dataset here (features like assembly, software, and other parameters)." />
<div class="figcaption">This is a GEO record. Appearance and amount of information can widely vary, particularly in summary and overall design (red). Inside each record, you will find multiple samples (green). These include information on any experimental or phenotypic differences (knock-out, wild type, treatments, among others). You will also find information on the processing of the dataset here (features like assembly, software, and other parameters).</div>
</div>
<p>Oftentimes it can be challenging to find the data you want in the format you need. While uploading data to GEO is a somewhat standardized procedure, looking for processed formats can be tricky.</p>
<p>At the bottom of each GEO record, you can find the <strong>Supplementary files</strong>. This typically includes a collection (zipped) of all supplementary files (which you can access individually by selecting a single sample).</p>
<div id="example-3" class="section level5 hasAnchor">
<h5 class="hasAnchor">Example 3<a href="#example-3" class="anchor-section" aria-label="Anchor link to header"></a></h5>
<p>In this example, we will explore a GEO record and obtain information about data processing from it.
Search for the dataset identified as <strong>GSE205807</strong>.</p>
<blockquote>
<p>What cell line was this dataset generated from?</p>
</blockquote>
<blockquote>
<p>What species and genome assembly is used for alignment?</p>
</blockquote>
<blockquote>
<p>What kind of processed files are present in each sample?</p>
</blockquote>
</div>
</div>
</div>
<div id="exercises-4" class="section level3 hasAnchor">
<h3 class="hasAnchor">EXERCISES<a href="#exercises-4" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="exercise-1.-finding-transcription-factor-targets-using-chip-seq" class="section level4 hasAnchor">
<h4 class="hasAnchor">Exercise 1. Finding transcription factor targets using ChIP-seq<a href="#exercise-1.-finding-transcription-factor-targets-using-chip-seq" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>ChIP-seq can be used to identify genes regulated by targeted transcription factors. In the example, we will explore a dataset obtained in the human liver cell line HepG2.</p>
<p>1.1 Download the set of peaks for the FOS transcription factor from the GEO dataset GSE104247.</p>
<blockquote>
<p>What genome assembly is this dataset aligned to? Apply the necessary changes to see it in IGV.</p>
</blockquote>
<p>1.2. Load the dataset in the IGV web app. Give a couple of examples of genes that are likely regulated by FOS in this cell line.</p>
<p>1.3. Pick a set of peaks obtained in your favorite transcription factor from the same GEO dataset! How is the set of peaks you observed before comparable?</p>
<div id="exercise-2.-atac-seq-and-dnase-seq-comparison" class="section level5 hasAnchor">
<h5 class="hasAnchor">Exercise 2. ATAC-seq and DNase-seq comparison<a href="#exercise-2.-atac-seq-and-dnase-seq-comparison" class="anchor-section" aria-label="Anchor link to header"></a></h5>
<p>Load a track in BED and BigWig format from ATAC-seq and DNase-seq obtained in the same cell line using the <em>Tracks</em> section from <em>ENCODE Other</em> (BED) and <em>ENCODE Signals - Other</em> (Tip: There is typically plenty of material included for the A549 cell line). Compare the called peaks and other features from each experiment (peak size, noise, coverage) and discuss the advantages one has over the other.</p>
</div>
<div id="exercise-3.-using-geo-to-look-for-dna-methylation-datasets" class="section level5 hasAnchor">
<h5 class="hasAnchor">Exercise 3. Using GEO to look for DNA methylation datasets<a href="#exercise-3.-using-geo-to-look-for-dna-methylation-datasets" class="anchor-section" aria-label="Anchor link to header"></a></h5>
<p>3.1. What filters would you apply to look for a dataset of DNA methylation array which includes rhesus macaque (<em>Macaca mulatta</em>) and gorilla (<em>Gorilla gorilla</em>)?</p>
<p>3.2. (optional) You want to find multiple transcriptomics datasets from <em>human tissues</em> infected with malaria parasites. What filters would you apply?</p>
</div>
<div id="exercise-4.-going-further" class="section level5 hasAnchor">
<h5 class="hasAnchor">Exercise 4. Going further<a href="#exercise-4.-going-further" class="anchor-section" aria-label="Anchor link to header"></a></h5>
<p>Let us see a different type of data - interaction data from ChIA-PET. Re-open the session <a href="IGV_sessions/ExampleATAC_ChIP_CTCF_hg38.json">here</a> we used for example 2. Then, load a ChIA-PET <em>loop</em> track from A549 (genome-wide chromatin interactions mediated by <em>CTCF</em>) using <strong>Tracks &gt; ENCODE Other</strong>.</p>
<p>CTCF is an important element in genome organisation and ChIA-PET is an assay used to measure chromatin interactions mediated by specific factors.</p>
<p>Observe interactions at chromosomal and gene level. What can you conclude from the comparison between ATAC-seq, CTCF ChIP-seq, and ChIA-PET?</p>
<blockquote>
<p>Is an interaction mediated by CTCF always associated with a CTCF peak? Why?</p>
</blockquote>
</div>
</div>
</div>
</div>
<div id="day-5-rna-seq---from-fastq-to-count-matrix" class="section level2 hasAnchor">
<h2 class="hasAnchor">Day 5: RNA-seq - from FASTQ to count matrix<a href="#day-5-rna-seq---from-fastq-to-count-matrix" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Today we will analyze the RNA sequencing data. Please refer to the intro section (<code>Where to start</code>) to log in the virtual machine.</p>
<p>After logging in, please check your current directory (<code>pwd</code>), and also please check your folder (<code>ls</code>). It should be empty for now.</p>
<div id="connecting-to-the-virtual-machine" class="section level3 hasAnchor">
<h3 class="hasAnchor">Connecting to the virtual machine<a href="#connecting-to-the-virtual-machine" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The tools that we will use today have already been preinstalled in a conda environment called <code>denbi</code>. You do not need to install them by yourself, but do remember to load them before using them.</p>
<ul>
<li>Follow the instructions in the Introduction (see left menu) to connect to the VM.</li>
<li>In the terminal, type <code>conda activate denbi</code></li>
</ul>
<p>After loading the <code>denbi</code> environment, you are ready to go.</p>
<p>Just in case you have problems in signing the virtual machines, you can find the output files here: <a href="https://drive.google.com/drive/folders/1EJu2F_-pFvw4v71h1QJHq_xOcK6nMAKi?usp=drive_link" class="uri">https://drive.google.com/drive/folders/1EJu2F_-pFvw4v71h1QJHq_xOcK6nMAKi?usp=drive_link</a></p>
<p>We will use the RNA-seq data from this paper: Nature volume 571, pages505–509 (2019). The raw sequence data is stored at: <a href="https://www.ebi.ac.uk/biostudies/arrayexpress/studies/E-MTAB-6798/sdrf" class="uri">https://www.ebi.ac.uk/biostudies/arrayexpress/studies/E-MTAB-6798/sdrf</a></p>
<p>Two samples are taken. One from the 2 week post-conception (2WPC) mouse brain and the other from 4WPC brain.</p>
<p>You can find the raw data here:</p>
<pre class="raw"><code>/vol/data/raw/mouse_brain_2wpb.fastq.gz


/vol/data/raw/mouse_brain_4wpb.fastq.gz</code></pre>
<p>We will start with these <code>fastq</code> files, align them to the reference genome and quantify the gene expression. You can find the workflow below.</p>
<div class="float">
<img src="figures/workflow.png" alt="Workflow for the RNA-seq analysis" />
<div class="figcaption">Workflow for the RNA-seq analysis</div>
</div>
</div>
<div id="quality-control" class="section level3 hasAnchor">
<h3 class="hasAnchor">Quality control<a href="#quality-control" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We retrieve the raw data and employ a software tool named <code>fastqc</code> to examine the quality of the sequence files.</p>
<p>We next examine the quality of the samples with <code>fastqc</code>.</p>
<pre class="fastqc"><code># Notice: only create folders/files under your own directoy. DO NOT create or delete anything outside your own directory.
mkdir path/to/your_fastqc_folder 

fastqc /vol/data/raw/mouse_brain_4wpb.fastq.gz -o path/to/your_fastqc_folder/

fastqc /vol/data/raw/mouse_brain_2wpb.fastq.gz -o path/to/your_fastqc_folder/
</code></pre>
<p>The start of the FastQC quality report provides basic statistics and an evaluation of various quality parameters.</p>
<p><img src="figures/fastqc_statistics.png" /></p>
<blockquote>
<p>Question: What information do you get from the QC report?</p>
</blockquote>
<p>In particular, we will look at the (1) per base quality, and (2) per sequence quality.</p>
<div class="float">
<img src="figures/per_base_quality.png" alt="Per base quality from the fastqc report. (Top) The result from the 2 week post-conception (2WPC) mouse brain. (Bottom) The result from the 4WPC mouse brain" />
<div class="figcaption">Per base quality from the fastqc report. (Top) The result from the 2 week post-conception (2WPC) mouse brain. (Bottom) The result from the 4WPC mouse brain</div>
</div>
<div class="float">
<img src="figures/per_sequence_quality.png" alt="Per sequence quality from the fastqc report. (Top)The result from the 2WPC mouse brain. (Bottom) The result from the 4WPC mouse brain" />
<div class="figcaption">Per sequence quality from the fastqc report. (Top)The result from the 2WPC mouse brain. (Bottom) The result from the 4WPC mouse brain</div>
</div>
<blockquote>
<p>Question: How does FastQC quantify the sequence quality?</p>
</blockquote>
<blockquote>
<p>Question: What do you find when you compare the two sequences?</p>
</blockquote>
</div>
<div id="alignment" class="section level3 hasAnchor">
<h3 class="hasAnchor">Alignment<a href="#alignment" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Alignment involves the process of mapping sequences to a reference genome.</p>
<p>We use <code>Bowtie2</code> to implement seuqence alignment. Common aligners include <code>BWA</code>, <code>TOPHAT</code>, and <code>STAR</code>. They differ in their accuracy, speed, and memory usage due to different computational algorithms.</p>
<pre><code># Notice: only create folders/files under your own directoy. DO NOT create or delete anything outside your own directory.

mkdir path/to/your_bam_folder


# Alignment (It will take for a while)
bowtie2 -p 2 --no-unal -x /vol/data/raw/chr6 -U /vol/data/raw/mouse_brain_2wpb.fastq.gz | samtools view -bS - &gt; path/to/your_bam_folder/mouse_brain_2wpb_chr6.bam 
bowtie2 -p 2 --no-unal -x /vol/data/raw/chr6 -U /vol/data/raw/mouse_brain_4wpb.fastq.gz | samtools view -bS - &gt; path/to/your_bam_folder/mouse_brain_4wpb_chr6.bam
</code></pre>
<blockquote>
<p>Question: How to view the BAM files? What do you find in your BAM files?
Ref: <a href="https://bookdown.org/content/24942ad6-9ed7-44e9-b214-1ea8ba9f0224/learning-the-bam-format.html" class="uri">https://bookdown.org/content/24942ad6-9ed7-44e9-b214-1ea8ba9f0224/learning-the-bam-format.html</a>
Hint: Use <code>samtools view</code> to view your BAM files.</p>
</blockquote>
<p>We then use <code>samtools flagstat</code> to evaluate the quality of the aligned bam file.</p>
<pre class="flagstat"><code># Notice: only create folders/files under your own directoy. DO NOT create or delete anything outside your own directory.
samtools flagstat path/to/your_bam_folder/mouse_brain_2wpb_chr6.bam  &gt; path/to/your_bam_folder/mouse_brain_flagstat_2wpb_chr6.txt
</code></pre>
<div class="float">
<img src="figures/flagstat.png" alt="Quality of the aligned bam file." />
<div class="figcaption">Quality of the aligned bam file.</div>
</div>
</div>
<div id="quantification" class="section level3 hasAnchor">
<h3 class="hasAnchor">Quantification<a href="#quantification" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Now, we utilise the BAM files to quantify the gene expression in our samples. We implement this with a tool named <code>htseq-count</code>. There are various methods exist for quantifying gene expression in <code>htseq-count</code>. For further illustration, please refer to the tutorial (<a href="https://htseq.readthedocs.io/en/release_0.11.1/count.html" class="uri">https://htseq.readthedocs.io/en/release_0.11.1/count.html</a>).</p>
<pre><code># Notice: only create folders/files under your own directoy. DO NOT create or delete anything outside your own directory.

mkdir path/to/your_count_folder
# Keep the high quality reads

samtools view -bq 40 path/to/your_bam_folder/mouse_brain_2wpb_chr6.bam &gt; path/to/your_bam_folder/mouse_brain_2wpb_chr6_Q40.bam

htseq-count -s no -m intersection-nonempty --nonunique all --format bam path/to/your_bam_folder/mouse_brain_2wpb_chr6_Q40.bam /vol/data/raw/mm39.ncbiRefSeq.chr6.gtf &gt; path/to/your_count_folder/mouse_brain_2wpb_chr6.txt
</code></pre>
<p>Please repeate the code above to get the count matrix from your mouse_brain_4wpb sample.</p>
<div class="float">
<img src="figures/count_matrix.png" alt="Gene expression profile in 2WPC mouse brain. The first column shows the gene name. The second column shows the count. Speical counts are highlighted in red" />
<div class="figcaption">Gene expression profile in 2WPC mouse brain. The first column shows the gene name. The second column shows the count. Speical counts are highlighted in red</div>
</div>
</div>
<div id="visualization-2" class="section level3 hasAnchor">
<h3 class="hasAnchor">Visualization<a href="#visualization-2" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We will see what the bam files and count matrix files look like.</p>
<p>To view the bam files with IGV, we need an index file.</p>
<pre class="visualization"><code># Notice: only create folders/files under your own directoy. DO NOT create or delete anything outside your own directory.
cd  path/to/your_bam_folder

samtools sort mouse_brain_2wpb_chr6_Q40.bam -o mouse_brain_2wpb_chr6_Q40_sorted.bam # sort the bam

samtools index mouse_brain_2wpb_chr6_Q40_sorted.bam # index the sorted bam</code></pre>
<p>To visualize the count matrix, you are encouraged to download the count matrix files to your local computer (using Cyberduck). Next, read the files in R, and visualize it (code below).</p>
<pre class="visualization"><code>
library(tidyverse)
setwd(&quot;path/to/your_count_folder&quot;)


tb_2wpb &lt;- read_tsv(&quot;mouse_brain_2wpb_chr6.txt&quot;, col_names = FALSE) %&gt;% 
  rename(Gene = X1, Count = X2) %&gt;% 
  mutate(Sample = &quot;2wpb&quot;)

tb_4wpb &lt;- read_tsv(&quot;mouse_brain_4wpb_chr6.txt&quot;, col_names = FALSE) %&gt;% 
  rename(Gene = X1, Count = X2) %&gt;%
  mutate(Sample = &quot;4wpb&quot;)

tb_merge &lt;- bind_rows(tb_2wpb, tb_4wpb)

tb_merge %&gt;% ggplot(mapping = aes(x = log10(Count + 1), fill = Sample)) +
  geom_density(alpha = 0.2)
</code></pre>
<p><img src="figures/count_distribution.png" /></p>
</div>
<div id="excercises" class="section level3 hasAnchor">
<h3 class="hasAnchor">Excercises<a href="#excercises" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<blockquote>
<ol style="list-style-type: decimal">
<li>Please explore your bam files with IGV. What do you find ?(Hint: download both .bam and .bai file —&gt; open IGV browser —&gt; click Genome &gt; Mouse (GRCmm39mm39) —&gt; click Tracks &gt; Local Files &gt; select the two files -&gt; select proper coordinate (e.g., chr6:28,527,661-28,527,699))</li>
</ol>
</blockquote>
<blockquote>
<ol start="2" style="list-style-type: decimal">
<li>Please explore your count matrix of these two samples. Do you find any genes that express differentially between the samples? Also, think about that, if you are a project leader, how would you design your experiment and what hypothesis testing strategy would you opt for? (Note: Formal analysis will be revealed next week)</li>
</ol>
</blockquote>
<blockquote>
<ol start="3" style="list-style-type: decimal">
<li>(Optional) Please try to run the whole workflow that you learnt today on mouse chromosome 8. The reference genome sequence can be found at /vol/data/raw. However, you are encouraged to retrieve the reference genome from the databases by yourself.</li>
</ol>
</blockquote>
<!--chapter:end:02-week.Rmd-->
</div>
</div>
</div>
<div id="project-cancer-hallmark-and-metabolic-pathways-in-cancer" class="section level1 hasAnchor">
<h1 class="hasAnchor">Project: Cancer Hallmark and Metabolic Pathways in Cancer<a href="#project-cancer-hallmark-and-metabolic-pathways-in-cancer" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<div id="introduction-1" class="section level2 hasAnchor">
<h2 class="hasAnchor">Introduction<a href="#introduction-1" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Tumor tissues are a complex mixture of distinct cell types. Cancer cells represent the most distinctive part
of tissue. These are normal cells which have undergone an oncogenic event, such as for example copy-number variations, in which portions of chromosomes are duplicated a large number of times, or chromosomal fusions, in which several portions of chromosomes are fused together and might induce the high expression of so-called oncogenes. However, in some cancer types, these tumor cells only represent a small fraction of the tumor tissue. Other non-tumor cells contribute to the <strong>tumor micro-environment</strong>, such as immune cells (e.g. tumor infiltrating lymphocytes, or TILs) or stromal cells. These different cell types form a complex system in which multiple interactions (e.g. through cytokine signalling) take place and promote the tumor development.</p>
<p>Despite this complexity, and the very distinct feature of different tumors (liquid, solid, sarcomas, carcinomas,..), a number of <strong>common cellular principles</strong> have been described, which can be identified in all cancer types; these are the so called <strong>cancer hallmarks</strong>. A first description of 6 hallmarks was made by Hanahan and Weinberg in 2000, and updated and extended to four further characteristics by the same authors in 2011 (see references below).</p>
<p><img src="hallmark_all.png" alt="Fig. 1 - Cancer hallmarks (adapted from Hanahan & Weinberg (2011))" width="700"></p>
<p>These hallmarks represent cellular mechanisms by which the tumor cells acquire a growth advantage over normal cells and evade cellular mechanisms, such as immune surveillance, meant to suppress cancer cells. On the other hand, various metabolic process are being altered in cancer cells to adapt the cell to the changed micro-environment. The best known example of such a metabolic changes is the <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4783224/"><em>Warburg-effect</em></a>, during which the cancer cell switch their energy production from an aerobe mode to an anaerobe-like mode, i.e. purely through glycolysis, even in the presence of oxygen (‘aerobe glycolysis’). However, in recent years, additional examples of the importance of metabolic changes in cancer have been described.</p>
<p>The activation of the 10 hallmarks can be detected by changes in the expression levels of certain genes that can be related to one or the other of these hallmarks or metabolic pathways. For example, angiogenesis, i.e. the development of new blood vessels in the tumor to provide nutrients, is induced by a number of well described genes such as <a href="https://www.genecards.org/cgi-bin/carddisp.pl?gene=EGFR&amp;keywords=EGFR"><em>EGFR</em></a>, along with a number of other genes which promote this process.</p>
<p>The goal of this project is to perform a <strong>pan-cancer analysis of these cancer hallmarks and metabolic pathways</strong>, using datasets of genes expression data from a large collection of tumor samples. We will use the TCGA (The Cancer Genome Atlas) collection of RNA-seq data, across a large number of tumor types and patients (hence <em>pan-cancer</em>) and identify the activity of these hallmarks by using a collection of gene sets which describe the activity of these pathways. Then, each group in this project will focus on a specific cancer type and compare these hallmarks between the cancer samples and matched normal control.</p>
<p>Hence, you will be conducting (1) a pan-cancer analysis to compare tumor types amongst themselves (similar to the other groups in the project), and (2) a cancer specific analysis to compare normal to cancer cells in this tumor type.</p>
<p>The 5 sub-projects correspond to a different tumor type for the cancer specific analysis</p>
<ol style="list-style-type: decimal">
<li>sub-project 1: BRCA = breast cancer;</li>
<li>sub-project 2: KIRC = kidney renal carcinoma;</li>
<li>sub-project 3: LUAD = lung adenocarcinoma;</li>
<li>sub-prohect 4: PRAD = prostate adenocarcinoma;</li>
<li>sub-project 5: THCA = Thyroid carcinoma</li>
<li>sub-project 6: COAD = colon adenocarcinoma;</li>
<li>sub-prohect 7: LIHC = liver hepatocarcinoma;</li>
<li>sub-project 8: LUSC = lung small cell carcinoma</li>
</ol>
</div>
<div id="objectives-and-workplan" class="section level2 hasAnchor">
<h2 class="hasAnchor">Objectives and workplan<a href="#objectives-and-workplan" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The objectives of your project are the following:</p>
<ol style="list-style-type: decimal">
<li><strong>Identify cancer hallmark pathways</strong>: we have provided a first list of genesets, which have been curated from the literature; check those and identify in the literature additional (for example metabolic pathways) that are relevant to cancer, and collect the corresponding gene sets. For this, you can use the following databases:</li>
</ol>
<ul>
<li><a href="https://www.gsea-msigdb.org/gsea/msigdb/">MSigDB</a>: this is a very large collection of gene sets organized by categories; you can browse by keyword</li>
<li><a href="http://geneontology.org/">Gene Ontology</a>: GO is a controlled vocabulary which describes all biological or molecular processed; you can query bey keyword, and then extract the genes which are annotated to this term (<em>make sure to select the right organism!</em>)</li>
<li><a href="https://www.genome.jp/kegg/pathway.html">KEGG</a>: Comprehensive database of biological pathways</li>
<li><a href="https://www.pathwaycommons.org/">Pathway Commons</a>: this is a very large database of pathways from which gene sets can be obtained</li>
</ul>
<ol start="2" style="list-style-type: decimal">
<li><strong>Compare gene sets</strong>: make a comparison of the cancer hallmarks gene sets and the selected metabolic gene sets; how many shared genes are there between these gene sets?</li>
</ol>
<ul>
<li>a compiled list of hallmark related datasets are provided (see dataset section); bring your metabolic pathways into a similar format</li>
<li>find a way to compare all these datasets, in order to identify groups of datasets with shared genes (identify a suitable metric to quantify the distance or similarity between 2 sets!)</li>
<li>display this comparison using a graphical representation</li>
</ul>
<ol start="3" style="list-style-type: decimal">
<li><strong>Perform a pan-cancer analysis</strong>: determine the expression profiles of these gene sets across all TCGA tumor datasets, and determine patterns and clusters of tumor samples</li>
</ol>
<ul>
<li>a data frame of expression values for many cancer samples is provided (see dataset section);</li>
<li>start with one pathway, and check the expression of the genes of this pathway compared to all genes for one sample; how could you summarize this into one number? This would represent the <strong>activity</strong> of this pathway in this sample</li>
<li>compute the activities of all pathways for all samples, and display this in an appropriate way</li>
<li>how can you identify patterns and clusters of samples / pathways?</li>
<li>compare this with the clinical informations provided in the datasets; can you find associations with specific clinical variables?</li>
</ul>
<ol start="4" style="list-style-type: decimal">
<li><strong>Compare Tumor to normal tissue through a focused analysis</strong>: perform a differential analysis for the select tumor type between normal tissue and tumor tissue</li>
</ol>
<ul>
<li>for the tumor type you selected, repeat a similar analysis as in (3) in tumor and normal samples</li>
<li>test which pathways have a significant different activity between the normal and tumor samples</li>
</ul>
<ol start="5" style="list-style-type: decimal">
<li><strong>Predict pathway activities using a regression model</strong>: predict the activity of a metabolic pathway using the other hallmarks/metabolic pathways as explanatory variables</li>
</ol>
<ul>
<li>select one pathway of interest</li>
<li>can you build a linear regression model to determine the activity of this pathway from the other pathways? Make sure to take into account the analysis performed in (2) !</li>
</ul>
<p><strong>Alternatively</strong>, you can try to perform a <strong><a href="https://www.datacamp.com/tutorial/logistic-regression-R">logistic regression model</a></strong> to perform a <strong>binary classification</strong> tumor vs. normal based on the activity of the pathways.</p>
<p>All these analysis steps should be summarized in a R markdown document, with appropriate description and graphical outputs!</p>
</div>
<div id="description-of-datasets" class="section level2 hasAnchor">
<h2 class="hasAnchor">Description of datasets<a href="#description-of-datasets" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>We have stored 4 datasets onto a public platform, figshare. These datasets represent</p>
<ol style="list-style-type: decimal">
<li><p>a data frame containing the <strong>gene expression data from RNA-seq</strong> for almost 10,000 TCGA cancer patients, representing 33 different tumor types; you can download the R-object <code>tcga_tumor_log2TPM.RDS</code> (beware, almost 1 Gb!) using this link: <a href="https://figshare.com/s/3c2bf4d766181a17a2e0">https://figshare.com/s/3c2bf4d766181a17a2e0</a></p></li>
<li><p>a data frame with <strong>37 clinical annotations</strong> for all TCGA samples present in the full data matrix; you can download the R-object <code>tcga_tumor_annotation.RDS</code> using this link : <a href="https://figshare.com/s/eed44ff6fdcf451cebd9">https://figshare.com/s/eed44ff6fdcf451cebd9</a></p></li>
<li><p>a R object containing, for 8 tumor types, the expression data of matched tumor and normal tissue; you can download this object <code>GKBioinfo_tumor_normal.rds</code> using this link <a href="https://figshare.com/s/5bd4cf0fc7a0af36b767">https://figshare.com/s/5bd4cf0fc7a0af36b767</a></p></li>
<li><p>a R object, containing a list of gene sets for the cancer hallmarks and additional pathways; you can download the object <code>hallmarks_genesets.rds</code> using this link <a href="https://figshare.com/s/4f917a4f41e90027dd6f">https://figshare.com/s/4f917a4f41e90027dd6f</a></p></li>
</ol>
<p>Once you have downloaded these 4 files, load them as below (make sure to replace the <code>path/to/your/directory/</code> part with the path to the location on your computer where you have stored the files!):</p>
<div class="sourceCode" id="cb328"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb328-1"><a href="#cb328-1" tabindex="-1"></a>tcga_exp <span class="ot">=</span> <span class="fu">readRDS</span>(<span class="st">&quot;path/to/your/directory/tcga_tumor_log2TPM.RDS&quot;</span>)</span>
<span id="cb328-2"><a href="#cb328-2" tabindex="-1"></a>tcga_annot <span class="ot">=</span> <span class="fu">readRDS</span>(<span class="st">&quot;path/to/your/directory/tcga_tumor_annotation.RDS&quot;</span>)</span>
<span id="cb328-3"><a href="#cb328-3" tabindex="-1"></a>tcga_tumor_norm <span class="ot">=</span> <span class="fu">readRDS</span>(<span class="st">&quot;path/to/your/directory/tcga_tumor_normal_datascience_proj_2020.RDS&quot;</span>)</span>
<span id="cb328-4"><a href="#cb328-4" tabindex="-1"></a>genesets <span class="ot">=</span> <span class="fu">readRDS</span>(<span class="st">&quot;path/to/your/directory/hallmarks_genesets.rds&quot;</span>)</span></code></pre></div>
<p>The first 2 objects are data frames</p>
<div class="sourceCode" id="cb329"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb329-1"><a href="#cb329-1" tabindex="-1"></a><span class="fu">dim</span>(tcga_exp)</span>
<span id="cb329-2"><a href="#cb329-2" tabindex="-1"></a>[<span class="dv">1</span>] <span class="dv">60498</span>  <span class="dv">9741</span></span>
<span id="cb329-3"><a href="#cb329-3" tabindex="-1"></a> </span>
<span id="cb329-4"><a href="#cb329-4" tabindex="-1"></a><span class="fu">dim</span>(tcga_annot)</span>
<span id="cb329-5"><a href="#cb329-5" tabindex="-1"></a>[<span class="dv">1</span>] <span class="dv">9741</span>   <span class="dv">37</span></span></code></pre></div>
<p>The third object is a list with 5 entries corresponding to 5 tumor types (BRCA = breast cancer; KIRC = kidney renal carcinoma; LUAD = lung adenocarcinoma; PRAD = prostate adenocarcinoma; THCA = Thyroid carcinoma); each entry of the list is itself a list with 3 entries, corresponding to (1) a data frame with expression values for tumor samples, (2) a data frame with expression values for normal samples from the same patients, (3) a data frame with clinical annotations for these patients.</p>
<p>Hence, if you want to access the expression matrix for tumor samples and normal sample for LUAD for example, you can do it through:</p>
<div class="sourceCode" id="cb330"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb330-1"><a href="#cb330-1" tabindex="-1"></a>luad <span class="ot">=</span> tcga_tumor_norm[[<span class="st">&quot;LUAD&quot;</span>]]</span>
<span id="cb330-2"><a href="#cb330-2" tabindex="-1"></a>luad.tumor <span class="ot">=</span> luad[[<span class="st">&quot;tumor&quot;</span>]]</span>
<span id="cb330-3"><a href="#cb330-3" tabindex="-1"></a>luad.norm <span class="ot">=</span> luad[[<span class="st">&quot;normal&quot;</span>]]</span>
<span id="cb330-4"><a href="#cb330-4" tabindex="-1"></a>luad.annot <span class="ot">=</span> luad[[<span class="st">&quot;clinical&quot;</span>]]</span>
<span id="cb330-5"><a href="#cb330-5" tabindex="-1"></a><span class="fu">dim</span>(luad.tumor)</span>
<span id="cb330-6"><a href="#cb330-6" tabindex="-1"></a>[<span class="dv">1</span>] <span class="dv">19624</span>    <span class="dv">58</span></span>
<span id="cb330-7"><a href="#cb330-7" tabindex="-1"></a><span class="fu">dim</span>(luad.norm)</span>
<span id="cb330-8"><a href="#cb330-8" tabindex="-1"></a>[<span class="dv">1</span>] <span class="dv">19624</span>    <span class="dv">58</span></span>
<span id="cb330-9"><a href="#cb330-9" tabindex="-1"></a><span class="fu">dim</span>(luad.annot)</span>
<span id="cb330-10"><a href="#cb330-10" tabindex="-1"></a>[<span class="dv">1</span>] <span class="dv">58</span> <span class="dv">37</span></span></code></pre></div>
<p>The last object is again a list</p>
<div class="sourceCode" id="cb331"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb331-1"><a href="#cb331-1" tabindex="-1"></a><span class="fu">class</span>(genesets)</span>
<span id="cb331-2"><a href="#cb331-2" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">&quot;list&quot;</span></span>
<span id="cb331-3"><a href="#cb331-3" tabindex="-1"></a><span class="fu">names</span>(genesets)</span>
<span id="cb331-4"><a href="#cb331-4" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">&quot;genesets&quot;</span>    <span class="st">&quot;description&quot;</span></span>
<span id="cb331-5"><a href="#cb331-5" tabindex="-1"></a><span class="fu">names</span>(genesets<span class="sc">$</span>genesets)</span>
<span id="cb331-6"><a href="#cb331-6" tabindex="-1"></a> [<span class="dv">1</span>] <span class="st">&quot;Angio_AACR&quot;</span>             <span class="st">&quot;Angio_ALL&quot;</span>              <span class="st">&quot;Angio_VEGF&quot;</span>            </span>
<span id="cb331-7"><a href="#cb331-7" tabindex="-1"></a> [<span class="dv">4</span>] <span class="st">&quot;Angio_N_Reg&quot;</span>            <span class="st">&quot;Angio_P_Reg&quot;</span>            <span class="st">&quot;Apop_AACR&quot;</span>             </span>
<span id="cb331-8"><a href="#cb331-8" tabindex="-1"></a> [<span class="dv">7</span>] <span class="st">&quot;Apop_ALL&quot;</span>               <span class="st">&quot;Apop_SURVIVAL&quot;</span>          <span class="st">&quot;Genome_AACR&quot;</span>           </span>
<span id="cb331-9"><a href="#cb331-9" tabindex="-1"></a>[<span class="dv">10</span>] <span class="st">&quot;Genome_DN_Reg&quot;</span>          <span class="st">&quot;Genome_REPAIR&quot;</span>          <span class="st">&quot;Growth_AACR&quot;</span>           </span>
<span id="cb331-10"><a href="#cb331-10" tabindex="-1"></a>[<span class="dv">13</span>] <span class="st">&quot;Growth_ALL&quot;</span>             <span class="st">&quot;Growth_CONTACT_Inh&quot;</span>     <span class="st">&quot;Growth_Tumor_Supp&quot;</span>     </span>
<span id="cb331-11"><a href="#cb331-11" tabindex="-1"></a>[<span class="dv">16</span>] <span class="st">&quot;Immune_AACR&quot;</span>            <span class="st">&quot;Immune_ALL&quot;</span>             <span class="st">&quot;Immune_UP&quot;</span>             </span>
<span id="cb331-12"><a href="#cb331-12" tabindex="-1"></a>[<span class="dv">19</span>] <span class="st">&quot;Immune_DN&quot;</span>              <span class="st">&quot;Immune_Ils&quot;</span>             <span class="st">&quot;Inflam_AACR&quot;</span>           </span>
<span id="cb331-13"><a href="#cb331-13" tabindex="-1"></a>[<span class="dv">22</span>] <span class="st">&quot;Inflam_PRO&quot;</span>             <span class="st">&quot;Inflam_NEG&quot;</span>             <span class="st">&quot;Meta_AACR&quot;</span>             </span>
<span id="cb331-14"><a href="#cb331-14" tabindex="-1"></a>[<span class="dv">25</span>] <span class="st">&quot;Meta_Glc&quot;</span>               <span class="st">&quot;Meta_Gln&quot;</span>               <span class="st">&quot;Meta_Chol&quot;</span>             </span>
<span id="cb331-15"><a href="#cb331-15" tabindex="-1"></a>[<span class="dv">28</span>] <span class="st">&quot;Meta_FA&quot;</span>                <span class="st">&quot;Meta_HIF1&quot;</span>              <span class="st">&quot;Metas_AACR&quot;</span>            </span>
<span id="cb331-16"><a href="#cb331-16" tabindex="-1"></a>[<span class="dv">31</span>] <span class="st">&quot;Metas_ALL&quot;</span>              <span class="st">&quot;Metas_EMT_MMP&quot;</span>          <span class="st">&quot;Metas_NEG&quot;</span>             </span>
<span id="cb331-17"><a href="#cb331-17" tabindex="-1"></a>[<span class="dv">34</span>] <span class="st">&quot;Metas_EMT&quot;</span>              <span class="st">&quot;Prol_AACR&quot;</span>              <span class="st">&quot;Prol_EGFR&quot;</span>             </span>
<span id="cb331-18"><a href="#cb331-18" tabindex="-1"></a>[<span class="dv">37</span>] <span class="st">&quot;Prol_MAPK&quot;</span>              <span class="st">&quot;Prol_PI3CK&quot;</span>             <span class="st">&quot;Prol_NOTCH&quot;</span>            </span>
<span id="cb331-19"><a href="#cb331-19" tabindex="-1"></a>[<span class="dv">40</span>] <span class="st">&quot;Prol_ERK&quot;</span>               <span class="st">&quot;Prol_RAS&quot;</span>               <span class="st">&quot;Prol_JAK_STAT&quot;</span>         </span>
<span id="cb331-20"><a href="#cb331-20" tabindex="-1"></a>[<span class="dv">43</span>] <span class="st">&quot;Prol_MTOR&quot;</span>              <span class="st">&quot;Prol_WNT&quot;</span>               <span class="st">&quot;Replication_AACR&quot;</span>      </span>
<span id="cb331-21"><a href="#cb331-21" tabindex="-1"></a>[<span class="dv">46</span>] <span class="st">&quot;Replication_Telomerase&quot;</span></span>
<span id="cb331-22"><a href="#cb331-22" tabindex="-1"></a><span class="do">## these are the first genes of the first geneset</span></span>
<span id="cb331-23"><a href="#cb331-23" tabindex="-1"></a><span class="fu">head</span>(genesets<span class="sc">$</span>genesets[[<span class="dv">1</span>]])</span>
<span id="cb331-24"><a href="#cb331-24" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">&quot;AAMP&quot;</span>   <span class="st">&quot;ACTA2&quot;</span>  <span class="st">&quot;ACVR1&quot;</span>  <span class="st">&quot;ACVRL1&quot;</span> <span class="st">&quot;ADGRG1&quot;</span> <span class="st">&quot;ADM&quot;</span>   </span>
<span id="cb331-25"><a href="#cb331-25" tabindex="-1"></a><span class="do">## the element description contains a description of the genesets</span></span>
<span id="cb331-26"><a href="#cb331-26" tabindex="-1"></a><span class="fu">head</span>(genesets<span class="sc">$</span>description)</span>
<span id="cb331-27"><a href="#cb331-27" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">&quot;na&quot;</span>                                  <span class="st">&quot;all angiogenesis inducing genes&quot;</span>    </span>
<span id="cb331-28"><a href="#cb331-28" tabindex="-1"></a>[<span class="dv">3</span>] <span class="st">&quot;VEGF pathway&quot;</span>                        <span class="st">&quot;negative regulators of angiogenesis&quot;</span></span>
<span id="cb331-29"><a href="#cb331-29" tabindex="-1"></a>[<span class="dv">5</span>] <span class="st">&quot;positive regulators of angiogenesis&quot;</span> <span class="st">&quot;na&quot;</span> </span></code></pre></div>
</div>
<div id="literature-review" class="section level2 hasAnchor">
<h2 class="hasAnchor">Literature review<a href="#literature-review" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>This list is meant as a <strong>starting point</strong> for your literature research; you should not restrict yourself to this list only!</p>
<div id="reviews-on-general-cancer-biology-and-cancer-metabolism" class="section level4 hasAnchor">
<h4 class="hasAnchor">Reviews on general cancer biology and cancer metabolism<a href="#reviews-on-general-cancer-biology-and-cancer-metabolism" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<ul>
<li><strong>Cancer biology -</strong> Hanahan, Douglas, and Robert A. Weinberg. “Hallmarks of cancer: the next generation.” cell 144, no. 5 (2011): 646-674. <a href="https://doi.org/10.1016/j.cell.2011.02.013" class="uri">https://doi.org/10.1016/j.cell.2011.02.013</a></li>
<li><strong>Cancer metabolism -</strong> De Berardinis, R. J., &amp; Chandel, N. S. (2016). Fundamentals of cancer metabolism. Science Advances, 2(5). <a href="https://doi.org/10.1126/sciadv.1600200" class="uri">https://doi.org/10.1126/sciadv.1600200</a></li>
<li><strong>Cancer metabolism -</strong> Pavlova, Natalya N., and Craig B. Thompson. “The emerging hallmarks of cancer metabolism.” Cell metabolism 23, no. 1 (2016): 27-47.https://doi.org/10.1016/j.cmet.2015.12.006</li>
</ul>
</div>
<div id="analysis-of-cancer-hallmarks-across-multiple-cancer-types-using-integrative-approches" class="section level4 hasAnchor">
<h4 class="hasAnchor">Analysis of cancer hallmarks across multiple cancer types using integrative approches<a href="#analysis-of-cancer-hallmarks-across-multiple-cancer-types-using-integrative-approches" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<ul>
<li><strong>Cancer immunology -</strong> Thorsson, Vésteinn, David L. Gibbs, Scott D. Brown, Denise Wolf, Dante S. Bortone, Tai-Hsien Ou Yang, Eduard Porta-Pardo et al. “The immune landscape of cancer.” Immunity 48, no. 4 (2018): 812-830. <a href="https://doi.org/10.1016/j.immuni.2018.03.023" class="uri">https://doi.org/10.1016/j.immuni.2018.03.023</a></li>
<li><strong>Cancer metabolism -</strong> Peng, Xinxin, Zhongyuan Chen, Farshad Farshidfar, Xiaoyan Xu, Philip L. Lorenzi, Yumeng Wang, Feixiong Cheng et al. “Molecular characterization and clinical relevance of metabolic expression subtypes in human cancers.” Cell reports 23, no. 1 (2018): 255-269. <a href="https://doi.org/10.1016/j.celrep.2018.03.077" class="uri">https://doi.org/10.1016/j.celrep.2018.03.077</a></li>
<li><strong>Cancer stemness/de-differenciation</strong> Malta, Tathiane M., Artem Sokolov, Andrew J. Gentles, Tomasz Burzykowski, Laila Poisson, John N. Weinstein, Bożena Kamińska et al. “Machine learning identifies stemness features associated with oncogenic dedifferentiation.” Cell 173, no. 2 (2018): 338-354. <a href="https://doi.org/10.1016/j.cell.2018.03.034" class="uri">https://doi.org/10.1016/j.cell.2018.03.034</a></li>
<li><strong>DNA damage in cancer -</strong> Knijnenburg, Theo A., Linghua Wang, Michael T. Zimmermann, Nyasha Chambwe, Galen F. Gao, Andrew D. Cherniack, Huihui Fan et al. “Genomic and molecular landscape of DNA damage repair deficiency across The Cancer Genome Atlas.” Cell reports 23, no. 1 (2018): 239-254. <a href="https://doi.org/10.1016/j.celrep.2018.03.076" class="uri">https://doi.org/10.1016/j.celrep.2018.03.076</a></li>
<li><strong>RAS signalling Cancer -</strong> Way, Gregory P., Francisco Sanchez-Vega, Konnor La, Joshua Armenia, Walid K. Chatila, Augustin Luna, Chris Sander et al. “Machine learning detects pan-cancer ras pathway activation in the cancer genome atlas.” Cell reports 23, no. 1 (2018): 172-180. <a href="https://doi.org/10.1016/j.celrep.2018.03.046" class="uri">https://doi.org/10.1016/j.celrep.2018.03.046</a></li>
</ul>
</div>
<div id="how-to-score-pathways" class="section level4 hasAnchor">
<h4 class="hasAnchor">How to score pathways ?<a href="#how-to-score-pathways" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<ul>
<li><p>Reimand, Jüri, Ruth Isserlin, Veronique Voisin, Mike Kucera, Christian Tannus-Lopes, Asha Rostamianfar, Lina Wadi et al. “Pathway enrichment analysis and visualization of omics data using g: Profiler, GSEA, Cytoscape and EnrichmentMap.” Nature protocols 14, no. 2 (2019): 482-517. <a href="https://doi.org/10.1038/s41596-018-0103-9" class="uri">https://doi.org/10.1038/s41596-018-0103-9</a></p></li>
<li><p>Paczkowska, Marta, Jonathan Barenboim, Nardnisa Sintupisut, Natalie S. Fox, Helen Zhu, Diala Abd-Rabbo, Miles W. Mee, Paul C. Boutros, and Jüri Reimand. “Integrative pathway enrichment analysis of multivariate omics data.” Nature communications 11, no. 1 (2020): 1-16. <a href="https://doi.org/10.1038/s41467-019-13983-9" class="uri">https://doi.org/10.1038/s41467-019-13983-9</a></p></li>
<li><p>Jia, Dongya, Mingyang Lu, Kwang Hwa Jung, Jun Hyoung Park, Linglin Yu, José N. Onuchic, Benny Abraham Kaipparettu, and Herbert Levine. “Elucidating cancer metabolic plasticity by coupling gene regulation with metabolic pathways.” Proceedings of the National Academy of Sciences 116, no. 9 (2019): 3909-3918. <a href="https://doi.org/10.1073/pnas.1816391116" class="uri">https://doi.org/10.1073/pnas.1816391116</a></p></li>
</ul>
</div>
</div>
<div id="how-to-structure-your-project" class="section level2 hasAnchor">
<h2 class="hasAnchor">How to structure your project<a href="#how-to-structure-your-project" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="project" class="section level3 hasAnchor">
<h3 class="hasAnchor">Project<a href="#project" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>You project <strong>MUST</strong> contain the following elements:
- <strong>descriptive statistics</strong> about the datasets
- <strong>graphical representations</strong>
- <strong>dimension reduction</strong> analysis (PCA, clustering or k-means)
- <strong>statistical tests</strong> (t-test, proportion tests etc)
- <strong>linear/logistic regression</strong> analysis, either uni- or multivariate</p>
<div id="data-cleanup" class="section level4 hasAnchor">
<h4 class="hasAnchor">Data cleanup<a href="#data-cleanup" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>You will be analyzing multiple genomic data sets together. It is
essential that you explore each dataset and clean it. Cleaning can refer
to many things:</p>
<ul>
<li>Removing missing values</li>
<li>Imputing missing values</li>
<li>Removing low variance columns/rows</li>
<li>Removing batch effects</li>
<li>Removing outlier samples (only if it is due to technical issues !!)</li>
<li>Making sure that data is in the correct format, for example, numbers
should be encoded as numeric and not as characters. Categorical
variables should be factors etc.</li>
<li>Re-ordering rows/columns in meaningful and useful ways</li>
</ul>
</div>
<div id="data-exploration" class="section level4 hasAnchor">
<h4 class="hasAnchor">Data exploration<a href="#data-exploration" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Now that you have cleaned data, explore your data to understand its
structure. Perform basic exploratory data analysis.</p>
<ul>
<li>Look at the distribution of the overall data, specific samples or
features.</li>
<li>Visualize the data distribution</li>
<li>Visualize the inter-dependencies among specific samples/features of
interest</li>
<li>Check some of your hypothesis like - is something high/low between
two conditions etc</li>
</ul>
</div>
<div id="data-reduction" class="section level4 hasAnchor">
<h4 class="hasAnchor">Data reduction<a href="#data-reduction" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>You have a high dimension matrix, that is, you have way more features
(~25000 genes) than observations.</p>
<ul>
<li>Try out methods to reduce the dimensionality of this data.</li>
<li>Cluster your samples to identify similar and dis-similar groups</li>
<li>Check how well the groups separate based on the features of your
interest</li>
</ul>
</div>
<div id="data-modelling" class="section level4 hasAnchor">
<h4 class="hasAnchor">Data modelling<a href="#data-modelling" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Predict activities of some pathways using the activities of other pathways.</p>
<!--chapter:end:03-project.Rmd-->
</div>
</div>
</div>
</div>
<div id="project-report-guide" class="section level1 unnumbered hasAnchor">
<h1 class="unnumbered hasAnchor">Project report guide<a href="#project-report-guide" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>At the end of this course, you will have to produce a report in RMarkdown. RMarkdown is an inherantly ready-to publish format. Meaning that if there are no mistakes on your text, it should simply produce a PDF in one go. Let’s see how.</p>
<div id="knitting" class="section level2 hasAnchor">
<h2 class="hasAnchor">Knitting<a href="#knitting" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Create a new RMarkdown document and start trying it out. You can already call this document <em>Report</em> if you would like to. You can also use a layout we have made available <a href="https://github.com/hdsu-bioquant/bioinfo_MoBiMA/tree/main/report_example">here</a>.</p>
<p>By clicking on the little blue yarn ball on top, you can <em>Knit</em> the document into a PDF (to PDF option can be found on the right drop arrow).</p>
<p>Tips to ensure you catch errors early on:</p>
<ul>
<li>Knit your document every so often! If you have too much code/text that you have not verified if it’s running or not, this might cause issues later on.</li>
<li>Read your errors when you have them: R will usually tell you where the error is located. Most times it’s a character error as some characters will have meanings in RMarkdown and that is not always clear at first.</li>
<li>Check for missing dependencies like <code>dplyr</code>, <code>tidyr</code>, or <code>ggplot2</code> if you are using these packages.</li>
<li>Comment your code!</li>
<li>Keep things short, simplify, and avoid having a wall of text/code that is hardly decipherable.</li>
</ul>
</div>
<div id="formatting" class="section level2 hasAnchor">
<h2 class="hasAnchor">Formatting<a href="#formatting" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>New chapter titles and subtitles are defined from <code>#</code> (larger title as in 1.) to <code>####</code> (smaller title as in 1.1.1.1). Further formatting can be found <a href="https://bookdown.org/yihui/rmarkdown-cookbook/formatting.html">here</a>.</p>
<p>Numerical equations and special characters have their own formatting. For example: <code>$\sqrt{13}$</code> is square-root of 13 <code>$\alpha=0.05$</code> is alpha=0.05</p>
<p>Here is an equation:</p>
<pre><code>\begin{equation} 
  f\left(k\right) = \binom{n}{k} p^k\left(1-p\right)^{n-k}
  (\#eq:binom)
\end{equation}</code></pre>
<p><span class="math display">\[\begin{equation}
  f\left(k\right) = \binom{n}{k} p^k\left(1-p\right)^{n-k}
  (\#eq:binom)
\end{equation}\]</span></p>
<p>You may refer to this equation using <code>\@ref(eq:binom)</code>.</p>
<div id="new-pages-and-paragraphs" class="section level3 hasAnchor">
<h3 class="hasAnchor">New pages and paragraphs<a href="#new-pages-and-paragraphs" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>You can use <code>\newpage</code> to create new pages and <code>\par</code> to make paragraphs.</p>
</div>
</div>
<div id="chunk-options" class="section level2 hasAnchor">
<h2 class="hasAnchor">Chunk options<a href="#chunk-options" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>There are a multitude of options to display the R chunks and their outputs. See all of them <a href="https://bookdown.org/yihui/rmarkdown-cookbook/chunk-options.html">here</a>. The ones will will most likely use are:</p>
<ul>
<li><code>echo=FALSE</code>: hides the code but not the results</li>
<li><code>results='hide'</code>: hides the results but not the code</li>
<li><code>include=FALSE</code>: does not include the chunk</li>
<li><code>eval=FALSE</code>: does not run the chunk</li>
</ul>
<p>Some of the ones below control figure placement and size:</p>
<ul>
<li><code>fig.height</code></li>
<li><code>fig.width</code></li>
<li><code>fig.align='center'</code></li>
<li><code>out.width</code> Extra: <code>results='hold', out.width='50%'</code>: will allow you to show 2 figure/plot outputs side by side.</li>
</ul>
</div>
<div id="external-images" class="section level2 hasAnchor">
<h2 class="hasAnchor">External images<a href="#external-images" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>When including external images, you need to indicate the location of such figure relative to your document: <code>![This is a Figure](figure_folder/avocado_image.jpeg)</code></p>
<div class="float">
<img src="figures/vangogh.jpeg" alt="This is a Figure. Source: https://www.metmuseum.org/de/art/collection/search/436535" />
<div class="figcaption">This is a Figure. Source: <a href="https://www.metmuseum.org/de/art/collection/search/436535" class="uri">https://www.metmuseum.org/de/art/collection/search/436535</a></div>
</div>
<p>For more additions like table, check the <a href="https://bookdown.org/yihui/rmarkdown-cookbook/tables.html">RMarkdown cookbook</a>.</p>
<p>Figures and tables <em>with captions</em> can also be cross-referenced from elsewhere using <code>\@ref(fig:chunk-label)</code> and <code>\@ref(tab:chunk-label)</code>, respectively.</p>
<p>Don’t miss Table @ref(tab:nice-tab). Reference tables using <code>\@ref(tab:nice-tab)</code></p>
<div class="sourceCode" id="cb333"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb333-1"><a href="#cb333-1" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(</span>
<span id="cb333-2"><a href="#cb333-2" tabindex="-1"></a>  <span class="fu">head</span>(pressure, <span class="dv">5</span>), <span class="at">caption =</span> <span class="st">&#39;Here is a nice table!&#39;</span>,</span>
<span id="cb333-3"><a href="#cb333-3" tabindex="-1"></a>  <span class="at">booktabs =</span> <span class="cn">TRUE</span></span>
<span id="cb333-4"><a href="#cb333-4" tabindex="-1"></a>)</span></code></pre></div>
<table>
<caption>(#tab:nice-tab)Here is a nice table!</caption>
<thead>
<tr class="header">
<th align="right">temperature</th>
<th align="right">pressure</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="right">0</td>
<td align="right">0.0002</td>
</tr>
<tr class="even">
<td align="right">20</td>
<td align="right">0.0012</td>
</tr>
<tr class="odd">
<td align="right">40</td>
<td align="right">0.0060</td>
</tr>
<tr class="even">
<td align="right">60</td>
<td align="right">0.0300</td>
</tr>
<tr class="odd">
<td align="right">80</td>
<td align="right">0.0900</td>
</tr>
</tbody>
</table>
</div>
<div id="references" class="section level2 hasAnchor">
<h2 class="hasAnchor">References<a href="#references" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Your report will very likely include some references. Each RMarkdown is accompanied by a <em>.bib</em> extension file. Indicate the name of that file in the beginning of the report (see example). You can see the one we have included in the example <a href="https://github.com/hdsu-bioquant/bioinfo_MoBiMA/tree/main/report_example">here</a> too. If you open that file in a text editor, you can see that entries are structured like this:</p>
<pre><code>@book{ggplot2,
    author = {Hadley Wickham},
    isbn = {978-3-319-24277-4},
    publisher = {Springer-Verlag New York},
    title = {ggplot2: Elegant Graphics for Data Analysis},
    url = {https://ggplot2.tidyverse.org},
    year = {2016},
    bdsk-url-1 = {https://ggplot2.tidyverse.org}}</code></pre>
<p>This citation format is compatible with citation managers like BibDesk (only available on macOS) or Mendeley. When you are using one of these, you are able to add in other references. Watch out for the reference key in your Markdown.</p>
<div id="citations" class="section level3 hasAnchor">
<h3 class="hasAnchor">Citations<a href="#citations" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Reference items present in your bibliography file(s) in the Markdown using <code>@key</code> or <code>[@key]</code>. For the previous example, key is actually <em>ggplot2</em>!</p>
</div>
<div id="footnotes" class="section level3 hasAnchor">
<h3 class="hasAnchor">Footnotes<a href="#footnotes" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Footnotes are put inside the square brackets after a caret <code>^[]</code>.</p>
<!--chapter:end:04-reportInstructions.Rmd-->
</div>
</div>
</div>
<!--bookdown:body:end-->
            </section>

          </div>
        </div>
      </div>
<!--bookdown:link_prev-->
<!--bookdown:link_next-->
    </div>
  </div>
<!--bookdown:config-->

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
