---
title: Bioinformatics Introductory Course - Master Molecular Biotechnology
author: "Biomedical Genomics group, IPMB"
site: bookdown::bookdown_site
output: bookdown::bs4_book

documentclass: book
bibliography: [book.bib, packages.bib]
# url: your book url like https://bookdown.org/yihui/bookdown
# cover-image: path to the social sharing image like images/cover.jpg
description: |
  RMarkdowns for the Data Analysis introduction course (SoSe2024), MoBi MA 
link-citations: yes
github-repo: rstudio/bookdown-demo
editor_options: 
  markdown: 
    wrap: sentence
---


![](./logos_IPMB_final_202309.jpg)  


# Introduction {.unnumbered}

R is a powerful programming language for the analysis of data.
R is very versatile, it is free, and very easy to understand.
That is because, unlike other programming languages, R is quite “human-friendly” and dynamic, thus it does not require compiling to be “machine-readable”.
All you have to do is type R commands in the console or script.

Through this course, you will learn:

> -   Basic R programming for data analysis (*before we start*, go though the sections present in this Introduction)
> -   How to use R to work on different datasets and apply statistics (*Week One*)
> -   How to annotate, analyse, obtain, pre-process, and visualize genomic data (*Week Two*)
> -   How to perform RNA-seq analysis (*Week Three*)

## Where to start {.unnumbered}


### Log into the server {.unnumbered}

We will use an **RStudio Server** running on a virtual machine.
Here are the instructions to connect:

You can find the list of students [here](https://docs.google.com/spreadsheets/d/1eMCr4b6UcOPIG8tJkfoEkMoZzJycWJPMYN5nMZWiU48/edit?usp=sharing).
Check your name, your user name on which you will connect.

#### if you are running machine with Linux or MacOS {.unnumbered}

-   open a terminal and type the following commant: `ssh <USER>@194.94.113.18 -p 30210 -L 8787:127.0.0.1:8787`; replace `<USER>` with your user name from the Google Sheet
-   in a browser, go to [this url](http://127.0.0.1:8787)
-   log in with your user name and the password that will be provided to you.

You can now start working in RStudio!

#### if you are running machine with Windows {.unnumbered}

Here, the user should install [Putty](https://www.putty.org/).
In the GUI, the following settings are needed:

-   In the main panel:
    -   Host name (or IP address): <USER>\@194.94.113.18
    -   Port: 30210 (check the right port in the Google Sheet)
    -   Connection type: SSH
-   In the panel "Connection / SSH / Tunnel":
    -   Source port: 8787
    -   Destination port: 127.0.0.1:8787
    -   Check options "Local" and "Auto"
-   Then click "Add"
-   Now click "Open"
-   in a browser, go to [this url](http://127.0.0.1:8787)
-   log in with your user name and the password that will be provided to you.

You can now start working in RStudio!

### Installing R and RStudio {.unnumbered}

Alternatively, you can also choose to install both R and RStudio.
Rstudio is the most commonly used IDE (or Integrated Development Environment) for R, and it will **use the R version installed on your computer**.
RStudio is meant to make R programming quite visual and way easier for you as a beginner.

Start by installing [R](https://cran.r-project.org/) and then [RStudio](https://posit.co/download/rstudio-desktop/).

RStudio is composed of 4 main windows/panels.\
The **Editor**, the **Environment**, the **Console**, and the **Files** window/panel.
Order can vary.

![In **Editor (1)**, you can find the area where you usually write most code, like a Rscript or a Rmarkdown. **Environment (2)** shows you your variables, history, among others. Here, you can see any variable you define in your session. You can use this to do a very basic inspection of objects you create or import. **Console (3)** is where all your code gets processed when you run it. The console remembers (which is why when you click arrow up or down, it shows you commands you ran before), but it does not keep! Therefore, ALWAYS write the meaningful code lines in a script/markdown (1). Lastly, **4** can be used for browsing documents (*Files*), see the plots you produce (*Plots*), searching/managing packages (*Packages*), or find help on functions (*Help*).](figures/RstudioLayout.png)

### Courses {.unnumbered}

It is recommended that you take an introductory class for R to get to know this language before you start coding here.
Start by going through [this Chapter on Getting Started with Data in R](https://moderndive.netlify.app/1-getting-started).
We will cover other basics on the next sections.

You can find more resources here:

> -   R Tutorial for Beginners at [guru99](https://www.guru99.com/r-tutorial.html)
> -   R courses at [Babraham Bioinformatics](https://www.bioinformatics.babraham.ac.uk/training.html#advancedrtidy)
> -   R for [Data Science](https://r4ds.had.co.nz)
> -   R markdown at [RStudio](https://rmarkdown.rstudio.com/lesson-1.html) and in this [cheatsheet](https://www.markdownguide.org/cheat-sheet/)
> -   [Youtube tutorial](https://www.youtube.com/watch?v=h29g21z0a68) on ggplot2 with one of its developers

### Installing and configuring Cyberduck {.unnumbered}

In order to access remote files (for example opening a pdf file), you will need to install an additional tool, called cyberduck (https://cyberduck.io/).

-   Install Cyberduck according to your operating system
-   Once installed, open Cyberduck
-   On the top left, click on Neue Verbindung(or new connection)

In the new window, make the following changes:

-   select sftp in the dropdown menu
-   in the server box, type 194.94.113.18
-   in the port box, type 30210
-   in Benutzername (or Username), type your username (like user1)
-   password box: (ask for it!)

Click on Verbinden (or Connect)

## Good practice {.unnumbered}

Two notes of advice about good practice now that your journey is about to start:

-   **Commenting**: Comment your code as you go with `#`! This way you avoid forgetting the meaning of a given line. This is also important if you want to share your code with others. Trust me, you will not remember what you are writing today in this class one week from now, so it will be best to have a reminder.\
-   **Clean and tidy code**: See that weird line in the middle of your script/markdown? This is meant to be a guideline about code length. If your code is so long that it crosses the line, please find a way to make it shorter. In addition, avoid making your chunks too long and make separate markdowns or scripts for each class or topic, it will keep everything more organised and easy to track.

## R Markdown {.unnumbered}

R Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents.
You will be working with R Markdown during this course.
For more details on using R Markdown see <http://rmarkdown.rstudio.com> or <https://www.markdownguide.org/basic-syntax/>.

R instructions in a markdown file are written in a "chunk", the one below.\
Chunks can be added using the `+C` icon at the top right corner of this panel/editor.\
There is also a keyboard shortcut for it.

```{r Print Hello World!}
# Sentences written after "#" are comments. Comments are ignored during the execution of the code.
# print "Hello world!" on the screen.
print("Hello world!")
```

> What happens in the editor if you remove the second parenthesis?
> Do you see the red cross appearing?
> You can click on it and read the comment.\
> And what happens if you press "enter" while the second parenthesis is missing?\

R is very smart and it will usually let know know when there is a simple syntax mistake like this one on your code.
R will also try to guess variable and function names from your input, all you have to do is *click TAB for auto-complete*.

## R Basics {.unnumbered}

### Variable assignment in R {.unnumbered}

You can start by testing one of the most basic tasks - variable assignment.
Variable assignments in R can be done using either the `<-` symbol, or the `=` symbol.

```{r Assigments, results='hide'}
a <- "Hello world!"
# or
hello = "Hello world!"
```

You can now run `print(a)` and `print(hello)`.
Note that after assignment both variables (a and hello) are listed in the "Environment" window of RStudio.

> R is case sensitive.
> `hello` is a known variable, but `Hello` is not!

You can remove variables too...

```{r rmvar, results='hide'}
rm(a)
```

### Classes {.unnumbered}

In the same way that a orange is a fruit or Malta is a country, any R variable corresponds to a class and holds a given structure.
Classes Let us say you generate a new variable (`name`).
How can you know more about it?

```{r}
name = "Marie"
```

The function `str()` gives you the **structure** of the variable.

```{r}
str(name)
```

where *chr* stands for character, the class of this variable.

The function `class()` gives you the **class** of the variable.

```{r}
class(name)
```

### Logical operations {.unnumbered}

We can perform tests on simple variables using the `==`, `>`,`<` operators:

```{r}
name == "Harry"
```

The output of this test will be either FALSE/TRUE.
You can test these using numbers too:

```{r}
33 > 23
```

### Vectors {.unnumbered}

We can make any vector using `c()`.
Try it with some numbers.

```{r}
var1 = c(3,7,12,2)
```

If you want to identify the specific number of this vector by position, you can use square brackets.
Like this:

```{r, results='hide'}
var1[1]     # outputs the first element
var1[2:3]   # outputs the 2nd to 3rd elements
var1[-3]    # outputs all except the 3rd element
```

### Matrixes {.unnumbered}

A matrix is a two-dimensional **array** of numbers.
They are defined using the `matrix()` function as follows: Take note of the difference between `byrow=FALSE` and `byrow=TRUE`.

```{r Defining matrices , results='hide'}
x <- c(1, 2, 3, 4, 5, 6)

X <- matrix(data = x, nrow = 2, ncol = 3, byrow = TRUE)
X <- matrix(data = x, nrow = 2, ncol = 3, byrow = FALSE)
```

> Do you understand all parameters ("nrow", "ncol", "byrow") of the `matrix()` function?
> What happens if you do not specify "byrow" as TRUE or FALSE?

You can ask for the **dimensions** of a matrix (and see later also for data frames) using the function `dim()`.
Do it for X.

```{r dim_matrix, results='hide'}
dim(X)
```

To access the elements of a matrix, it is similar to the vector but with 2 indexes:

```{r Accessing Matrices, results='hide'}
X[1, 2]	                    # [1] 3  -> element in the first row and second column
X[1:2, 1]	                  # [1] 1 2	
X[2, ]	                    # [1] 2 4 6	 -> all elements of the second row
length(X[1:2, 1])	          # [1] 2  -> X[1:2, 1] is a vector of length 2
```

### Dataframes {.unnumbered}

Data frames are like matrices, but they can contain **multiple types** of values mixed.\
We can create a data frame using the `data.frame()` function.
We can also convert a matrix into a data frame with the `as.data.frame()` function.

We can prepare three vectors of the same length (for example 4 elements) and create a data frame:

```{r student_gen}
Name <- c("Leah", "Alice", "Jonas", "Paula")
Age <- c(21, 22, 20, 22)
Course <- c("Mathematics", "Physics", "Medicine", "Biology")
Place_of_birth <- c("USA", "Germany", "Germany", "France")

Students <- data.frame(First_Name = Name, 
                       Age = Age, 
                       Course = Course,
                       Place_of_birth = Place_of_birth)
Students
```

We can again access specific elements in a similar manner to the matrix before.
Try some of these examples:

```{r get_elements_df, results='hide'}
# Columns
Students$Age 
Students[["Age"]]

# Rows or columns
Students[, 3]
Students[3,]

# Elements
Students$Course[3]    # access element 3 of the column "Course"
Students[2:3, 3]      # select elements 2 and 3 of column 3 
```

### Apply, sapply, lapply {.unnumbered}

`Apply()` is used to repeat the same operation over all columns/rows of a **matrix** or **data frame**.
Therefore, we need to specify three parameters:\
1.
The matrix on which the operation should be performed\
2.
Whether to repeat the operation over rows, which is defined by a "1", or columns (use a "2" instead)\
3.
The operation that should be performed.\

Let's have a look at an example to make this more clear: We want to determine the minimum for each row of the matrix "Mat".
Therefore, we can use the function `min()`.

```{r apply for matrices}
#First, we build the matrix 
m <- rnorm(30)   # generate 30 random numbers for the matrix
Mat <- matrix(data = m, nrow = 6)
Mat

#Next, we want to determine the minimum value of each row: 
apply(Mat, 1, min)
```

> What happens if you change the "1" in the apply function to "2"?
> Try it!

> Remember that you can use `help(apply)` or `?apply` to get help on this function!

We can perform a loop over all elements of a **vector** or **list** using the `sapply()` function.\
`sapply()` needs two information:\
1.
which vector do we consider?\
2.
which function do we want to apply to each element of this vector?\
`sapply()` will return a **vector** containing the results.

For instance, we can calculate the square root of every element of a vector using the function `sqrt()`:

```{r sapply for vectors}
x <- c(1:5)
sapply(x, sqrt)
```

Instead of using built-in functions, we can also write our own little function and combine it with `sapply()`, `apply()` or `lapply()` (see below):

```{r}
# sapply()
z <- c(1:5)
sapply(z, function(x) {x*2}) 

# apply()
Mat 
apply(Mat, 1, function(x) {sum(x*2)}) #the operation is performed on every row of "Mat"; every element is multiplied by two and the sum of the row is calculated
```

Besides, we can use `sapply()` for a **list**, for example to calculate the length of every element using `length()`:

```{r lapply for lists}
List1 <- list(color = c("blue", "red"), size = 5, state = c(TRUE, FALSE, TRUE, TRUE))

sapply(List1, length)
```

As you can see, this returns a vector with the length of every list element.

However, sometimes it can be useful to keep the results stored in a list.
Therefore, we can use `lapply()` instead.
Let's have a look at the difference between `sapply()` and `lapply()` using the example from above:

```{r lapply}
lapply(List1, length)
```

> Do you understand the difference?

### For Loops {.unnumbered}

Besides the apply-family, we can use **for loops** for iterating over a sequence:

```{r For loops}
x <- c("a", "b", "c")

for (i in x) {
  print(i)
}

#or 
for (i in 1:4) {
  print(i*3)
}
```

### If Statements {.unnumbered}

You can use an **if statement** to execute a block of code only, if a condition is TRUE.

```{r If statements}
a <- 5
b <- 10

if (a < b) {
  print("a is smaller than b")
}
```

> Modify "a" or "b" and see how the output changes!

We can also add **else if** to this statement.
In that case, if the **if** condition is FALSE, the **else if** condition will be tried:

```{r else if}
a <- 10
b <- 10

if (a < b) {
  print("a is smaller than b")
} else if (a == b) {
  print("a is equal to b")
}
```

Besides, we can add an **else** statement to be executed when all previous conditions are not TRUE.

```{r else}
a <- 10
b <- 5

if (a < b) {
  print("a is smaller than b")
} else if (a == b) {
  print("a is equal to b")
} else {
  print("a is greater than b")
}

```

> Do you understand how this works?
> What would happen if a = 10 and b = 10?
> Try it!

## Data wrangling with tidyverse {.unnumbered}

### Introducing tidyverse {.unnumbered}

**Tidyverse** is a set of packages often used to manipulate data.
You can think of tidyverse as a variation of the original R, which is meant to be a grammar specific to data manipulation.

You can install these through:

```{r, results='hide', eval=F}
install.packages("tidyverse")
```

And load it using

```{r, results='hide', warning=FALSE}
library(tidyverse)
```

You can find R packages in CRAN.
CRAN is a large repository containing almost 20 thousand packages!
For this course, we will just use CRAN packages.
To install them, you will use `install.packages()`.

We will learn a couple of functions from this framework which will make the next tasks a lot easier.
Tidyverse works with the *tibble* class, which is very similar to a the dataframe class, only faster and tidier.

Start by loading a default tibble including sleeping patterns from different mammals.
It also includes information on the animal order, genus, diet, or body weight.
To do so, run the following:

```{r}
data(msleep)
```

We can start by inspecting `msleep`.

```{r, results='hide'}
glimpse(msleep)
```

We can print the first rows of a dataframe using `head()`.
Try it out.

```{r}
head(msleep)
```

### Filter based on conditions {.unnumbered}

Now that we know how this dataset looks like, we can start doing more with it.
Let us see the information for the domestic pig.\
We can do it using `filter()` on the **genus** column:

```{r}
msleep %>%
  filter(genus == "Sus")
```

We also use the `%>%` operator here.
This is a **pipe operator** and you will use it very often.

To understand piping, let's imagine you are making a three layer cake in a factory.
Instead of manually carrying the cake from one station to another to add layers and frosting, you have a conveyor belt transporting the cake through the stations automatically to make the process much more efficient.
Piping avoids that you have to create multiple intermediate objects to achieve your final result when performing multiple functions.

> Try filtering other columns based on numeric conditions.
> Use `filter()` to see which 7 animals in the datasets have a body weight superior to 200 (kg).

### Select specific columns {.unnumbered}

Now, let's say you want to *select* only specific columns.
Use the `select()` function to select the columns `name`, `order`, and `bodywt` (body weight).

```{r}
body_weights = msleep %>%
  select(name, order, bodywt)

head(body_weights)
```

### Operations on groups {.unnumbered}

We can assess the average (`mean()`) body weight by animal order present.
To do so, we will use `summarise()` and `group_by()` functions on the *body_weights* we generated previously.

-   `group_by()`: Assigns a group to a given set of rows (observations).\
-   `summarise()`: Makes calculations for a given group based on the row values assigned to it.

```{r}
body_weights %>% 
  group_by(order) %>%
  summarise(mean_bweight = mean(bodywt, na.rm = T)) # na.rm = T will exclude missing values (NA)
```

### Reshaping data {.unnumbered}

On tidyverse, he objects you are working on are sometimes not in a tidy format.
There are three rules which make a dataset tidy:

-   Each variable has its own column.
-   Each observation has its own row.
-   Each value has its own cell.

![Here is the comparison between object types. Tidy objects (right) require an extra step of processing.](figures/tidyexp.png) [Image source](https://r4ds.hadley.nz/data-tidy.html)

How can we convert them then?
The mammalian sleep dataset we used earlier is already in a tidy format, so we will generate an example.

You are testing a new diet on 7 cats, and thus registering their weight before and after 1 month on the diet.

```{r}
cat.weights <- data.frame(CatName=c('Muffin', 'Lily', 'Mittens', 'Oreo', 'Loki', 'Fluffy', 'Honey'),
                          month0=c(4.3, 4.5, 5, 4.4, 3.8, 5.5, 4.5),
                          month1=c(4.4, 4.5, 4.9, 4.3, 3.9, 5.4, 4.6))

cat.weights
```

In this example, you can see that we have 2 different variables for the same unit (weight), these are our *gatherable* columns.
We can then make this dataframe tidy using the `gather()` from the `tidyr` package:

```{r, warning=FALSE}
library(tidyr)

cat.weights %>% 
  gather(key = Time, # column name given to the gathered column names
         value = Weight, # column name will be given to the values
         month0:month1) # columns to gather
```

Do you understand the differences?
Note that now each name shows up twice.

## Plotting using ggplot2 {.unnumbered}

One of the many advantages of tidyverse is how compatible it is with `ggplot2`.
Thanks to the wonders of the `%>%` operator, you can generate a basic plot instantly from a gathered/tidy tibble without creating intermediate objects.

**The structure of any `ggplot()` is always based on two essential elements: the aesthetics (`aes()`) and the plot layers (named as `geom_something()`. `aes()` can found within the `ggplot()` function or on the `geom_something()`.** There are multiple options for `geom_X()` depending on the information you are plotting.
You can find all possibilities [here](https://ggplot2.tidyverse.org/reference/).

The `aes()` function can harbor information on axis (x, y) and other aesthetics.
Let us see an example within the scatter plot next.

In this course we will work with `ggplot()`, but plots can also be generated by another framework.
The alternative to `ggplot()` is `plot()`.
Beware when looking for solutions for your data visualization problems.

### Scatter plot {.unnumbered}

For example, if we want to see the relationship between the sleep time (total, in hours) and the body weights, we use this formula:

```{r, fig.height=4, fig.width=4}
ggplot(msleep,                    # start by picking the dataframe to plot
       aes(x = bodywt,            # column to use for x-axis
           y = sleep_total)) +    # column to use for y-axis
  geom_point()                    # plot layer for points
```

> The x-axis looks a little squished, so it will be hard to view many of the trends.
> We can then log-transform this scale using the `log()` function.
> Try it out!

We can try to play with the other aesthetics of the plot.
For example, we can use the colour easthetic to colour the dots based on the *vore*/diet.

```{r, fig.height=4, fig.width=5}
ggplot(msleep,                    
       aes(x = log(bodywt),            # column to use for x-axis
           y = sleep_total,            # column to use for y-axis
           colour = vore)) +           # column used for colouring dots
  geom_point() 
```

### Histogram {.unnumbered}

Let us start by making an histogram, as we will work with distributions over this course.
If we want to see the global distribution of sleep time (total, in hours) for all animals in the dataset, we will use `geom_histogram()`.
Like this:

```{r, message=FALSE, fig.height=4, fig.width=5}
ggplot(msleep,                    # start by picking the dataframe to plot
       aes(x = sleep_total)) +    # column to use for y-axis
  geom_histogram()                # plot layer for histogram
```

In some cases, you will not have the data on a tidy tibble and rather on a vector for example.
For such cases, the structure would be slightly different.
Check how you can plot the distribution of a random set of numbers.

```{r}
rand.numbers = rnorm(100) # we will learn this function later on

ggplot() +   
  geom_histogram(aes(x = rand.numbers),  # aes goes in the plot layer for histogram
                 bins = 50)
```

> Try changing the option `bins`.
> What do you see?

### Boxplot {.unnumbered}

In this next example, we are creating a boxplot with the sleep time (total, in hours) for each dietary group (`vore`).

```{r, fig.height=3, fig.width=4}
ggplot(msleep, # we use na.omit here to hide NAs from the plot
       aes(x = vore,
           y = sleep_total)) +
  geom_boxplot()
```

See the NA on the x-axis?
In the course (week one) we will learn how to clean the data to avoid missing values like these ones.

> Try some of the other geom\_ layers available.
> As an example, `geom_violin()` can be used in a similar way to `geom_boxplot()`.

## Common errors {.unnumbered}

`Error in library(viridis) : there is no package called ‘viridis’`:\
**Solution:** install the required package using `install.packages("package")` and load it using `library(package)`.

`Error in gather : could not find function "gather"`:\
**Solution:** load the package the function belongs to.
You can see it by running `??gather`

`Error in fortify(): ! data must be a <data.frame>, or an object coercible by fortify(), not a <uneval> object.`:\
**Solution:** `aes()` is within `ggplot()` when it should be on `geom_...()`.

## Introductory exercises {.unnumbered}

### Exercise 1: Vectors {.unnumbered}

1.  Create a vector named "test_scores" containing the test scores 75, 90, 65, 68 and 83 of students.

2.  Create a vector "passing" that contains TRUE for test scores above 70 and FALSE for scores equal or below 70.
    *Hint*: Remember that you can perform tests on vector elements.
    Use `sum()` to check how many students passed the test.

3.  *(expert)* Can you create a vector "high_scores" that contains only scores above 70, using the results obtained in 1.
    and 2.?

### Exercise 2: Matrices {.unnumbered}

1.  Create a matrix that looks like this:

```{r}
#      [,1] [,2] [,3]
# [1,]    2    4    6
# [2,]    8    10   12
# [3,]    3    5    7
# [4,]    2    9    11
```

2.  Select different parts of the matrix, for example:

    -   the element of the second row and first column
    -   the second and third element of the second column
    -   all elements of the last column
    -   Try something else!

3.  Add a new column to the matrix using `cbind()`.

4.  Can you name the columns of the matrix as "A", "B", "C" and "D" using `colnames()`?

### Exercise 3: Data Frames {.unnumbered}

1.  Create a data frame from these vectors, with three columns and five rows:

```{r Exercise 4 Data frames}
Product_name <- c("orange", "strawberry", "broccoli", "blueberry", "cucumber")
Fruit <- c(TRUE, TRUE, FALSE, TRUE, FALSE)
Color <- c("orange", "red", "green", "blue", "green")
```

2.  Select the column "Fruit".
    How many Fruits are there?
    *Hint*: Remember the function `sum()`.

3.  Add a column named "Berry" containing TRUE (for products that are berries) or FALSE to the data frame.
