<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>4 3. Normal/Gaussian distribution | Bioinformatics introductory course - MSc Molekulare Biotechnologie</title>
  <meta name="description" content="RMarkdowns for the Data Analysis introduction course (SoSe2024), MoBi MA" />
  <meta name="generator" content="bookdown 0.37 and GitBook 2.6.7" />

  <meta property="og:title" content="4 3. Normal/Gaussian distribution | Bioinformatics introductory course - MSc Molekulare Biotechnologie" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="RMarkdowns for the Data Analysis introduction course (SoSe2024), MoBi MA" />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="4 3. Normal/Gaussian distribution | Bioinformatics introductory course - MSc Molekulare Biotechnologie" />
  
  <meta name="twitter:description" content="RMarkdowns for the Data Analysis introduction course (SoSe2024), MoBi MA" />
  

<meta name="author" content="Biomedical Genomics group, IPMB" />


<meta name="date" content="2024-03-01" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="week-1-probability-distributions.html"/>
<link rel="next" href="exercises-2.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>
<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Bioinformatics Introduction course (MoBi MA)</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Introduction</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#where-to-start"><i class="fa fa-check"></i>Where to start</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#r"><i class="fa fa-check"></i>R</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#rstudio"><i class="fa fa-check"></i>RStudio</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#variable-assignment-in-r"><i class="fa fa-check"></i>Variable assignment in R</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#classes"><i class="fa fa-check"></i>Classes</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#logical-operations"><i class="fa fa-check"></i>Logical operations</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#vectors"><i class="fa fa-check"></i>Vectors</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#matrixes"><i class="fa fa-check"></i>Matrixes</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#dataframes"><i class="fa fa-check"></i>Dataframes</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#r-markdown"><i class="fa fa-check"></i>R Markdown</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#advanced-data-manipulation-with-tidyverse"><i class="fa fa-check"></i>Advanced data manipulation with tidyverse</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#filter-based-on-conditions"><i class="fa fa-check"></i>Filter based on conditions</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#select-specific-columns"><i class="fa fa-check"></i>Select specific columns</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#operations-on-groups"><i class="fa fa-check"></i>Operations on groups</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#pretty-plots-using-ggplot2"><i class="fa fa-check"></i>Pretty plots using ggplot2</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#histogram"><i class="fa fa-check"></i>Histogram</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#scatter-plot"><i class="fa fa-check"></i>Scatter plot</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#boxplot"><i class="fa fa-check"></i>Boxplot</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="1" data-path="week-1-descriptive-statistics-and-data-types.html"><a href="week-1-descriptive-statistics-and-data-types.html"><i class="fa fa-check"></i><b>1</b> WEEK 1: Descriptive statistics and data types</a>
<ul>
<li class="chapter" data-level="1.1" data-path="week-1-descriptive-statistics-and-data-types.html"><a href="week-1-descriptive-statistics-and-data-types.html#data-features-and-where-to-find-them"><i class="fa fa-check"></i><b>1.1</b> Data features and where to find them</a>
<ul>
<li class="chapter" data-level="1.1.1" data-path="week-1-descriptive-statistics-and-data-types.html"><a href="week-1-descriptive-statistics-and-data-types.html#load-the-data"><i class="fa fa-check"></i><b>1.1.1</b> Load the data</a></li>
<li class="chapter" data-level="1.1.2" data-path="week-1-descriptive-statistics-and-data-types.html"><a href="week-1-descriptive-statistics-and-data-types.html#dimensions-and-naming"><i class="fa fa-check"></i><b>1.1.2</b> Dimensions and naming</a></li>
<li class="chapter" data-level="1.1.3" data-path="week-1-descriptive-statistics-and-data-types.html"><a href="week-1-descriptive-statistics-and-data-types.html#numerical-features"><i class="fa fa-check"></i><b>1.1.3</b> Numerical features</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="week-1-descriptive-statistics-and-data-types.html"><a href="week-1-descriptive-statistics-and-data-types.html#data-cleaning"><i class="fa fa-check"></i><b>1.2</b> Data cleaning</a>
<ul>
<li class="chapter" data-level="1.2.1" data-path="week-1-descriptive-statistics-and-data-types.html"><a href="week-1-descriptive-statistics-and-data-types.html#filter"><i class="fa fa-check"></i><b>1.2.1</b> filter()</a></li>
<li class="chapter" data-level="1.2.2" data-path="week-1-descriptive-statistics-and-data-types.html"><a href="week-1-descriptive-statistics-and-data-types.html#is.na"><i class="fa fa-check"></i><b>1.2.2</b> is.na()</a></li>
<li class="chapter" data-level="1.2.3" data-path="week-1-descriptive-statistics-and-data-types.html"><a href="week-1-descriptive-statistics-and-data-types.html#mutate"><i class="fa fa-check"></i><b>1.2.3</b> mutate()</a></li>
<li class="chapter" data-level="1.2.4" data-path="week-1-descriptive-statistics-and-data-types.html"><a href="week-1-descriptive-statistics-and-data-types.html#accross"><i class="fa fa-check"></i><b>1.2.4</b> accross()</a></li>
<li class="chapter" data-level="1.2.5" data-path="week-1-descriptive-statistics-and-data-types.html"><a href="week-1-descriptive-statistics-and-data-types.html#in"><i class="fa fa-check"></i><b>1.2.5</b> %in%</a></li>
<li class="chapter" data-level="1.2.6" data-path="week-1-descriptive-statistics-and-data-types.html"><a href="week-1-descriptive-statistics-and-data-types.html#ready-for-the-cleaning"><i class="fa fa-check"></i><b>1.2.6</b> Ready for the cleaning!</a></li>
</ul></li>
<li class="chapter" data-level="1.3" data-path="week-1-descriptive-statistics-and-data-types.html"><a href="week-1-descriptive-statistics-and-data-types.html#visualizing-data-distribution"><i class="fa fa-check"></i><b>1.3</b> Visualizing data distribution</a>
<ul>
<li class="chapter" data-level="1.3.1" data-path="week-1-descriptive-statistics-and-data-types.html"><a href="week-1-descriptive-statistics-and-data-types.html#histograms"><i class="fa fa-check"></i><b>1.3.1</b> Histograms</a></li>
<li class="chapter" data-level="1.3.2" data-path="week-1-descriptive-statistics-and-data-types.html"><a href="week-1-descriptive-statistics-and-data-types.html#density-plots"><i class="fa fa-check"></i><b>1.3.2</b> Density plots</a></li>
<li class="chapter" data-level="1.3.3" data-path="week-1-descriptive-statistics-and-data-types.html"><a href="week-1-descriptive-statistics-and-data-types.html#boxplots"><i class="fa fa-check"></i><b>1.3.3</b> Boxplots</a></li>
<li class="chapter" data-level="1.3.4" data-path="week-1-descriptive-statistics-and-data-types.html"><a href="week-1-descriptive-statistics-and-data-types.html#qq-plots"><i class="fa fa-check"></i><b>1.3.4</b> QQ-plots</a></li>
</ul></li>
<li class="chapter" data-level="1.4" data-path="week-1-descriptive-statistics-and-data-types.html"><a href="week-1-descriptive-statistics-and-data-types.html#correlation"><i class="fa fa-check"></i><b>1.4</b> Correlation</a>
<ul>
<li class="chapter" data-level="1.4.1" data-path="week-1-descriptive-statistics-and-data-types.html"><a href="week-1-descriptive-statistics-and-data-types.html#measuring-the-centrality-in-data"><i class="fa fa-check"></i><b>1.4.1</b> Measuring the centrality in data</a></li>
<li class="chapter" data-level="1.4.2" data-path="week-1-descriptive-statistics-and-data-types.html"><a href="week-1-descriptive-statistics-and-data-types.html#association-between-variables"><i class="fa fa-check"></i><b>1.4.2</b> Association between variables</a></li>
</ul></li>
<li class="chapter" data-level="1.5" data-path="week-1-descriptive-statistics-and-data-types.html"><a href="week-1-descriptive-statistics-and-data-types.html#exercises"><i class="fa fa-check"></i><b>1.5</b> Exercises</a>
<ul>
<li class="chapter" data-level="1.5.1" data-path="week-1-descriptive-statistics-and-data-types.html"><a href="week-1-descriptive-statistics-and-data-types.html#exercise-1-data-features"><i class="fa fa-check"></i><b>1.5.1</b> Exercise 1: Data features</a></li>
<li class="chapter" data-level="1.5.2" data-path="week-1-descriptive-statistics-and-data-types.html"><a href="week-1-descriptive-statistics-and-data-types.html#exercise-2-visualization"><i class="fa fa-check"></i><b>1.5.2</b> Exercise 2: Visualization</a></li>
<li class="chapter" data-level="1.5.3" data-path="week-1-descriptive-statistics-and-data-types.html"><a href="week-1-descriptive-statistics-and-data-types.html#exercise-3-correlation"><i class="fa fa-check"></i><b>1.5.3</b> Exercise 3: Correlation</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="week-1-dimensionality-reduction-and-unsupervised-learning.html"><a href="week-1-dimensionality-reduction-and-unsupervised-learning.html"><i class="fa fa-check"></i><b>2</b> WEEK 1: Dimensionality reduction and unsupervised learning</a>
<ul>
<li class="chapter" data-level="2.1" data-path="week-1-dimensionality-reduction-and-unsupervised-learning.html"><a href="week-1-dimensionality-reduction-and-unsupervised-learning.html#preparing-the-data"><i class="fa fa-check"></i><b>2.1</b> Preparing the data</a>
<ul>
<li class="chapter" data-level="2.1.1" data-path="week-1-dimensionality-reduction-and-unsupervised-learning.html"><a href="week-1-dimensionality-reduction-and-unsupervised-learning.html#load-data"><i class="fa fa-check"></i><b>2.1.1</b> Load data</a></li>
<li class="chapter" data-level="2.1.2" data-path="week-1-dimensionality-reduction-and-unsupervised-learning.html"><a href="week-1-dimensionality-reduction-and-unsupervised-learning.html#data-transformation"><i class="fa fa-check"></i><b>2.1.2</b> Data transformation</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="week-1-dimensionality-reduction-and-unsupervised-learning.html"><a href="week-1-dimensionality-reduction-and-unsupervised-learning.html#k-means-clustering"><i class="fa fa-check"></i><b>2.2</b> k-means clustering</a>
<ul>
<li class="chapter" data-level="2.2.1" data-path="week-1-dimensionality-reduction-and-unsupervised-learning.html"><a href="week-1-dimensionality-reduction-and-unsupervised-learning.html#performing-k-means"><i class="fa fa-check"></i><b>2.2.1</b> Performing k-means</a></li>
<li class="chapter" data-level="2.2.2" data-path="week-1-dimensionality-reduction-and-unsupervised-learning.html"><a href="week-1-dimensionality-reduction-and-unsupervised-learning.html#quality-of-the-clustering"><i class="fa fa-check"></i><b>2.2.2</b> Quality of the clustering</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="week-1-dimensionality-reduction-and-unsupervised-learning.html"><a href="week-1-dimensionality-reduction-and-unsupervised-learning.html#hierarchical-clustering"><i class="fa fa-check"></i><b>2.3</b> Hierarchical clustering</a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="week-1-dimensionality-reduction-and-unsupervised-learning.html"><a href="week-1-dimensionality-reduction-and-unsupervised-learning.html#determine-the-most-variable-genes"><i class="fa fa-check"></i><b>2.3.1</b> Determine the most variable genes</a></li>
<li class="chapter" data-level="2.3.2" data-path="week-1-dimensionality-reduction-and-unsupervised-learning.html"><a href="week-1-dimensionality-reduction-and-unsupervised-learning.html#computing-the-correlation-between-all-patients"><i class="fa fa-check"></i><b>2.3.2</b> Computing the correlation between all patients</a></li>
<li class="chapter" data-level="2.3.3" data-path="week-1-dimensionality-reduction-and-unsupervised-learning.html"><a href="week-1-dimensionality-reduction-and-unsupervised-learning.html#including-clinical-annotations-in-the-heatmap"><i class="fa fa-check"></i><b>2.3.3</b> Including clinical annotations in the heatmap</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="week-1-dimensionality-reduction-and-unsupervised-learning.html"><a href="week-1-dimensionality-reduction-and-unsupervised-learning.html#principal-component-analysis"><i class="fa fa-check"></i><b>2.4</b> Principal component analysis</a></li>
<li class="chapter" data-level="2.5" data-path="week-1-dimensionality-reduction-and-unsupervised-learning.html"><a href="week-1-dimensionality-reduction-and-unsupervised-learning.html#exercises-1"><i class="fa fa-check"></i><b>2.5</b> EXERCISES</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="week-1-probability-distributions.html"><a href="week-1-probability-distributions.html"><i class="fa fa-check"></i><b>3</b> WEEK 1: Probability distributions</a>
<ul>
<li class="chapter" data-level="3.1" data-path="week-1-probability-distributions.html"><a href="week-1-probability-distributions.html#probability-distributions"><i class="fa fa-check"></i><b>3.1</b> Probability distributions</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="week-1-probability-distributions.html"><a href="week-1-probability-distributions.html#practice-on-the-various-functions"><i class="fa fa-check"></i><b>3.1.1</b> Practice on the various functions</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="normalgaussian-distribution.html"><a href="normalgaussian-distribution.html"><i class="fa fa-check"></i><b>4</b> 3. Normal/Gaussian distribution</a>
<ul>
<li class="chapter" data-level="4.0.1" data-path="normalgaussian-distribution.html"><a href="normalgaussian-distribution.html#visualization"><i class="fa fa-check"></i><b>4.0.1</b> 3.1 Visualization</a></li>
<li class="chapter" data-level="4.0.2" data-path="normalgaussian-distribution.html"><a href="normalgaussian-distribution.html#application-on-a-real-dataset"><i class="fa fa-check"></i><b>4.0.2</b> 3.2 Application on a real dataset</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="exercises-2.html"><a href="exercises-2.html"><i class="fa fa-check"></i><b>5</b> Exercises</a></li>
<li class="chapter" data-level="6" data-path="poisson-distribution.html"><a href="poisson-distribution.html"><i class="fa fa-check"></i><b>6</b> 4. Poisson distribution</a>
<ul>
<li class="chapter" data-level="6.0.1" data-path="poisson-distribution.html"><a href="poisson-distribution.html#practical-example-of-a-poisson-distribution"><i class="fa fa-check"></i><b>6.0.1</b> 4.1 Practical example of a Poisson distribution</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i><b>7</b> 5. References</a>
<ul>
<li class="chapter" data-level="7.1" data-path="references.html"><a href="references.html#confidence-interval"><i class="fa fa-check"></i><b>7.1</b> Confidence interval</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="week-1-hypothesis-testing.html"><a href="week-1-hypothesis-testing.html"><i class="fa fa-check"></i><b>8</b> WEEK 1: Hypothesis testing</a>
<ul>
<li class="chapter" data-level="8.1" data-path="week-1-hypothesis-testing.html"><a href="week-1-hypothesis-testing.html#students-t-test"><i class="fa fa-check"></i><b>8.1</b> Student’s t-test</a></li>
<li class="chapter" data-level="8.2" data-path="week-1-hypothesis-testing.html"><a href="week-1-hypothesis-testing.html#wilcoxon-test"><i class="fa fa-check"></i><b>8.2</b> Wilcoxon test</a></li>
<li class="chapter" data-level="8.3" data-path="week-1-hypothesis-testing.html"><a href="week-1-hypothesis-testing.html#chi-squared-test"><i class="fa fa-check"></i><b>8.3</b> Chi-squared test</a></li>
<li class="chapter" data-level="8.4" data-path="week-1-hypothesis-testing.html"><a href="week-1-hypothesis-testing.html#fishers-exact-test"><i class="fa fa-check"></i><b>8.4</b> Fisher’s exact test</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="week-1-multiple-testing-and-regression.html"><a href="week-1-multiple-testing-and-regression.html"><i class="fa fa-check"></i><b>9</b> WEEK 1: Multiple testing and regression</a>
<ul>
<li class="chapter" data-level="9.1" data-path="week-1-multiple-testing-and-regression.html"><a href="week-1-multiple-testing-and-regression.html#multiple-testing"><i class="fa fa-check"></i><b>9.1</b> Multiple testing</a></li>
<li class="chapter" data-level="9.2" data-path="week-1-multiple-testing-and-regression.html"><a href="week-1-multiple-testing-and-regression.html#regression"><i class="fa fa-check"></i><b>9.2</b> Regression</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="week-2-annotathon.html"><a href="week-2-annotathon.html"><i class="fa fa-check"></i><b>10</b> WEEK 2: Annotathon!</a></li>
<li class="chapter" data-level="11" data-path="week-2-data-formats-and-where-to-find-them.html"><a href="week-2-data-formats-and-where-to-find-them.html"><i class="fa fa-check"></i><b>11</b> WEEK 2: Data formats and where to find them</a></li>
<li class="chapter" data-level="12" data-path="week-2-rna-seq---from-fastq-to-count-matrix.html"><a href="week-2-rna-seq---from-fastq-to-count-matrix.html"><i class="fa fa-check"></i><b>12</b> WEEK 2: RNA-seq - from FASTQ to count matrix</a></li>
<li class="chapter" data-level="" data-path="references-1.html"><a href="references-1.html"><i class="fa fa-check"></i>References</a></li>
<li class="chapter" data-level="13" data-path="instructions.html"><a href="instructions.html"><i class="fa fa-check"></i><b>13</b> instructions</a>
<ul>
<li class="chapter" data-level="13.1" data-path="instructions.html"><a href="instructions.html#chapters-and-sub-chapters"><i class="fa fa-check"></i><b>13.1</b> Chapters and sub-chapters</a></li>
<li class="chapter" data-level="13.2" data-path="instructions.html"><a href="instructions.html#captioned-figures-and-tables"><i class="fa fa-check"></i><b>13.2</b> Captioned figures and tables</a></li>
<li class="chapter" data-level="13.3" data-path="instructions.html"><a href="instructions.html#footnotes"><i class="fa fa-check"></i><b>13.3</b> Footnotes</a></li>
<li class="chapter" data-level="13.4" data-path="instructions.html"><a href="instructions.html#citations"><i class="fa fa-check"></i><b>13.4</b> Citations</a></li>
<li class="chapter" data-level="13.5" data-path="instructions.html"><a href="instructions.html#equations"><i class="fa fa-check"></i><b>13.5</b> Equations</a></li>
<li class="chapter" data-level="13.6" data-path="instructions.html"><a href="instructions.html#theorems-and-proofs"><i class="fa fa-check"></i><b>13.6</b> Theorems and proofs</a></li>
<li class="chapter" data-level="13.7" data-path="instructions.html"><a href="instructions.html#callout-blocks"><i class="fa fa-check"></i><b>13.7</b> Callout blocks</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Bioinformatics introductory course - MSc Molekulare Biotechnologie</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="normalgaussian-distribution" class="section level1 hasAnchor" number="4">
<h1><span class="header-section-number">4</span> 3. Normal/Gaussian distribution<a href="normalgaussian-distribution.html#normalgaussian-distribution" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>The normal or the Gaussian distribution is given as:</p>
<p><span class="math display">\[P(x) = \frac{1}{{\sigma \sqrt {2\pi } }} \cdot e ^ \frac{-(x- \mu)^2}{{2\sigma ^2 }} \]</span> where <span class="math inline">\(\mu\)</span> is the mean of the distribution and <span class="math inline">\(\sigma\)</span> is the standard deviation.</p>
<p>The <strong>standard normal distribution</strong> is a special case of <strong>normal distribution</strong> where the values for <span class="math inline">\(\mu = 0\)</span> and <span class="math inline">\(\sigma = 1\)</span>. Thus, the above equation for the Normal distribution simplifies to:</p>
<p><span class="math display">\[P(x) = \frac{1}{{\sqrt {2\pi } }} \cdot e ^ \frac{-x^2}{2} \]</span> Now for any <span class="math inline">\(x\)</span> we can easily solve this equation since <span class="math inline">\(\pi\)</span> and <span class="math inline">\(e\)</span> are known constants.</p>
<hr />
<div id="visualization" class="section level3 hasAnchor" number="4.0.1">
<h3><span class="header-section-number">4.0.1</span> 3.1 Visualization<a href="normalgaussian-distribution.html#visualization" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Let’s generate three random normal distributions with different means and standard deviations and visualize them together</p>
<div class="sourceCode" id="cb160"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb160-1"><a href="normalgaussian-distribution.html#cb160-1" tabindex="-1"></a><span class="do">## Use the function `dnorm()` to plot the density distribution</span></span>
<span id="cb160-2"><a href="normalgaussian-distribution.html#cb160-2" tabindex="-1"></a>x <span class="ot">=</span> <span class="fu">seq</span>(<span class="sc">-</span><span class="dv">10</span>, <span class="dv">30</span>, <span class="at">by=</span>.<span class="dv">1</span>)</span>
<span id="cb160-3"><a href="normalgaussian-distribution.html#cb160-3" tabindex="-1"></a>d1 <span class="ot">=</span> <span class="fu">dnorm</span>(x, <span class="at">mean=</span><span class="dv">0</span>, <span class="at">sd=</span><span class="dv">1</span>)</span>
<span id="cb160-4"><a href="normalgaussian-distribution.html#cb160-4" tabindex="-1"></a>d2 <span class="ot">=</span> <span class="fu">dnorm</span>(x, <span class="at">mean=</span><span class="dv">10</span>, <span class="at">sd=</span><span class="dv">1</span>)</span>
<span id="cb160-5"><a href="normalgaussian-distribution.html#cb160-5" tabindex="-1"></a>d3 <span class="ot">=</span> <span class="fu">dnorm</span>(x, <span class="at">mean=</span><span class="dv">20</span>, <span class="at">sd=</span><span class="dv">1</span>)</span>
<span id="cb160-6"><a href="normalgaussian-distribution.html#cb160-6" tabindex="-1"></a></span>
<span id="cb160-7"><a href="normalgaussian-distribution.html#cb160-7" tabindex="-1"></a><span class="co"># Compare with the histogram build from 1000 random number drawn from the standard normal distribution</span></span>
<span id="cb160-8"><a href="normalgaussian-distribution.html#cb160-8" tabindex="-1"></a>r1 <span class="ot">=</span> <span class="fu">rnorm</span>(<span class="at">n=</span><span class="dv">1000</span>, <span class="at">mean=</span><span class="dv">0</span>, <span class="at">sd =</span><span class="dv">1</span>)  <span class="co"># random distributions of values</span></span>
<span id="cb160-9"><a href="normalgaussian-distribution.html#cb160-9" tabindex="-1"></a>r2 <span class="ot">=</span> <span class="fu">rnorm</span>(<span class="at">n=</span><span class="dv">1000</span>, <span class="at">mean=</span><span class="dv">10</span>, <span class="at">sd =</span><span class="dv">1</span>)</span>
<span id="cb160-10"><a href="normalgaussian-distribution.html#cb160-10" tabindex="-1"></a>r3 <span class="ot">=</span> <span class="fu">rnorm</span>(<span class="at">n=</span><span class="dv">1000</span>, <span class="at">mean=</span><span class="dv">20</span>, <span class="at">sd =</span><span class="dv">1</span>)</span>
<span id="cb160-11"><a href="normalgaussian-distribution.html#cb160-11" tabindex="-1"></a></span>
<span id="cb160-12"><a href="normalgaussian-distribution.html#cb160-12" tabindex="-1"></a><span class="co"># Histogram visualization</span></span>
<span id="cb160-13"><a href="normalgaussian-distribution.html#cb160-13" tabindex="-1"></a><span class="fu">hist</span>(r1,<span class="at">breaks=</span><span class="fu">seq</span>(<span class="sc">-</span><span class="dv">10</span>,<span class="dv">30</span>,<span class="at">by=</span>.<span class="dv">5</span>),<span class="at">probability =</span> <span class="cn">TRUE</span>,<span class="at">ylim=</span><span class="fu">c</span>(<span class="dv">0</span>,.<span class="dv">5</span>),<span class="at">main=</span><span class="st">&quot;&quot;</span>,<span class="at">xlab=</span><span class="st">&quot;&quot;</span>)</span>
<span id="cb160-14"><a href="normalgaussian-distribution.html#cb160-14" tabindex="-1"></a><span class="fu">hist</span>(r2,<span class="at">breaks=</span><span class="fu">seq</span>(<span class="sc">-</span><span class="dv">10</span>,<span class="dv">30</span>,<span class="at">by=</span>.<span class="dv">5</span>),<span class="at">probability =</span> <span class="cn">TRUE</span>,<span class="at">add=</span><span class="cn">TRUE</span>)</span>
<span id="cb160-15"><a href="normalgaussian-distribution.html#cb160-15" tabindex="-1"></a><span class="fu">hist</span>(r3,<span class="at">breaks=</span><span class="fu">seq</span>(<span class="sc">-</span><span class="dv">10</span>,<span class="dv">30</span>,<span class="at">by=</span>.<span class="dv">5</span>),<span class="at">probability =</span> <span class="cn">TRUE</span>,<span class="at">add=</span><span class="cn">TRUE</span>)</span>
<span id="cb160-16"><a href="normalgaussian-distribution.html#cb160-16" tabindex="-1"></a></span>
<span id="cb160-17"><a href="normalgaussian-distribution.html#cb160-17" tabindex="-1"></a><span class="co"># Line with the density distribution as above</span></span>
<span id="cb160-18"><a href="normalgaussian-distribution.html#cb160-18" tabindex="-1"></a><span class="fu">lines</span>(x,d1,<span class="at">col=</span><span class="st">&#39;red&#39;</span>,<span class="at">type=</span><span class="st">&#39;l&#39;</span>,<span class="at">lwd=</span><span class="dv">2</span>)</span>
<span id="cb160-19"><a href="normalgaussian-distribution.html#cb160-19" tabindex="-1"></a><span class="fu">lines</span>(x,d2,<span class="at">col=</span><span class="st">&#39;blue&#39;</span>,<span class="at">lwd=</span><span class="dv">2</span>)</span>
<span id="cb160-20"><a href="normalgaussian-distribution.html#cb160-20" tabindex="-1"></a><span class="fu">lines</span>(x,d3,<span class="at">col=</span><span class="st">&#39;green&#39;</span>,<span class="at">lwd=</span><span class="dv">2</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-88-1.png" width="672" /></p>
<blockquote>
<p>Play with the mean and sd parameters and visualize the distributions (plain lines) as well as the corresponding histograms. What happens when you remove the parameter <code>probability=TRUE</code> in the histogram function?</p>
</blockquote>
<hr />
</div>
<div id="application-on-a-real-dataset" class="section level3 hasAnchor" number="4.0.2">
<h3><span class="header-section-number">4.0.2</span> 3.2 Application on a real dataset<a href="normalgaussian-distribution.html#application-on-a-real-dataset" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Now we will use the Normal distribution to make predictions about gene expression of TP53 in lung cancer. TP53 is the most commonly mutated gene across multiple cancer types especially in lung cancers. We will read a table containing measurements of TP53 expression levels in 586 patients.</p>
<div class="sourceCode" id="cb161"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb161-1"><a href="normalgaussian-distribution.html#cb161-1" tabindex="-1"></a>tp53.exp <span class="ot">=</span> <span class="fu">read.table</span>(<span class="st">&quot;https://www.dropbox.com/s/rwopdr8ycmdg8bd/TP53_expression_LungAdeno.txt?dl=1&quot;</span>, <span class="at">header=</span>T, <span class="at">sep=</span><span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>)[,<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]</span>
<span id="cb161-2"><a href="normalgaussian-distribution.html#cb161-2" tabindex="-1"></a><span class="fu">summary</span>(tp53.exp)</span></code></pre></div>
<pre><code>##    Samples          TP53_expression 
##  Length:586         Min.   :  92.6  
##  Class :character   1st Qu.: 911.8  
##  Mode  :character   Median :1313.3  
##                     Mean   :1380.8  
##                     3rd Qu.:1778.1  
##                     Max.   :4703.9  
##                     NA&#39;s   :69</code></pre>
<div id="data-cleaning-and-central-values" class="section level4 hasAnchor" number="4.0.2.1">
<h4><span class="header-section-number">4.0.2.1</span> 3.2.1 Data cleaning and central values<a href="normalgaussian-distribution.html#data-cleaning-and-central-values" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>We will remove all the missing values and calculate the mean and standard deviation for the TP53 gene expression.</p>
<div class="sourceCode" id="cb163"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb163-1"><a href="normalgaussian-distribution.html#cb163-1" tabindex="-1"></a>tp53.exp <span class="ot">=</span> tp53.exp[<span class="sc">!</span><span class="fu">is.na</span>(tp53.exp<span class="sc">$</span>TP53_expression),] <span class="co"># ! means NOT. Here, only if is.na(xx) is TRUE, !is.na(xx) will be FALSE and not included.</span></span>
<span id="cb163-2"><a href="normalgaussian-distribution.html#cb163-2" tabindex="-1"></a>m.tp53 <span class="ot">=</span> <span class="fu">mean</span>(tp53.exp<span class="sc">$</span>TP53_expression) <span class="co"># mean</span></span>
<span id="cb163-3"><a href="normalgaussian-distribution.html#cb163-3" tabindex="-1"></a>s.tp53 <span class="ot">=</span> <span class="fu">sd</span>(tp53.exp<span class="sc">$</span>TP53_expression) <span class="co"># standard deviation</span></span>
<span id="cb163-4"><a href="normalgaussian-distribution.html#cb163-4" tabindex="-1"></a>m.tp53;s.tp53</span></code></pre></div>
<pre><code>## [1] 1380.822</code></pre>
<pre><code>## [1] 719.5934</code></pre>
</div>
<div id="modeling-of-the-data-using-a-normal-distribution" class="section level4 hasAnchor" number="4.0.2.2">
<h4><span class="header-section-number">4.0.2.2</span> 3.2.2 Modeling of the data using a normal distribution<a href="normalgaussian-distribution.html#modeling-of-the-data-using-a-normal-distribution" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Let’s see how well a normal distribution with <span class="math inline">\(\mu = 1380.822\)</span> (m.tp53) and <span class="math inline">\(\sigma = 719.5934\)</span> (s.tp53) can approximate the <strong>real</strong> distribution of TP53 expression. We assume that the population mean and standard deviation is similar as calculated above since we cannot measure the expression of TP53 in each and every lung cancer patient in the world.</p>
<div class="sourceCode" id="cb166"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb166-1"><a href="normalgaussian-distribution.html#cb166-1" tabindex="-1"></a><span class="co"># distribution of the measured data</span></span>
<span id="cb166-2"><a href="normalgaussian-distribution.html#cb166-2" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">density</span>(tp53.exp<span class="sc">$</span>TP53_expression), <span class="at">main =</span><span class="st">&quot;&quot;</span>, <span class="at">xlim=</span><span class="fu">c</span>(<span class="sc">-</span><span class="dv">1500</span>, <span class="dv">6000</span>),<span class="at">lwd=</span><span class="dv">3</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-91-1.png" width="672" /></p>
<div class="sourceCode" id="cb167"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb167-1"><a href="normalgaussian-distribution.html#cb167-1" tabindex="-1"></a><span class="co"># Make a normal distribution with the above parameters</span></span>
<span id="cb167-2"><a href="normalgaussian-distribution.html#cb167-2" tabindex="-1"></a>x <span class="ot">=</span> <span class="fu">seq</span>(<span class="dv">0</span>,<span class="dv">5000</span>,<span class="at">by=</span><span class="dv">5</span>)</span>
<span id="cb167-3"><a href="normalgaussian-distribution.html#cb167-3" tabindex="-1"></a>d.pred <span class="ot">=</span> <span class="fu">dnorm</span>(x, <span class="at">mean =</span> m.tp53, <span class="at">sd =</span> s.tp53)</span>
<span id="cb167-4"><a href="normalgaussian-distribution.html#cb167-4" tabindex="-1"></a></span>
<span id="cb167-5"><a href="normalgaussian-distribution.html#cb167-5" tabindex="-1"></a><span class="co"># Now plot both, predicted and measured data</span></span>
<span id="cb167-6"><a href="normalgaussian-distribution.html#cb167-6" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">density</span>(tp53.exp<span class="sc">$</span>TP53_expression), <span class="at">main =</span><span class="st">&quot;&quot;</span>, <span class="at">xlim=</span><span class="fu">c</span>(<span class="sc">-</span><span class="dv">1500</span>, <span class="dv">6000</span>),<span class="at">lwd=</span><span class="dv">3</span>);<span class="fu">lines</span>(x,d.pred, <span class="at">col=</span><span class="st">&quot;red&quot;</span>,<span class="at">lwd=</span><span class="dv">3</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-92-1.png" width="672" /></p>
</div>
<div id="data-prediction-using-the-model-normal-distribution" class="section level4 hasAnchor" number="4.0.2.3">
<h4><span class="header-section-number">4.0.2.3</span> 3.2.3 Data prediction using the model (normal distribution)<a href="normalgaussian-distribution.html#data-prediction-using-the-model-normal-distribution" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Using a normal distribution with <span class="math inline">\(\mu = 1380.822\)</span> (m.tp53) and <span class="math inline">\(\sigma = 719.5934\)</span> (s.tp53), we will ask the following questions -</p>
<p>– <strong>(Q1)</strong> What is the probability of observing the expression of TP53 to be <strong>less</strong> than 1000?</p>
<p>– <strong>(Q2)</strong> What is the probability of observing the expression of TP53 to be <strong>greater</strong> than 1000?</p>
<div class="sourceCode" id="cb168"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb168-1"><a href="normalgaussian-distribution.html#cb168-1" tabindex="-1"></a><span class="co"># Q1</span></span>
<span id="cb168-2"><a href="normalgaussian-distribution.html#cb168-2" tabindex="-1"></a><span class="fu">pnorm</span>(<span class="at">q=</span><span class="dv">1000</span>, <span class="at">mean =</span>m.tp53, <span class="at">sd =</span> s.tp53) <span class="co"># returns the cumulative probability</span></span></code></pre></div>
<pre><code>## [1] 0.2983271</code></pre>
<div class="sourceCode" id="cb170"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb170-1"><a href="normalgaussian-distribution.html#cb170-1" tabindex="-1"></a><span class="co"># Q2</span></span>
<span id="cb170-2"><a href="normalgaussian-distribution.html#cb170-2" tabindex="-1"></a><span class="dv">1</span> <span class="sc">-</span> <span class="fu">pnorm</span>(<span class="at">q=</span><span class="dv">1000</span>, <span class="at">mean =</span>m.tp53, <span class="at">sd =</span> s.tp53) </span></code></pre></div>
<pre><code>## [1] 0.7016729</code></pre>
<div class="sourceCode" id="cb172"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb172-1"><a href="normalgaussian-distribution.html#cb172-1" tabindex="-1"></a><span class="co"># is same as</span></span>
<span id="cb172-2"><a href="normalgaussian-distribution.html#cb172-2" tabindex="-1"></a><span class="fu">pnorm</span>(<span class="at">q=</span><span class="dv">1000</span>, <span class="at">mean =</span>m.tp53, <span class="at">sd =</span> s.tp53, <span class="at">lower.tail =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<pre><code>## [1] 0.7016729</code></pre>
<div class="sourceCode" id="cb174"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb174-1"><a href="normalgaussian-distribution.html#cb174-1" tabindex="-1"></a><span class="co"># or pnorm(q=1000, mean =m.tp53, sd = s.tp53, lower.tail = F). FALSE or F is equivalent</span></span></code></pre></div>
</div>
<div id="evaluating-the-quality-of-the-predictions" class="section level4 hasAnchor" number="4.0.2.4">
<h4><span class="header-section-number">4.0.2.4</span> 3.2.4 Evaluating the quality of the predictions<a href="normalgaussian-distribution.html#evaluating-the-quality-of-the-predictions" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Let’s check how good these predictions are compared to <strong>real</strong> data.</p>
<div class="sourceCode" id="cb175"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb175-1"><a href="normalgaussian-distribution.html#cb175-1" tabindex="-1"></a><span class="co"># Q1: What is the probability of observing the expression of TP53 to be **less** than 1000?</span></span>
<span id="cb175-2"><a href="normalgaussian-distribution.html#cb175-2" tabindex="-1"></a><span class="fu">sum</span>(tp53.exp<span class="sc">$</span>TP53_expression <span class="sc">&lt;</span> <span class="dv">1000</span>)<span class="sc">/</span><span class="fu">nrow</span>(tp53.exp)</span></code></pre></div>
<pre><code>## [1] 0.2978723</code></pre>
<div class="sourceCode" id="cb177"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb177-1"><a href="normalgaussian-distribution.html#cb177-1" tabindex="-1"></a><span class="co"># Q2: What is the probability of observing the expression of TP53 to be **greater** than 1000?</span></span>
<span id="cb177-2"><a href="normalgaussian-distribution.html#cb177-2" tabindex="-1"></a><span class="fu">sum</span>(tp53.exp<span class="sc">$</span>TP53_expression <span class="sc">&gt;</span> <span class="dv">1000</span>)<span class="sc">/</span><span class="fu">nrow</span>(tp53.exp)</span></code></pre></div>
<pre><code>## [1] 0.7021277</code></pre>
<blockquote>
<p>I would say those predictions are pretty good !! Now, let’s try to break this model. Re-execute the code above with different <span class="math inline">\(q\)</span> values <code>q=100, q=500, q=4000, q=4500</code> etc. At what values do you think the model would not perform well. HINT: Look at the tails of the distribution!</p>
</blockquote>
<div class="sourceCode" id="cb179"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb179-1"><a href="normalgaussian-distribution.html#cb179-1" tabindex="-1"></a><span class="co"># What is the probability of observing the expression of TP53 to be less than q?</span></span>
<span id="cb179-2"><a href="normalgaussian-distribution.html#cb179-2" tabindex="-1"></a>q <span class="ot">=</span> <span class="fu">c</span>(<span class="dv">100</span>,<span class="dv">500</span>,<span class="dv">1000</span>,<span class="dv">4000</span>,<span class="dv">4500</span>)</span>
<span id="cb179-3"><a href="normalgaussian-distribution.html#cb179-3" tabindex="-1"></a></span>
<span id="cb179-4"><a href="normalgaussian-distribution.html#cb179-4" tabindex="-1"></a><span class="co"># The function sapply() is used to calculate and store the predicted (a) and real (b) values</span></span>
<span id="cb179-5"><a href="normalgaussian-distribution.html#cb179-5" tabindex="-1"></a>model_mat <span class="ot">=</span><span class="fu">sapply</span>(q,<span class="cf">function</span>(x) {</span>
<span id="cb179-6"><a href="normalgaussian-distribution.html#cb179-6" tabindex="-1"></a>   a <span class="ot">=</span> <span class="fu">pnorm</span>(<span class="at">q=</span>x, <span class="at">mean =</span>m.tp53, <span class="at">sd =</span> s.tp53) </span>
<span id="cb179-7"><a href="normalgaussian-distribution.html#cb179-7" tabindex="-1"></a>   b <span class="ot">=</span> <span class="fu">sum</span>(tp53.exp<span class="sc">$</span>TP53_expression <span class="sc">&lt;</span> x)<span class="sc">/</span><span class="fu">nrow</span>(tp53.exp)</span>
<span id="cb179-8"><a href="normalgaussian-distribution.html#cb179-8" tabindex="-1"></a>   <span class="fu">return</span>(<span class="fu">c</span>(a,b))</span>
<span id="cb179-9"><a href="normalgaussian-distribution.html#cb179-9" tabindex="-1"></a>})</span>
<span id="cb179-10"><a href="normalgaussian-distribution.html#cb179-10" tabindex="-1"></a><span class="co"># Change into a dataframe and give row names</span></span>
<span id="cb179-11"><a href="normalgaussian-distribution.html#cb179-11" tabindex="-1"></a>model_mat <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(model_mat,<span class="at">row.names =</span> <span class="fu">c</span>(<span class="st">&quot;Predicted&quot;</span>,<span class="st">&quot;Measured&quot;</span>))</span>
<span id="cb179-12"><a href="normalgaussian-distribution.html#cb179-12" tabindex="-1"></a><span class="co"># Set column names</span></span>
<span id="cb179-13"><a href="normalgaussian-distribution.html#cb179-13" tabindex="-1"></a><span class="fu">colnames</span>(model_mat) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;q=100&quot;</span>, <span class="st">&quot;q=500&quot;</span>, <span class="st">&quot;q=1000&quot;</span>, <span class="st">&quot;q=4000&quot;</span>, <span class="st">&quot;q=4500&quot;</span>)</span>
<span id="cb179-14"><a href="normalgaussian-distribution.html#cb179-14" tabindex="-1"></a><span class="co"># Display the table (dataframe)</span></span>
<span id="cb179-15"><a href="normalgaussian-distribution.html#cb179-15" tabindex="-1"></a>model_mat</span></code></pre></div>
<pre><code>##                 q=100     q=500    q=1000    q=4000    q=4500
## Predicted 0.037544169 0.1104658 0.2983271 0.9998636 0.9999927
## Measured  0.001934236 0.1044487 0.2978723 0.9922631 0.9961315</code></pre>
<p>Again, using a normal distribution with <span class="math inline">\(\mu = 1380.822\)</span> and <span class="math inline">\(\sigma = 719.5934\)</span>, what if we ask these kinds of questions:</p>
<div class="sourceCode" id="cb181"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb181-1"><a href="normalgaussian-distribution.html#cb181-1" tabindex="-1"></a><span class="co"># Q1: What is the value of TP53 expression at the 10% quantile?</span></span>
<span id="cb181-2"><a href="normalgaussian-distribution.html#cb181-2" tabindex="-1"></a><span class="fu">qnorm</span>(<span class="at">p =</span> <span class="fl">0.1</span>, <span class="at">mean =</span> m.tp53, <span class="at">sd =</span> s.tp53)</span></code></pre></div>
<pre><code>## [1] 458.6258</code></pre>
<div class="sourceCode" id="cb183"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb183-1"><a href="normalgaussian-distribution.html#cb183-1" tabindex="-1"></a><span class="co"># Q2: What is the value of TP53 expression at the 90% quantile?</span></span>
<span id="cb183-2"><a href="normalgaussian-distribution.html#cb183-2" tabindex="-1"></a><span class="fu">qnorm</span>(<span class="at">p =</span> <span class="fl">0.9</span>, <span class="at">mean =</span> m.tp53, <span class="at">sd =</span> s.tp53)</span></code></pre></div>
<pre><code>## [1] 2303.018</code></pre>
<p>Let’s check how good these predictions are compared to our real data.</p>
<div class="sourceCode" id="cb185"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb185-1"><a href="normalgaussian-distribution.html#cb185-1" tabindex="-1"></a><span class="fu">quantile</span>(tp53.exp<span class="sc">$</span>TP53_expression, <span class="at">probs =</span> <span class="fu">c</span>(<span class="fl">0.1</span>, <span class="fl">0.9</span>))</span></code></pre></div>
<pre><code>##       10%       90% 
##  478.2844 2240.1913</code></pre>
<p>Again the predictions are pretty good!</p>
</div>
<div id="graphical-visualization" class="section level4 hasAnchor" number="4.0.2.5">
<h4><span class="header-section-number">4.0.2.5</span> 3.2.5 Graphical visualization<a href="normalgaussian-distribution.html#graphical-visualization" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>We can also visualize all of this on a simple graph:</p>
<div class="sourceCode" id="cb187"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb187-1"><a href="normalgaussian-distribution.html#cb187-1" tabindex="-1"></a><span class="co"># Model prediction:</span></span>
<span id="cb187-2"><a href="normalgaussian-distribution.html#cb187-2" tabindex="-1"></a>x <span class="ot">=</span> <span class="fu">seq</span>(<span class="dv">0</span>,<span class="dv">5000</span>,<span class="at">by=</span><span class="dv">5</span>)</span>
<span id="cb187-3"><a href="normalgaussian-distribution.html#cb187-3" tabindex="-1"></a>d.pred <span class="ot">=</span> <span class="fu">dnorm</span>(x,<span class="at">mean =</span> m.tp53, <span class="at">sd =</span> s.tp53)</span>
<span id="cb187-4"><a href="normalgaussian-distribution.html#cb187-4" tabindex="-1"></a></span>
<span id="cb187-5"><a href="normalgaussian-distribution.html#cb187-5" tabindex="-1"></a><span class="co"># Model and measured data:</span></span>
<span id="cb187-6"><a href="normalgaussian-distribution.html#cb187-6" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">density</span>(tp53.exp<span class="sc">$</span>TP53_expression), <span class="at">main =</span><span class="st">&quot;&quot;</span>, <span class="at">xlim=</span><span class="fu">c</span>(<span class="sc">-</span><span class="dv">1500</span>, <span class="dv">6000</span>),<span class="at">lwd=</span><span class="dv">3</span>);<span class="fu">lines</span>(x,d.pred, <span class="at">col=</span><span class="st">&quot;red&quot;</span>,<span class="at">lwd=</span><span class="dv">3</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-98-1.png" width="672" /></p>
<div class="sourceCode" id="cb188"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb188-1"><a href="normalgaussian-distribution.html#cb188-1" tabindex="-1"></a><span class="co"># Model and measured data and predicted versus measured 0.1 and 0.9 quantiles:</span></span>
<span id="cb188-2"><a href="normalgaussian-distribution.html#cb188-2" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">density</span>(tp53.exp<span class="sc">$</span>TP53_expression), <span class="at">main =</span><span class="st">&quot;&quot;</span>, <span class="at">xlim=</span><span class="fu">c</span>(<span class="sc">-</span><span class="dv">1500</span>, <span class="dv">6000</span>),<span class="at">lwd=</span><span class="dv">3</span>);<span class="fu">lines</span>(x,d.pred, <span class="at">col=</span><span class="st">&quot;red&quot;</span>,<span class="at">lwd=</span><span class="dv">3</span>);<span class="fu">abline</span>(<span class="at">v=</span><span class="fu">quantile</span>(tp53.exp<span class="sc">$</span>TP53_expression, <span class="at">probs =</span> <span class="fu">c</span>(<span class="fl">0.1</span>, <span class="fl">0.9</span>)));<span class="fu">abline</span>(<span class="at">v=</span> <span class="fu">c</span>(<span class="fu">qnorm</span>(<span class="at">p =</span> <span class="fl">0.1</span>, <span class="at">mean =</span> m.tp53, <span class="at">sd =</span> s.tp53), <span class="fu">qnorm</span>(<span class="at">p =</span> <span class="fl">0.9</span>, <span class="at">mean =</span> m.tp53, <span class="at">sd =</span> s.tp53)),<span class="at">lty=</span><span class="dv">3</span>,<span class="at">lwd=</span><span class="dv">2</span>,<span class="at">col=</span><span class="st">&#39;red&#39;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-98-2.png" width="672" /></p>
<div class="sourceCode" id="cb189"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb189-1"><a href="normalgaussian-distribution.html#cb189-1" tabindex="-1"></a><span class="co"># Appreciate the quality of the predictions! Compare the black and red vertical lines.</span></span></code></pre></div>
<blockquote>
<p>Re-execute the code above with <code>p=0.25, p=0.5, p=0.75 etc</code> and check how good the predictions are.</p>
</blockquote>
</div>
<div id="graphical-visualization-using-a-q-q-plot" class="section level4 hasAnchor" number="4.0.2.6">
<h4><span class="header-section-number">4.0.2.6</span> 3.2.6 Graphical visualization using a Q-Q plot<a href="normalgaussian-distribution.html#graphical-visualization-using-a-q-q-plot" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Now, let’s plot the sample quantiles against theoretical quantiles to check the similarity between the two. This is called a quantile - quantile plot or a Q-Q plot, which you are familiar with (see Exercises Sheet 1).</p>
<div class="sourceCode" id="cb190"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb190-1"><a href="normalgaussian-distribution.html#cb190-1" tabindex="-1"></a>q <span class="ot">=</span> <span class="fu">seq</span>(<span class="dv">0</span>,<span class="dv">1</span>,<span class="fl">0.01</span>) <span class="co"># Creating a vector of quantiles</span></span>
<span id="cb190-2"><a href="normalgaussian-distribution.html#cb190-2" tabindex="-1"></a></span>
<span id="cb190-3"><a href="normalgaussian-distribution.html#cb190-3" tabindex="-1"></a><span class="co"># Find values corresponding to these quantiles in the real data</span></span>
<span id="cb190-4"><a href="normalgaussian-distribution.html#cb190-4" tabindex="-1"></a>q.observed <span class="ot">=</span> <span class="fu">quantile</span>(tp53.exp<span class="sc">$</span>TP53_expression, <span class="at">probs =</span> q)</span>
<span id="cb190-5"><a href="normalgaussian-distribution.html#cb190-5" tabindex="-1"></a></span>
<span id="cb190-6"><a href="normalgaussian-distribution.html#cb190-6" tabindex="-1"></a><span class="co"># Find values corresponding to these quantiles in the theoretical normal distribution</span></span>
<span id="cb190-7"><a href="normalgaussian-distribution.html#cb190-7" tabindex="-1"></a>q.theoretical <span class="ot">=</span> <span class="fu">qnorm</span>(<span class="at">p =</span> q, <span class="at">mean =</span> m.tp53, <span class="at">sd =</span> s.tp53)</span>
<span id="cb190-8"><a href="normalgaussian-distribution.html#cb190-8" tabindex="-1"></a></span>
<span id="cb190-9"><a href="normalgaussian-distribution.html#cb190-9" tabindex="-1"></a><span class="co"># Correlate the above two values</span></span>
<span id="cb190-10"><a href="normalgaussian-distribution.html#cb190-10" tabindex="-1"></a><span class="fu">plot</span>(<span class="at">x=</span>q.theoretical, <span class="at">y=</span>q.observed);<span class="fu">abline</span>(<span class="dv">0</span>,<span class="dv">1</span>,<span class="at">lty=</span><span class="dv">1</span>,<span class="at">lwd=</span><span class="dv">2</span>,<span class="at">col=</span><span class="st">&#39;red&#39;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-99-1.png" width="672" /></p>
<hr />
</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="week-1-probability-distributions.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="exercises-2.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/USERNAME/REPO/edit/BRANCH/03-ProbDist.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["_main.pdf", "_main.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
