<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>1 WEEK ONE | Bioinformatics introductory course - MSc Molekulare Biotechnologie</title>
  <meta name="description" content="RMarkdowns for the Data Analysis introduction course (SoSe2024), MoBi MA" />
  <meta name="generator" content="bookdown 0.37 and GitBook 2.6.7" />

  <meta property="og:title" content="1 WEEK ONE | Bioinformatics introductory course - MSc Molekulare Biotechnologie" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="RMarkdowns for the Data Analysis introduction course (SoSe2024), MoBi MA" />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="1 WEEK ONE | Bioinformatics introductory course - MSc Molekulare Biotechnologie" />
  
  <meta name="twitter:description" content="RMarkdowns for the Data Analysis introduction course (SoSe2024), MoBi MA" />
  

<meta name="author" content="Biomedical Genomics group, IPMB" />


<meta name="date" content="2024-03-08" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="index.html"/>
<link rel="next" href="week-two.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Bioinformatics Introduction course (MoBi MA)</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Introduction</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#where-to-start"><i class="fa fa-check"></i>Where to start</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#installing-r-and-rstudio"><i class="fa fa-check"></i>Installing R and RStudio</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#courses"><i class="fa fa-check"></i>Courses</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#good-practice"><i class="fa fa-check"></i>Good practice</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#r-markdown"><i class="fa fa-check"></i>R Markdown</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#r-basics"><i class="fa fa-check"></i>R Basics</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#variable-assignment-in-r"><i class="fa fa-check"></i>Variable assignment in R</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#classes"><i class="fa fa-check"></i>Classes</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#logical-operations"><i class="fa fa-check"></i>Logical operations</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#vectors"><i class="fa fa-check"></i>Vectors</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#matrixes"><i class="fa fa-check"></i>Matrixes</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#dataframes"><i class="fa fa-check"></i>Dataframes</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#apply-sapply-lapply"><i class="fa fa-check"></i>Apply, sapply, lapply</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#for-loops"><i class="fa fa-check"></i>For Loops</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#if-statements"><i class="fa fa-check"></i>If Statements</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#data-wrangling-with-tidyverse"><i class="fa fa-check"></i>Data wrangling with tidyverse</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#introducing-tidyverse"><i class="fa fa-check"></i>Introducing tidyverse</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#filter-based-on-conditions"><i class="fa fa-check"></i>Filter based on conditions</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#select-specific-columns"><i class="fa fa-check"></i>Select specific columns</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#operations-on-groups"><i class="fa fa-check"></i>Operations on groups</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#reshaping-data"><i class="fa fa-check"></i>Reshaping data</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#plotting-using-ggplot2"><i class="fa fa-check"></i>Plotting using ggplot2</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#scatter-plot"><i class="fa fa-check"></i>Scatter plot</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#histogram"><i class="fa fa-check"></i>Histogram</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#boxplot"><i class="fa fa-check"></i>Boxplot</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#common-errors"><i class="fa fa-check"></i>Common errors</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#introductory-exercises"><i class="fa fa-check"></i>Introductory exercises</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#exercise-1-vectors"><i class="fa fa-check"></i>Exercise 1: Vectors</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#exercise-2-matrices"><i class="fa fa-check"></i>Exercise 2: Matrices</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#exercise-3-data-frames"><i class="fa fa-check"></i>Exercise 3: Data Frames</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="1" data-path="week-one.html"><a href="week-one.html"><i class="fa fa-check"></i><b>1</b> WEEK ONE</a>
<ul>
<li class="chapter" data-level="1.1" data-path="week-one.html"><a href="week-one.html#day-1-descriptive-statistics-and-data-types"><i class="fa fa-check"></i><b>1.1</b> Day 1: Descriptive statistics and data types</a>
<ul>
<li class="chapter" data-level="1.1.1" data-path="week-one.html"><a href="week-one.html#data-features-and-where-to-find-them"><i class="fa fa-check"></i><b>1.1.1</b> Data features and where to find them</a></li>
<li class="chapter" data-level="1.1.2" data-path="week-one.html"><a href="week-one.html#data-cleaning"><i class="fa fa-check"></i><b>1.1.2</b> Data cleaning</a></li>
<li class="chapter" data-level="1.1.3" data-path="week-one.html"><a href="week-one.html#visualizing-data-distribution"><i class="fa fa-check"></i><b>1.1.3</b> Visualizing data distribution</a></li>
<li class="chapter" data-level="1.1.4" data-path="week-one.html"><a href="week-one.html#correlation"><i class="fa fa-check"></i><b>1.1.4</b> Correlation</a></li>
<li class="chapter" data-level="1.1.5" data-path="week-one.html"><a href="week-one.html#exercises"><i class="fa fa-check"></i><b>1.1.5</b> EXERCISES</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="week-one.html"><a href="week-one.html#day-2-dimensionality-reduction-and-unsupervised-learning"><i class="fa fa-check"></i><b>1.2</b> Day 2: Dimensionality reduction and unsupervised learning</a>
<ul>
<li class="chapter" data-level="1.2.1" data-path="week-one.html"><a href="week-one.html#preparing-the-data"><i class="fa fa-check"></i><b>1.2.1</b> Preparing the data</a></li>
<li class="chapter" data-level="1.2.2" data-path="week-one.html"><a href="week-one.html#k-means-clustering"><i class="fa fa-check"></i><b>1.2.2</b> k-means clustering</a></li>
<li class="chapter" data-level="1.2.3" data-path="week-one.html"><a href="week-one.html#hierarchical-clustering"><i class="fa fa-check"></i><b>1.2.3</b> Hierarchical clustering</a></li>
<li class="chapter" data-level="1.2.4" data-path="week-one.html"><a href="week-one.html#principal-component-analysis"><i class="fa fa-check"></i><b>1.2.4</b> Principal component analysis</a></li>
<li class="chapter" data-level="1.2.5" data-path="week-one.html"><a href="week-one.html#exercises-1"><i class="fa fa-check"></i><b>1.2.5</b> EXERCISES</a></li>
</ul></li>
<li class="chapter" data-level="1.3" data-path="week-one.html"><a href="week-one.html#day-3-probability-distributions"><i class="fa fa-check"></i><b>1.3</b> Day 3: Probability distributions</a>
<ul>
<li class="chapter" data-level="1.3.1" data-path="week-one.html"><a href="week-one.html#probability-distributions"><i class="fa fa-check"></i><b>1.3.1</b> Probability distributions</a></li>
<li class="chapter" data-level="1.3.2" data-path="week-one.html"><a href="week-one.html#confidence-interval"><i class="fa fa-check"></i><b>1.3.2</b> Confidence interval</a></li>
<li class="chapter" data-level="1.3.3" data-path="week-one.html"><a href="week-one.html#exercises-2"><i class="fa fa-check"></i><b>1.3.3</b> EXERCISES</a></li>
</ul></li>
<li class="chapter" data-level="1.4" data-path="week-one.html"><a href="week-one.html#day-4-hypothesis-testing"><i class="fa fa-check"></i><b>1.4</b> Day 4: Hypothesis testing</a>
<ul>
<li class="chapter" data-level="1.4.1" data-path="week-one.html"><a href="week-one.html#tests-of-mean"><i class="fa fa-check"></i><b>1.4.1</b> Tests of mean</a></li>
<li class="chapter" data-level="1.4.2" data-path="week-one.html"><a href="week-one.html#proportion-tests"><i class="fa fa-check"></i><b>1.4.2</b> Proportion tests</a></li>
<li class="chapter" data-level="1.4.3" data-path="week-one.html"><a href="week-one.html#exercises-3"><i class="fa fa-check"></i><b>1.4.3</b> EXERCISES</a></li>
</ul></li>
<li class="chapter" data-level="1.5" data-path="week-one.html"><a href="week-one.html#day-5-multiple-testing-and-regression"><i class="fa fa-check"></i><b>1.5</b> Day 5: Multiple testing and regression</a>
<ul>
<li class="chapter" data-level="1.5.1" data-path="week-one.html"><a href="week-one.html#multiple-testing"><i class="fa fa-check"></i><b>1.5.1</b> Multiple testing</a></li>
<li class="chapter" data-level="1.5.2" data-path="week-one.html"><a href="week-one.html#regression"><i class="fa fa-check"></i><b>1.5.2</b> Regression</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="week-two.html"><a href="week-two.html"><i class="fa fa-check"></i><b>2</b> WEEK TWO</a>
<ul>
<li class="chapter" data-level="2.1" data-path="week-two.html"><a href="week-two.html#days-1-to-3-annotathon"><i class="fa fa-check"></i><b>2.1</b> Days 1 to 3: Annotathon!</a></li>
<li class="chapter" data-level="2.2" data-path="week-two.html"><a href="week-two.html#day-4-data-formats-and-where-to-find-them"><i class="fa fa-check"></i><b>2.2</b> Day 4: Data formats and where to find them</a></li>
<li class="chapter" data-level="2.3" data-path="week-two.html"><a href="week-two.html#day-5-rna-seq---from-fastq-to-count-matrix"><i class="fa fa-check"></i><b>2.3</b> Day 5: RNA-seq - from FASTQ to count matrix</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="project-report-guide.html"><a href="project-report-guide.html"><i class="fa fa-check"></i><b>3</b> Project report guide</a>
<ul>
<li class="chapter" data-level="3.1" data-path="project-report-guide.html"><a href="project-report-guide.html#knitting"><i class="fa fa-check"></i><b>3.1</b> Knitting</a></li>
<li class="chapter" data-level="3.2" data-path="project-report-guide.html"><a href="project-report-guide.html#formatting"><i class="fa fa-check"></i><b>3.2</b> Formatting</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="project-report-guide.html"><a href="project-report-guide.html#new-pages-and-paragraphs"><i class="fa fa-check"></i><b>3.2.1</b> New pages and paragraphs</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="project-report-guide.html"><a href="project-report-guide.html#chunk-options"><i class="fa fa-check"></i><b>3.3</b> Chunk options</a></li>
<li class="chapter" data-level="3.4" data-path="project-report-guide.html"><a href="project-report-guide.html#external-images"><i class="fa fa-check"></i><b>3.4</b> External images</a></li>
<li class="chapter" data-level="3.5" data-path="project-report-guide.html"><a href="project-report-guide.html#references"><i class="fa fa-check"></i><b>3.5</b> References</a>
<ul>
<li class="chapter" data-level="3.5.1" data-path="project-report-guide.html"><a href="project-report-guide.html#citations"><i class="fa fa-check"></i><b>3.5.1</b> Citations</a></li>
<li class="chapter" data-level="3.5.2" data-path="project-report-guide.html"><a href="project-report-guide.html#footnotes"><i class="fa fa-check"></i><b>3.5.2</b> Footnotes</a></li>
</ul></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Bioinformatics introductory course - MSc Molekulare Biotechnologie</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="week-one" class="section level1 hasAnchor" number="1">
<h1><span class="header-section-number">1</span> WEEK ONE<a href="week-one.html#week-one" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<div id="day-1-descriptive-statistics-and-data-types" class="section level2 hasAnchor" number="1.1">
<h2><span class="header-section-number">1.1</span> Day 1: Descriptive statistics and data types<a href="week-one.html#day-1-descriptive-statistics-and-data-types" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Today, you will learn how to perform basic tasks on a dataframe/tibble, descriptive statistics, perform data cleaning, and plotting.</p>
<div id="data-features-and-where-to-find-them" class="section level3 hasAnchor" number="1.1.1">
<h3><span class="header-section-number">1.1.1</span> Data features and where to find them<a href="week-one.html#data-features-and-where-to-find-them" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="load-the-data" class="section level4 hasAnchor" number="1.1.1.1">
<h4><span class="header-section-number">1.1.1.1</span> Load the data<a href="week-one.html#load-the-data" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>The diabetes dataset, which we will be using in this practical class will be downloaded from an online repository.
We will load that into R and have a sneak peek into how it looks like with the console.
In the following you will see several functions that give us information about our dataset.</p>
<div class="sourceCode" id="cb72"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb72-1"><a href="week-one.html#cb72-1" tabindex="-1"></a>dat <span class="ot">=</span> <span class="fu">as_tibble</span>(<span class="fu">read.delim</span>(<span class="st">&#39;https://tinyurl.com/y4fark9g&#39;</span>)) <span class="co"># Load the dataset</span></span>
<span id="cb72-2"><a href="week-one.html#cb72-2" tabindex="-1"></a><span class="fu">head</span>(dat, <span class="dv">10</span>) <span class="co"># Look at the first 10 lines of the table</span></span></code></pre></div>
<pre><code>## # A tibble: 10 × 19
##       id  chol stab.glu   hdl ratio glyhb location     age gender height weight
##    &lt;int&gt; &lt;int&gt;    &lt;int&gt; &lt;int&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;      &lt;int&gt; &lt;chr&gt;   &lt;int&gt;  &lt;int&gt;
##  1  1000   203       82    56  3.60  4.31 Buckingham    46 female     62    121
##  2  1001   165       97    24  6.90  4.44 Buckingham    29 female     64    218
##  3  1002   228       92    37  6.20  4.64 Buckingham    58 female     61    256
##  4  1003    78       93    12  6.5   4.63 Buckingham    67 male       67    119
##  5  1005   249       90    28  8.90  7.72 Buckingham    64 male       68    183
##  6  1008   248       94    69  3.60  4.81 Buckingham    34 male       71    190
##  7  1011   195       92    41  4.80  4.84 Buckingham    30 male       69    191
##  8  1015   227       75    44  5.20  3.94 Buckingham    37 male       59    170
##  9  1016   177       87    49  3.60  4.84 Buckingham    45 male       69    166
## 10  1022   263       89    40  6.60  5.78 Buckingham    55 female     63    202
## # ℹ 8 more variables: frame &lt;chr&gt;, bp.1s &lt;int&gt;, bp.1d &lt;int&gt;, bp.2s &lt;int&gt;,
## #   bp.2d &lt;int&gt;, waist &lt;int&gt;, hip &lt;int&gt;, time.ppn &lt;int&gt;</code></pre>
</div>
<div id="dimensions-and-naming" class="section level4 hasAnchor" number="1.1.1.2">
<h4><span class="header-section-number">1.1.1.2</span> Dimensions and naming<a href="week-one.html#dimensions-and-naming" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p><strong>1.</strong> What is the dimension of our dataset (i.e. how many rows/columns are there in our data)</p>
<div class="sourceCode" id="cb74"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb74-1"><a href="week-one.html#cb74-1" tabindex="-1"></a><span class="co"># Dimension</span></span>
<span id="cb74-2"><a href="week-one.html#cb74-2" tabindex="-1"></a><span class="fu">dim</span>(dat)</span></code></pre></div>
<pre><code>## [1] 403  19</code></pre>
<div class="sourceCode" id="cb76"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb76-1"><a href="week-one.html#cb76-1" tabindex="-1"></a><span class="co"># Number of columns</span></span>
<span id="cb76-2"><a href="week-one.html#cb76-2" tabindex="-1"></a><span class="fu">ncol</span>(dat)</span>
<span id="cb76-3"><a href="week-one.html#cb76-3" tabindex="-1"></a><span class="co"># Number of rows</span></span>
<span id="cb76-4"><a href="week-one.html#cb76-4" tabindex="-1"></a><span class="fu">nrow</span>(dat)</span></code></pre></div>
<p><strong>2.</strong> What are the column names of our dataset</p>
<div class="sourceCode" id="cb77"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb77-1"><a href="week-one.html#cb77-1" tabindex="-1"></a><span class="fu">colnames</span>(dat) <span class="co"># Similarly rownames() for rows</span></span></code></pre></div>
<pre><code>##  [1] &quot;id&quot;       &quot;chol&quot;     &quot;stab.glu&quot; &quot;hdl&quot;      &quot;ratio&quot;    &quot;glyhb&quot;   
##  [7] &quot;location&quot; &quot;age&quot;      &quot;gender&quot;   &quot;height&quot;   &quot;weight&quot;   &quot;frame&quot;   
## [13] &quot;bp.1s&quot;    &quot;bp.1d&quot;    &quot;bp.2s&quot;    &quot;bp.2d&quot;    &quot;waist&quot;    &quot;hip&quot;     
## [19] &quot;time.ppn&quot;</code></pre>
<p>Probably you are confused about what these column names mean.
For more description on these values <a href="https://biostat.app.vumc.org/wiki/pub/Main/DataSets/Cdiabetes.html">look here</a></p>
</div>
<div id="numerical-features" class="section level4 hasAnchor" number="1.1.1.3">
<h4><span class="header-section-number">1.1.1.3</span> Numerical features<a href="week-one.html#numerical-features" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p><strong>3.</strong> How do we extract the minimum and maximum age of patients in our dataset?</p>
<div class="sourceCode" id="cb79"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb79-1"><a href="week-one.html#cb79-1" tabindex="-1"></a><span class="fu">min</span>(dat<span class="sc">$</span>age)</span>
<span id="cb79-2"><a href="week-one.html#cb79-2" tabindex="-1"></a><span class="fu">max</span>(dat<span class="sc">$</span>age)</span>
<span id="cb79-3"><a href="week-one.html#cb79-3" tabindex="-1"></a><span class="fu">range</span>(dat<span class="sc">$</span>age)</span></code></pre></div>
<blockquote>
<p>Can you find out the same for height and weight?</p>
</blockquote>
<p><strong>4.</strong> How does the overall summary of our entire dataset look like?</p>
<div class="sourceCode" id="cb80"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb80-1"><a href="week-one.html#cb80-1" tabindex="-1"></a><span class="fu">summary</span>(dat)</span></code></pre></div>
<blockquote>
<p>Can you explain what you see after you run the <code>summary()</code> function?</p>
</blockquote>
<p>Feel free to play around with this syntax until you feel comfortable with it.
You can open a window with <code>View(dat)</code> to compare your results.</p>
<hr />
</div>
</div>
<div id="data-cleaning" class="section level3 hasAnchor" number="1.1.2">
<h3><span class="header-section-number">1.1.2</span> Data cleaning<a href="week-one.html#data-cleaning" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Very often the first thing one needs to do before any data science project is to clean up the raw data and transform it into a format that is readily understood and easy to use for all downstream analysis.
This process usually involves: –</p>
<ul>
<li>Removing empty value rows/columns</li>
<li>Removing unused or unnecessary rows/columns</li>
<li>Reordering the data matrix</li>
<li>Keeping columns uniformly numeric (age, weight etc) or string (names, places etc) or logical (TRUE/FALSE, 1/0)</li>
<li>Handling strange caveats which are data specific like replacing <code>,</code> or <code>.</code>, or <code>;</code> from numbers etc</li>
</ul>
<p>Lets do some clean up of our own diabetes data</p>
<ol style="list-style-type: decimal">
<li>We will make the <code>id</code> column the row names for the dataset;</li>
<li>We will remove the <code>bp.2s</code> and <code>bp.2d</code> columns as it has mostly missing values (see summary above);</li>
<li>We will also remove the column <code>time.ppn</code> which will not be required in our analysis;</li>
<li>We will reorder the columns of the data such that all the qualitative and quantitative values are separated.</li>
</ol>
<p>To perform this cleanup, we need a couple of important functions, that we will first discuss:</p>
<ul>
<li><code>filter</code></li>
<li><code>is.na</code></li>
<li><code>mutate</code></li>
<li><code>accross</code></li>
<li><code>%in%</code></li>
</ul>
<div id="filter" class="section level4 hasAnchor" number="1.1.2.1">
<h4><span class="header-section-number">1.1.2.1</span> filter()<a href="week-one.html#filter" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p><code>filter()</code> is used on a dataset to filter rows satisfying a condition you specify like we saw previously (Introduction).
Let’s look at an example. We are only filtering for senior individuals in our dataset.</p>
<div class="sourceCode" id="cb81"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb81-1"><a href="week-one.html#cb81-1" tabindex="-1"></a>dat_seniors <span class="ot">=</span> dat <span class="sc">%&gt;%</span></span>
<span id="cb81-2"><a href="week-one.html#cb81-2" tabindex="-1"></a>  <span class="fu">filter</span>(age <span class="sc">&lt;=</span> <span class="dv">65</span>)</span></code></pre></div>
<p>We can also filter based on other conditions, like location, sex, among others.
In some cases, we can also use <code>which()</code> to filter values. The syntax is different…</p>
<div class="sourceCode" id="cb82"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb82-1"><a href="week-one.html#cb82-1" tabindex="-1"></a>dat[dat<span class="sc">$</span>age <span class="sc">&lt;=</span> <span class="dv">65</span>,]</span></code></pre></div>
<p>… but it works for vectors and other classes. Let’s see the next example.</p>
<div class="sourceCode" id="cb83"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb83-1"><a href="week-one.html#cb83-1" tabindex="-1"></a><span class="co"># number of animals you have</span></span>
<span id="cb83-2"><a href="week-one.html#cb83-2" tabindex="-1"></a>number <span class="ot">=</span> <span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">4</span>,<span class="dv">5</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">5</span>)</span>
<span id="cb83-3"><a href="week-one.html#cb83-3" tabindex="-1"></a><span class="co"># Let&#39;s create a different vector (of the same length)</span></span>
<span id="cb83-4"><a href="week-one.html#cb83-4" tabindex="-1"></a>animals <span class="ot">=</span> <span class="fu">c</span>(<span class="st">&quot;cat&quot;</span>, <span class="st">&quot;dog&quot;</span>, <span class="st">&quot;cow&quot;</span>, <span class="st">&quot;parrot&quot;</span>, <span class="st">&quot;zebra&quot;</span>, <span class="st">&quot;sparrow&quot;</span>, <span class="st">&quot;lizard&quot;</span>)</span>
<span id="cb83-5"><a href="week-one.html#cb83-5" tabindex="-1"></a><span class="co"># Let&#39;s use the &quot;which()&quot; function now</span></span>
<span id="cb83-6"><a href="week-one.html#cb83-6" tabindex="-1"></a>animals[<span class="fu">which</span>(number <span class="sc">&gt;</span> <span class="dv">2</span>)]</span></code></pre></div>
<pre><code>## [1] &quot;dog&quot;    &quot;cow&quot;    &quot;parrot&quot; &quot;lizard&quot;</code></pre>
<p>We selected all animals from the “animals” vector that correspond to more than three individuals in the “number” vector.</p>
</div>
<div id="is.na" class="section level4 hasAnchor" number="1.1.2.2">
<h4><span class="header-section-number">1.1.2.2</span> is.na()<a href="week-one.html#is.na" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p><code>is.na()</code> is used to determine if NA values are present in a given object. We can try a simple example with one variable being assigned as NA.</p>
<div class="sourceCode" id="cb85"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb85-1"><a href="week-one.html#cb85-1" tabindex="-1"></a>x <span class="ot">=</span> <span class="dv">2</span></span>
<span id="cb85-2"><a href="week-one.html#cb85-2" tabindex="-1"></a><span class="fu">is.na</span>(x)</span></code></pre></div>
<pre><code>## [1] FALSE</code></pre>
<div class="sourceCode" id="cb87"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb87-1"><a href="week-one.html#cb87-1" tabindex="-1"></a>y <span class="ot">=</span> <span class="cn">NA</span></span>
<span id="cb87-2"><a href="week-one.html#cb87-2" tabindex="-1"></a><span class="fu">is.na</span>(y)</span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<p>We can do this with vectors obtained from <code>dat</code>. What class is the output in?</p>
<div class="sourceCode" id="cb89"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb89-1"><a href="week-one.html#cb89-1" tabindex="-1"></a><span class="fu">is.na</span>(dat<span class="sc">$</span>glyhb)</span></code></pre></div>
</div>
<div id="mutate" class="section level4 hasAnchor" number="1.1.2.3">
<h4><span class="header-section-number">1.1.2.3</span> mutate()<a href="week-one.html#mutate" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p><code>mutate()</code> is often used to create a new column based on another column of the dataframe. Let us use this function to <em>mutate</em> two new columns including the weight in kilograms and the height in centimeters. <strong>The conversion from pounds to kilograms can be done by multiplying weight in pounds by 0.454. To covert height to centimeters we only need to multiply height (inches) by 2.54</strong>.</p>
<div class="sourceCode" id="cb90"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb90-1"><a href="week-one.html#cb90-1" tabindex="-1"></a>dat <span class="sc">%&gt;%</span></span>
<span id="cb90-2"><a href="week-one.html#cb90-2" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">weight.kg =</span> weight <span class="sc">*</span> <span class="fl">0.454</span>,        <span class="co"># you can generate both columns using the same mutate!</span></span>
<span id="cb90-3"><a href="week-one.html#cb90-3" tabindex="-1"></a>         <span class="at">height.cm =</span> height <span class="sc">*</span> <span class="fl">2.54</span>) <span class="sc">%&gt;%</span>     <span class="co"># we do not need to save this output</span></span>
<span id="cb90-4"><a href="week-one.html#cb90-4" tabindex="-1"></a>  <span class="fu">select</span>(id, weight.kg, height.cm)</span></code></pre></div>
<pre><code>## # A tibble: 403 × 3
##       id weight.kg height.cm
##    &lt;int&gt;     &lt;dbl&gt;     &lt;dbl&gt;
##  1  1000      54.9      157.
##  2  1001      99.0      163.
##  3  1002     116.       155.
##  4  1003      54.0      170.
##  5  1005      83.1      173.
##  6  1008      86.3      180.
##  7  1011      86.7      175.
##  8  1015      77.2      150.
##  9  1016      75.4      175.
## 10  1022      91.7      160.
## # ℹ 393 more rows</code></pre>
</div>
<div id="accross" class="section level4 hasAnchor" number="1.1.2.4">
<h4><span class="header-section-number">1.1.2.4</span> accross()<a href="week-one.html#accross" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p><code>accross()</code> is very often used together with <code>mutate()</code> and another helper function, like <code>everywhere()</code>, <code>starts_with()</code>, <code>ends_with()</code>, or <code>contains()</code>.
Later, we will use <code>accross()</code> together with the other functions we learned previously to remove NAs like this:</p>
<div class="sourceCode" id="cb92"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb92-1"><a href="week-one.html#cb92-1" tabindex="-1"></a>dat <span class="sc">%&gt;%</span></span>
<span id="cb92-2"><a href="week-one.html#cb92-2" tabindex="-1"></a>  <span class="fu">rowwise</span>() <span class="sc">%&gt;%</span></span>
<span id="cb92-3"><a href="week-one.html#cb92-3" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">na_count =</span> <span class="fu">sum</span>(<span class="fu">is.na</span>(<span class="fu">across</span>(<span class="fu">everything</span>()))))</span></code></pre></div>
<p>There is much to unpack here:</p>
<ul>
<li><code>rowwise()</code> ensures that the next operations are applied by row.<br />
</li>
<li><code>mutate()</code> adds a new column called na_count to the dataframe.<br />
</li>
<li><code>across(everything())</code> selects all columns in the current row.<br />
</li>
<li><code>sum(is.na(...))</code> calculates the sum of missing values for each row.</li>
</ul>
<blockquote>
<p>Try to run the previous example without <code>rowwise()</code>. What does it look like?</p>
</blockquote>
</div>
<div id="in" class="section level4 hasAnchor" number="1.1.2.5">
<h4><span class="header-section-number">1.1.2.5</span> %in%<a href="week-one.html#in" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>This is an operator to check which elements of a first vector are inside a second vector.</p>
<div class="sourceCode" id="cb93"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb93-1"><a href="week-one.html#cb93-1" tabindex="-1"></a><span class="fu">c</span>(<span class="st">&#39;Frodo&#39;</span>,<span class="st">&#39;Sam&#39;</span>) <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">&#39;Pippin&#39;</span>,<span class="st">&#39;Sam&#39;</span>,<span class="st">&#39;Frodo&#39;</span>,<span class="st">&#39;Merry&#39;</span>)</span></code></pre></div>
<pre><code>## [1] TRUE TRUE</code></pre>
</div>
<div id="ready-for-the-cleaning" class="section level4 hasAnchor" number="1.1.2.6">
<h4><span class="header-section-number">1.1.2.6</span> Ready for the cleaning!<a href="week-one.html#ready-for-the-cleaning" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>The first column of the dataframe is the column with the name “id”.
The rows are just numbered, without names.
We are going to rename the rows using the column “id”. The function <code>column_to_rownames()</code> allows us to do this efficiently.</p>
<div class="sourceCode" id="cb95"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb95-1"><a href="week-one.html#cb95-1" tabindex="-1"></a><span class="co"># set the row names using the column id</span></span>
<span id="cb95-2"><a href="week-one.html#cb95-2" tabindex="-1"></a>dat <span class="ot">=</span> dat <span class="sc">%&gt;%</span></span>
<span id="cb95-3"><a href="week-one.html#cb95-3" tabindex="-1"></a>  <span class="fu">column_to_rownames</span>(<span class="at">var =</span> <span class="st">&#39;id&#39;</span>)</span></code></pre></div>
<blockquote>
<p>Keep in mind that rownames must be unique!</p>
</blockquote>
<p>The <em>na_count</em> column will then include the number of NAs per row. Do you understand how it works?
We finally apply <code>filter</code> again to keep only rows with less than or 2 NAs.</p>
<div class="sourceCode" id="cb96"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb96-1"><a href="week-one.html#cb96-1" tabindex="-1"></a>dat <span class="ot">=</span> dat <span class="sc">%&gt;%</span></span>
<span id="cb96-2"><a href="week-one.html#cb96-2" tabindex="-1"></a>  <span class="fu">rowwise</span>() <span class="sc">%&gt;%</span></span>
<span id="cb96-3"><a href="week-one.html#cb96-3" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">na_count =</span> <span class="fu">sum</span>(<span class="fu">is.na</span>(<span class="fu">across</span>(<span class="fu">everything</span>())))) <span class="sc">%&gt;%</span></span>
<span id="cb96-4"><a href="week-one.html#cb96-4" tabindex="-1"></a>  <span class="fu">filter</span>(na_count <span class="sc">&lt;=</span> <span class="dv">2</span>)</span></code></pre></div>
<p>We will also remove the <em>na_count</em> and some problematic columns (<em>bp.2s</em>, <em>bp.2d</em> and <em>time.ppn</em>) by <strong>selecting the ones which are not these</strong>. We can do this using <code>!</code>, as this character can be used to invert results. Let us try it with <code>select()</code>.</p>
<div class="sourceCode" id="cb97"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb97-1"><a href="week-one.html#cb97-1" tabindex="-1"></a>dat <span class="ot">=</span> dat <span class="sc">%&gt;%</span></span>
<span id="cb97-2"><a href="week-one.html#cb97-2" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">!</span><span class="fu">c</span>(na_count, time.ppn, bp<span class="fl">.2</span>d, bp<span class="fl">.2</span>s))</span></code></pre></div>
<p>Next, we can re-order the remaining columns, in order to put the categorical columns first, and numerical columns after. We can use <code>select</code> to order columns too, but we need to combine it with <code>where()</code> and functions which verify the class of the columns, like <code>is.character()</code> or <code>is.numeric()</code>.</p>
<p>Here is a simple example:</p>
<div class="sourceCode" id="cb98"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb98-1"><a href="week-one.html#cb98-1" tabindex="-1"></a><span class="co"># Create a character and numeric </span></span>
<span id="cb98-2"><a href="week-one.html#cb98-2" tabindex="-1"></a>name <span class="ot">=</span> <span class="fu">c</span>(<span class="st">&quot;Antonia&quot;</span>)</span>
<span id="cb98-3"><a href="week-one.html#cb98-3" tabindex="-1"></a>age <span class="ot">=</span> <span class="fu">c</span>(<span class="dv">23</span>)</span>
<span id="cb98-4"><a href="week-one.html#cb98-4" tabindex="-1"></a></span>
<span id="cb98-5"><a href="week-one.html#cb98-5" tabindex="-1"></a><span class="co"># Verify if the previous object are from the character/numeric classes</span></span>
<span id="cb98-6"><a href="week-one.html#cb98-6" tabindex="-1"></a><span class="fu">is.character</span>(name)</span>
<span id="cb98-7"><a href="week-one.html#cb98-7" tabindex="-1"></a><span class="fu">is.character</span>(age)</span>
<span id="cb98-8"><a href="week-one.html#cb98-8" tabindex="-1"></a><span class="fu">is.numeric</span>(age)</span></code></pre></div>
<p>And here we can apply the same principle to the re-ordering:</p>
<div class="sourceCode" id="cb99"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb99-1"><a href="week-one.html#cb99-1" tabindex="-1"></a>dat <span class="ot">&lt;-</span> dat <span class="sc">%&gt;%</span></span>
<span id="cb99-2"><a href="week-one.html#cb99-2" tabindex="-1"></a>  <span class="fu">select</span>(</span>
<span id="cb99-3"><a href="week-one.html#cb99-3" tabindex="-1"></a>    <span class="co"># Select categorical columns</span></span>
<span id="cb99-4"><a href="week-one.html#cb99-4" tabindex="-1"></a>    <span class="fu">where</span>(is.character), </span>
<span id="cb99-5"><a href="week-one.html#cb99-5" tabindex="-1"></a>    <span class="co"># Select numerical columns</span></span>
<span id="cb99-6"><a href="week-one.html#cb99-6" tabindex="-1"></a>    <span class="fu">where</span>(is.numeric)</span>
<span id="cb99-7"><a href="week-one.html#cb99-7" tabindex="-1"></a>  )</span>
<span id="cb99-8"><a href="week-one.html#cb99-8" tabindex="-1"></a></span>
<span id="cb99-9"><a href="week-one.html#cb99-9" tabindex="-1"></a><span class="co"># OR you can use the indexes too, but if you more than 10-20 columns, that is not ideal</span></span>
<span id="cb99-10"><a href="week-one.html#cb99-10" tabindex="-1"></a><span class="co"># dat = dat[,c(8,6,11,9,10,14,15,2,5,1,3,4,12,13)]</span></span></code></pre></div>
<p>Now lets look at our cleaned data:</p>
<div class="sourceCode" id="cb100"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb100-1"><a href="week-one.html#cb100-1" tabindex="-1"></a><span class="fu">summary</span>(dat)</span></code></pre></div>
<pre><code>##    location            gender             frame                chol      
##  Length:377         Length:377         Length:377         Min.   : 78.0  
##  Class :character   Class :character   Class :character   1st Qu.:179.0  
##  Mode  :character   Mode  :character   Mode  :character   Median :204.0  
##                                                           Mean   :208.2  
##                                                           3rd Qu.:230.0  
##                                                           Max.   :443.0  
##                                                                          
##     stab.glu          hdl             ratio            glyhb       
##  Min.   : 48.0   Min.   : 12.00   Min.   : 1.500   Min.   : 2.680  
##  1st Qu.: 81.0   1st Qu.: 38.00   1st Qu.: 3.200   1st Qu.: 4.390  
##  Median : 90.0   Median : 46.00   Median : 4.200   Median : 4.860  
##  Mean   :107.3   Mean   : 50.36   Mean   : 4.538   Mean   : 5.594  
##  3rd Qu.:108.0   3rd Qu.: 59.00   3rd Qu.: 5.400   3rd Qu.: 5.622  
##  Max.   :385.0   Max.   :120.00   Max.   :19.300   Max.   :16.110  
##                                                    NA&#39;s   :3       
##       age           height          weight          bp.1s      
##  Min.   :19.0   Min.   :52.00   Min.   : 99.0   Min.   : 90.0  
##  1st Qu.:34.0   1st Qu.:63.00   1st Qu.:151.0   1st Qu.:122.0  
##  Median :45.0   Median :66.00   Median :174.0   Median :136.0  
##  Mean   :46.9   Mean   :66.02   Mean   :178.1   Mean   :137.4  
##  3rd Qu.:60.0   3rd Qu.:69.00   3rd Qu.:200.0   3rd Qu.:148.0  
##  Max.   :92.0   Max.   :76.00   Max.   :325.0   Max.   :250.0  
##                                                                
##      bp.1d            waist            hip       
##  Min.   : 48.00   Min.   :26.00   Min.   :30.00  
##  1st Qu.: 75.00   1st Qu.:33.00   1st Qu.:39.00  
##  Median : 82.00   Median :37.00   Median :42.00  
##  Mean   : 83.69   Mean   :37.95   Mean   :43.08  
##  3rd Qu.: 92.00   3rd Qu.:41.25   3rd Qu.:46.00  
##  Max.   :124.00   Max.   :56.00   Max.   :64.00  
##                   NA&#39;s   :1       NA&#39;s   :1</code></pre>
<p>Hold up, the ordering and selection of columns looks right, but it seems that there are certain rows that have missing values still (like <code>glyhb</code> column has 3 <code>NA</code> values still).
Lets remove all rows with any missing value using <code>na.omit()</code>.
Remember, 1 row = 1 patient.</p>
<div class="sourceCode" id="cb102"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb102-1"><a href="week-one.html#cb102-1" tabindex="-1"></a>dat <span class="ot">=</span> dat <span class="sc">%&gt;%</span></span>
<span id="cb102-2"><a href="week-one.html#cb102-2" tabindex="-1"></a>  <span class="fu">na.omit</span>()</span></code></pre></div>
<blockquote>
<p>How many patients were removed because they were associated with missing values?</p>
</blockquote>
<p>Now our cleaned data has no missing values, columns are cleanly ordered and each column is in the right format</p>
<div class="sourceCode" id="cb103"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb103-1"><a href="week-one.html#cb103-1" tabindex="-1"></a><span class="fu">summary</span>(dat)</span></code></pre></div>
<pre><code>##    location            gender             frame                chol      
##  Length:367         Length:367         Length:367         Min.   : 78.0  
##  Class :character   Class :character   Class :character   1st Qu.:179.0  
##  Mode  :character   Mode  :character   Mode  :character   Median :204.0  
##                                                           Mean   :207.5  
##                                                           3rd Qu.:229.0  
##                                                           Max.   :443.0  
##     stab.glu          hdl             ratio            glyhb       
##  Min.   : 48.0   Min.   : 12.00   Min.   : 1.500   Min.   : 2.680  
##  1st Qu.: 81.0   1st Qu.: 38.00   1st Qu.: 3.200   1st Qu.: 4.390  
##  Median : 90.0   Median : 46.00   Median : 4.200   Median : 4.860  
##  Mean   :107.3   Mean   : 50.28   Mean   : 4.536   Mean   : 5.602  
##  3rd Qu.:108.0   3rd Qu.: 59.00   3rd Qu.: 5.400   3rd Qu.: 5.630  
##  Max.   :385.0   Max.   :120.00   Max.   :19.300   Max.   :16.110  
##       age            height          weight          bp.1s      
##  Min.   :19.00   Min.   :52.00   Min.   : 99.0   Min.   : 90.0  
##  1st Qu.:34.00   1st Qu.:63.00   1st Qu.:151.0   1st Qu.:121.5  
##  Median :45.00   Median :66.00   Median :174.0   Median :136.0  
##  Mean   :46.68   Mean   :66.05   Mean   :178.1   Mean   :137.1  
##  3rd Qu.:60.00   3rd Qu.:69.00   3rd Qu.:200.0   3rd Qu.:148.0  
##  Max.   :92.00   Max.   :76.00   Max.   :325.0   Max.   :250.0  
##      bp.1d           waist            hip       
##  Min.   : 48.0   Min.   :26.00   Min.   :30.00  
##  1st Qu.: 75.0   1st Qu.:33.00   1st Qu.:39.00  
##  Median : 82.0   Median :37.00   Median :42.00  
##  Mean   : 83.4   Mean   :37.93   Mean   :43.04  
##  3rd Qu.: 92.0   3rd Qu.:41.50   3rd Qu.:46.00  
##  Max.   :124.0   Max.   :56.00   Max.   :64.00</code></pre>
<blockquote>
<p>Can you identify which types of data (continuous, discrete etc) each column above represents and why?</p>
</blockquote>
<hr />
</div>
</div>
<div id="visualizing-data-distribution" class="section level3 hasAnchor" number="1.1.3">
<h3><span class="header-section-number">1.1.3</span> Visualizing data distribution<a href="week-one.html#visualizing-data-distribution" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>In this section you will also learn the essential functions to plot data in an intuitive and useful way using the <code>ggplot2</code> package, just like in the introductory section to tidyverse.</p>
<div id="histograms" class="section level4 hasAnchor" number="1.1.3.1">
<h4><span class="header-section-number">1.1.3.1</span> Histograms<a href="week-one.html#histograms" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>We can plot the column “stab.glu” as a histogram using the <code>hist()</code> function:</p>
<div class="sourceCode" id="cb105"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb105-1"><a href="week-one.html#cb105-1" tabindex="-1"></a><span class="fu">ggplot</span>(dat,</span>
<span id="cb105-2"><a href="week-one.html#cb105-2" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> stab.glu)) <span class="sc">+</span></span>
<span id="cb105-3"><a href="week-one.html#cb105-3" tabindex="-1"></a>  <span class="fu">geom_histogram</span>() <span class="sc">+</span></span>
<span id="cb105-4"><a href="week-one.html#cb105-4" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&quot;Stabilized Glucose concentration in blood&quot;</span>,  <span class="co"># add labels to the x-axis</span></span>
<span id="cb105-5"><a href="week-one.html#cb105-5" tabindex="-1"></a>       <span class="at">title =</span> <span class="st">&quot;Glucose concentration&quot;</span>)                  <span class="co"># add title</span></span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-47-1.png" width="672" /></p>
<blockquote>
<p>Add the parameter <code>bins = 50</code> in the above lines of code (inside <code>geom_histogram</code>) and see what happens.
Try different values for <code>bins</code> like <code>10, 20, 75, 100</code>. Can you interpret the differences?
Is this a good or bad thing about histograms?</p>
</blockquote>
</div>
<div id="density-plots" class="section level4 hasAnchor" number="1.1.3.2">
<h4><span class="header-section-number">1.1.3.2</span> Density plots<a href="week-one.html#density-plots" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>For density plots, we use the <code>geom_density()</code> function to estimate the probability density function for a given variable.</p>
<div class="sourceCode" id="cb106"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb106-1"><a href="week-one.html#cb106-1" tabindex="-1"></a><span class="fu">ggplot</span>(dat,</span>
<span id="cb106-2"><a href="week-one.html#cb106-2" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> stab.glu)) <span class="sc">+</span></span>
<span id="cb106-3"><a href="week-one.html#cb106-3" tabindex="-1"></a>  <span class="fu">geom_density</span>() <span class="sc">+</span></span>
<span id="cb106-4"><a href="week-one.html#cb106-4" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&quot;Stabilized Glucose concentration in blood&quot;</span>,  <span class="co"># add labels to the x-axis</span></span>
<span id="cb106-5"><a href="week-one.html#cb106-5" tabindex="-1"></a>       <span class="at">title =</span> <span class="st">&quot;Glucose concentration&quot;</span>)                  <span class="co"># add title</span></span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-48-1.png" width="672" /></p>
</div>
<div id="boxplots" class="section level4 hasAnchor" number="1.1.3.3">
<h4><span class="header-section-number">1.1.3.3</span> Boxplots<a href="week-one.html#boxplots" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>The <code>boxplot()</code> function produces a boxplot for a given variable:</p>
<div class="sourceCode" id="cb107"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb107-1"><a href="week-one.html#cb107-1" tabindex="-1"></a><span class="fu">ggplot</span>(dat,</span>
<span id="cb107-2"><a href="week-one.html#cb107-2" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> stab.glu)) <span class="sc">+</span></span>
<span id="cb107-3"><a href="week-one.html#cb107-3" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb107-4"><a href="week-one.html#cb107-4" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span><span class="st">&quot;Stabilized Glucose concentration in blood&quot;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-49-1.png" width="384" /></p>
<blockquote>
<p>Can you explain all features of this graph, such as upper/lower whisker, 25% quantile, …?</p>
</blockquote>
</div>
<div id="qq-plots" class="section level4 hasAnchor" number="1.1.3.4">
<h4><span class="header-section-number">1.1.3.4</span> QQ-plots<a href="week-one.html#qq-plots" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>We can use <strong>QQ-plots</strong> to either (1) compare two distributions, or (2) compare a distribution with a theoretical distribution (typically the normal distribution).</p>
<p>We can for example compare the distribution of the blood pressure values to check if they are normally distributed</p>
<div class="sourceCode" id="cb108"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb108-1"><a href="week-one.html#cb108-1" tabindex="-1"></a><span class="do">## Let&#39;s first make a histogram</span></span>
<span id="cb108-2"><a href="week-one.html#cb108-2" tabindex="-1"></a><span class="fu">ggplot</span>(dat,</span>
<span id="cb108-3"><a href="week-one.html#cb108-3" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> bp<span class="fl">.1</span>s)) <span class="sc">+</span></span>
<span id="cb108-4"><a href="week-one.html#cb108-4" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins =</span> <span class="dv">50</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-50-1.png" width="672" /></p>
<p>Now we can use the function <code>geom_qq()</code> to generate the <strong>QQ-plot</strong> of this distribution against the standard normal distribution:</p>
<div class="sourceCode" id="cb109"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb109-1"><a href="week-one.html#cb109-1" tabindex="-1"></a><span class="fu">ggplot</span>(dat,</span>
<span id="cb109-2"><a href="week-one.html#cb109-2" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">sample =</span> bp<span class="fl">.1</span>s)) <span class="sc">+</span>     <span class="co"># we use sample= inside aes for the QQ-plot</span></span>
<span id="cb109-3"><a href="week-one.html#cb109-3" tabindex="-1"></a>  <span class="fu">geom_qq</span>()                       <span class="co"># creates the QQ-plot</span></span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-51-1.png" width="288" /></p>
<p>Using the additional command <code>geom_qq_line()</code>, we can add a straight line that goes through the first and third quartile:</p>
<div class="sourceCode" id="cb110"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb110-1"><a href="week-one.html#cb110-1" tabindex="-1"></a><span class="fu">ggplot</span>(dat,</span>
<span id="cb110-2"><a href="week-one.html#cb110-2" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">sample =</span> bp<span class="fl">.1</span>s)) <span class="sc">+</span>     <span class="co"># we use sample= inside aes for the QQ-plot</span></span>
<span id="cb110-3"><a href="week-one.html#cb110-3" tabindex="-1"></a>  <span class="fu">geom_qq</span>() <span class="sc">+</span></span>
<span id="cb110-4"><a href="week-one.html#cb110-4" tabindex="-1"></a>  <span class="fu">geom_qq_line</span>(<span class="at">colour =</span> <span class="st">&#39;red&#39;</span>)                  <span class="co"># adds in the QQ-line on top</span></span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-52-1.png" width="288" /></p>
<blockquote>
<p>So, is the distribution normal??</p>
</blockquote>
<p>Now let’s compare the quantiles of the cholesterol values by biological sex.
<strong>Notes on <code>ggplot()</code> here:</strong> Rather than <code>ggplot(dataset, aes(...))</code> we use <code>ggplot() + geom_xx(aes(...))</code> for situations where the data we wish to plot is not in a dataframe.</p>
<div class="sourceCode" id="cb111"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb111-1"><a href="week-one.html#cb111-1" tabindex="-1"></a><span class="co"># We can use &quot;filter()&quot; to filter the cholesterol values for men and women</span></span>
<span id="cb111-2"><a href="week-one.html#cb111-2" tabindex="-1"></a>dat.male <span class="ot">=</span> dat <span class="sc">%&gt;%</span></span>
<span id="cb111-3"><a href="week-one.html#cb111-3" tabindex="-1"></a>  <span class="fu">filter</span>(gender <span class="sc">==</span> <span class="st">&#39;male&#39;</span>)</span>
<span id="cb111-4"><a href="week-one.html#cb111-4" tabindex="-1"></a></span>
<span id="cb111-5"><a href="week-one.html#cb111-5" tabindex="-1"></a>dat.female <span class="ot">=</span> dat <span class="sc">%&gt;%</span></span>
<span id="cb111-6"><a href="week-one.html#cb111-6" tabindex="-1"></a>  <span class="fu">filter</span>(gender <span class="sc">==</span> <span class="st">&#39;female&#39;</span>)</span>
<span id="cb111-7"><a href="week-one.html#cb111-7" tabindex="-1"></a></span>
<span id="cb111-8"><a href="week-one.html#cb111-8" tabindex="-1"></a><span class="co"># Compute the quantiles (note the &quot;na.rm&quot; option to ignore missing NA values!)</span></span>
<span id="cb111-9"><a href="week-one.html#cb111-9" tabindex="-1"></a>q.male <span class="ot">=</span> <span class="fu">quantile</span>(dat.male<span class="sc">$</span>bp<span class="fl">.1</span>s, </span>
<span id="cb111-10"><a href="week-one.html#cb111-10" tabindex="-1"></a>                  <span class="at">probs=</span><span class="fu">seq</span>(<span class="dv">0</span>,<span class="dv">1</span>,<span class="at">by=</span><span class="fl">0.05</span>), </span>
<span id="cb111-11"><a href="week-one.html#cb111-11" tabindex="-1"></a>                  <span class="at">na.rm=</span><span class="cn">TRUE</span>)</span>
<span id="cb111-12"><a href="week-one.html#cb111-12" tabindex="-1"></a>q.female <span class="ot">=</span> <span class="fu">quantile</span>(dat.female<span class="sc">$</span>bp<span class="fl">.1</span>s, </span>
<span id="cb111-13"><a href="week-one.html#cb111-13" tabindex="-1"></a>                    <span class="at">probs=</span><span class="fu">seq</span>(<span class="dv">0</span>,<span class="dv">1</span>,<span class="at">by=</span><span class="fl">0.05</span>),</span>
<span id="cb111-14"><a href="week-one.html#cb111-14" tabindex="-1"></a>                    <span class="at">na.rm=</span><span class="cn">TRUE</span>)</span>
<span id="cb111-15"><a href="week-one.html#cb111-15" tabindex="-1"></a></span>
<span id="cb111-16"><a href="week-one.html#cb111-16" tabindex="-1"></a><span class="co"># Now plot against each other!</span></span>
<span id="cb111-17"><a href="week-one.html#cb111-17" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb111-18"><a href="week-one.html#cb111-18" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> q.male, <span class="at">y =</span> q.female)) <span class="sc">+</span></span>
<span id="cb111-19"><a href="week-one.html#cb111-19" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">&quot;Quantiles&quot;</span>, <span class="at">x =</span> <span class="st">&quot;Male quantiles&quot;</span>, <span class="at">y =</span> <span class="st">&quot;Female quantiles&quot;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-53-1.png" width="288" /></p>
<hr />
</div>
</div>
<div id="correlation" class="section level3 hasAnchor" number="1.1.4">
<h3><span class="header-section-number">1.1.4</span> Correlation<a href="week-one.html#correlation" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="measuring-the-centrality-in-data" class="section level4 hasAnchor" number="1.1.4.1">
<h4><span class="header-section-number">1.1.4.1</span> Measuring the centrality in data<a href="week-one.html#measuring-the-centrality-in-data" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Before you begin, think back to the mean, median and quantiles we saw on the boxplot. Do you remember what these terms mean? How does an asymmetrical distribution influence mean and median?
We have already seen that the <code>summary()</code> and <code>quantile()</code> functions in R can compute the mean, median and quantiles of any given data.</p>
<div class="sourceCode" id="cb112"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb112-1"><a href="week-one.html#cb112-1" tabindex="-1"></a><span class="fu">mean</span>(dat<span class="sc">$</span>stab.glu) </span>
<span id="cb112-2"><a href="week-one.html#cb112-2" tabindex="-1"></a><span class="fu">median</span>(dat<span class="sc">$</span>stab.glu) </span>
<span id="cb112-3"><a href="week-one.html#cb112-3" tabindex="-1"></a><span class="fu">quantile</span>(dat<span class="sc">$</span>stab.glu) </span></code></pre></div>
<blockquote>
<p>Calculate the mean and median of other continuous numeric data in the diabetes dataset and measure the difference between them. (a) Why is there a difference between the mean and median? (b) Why do you think there are larger differences for some and almost no difference for others?</p>
</blockquote>
</div>
<div id="association-between-variables" class="section level4 hasAnchor" number="1.1.4.2">
<h4><span class="header-section-number">1.1.4.2</span> Association between variables<a href="week-one.html#association-between-variables" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Often a common step during any data analysis project is to find associations between variables present in the dataset. Such associations helps us to decipher the underlying structure in the data.</p>
<p>For instance, in our diabetes dataset we would expect a high correlation between free blood glucose levels and glycosylated blood levels or between waist and hip sizes. One of the most common ways of measuring associations is <em>correlations</em>.</p>
<p>Let us start by producing a <strong>scatter plot</strong> between a pair of variables:</p>
<div class="sourceCode" id="cb113"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb113-1"><a href="week-one.html#cb113-1" tabindex="-1"></a><span class="fu">ggplot</span>(dat,</span>
<span id="cb113-2"><a href="week-one.html#cb113-2" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> stab.glu, <span class="at">y =</span> glyhb)) <span class="sc">+</span></span>
<span id="cb113-3"><a href="week-one.html#cb113-3" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb113-4"><a href="week-one.html#cb113-4" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x=</span><span class="st">&#39;Stabilized glucose&#39;</span>, <span class="at">y=</span><span class="st">&#39;Glycosylated hemoglobin&#39;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-55-1.png" width="672" /></p>
<blockquote>
<p>Do you suspect that the two variables have a relationship? Do the scatter plot for other pairs of numerical variables!</p>
</blockquote>
<p>We now can compute the correlation of the two variables. We can compute the <strong>Pearson correlation</strong> or the <strong>Spearman correlation</strong>:</p>
<div class="sourceCode" id="cb114"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb114-1"><a href="week-one.html#cb114-1" tabindex="-1"></a><span class="do">## compute the Pearson correlation</span></span>
<span id="cb114-2"><a href="week-one.html#cb114-2" tabindex="-1"></a><span class="fu">cor</span>(dat<span class="sc">$</span>stab.glu, dat<span class="sc">$</span>glyhb, <span class="at">method=</span><span class="st">&#39;pearson&#39;</span>)</span></code></pre></div>
<pre><code>## [1] 0.7411355</code></pre>
<div class="sourceCode" id="cb116"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb116-1"><a href="week-one.html#cb116-1" tabindex="-1"></a><span class="do">## compute the Spearman correlation</span></span>
<span id="cb116-2"><a href="week-one.html#cb116-2" tabindex="-1"></a><span class="fu">cor</span>(dat<span class="sc">$</span>stab.glu, dat<span class="sc">$</span>glyhb, <span class="at">method=</span><span class="st">&#39;spearman&#39;</span>)</span></code></pre></div>
<pre><code>## [1] 0.5214866</code></pre>
<p>The Spearman correlation seems much lower, right? To understand why, we can do a scatter plot between the <strong>ranks</strong> of the two variables:</p>
<div class="sourceCode" id="cb118"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb118-1"><a href="week-one.html#cb118-1" tabindex="-1"></a><span class="fu">ggplot</span>(dat,</span>
<span id="cb118-2"><a href="week-one.html#cb118-2" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">rank</span>(stab.glu), <span class="at">y =</span> <span class="fu">rank</span>(glyhb))) <span class="sc">+</span></span>
<span id="cb118-3"><a href="week-one.html#cb118-3" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb118-4"><a href="week-one.html#cb118-4" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x=</span><span class="st">&#39;Rank - Stabilized glucose&#39;</span>, <span class="at">y=</span><span class="st">&#39;Rank - Glycosylated hemoglobin&#39;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-57-1.png" width="672" /></p>
<p>Do you understand the usage of ranks here? Run `rank()`` on a vector like <em>c(3,5,10,1,23)</em> to see how the output looks like.</p>
<blockquote>
<p>Associations are among the simplest forms of structure in the data! It is important to remember that <em>Association does not imply correlation</em> and <em>Correlation does not imply causation</em>. Take a look at this page to view few common logical fallacies. <a href="https://en.wikipedia.org/wiki/Fallacy">see here</a></p>
</blockquote>
<hr />
</div>
</div>
<div id="exercises" class="section level3 hasAnchor" number="1.1.5">
<h3><span class="header-section-number">1.1.5</span> EXERCISES<a href="week-one.html#exercises" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="exercise-1-data-features" class="section level4 hasAnchor" number="1.1.5.1">
<h4><span class="header-section-number">1.1.5.1</span> Exercise 1: Data features<a href="week-one.html#exercise-1-data-features" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<ol style="list-style-type: decimal">
<li><p>Try to obtain the same result of the <code>head()</code> function by using slicing on the dataset “dat” using row indexes.</p></li>
<li><p>Print out the last element in the last column of “dat” using the <code>dim()</code> function instead of using numerals.</p></li>
</ol>
</div>
<div id="exercise-2-visualization-and-correlation" class="section level4 hasAnchor" number="1.1.5.2">
<h4><span class="header-section-number">1.1.5.2</span> Exercise 2: Visualization and correlation<a href="week-one.html#exercise-2-visualization-and-correlation" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<ol style="list-style-type: decimal">
<li><p>Visualize the cholesterol levels of all patients with a histogram using the <code>geom_histogram()</code> function.</p></li>
<li><p>Visualize the cholesterol levels of all <strong>male</strong> patients with a histogram using <code>geom_histogram()</code>.
<em>(expert):</em> Mark the median, first and third quartile with vertical lines using <code>geom_vline()</code>. The values are defined using <code>xintercept</code> (as in which value of x should the vertical line intercept).
Then mark median, first and third quantile for <strong>female</strong> patients in the same graph with a different color.
What can you tell from the differences in these values?</p></li>
<li><p>Is there an association between “hip” and “waist” on the data frame “dat”? Use the <code>geom_point()</code> function to do a scatter plot of the <strong>values</strong> and of the <strong>ranks</strong> (as determined by the <code>rank()</code> function). Compute both the <strong>Pearson</strong> and <strong>Spearman</strong> correlation values.</p></li>
</ol>
</div>
<div id="going-further-expert" class="section level4 hasAnchor" number="1.1.5.3">
<h4><span class="header-section-number">1.1.5.3</span> Going further <em>(expert)</em><a href="week-one.html#going-further-expert" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<ol style="list-style-type: decimal">
<li><p>Select only the numerical columns (using <code>select(where(is.numeric))</code>), and apply the function <code>cor()</code> <strong>directly</strong>: <code>%&gt;% cor()</code> . What happens? What is the output? Store the result of this command in a new variable named “all.cor”. Plot a heatmap of these results using the <code>pheatmap()</code> function from the “pheatmap” package. Remember that you first have to install and then activate this package using <code>library("pheatmap")</code>.</p></li>
<li><p>Find the <strong>highest</strong> and <strong>lowest</strong> Pearson correlation value of the result from exercise 2.2. To which pair of variables do they correspond? Plot the corresponding scatter plots using <code>geom_point()</code>! <em>Hint</em>: Before finding the highest Pearson correlation value, use the <code>diag()</code> function to set all diagonal values (=1) of “all.cor” to NA.</p></li>
</ol>
</div>
</div>
</div>
<div id="day-2-dimensionality-reduction-and-unsupervised-learning" class="section level2 hasAnchor" number="1.2">
<h2><span class="header-section-number">1.2</span> Day 2: Dimensionality reduction and unsupervised learning<a href="week-one.html#day-2-dimensionality-reduction-and-unsupervised-learning" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="preparing-the-data" class="section level3 hasAnchor" number="1.2.1">
<h3><span class="header-section-number">1.2.1</span> Preparing the data<a href="week-one.html#preparing-the-data" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><strong>Unsupervised clustering</strong> is one of the most basic data analysis techniques. It allows to identify groups (or clusters) or observations (here: patients) or variables. <em>Unsupervised</em> means that you are not using any prior knowledge about groups of variables or associations. <strong>K-means clustering</strong> is a good example of unsupervised learning because the method categorizes sample based uniquely on the data.</p>
<p>In this part, we will use a dataset of gene expression data from the TCGA (The Cancer Genome Atlas) project. This project has sequenced several thousand samples from cancer patients of more than 30 cancer types. We will use a subset of this data, containing 200 samples (=patients, as columns) , for which the expression of 300 genes (= rows) has been measured.</p>
<div id="load-data" class="section level4 hasAnchor" number="1.2.1.1">
<h4><span class="header-section-number">1.2.1.1</span> Load data<a href="week-one.html#load-data" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>We will start by reading the gene expression data. The columns are the samples and the rows are the genes. This is matrix, which allows some numerical operations to be conducted directly.</p>
<div class="sourceCode" id="cb119"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb119-1"><a href="week-one.html#cb119-1" tabindex="-1"></a>brca.exp <span class="ot">=</span> <span class="fu">readRDS</span>(<span class="fu">url</span>(<span class="st">&#39;https://www.dropbox.com/s/qububmfvtv443mq/brca.exp.rds?dl=1&#39;</span>))</span>
<span id="cb119-2"><a href="week-one.html#cb119-2" tabindex="-1"></a><span class="fu">dim</span>(brca.exp)</span></code></pre></div>
<pre><code>## [1] 100 200</code></pre>
<p><strong>WARNING</strong>: If you have problem loading the data, please download <a href="https://www.dropbox.com/s/qububmfvtv443mq/brca.exp.rds?dl=1">this file</a>, store it on your disk, and open it with the following command:</p>
<div class="sourceCode" id="cb121"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb121-1"><a href="week-one.html#cb121-1" tabindex="-1"></a><span class="co">#brca.exp = readRDS(&quot;xxxx&quot;) # xxxx should be replaced with the path to the downloaded file in your device</span></span></code></pre></div>
<p>Next we will load the clinical annotation file for this gene expression data and explore it</p>
<div class="sourceCode" id="cb122"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb122-1"><a href="week-one.html#cb122-1" tabindex="-1"></a>brca.anno <span class="ot">=</span> <span class="fu">readRDS</span>(<span class="fu">url</span>(<span class="st">&#39;https://www.dropbox.com/s/9xlivejqkj77llc/brca.anno.rds?dl=1&#39;</span>))</span>
<span id="cb122-2"><a href="week-one.html#cb122-2" tabindex="-1"></a><span class="fu">head</span>(brca.anno)</span></code></pre></div>
<pre><code>##                 Age ER_status HER2_status Classification
## TCGA-BH-A1EO-01  68  Positive    Negative      Luminal A
## TCGA-E2-A14N-01  37  Negative    Negative     Basal-like
## TCGA-AN-A0FF-01  32  Positive    Negative      Luminal B
## TCGA-A2-A04V-01  39  Positive    Negative      Luminal A
## TCGA-AN-A0XP-01  69  Positive    Negative      Luminal A
## TCGA-C8-A12U-01  46  Positive    Negative      Luminal B</code></pre>
<p>Same here: if you have issues running the previous <code>readRDS</code> command, download <a href="https://www.dropbox.com/s/z6bzwzgzdhky1qz/brca.anno.rds?dl=1">this file</a>, save it on your disk and load it with</p>
<div class="sourceCode" id="cb124"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb124-1"><a href="week-one.html#cb124-1" tabindex="-1"></a><span class="do">### brca.anno = readRDS(xxx)</span></span></code></pre></div>
<p>You can check the number of samples for each tumor type using the <code>table()</code> function, applied to a specific column (here, there is only one column…)</p>
<div class="sourceCode" id="cb125"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb125-1"><a href="week-one.html#cb125-1" tabindex="-1"></a><span class="fu">table</span>(brca.anno<span class="sc">$</span>HER2_status)</span></code></pre></div>
<pre><code>## 
## Equivocal  Negative  Positive 
##         4       174        22</code></pre>
</div>
<div id="data-transformation" class="section level4 hasAnchor" number="1.2.1.2">
<h4><span class="header-section-number">1.2.1.2</span> Data transformation<a href="week-one.html#data-transformation" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>You will see that the distribution of the data is extremely squeezed due to <strong>outliers</strong> with very high or low values. We will need to make the data more homogeneous, so that our downstream analysis is not affected by these very large magnitude numbers.</p>
<p>We will carry out the following data processing steps. Some of these steps use rather arbitrary values, which come from visually inspecting the data!</p>
<ol style="list-style-type: decimal">
<li><p><strong>Thresholding:</strong> cap the values to the 95th percentile</p></li>
<li><p><strong>Homogenization:</strong> base-2 logarithmic transformation of the entire dataset</p></li>
<li><p><strong>Scaling:</strong> standardize the data so that across genes the mean = 0 and variance = 1.</p></li>
</ol>
<p><strong>Before we start modifying the data, we will store the original data frame into a variable, so that in case of problems we can revert back to the initial data!!</strong></p>
<div class="sourceCode" id="cb127"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb127-1"><a href="week-one.html#cb127-1" tabindex="-1"></a>brca.exp.original <span class="ot">=</span> brca.exp <span class="co"># keeps the original as matrix</span></span></code></pre></div>
<p><strong>Thresholding</strong></p>
<div class="sourceCode" id="cb128"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb128-1"><a href="week-one.html#cb128-1" tabindex="-1"></a><span class="do">## what is the value of the 95th percent percentile?</span></span>
<span id="cb128-2"><a href="week-one.html#cb128-2" tabindex="-1"></a>q95 <span class="ot">=</span> <span class="fu">quantile</span>(brca.exp,<span class="at">probs=</span><span class="fl">0.95</span>)</span>
<span id="cb128-3"><a href="week-one.html#cb128-3" tabindex="-1"></a></span>
<span id="cb128-4"><a href="week-one.html#cb128-4" tabindex="-1"></a><span class="do">## set all values above to this value</span></span>
<span id="cb128-5"><a href="week-one.html#cb128-5" tabindex="-1"></a>brca.exp[brca.exp<span class="sc">&gt;</span>q95] <span class="ot">=</span> q95</span></code></pre></div>
<p><strong>Homogenization and Scaling</strong></p>
<p>We will perform this step by log-transforming the data. We are able to use this operation because the data is still in a matrix.</p>
<div class="sourceCode" id="cb129"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb129-1"><a href="week-one.html#cb129-1" tabindex="-1"></a>brca.exp <span class="ot">=</span> <span class="fu">log2</span>(brca.exp<span class="sc">+</span><span class="dv">1</span>)</span></code></pre></div>
<blockquote>
<p>Why do we add +1 ?</p>
</blockquote>
<p>Next, we will scale the data and plot its distribution. To do this efficient, we need to convert the data to tibble first, make it tidy, and then plot.</p>
<p>Conversion to tibble can be done using <code>as_tibble(brca.exp, rownames = NA)</code>, where <code>rownames = NA</code> is meant to keep the original rownames (in this case, gene names) in the new tibble, although they are invisible.</p>
<p>Check this out:</p>
<div class="sourceCode" id="cb130"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb130-1"><a href="week-one.html#cb130-1" tabindex="-1"></a><span class="fu">rownames</span>(<span class="fu">as_tibble</span>(brca.exp, <span class="at">rownames =</span> <span class="cn">NA</span>))[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code></pre></div>
<pre><code>## [1] &quot;TFF1&quot;    &quot;C4orf7&quot;  &quot;AGR3&quot;    &quot;GABRP&quot;   &quot;C1orf64&quot;</code></pre>
<p>In addition, <code>gather(key = "sample", value = "expression")</code> converts the tibble to a long format, where “sample” represents the original column names, and “expression” represents the values present in the initial matrix.</p>
<div class="sourceCode" id="cb132"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb132-1"><a href="week-one.html#cb132-1" tabindex="-1"></a><span class="do">## scaling</span></span>
<span id="cb132-2"><a href="week-one.html#cb132-2" tabindex="-1"></a>brca.exp <span class="ot">=</span> <span class="fu">scale</span>(brca.exp)</span>
<span id="cb132-3"><a href="week-one.html#cb132-3" tabindex="-1"></a></span>
<span id="cb132-4"><a href="week-one.html#cb132-4" tabindex="-1"></a><span class="do">## plotting the density</span></span>
<span id="cb132-5"><a href="week-one.html#cb132-5" tabindex="-1"></a><span class="fu">as_tibble</span>(brca.exp, <span class="at">rownames =</span> <span class="cn">NA</span>) <span class="sc">%&gt;%</span></span>
<span id="cb132-6"><a href="week-one.html#cb132-6" tabindex="-1"></a>  <span class="fu">gather</span>(<span class="at">key =</span> <span class="st">&quot;sample&quot;</span>, </span>
<span id="cb132-7"><a href="week-one.html#cb132-7" tabindex="-1"></a>         <span class="at">value =</span> <span class="st">&quot;expression&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb132-8"><a href="week-one.html#cb132-8" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> expression)) <span class="sc">+</span></span>
<span id="cb132-9"><a href="week-one.html#cb132-9" tabindex="-1"></a>  <span class="fu">geom_histogram</span>() <span class="sc">+</span></span>
<span id="cb132-10"><a href="week-one.html#cb132-10" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">&quot;Transformed data&quot;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-67-1.png" width="672" /></p>
<p>Compare to the density plot before these pre-processing steps using the same strategy.</p>
<div class="sourceCode" id="cb133"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb133-1"><a href="week-one.html#cb133-1" tabindex="-1"></a><span class="fu">as_tibble</span>(brca.exp.original, <span class="at">rownames =</span> <span class="cn">NA</span>) <span class="sc">%&gt;%</span></span>
<span id="cb133-2"><a href="week-one.html#cb133-2" tabindex="-1"></a>  <span class="fu">gather</span>(<span class="at">key =</span> <span class="st">&quot;sample&quot;</span>, <span class="at">value =</span> <span class="st">&quot;expression&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb133-3"><a href="week-one.html#cb133-3" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> expression)) <span class="sc">+</span></span>
<span id="cb133-4"><a href="week-one.html#cb133-4" tabindex="-1"></a>  <span class="fu">geom_histogram</span>() <span class="sc">+</span></span>
<span id="cb133-5"><a href="week-one.html#cb133-5" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">&quot;Untransformed data&quot;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-68-1.png" width="672" /></p>
<hr />
</div>
</div>
<div id="k-means-clustering" class="section level3 hasAnchor" number="1.2.2">
<h3><span class="header-section-number">1.2.2</span> k-means clustering<a href="week-one.html#k-means-clustering" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Another widely used method for grouping observations is the k-means clustering. Now we will cluster our dataset using k-means and explore the underlying structure of the data. In this dataset, different clusters could represent different batches, different tumour subtypes, among other features.</p>
<div id="performing-k-means" class="section level4 hasAnchor" number="1.2.2.1">
<h4><span class="header-section-number">1.2.2.1</span> Performing k-means<a href="week-one.html#performing-k-means" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>We use the function <code>kmeans()</code> in R on our matrix. You can check the options and usage in the help panel on the right. The parameter <code>centers</code> indicates how many clusters are requested.</p>
<div class="sourceCode" id="cb134"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb134-1"><a href="week-one.html#cb134-1" tabindex="-1"></a>km <span class="ot">=</span> <span class="fu">kmeans</span>(<span class="at">x=</span><span class="fu">t</span>(brca.exp), </span>
<span id="cb134-2"><a href="week-one.html#cb134-2" tabindex="-1"></a>            <span class="at">centers =</span> <span class="dv">2</span>, </span>
<span id="cb134-3"><a href="week-one.html#cb134-3" tabindex="-1"></a>            <span class="at">nstart =</span> <span class="dv">10</span>)</span></code></pre></div>
<blockquote>
<p>Just type <code>km</code> in your console and check all results generated. Play around with the <code>centers</code> parameter. See cluster assignments by typing <code>table(km$cluster)</code></p>
</blockquote>
</div>
<div id="quality-of-the-clustering" class="section level4 hasAnchor" number="1.2.2.2">
<h4><span class="header-section-number">1.2.2.2</span> Quality of the clustering<a href="week-one.html#quality-of-the-clustering" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>We can judge the quality of the clustering by computing the <strong>intra</strong>-cluster distances, i.e. the sum (squared) of all distances between pairs of objects belonging to the same cluster. This is called the <strong>within sum of squares (WSS)</strong>. The better the clustering, the smaller WSS should be. However, it also automatically decreases with increasing k.</p>
<blockquote>
<p>What would be WSS if we request a number of clusters equal to the number of data points? You can check what the WSS is for a particular clustering by typing</p>
</blockquote>
<div class="sourceCode" id="cb135"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb135-1"><a href="week-one.html#cb135-1" tabindex="-1"></a>km<span class="sc">$</span>tot.withinss</span></code></pre></div>
<pre><code>## [1] 4123.459</code></pre>
<blockquote>
<p>run k-means for k=2 to k=7 clusters, and for each k check the WSS value. How does WSS evolve with increasing k?</p>
</blockquote>
<p>We can also run a little loop to test different k. Loops are very important structures in any programming language.
We can test a simple scenario before. Check how the output of this simple loop looks like.</p>
<div class="sourceCode" id="cb137"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb137-1"><a href="week-one.html#cb137-1" tabindex="-1"></a>k_to_test <span class="ot">=</span> <span class="fu">c</span>(<span class="dv">2</span><span class="sc">:</span><span class="dv">7</span>)</span>
<span id="cb137-2"><a href="week-one.html#cb137-2" tabindex="-1"></a></span>
<span id="cb137-3"><a href="week-one.html#cb137-3" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(k_to_test)) {</span>
<span id="cb137-4"><a href="week-one.html#cb137-4" tabindex="-1"></a>  <span class="fu">print</span>(i) <span class="co"># We can print the indexes</span></span>
<span id="cb137-5"><a href="week-one.html#cb137-5" tabindex="-1"></a>}</span></code></pre></div>
<pre><code>## [1] 1
## [1] 2
## [1] 3
## [1] 4
## [1] 5
## [1] 6</code></pre>
<div class="sourceCode" id="cb139"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb139-1"><a href="week-one.html#cb139-1" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(k_to_test)) {</span>
<span id="cb139-2"><a href="week-one.html#cb139-2" tabindex="-1"></a>  <span class="fu">print</span>(k_to_test[i]) <span class="co"># or the actual elements</span></span>
<span id="cb139-3"><a href="week-one.html#cb139-3" tabindex="-1"></a>}</span></code></pre></div>
<pre><code>## [1] 2
## [1] 3
## [1] 4
## [1] 5
## [1] 6
## [1] 7</code></pre>
<blockquote>
<p>Do you understand the difference between the 2 previous for loops? Try to make your own for loop.</p>
</blockquote>
<p>Now we can make one to test k from 2 to 7:</p>
<div class="sourceCode" id="cb141"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb141-1"><a href="week-one.html#cb141-1" tabindex="-1"></a>km_wws <span class="ot">=</span> <span class="fu">numeric</span>() <span class="co"># we start by creating an empty vector</span></span>
<span id="cb141-2"><a href="week-one.html#cb141-2" tabindex="-1"></a></span>
<span id="cb141-3"><a href="week-one.html#cb141-3" tabindex="-1"></a><span class="co"># To write in the position 1, we use i</span></span>
<span id="cb141-4"><a href="week-one.html#cb141-4" tabindex="-1"></a><span class="co"># to find the 1st element, we use k_to_test[i]</span></span>
<span id="cb141-5"><a href="week-one.html#cb141-5" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(k_to_test)) {</span>
<span id="cb141-6"><a href="week-one.html#cb141-6" tabindex="-1"></a>  km_wws[i] <span class="ot">=</span> <span class="fu">kmeans</span>(<span class="at">x=</span><span class="fu">t</span>(brca.exp), </span>
<span id="cb141-7"><a href="week-one.html#cb141-7" tabindex="-1"></a>                     <span class="at">centers =</span> k_to_test[i])<span class="sc">$</span>tot.withinss</span>
<span id="cb141-8"><a href="week-one.html#cb141-8" tabindex="-1"></a>}</span>
<span id="cb141-9"><a href="week-one.html#cb141-9" tabindex="-1"></a></span>
<span id="cb141-10"><a href="week-one.html#cb141-10" tabindex="-1"></a><span class="co"># We can plot the k against WSS using geom_line</span></span>
<span id="cb141-11"><a href="week-one.html#cb141-11" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb141-12"><a href="week-one.html#cb141-12" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">x =</span> k_to_test, <span class="at">y =</span> km_wws)) <span class="sc">+</span></span>
<span id="cb141-13"><a href="week-one.html#cb141-13" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x=</span><span class="st">&quot;Number of clusters K&quot;</span>,</span>
<span id="cb141-14"><a href="week-one.html#cb141-14" tabindex="-1"></a>       <span class="at">y=</span><span class="st">&quot;Total within-clusters sum of squares&quot;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-72-1.png" width="672" /></p>
<blockquote>
<p>Do you see an obvious “elbow” or “kink” in the curve?? Another criteria for the quality of the clustering is the <strong>silhouette</strong> method.</p>
</blockquote>
<p>To run the silhouette method, we need to compute the pairwise distances between all objects (i.e. patients) in the data matrix. This is done with the <code>dist</code> function, which can take different metrics (euclidean, …)</p>
<div class="sourceCode" id="cb142"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb142-1"><a href="week-one.html#cb142-1" tabindex="-1"></a><span class="do">## compute the patient-patient distance matrix (this is why we transpose using the `t()` function)</span></span>
<span id="cb142-2"><a href="week-one.html#cb142-2" tabindex="-1"></a>D <span class="ot">=</span> <span class="fu">dist</span>(<span class="fu">t</span>(brca.exp))</span></code></pre></div>
<p>We now compute the silhouette for a specific k-means clustering:</p>
<div class="sourceCode" id="cb143"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb143-1"><a href="week-one.html#cb143-1" tabindex="-1"></a><span class="fu">library</span>(cluster)</span></code></pre></div>
<pre><code>## 
## Attaching package: &#39;cluster&#39;</code></pre>
<pre><code>## The following object is masked _by_ &#39;.GlobalEnv&#39;:
## 
##     animals</code></pre>
<div class="sourceCode" id="cb146"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb146-1"><a href="week-one.html#cb146-1" tabindex="-1"></a>km <span class="ot">=</span> <span class="fu">kmeans</span>(<span class="at">x=</span><span class="fu">t</span>(brca.exp), <span class="at">centers =</span> <span class="dv">3</span>, <span class="at">nstart =</span> <span class="dv">10</span>)</span>
<span id="cb146-2"><a href="week-one.html#cb146-2" tabindex="-1"></a>s <span class="ot">=</span> <span class="fu">silhouette</span>(km<span class="sc">$</span>cluster,D)</span>
<span id="cb146-3"><a href="week-one.html#cb146-3" tabindex="-1"></a></span>
<span id="cb146-4"><a href="week-one.html#cb146-4" tabindex="-1"></a><span class="co"># Let us use the basic R function plot() to see the results</span></span>
<span id="cb146-5"><a href="week-one.html#cb146-5" tabindex="-1"></a><span class="fu">plot</span>(s)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-74-1.png" width="672" /></p>
<hr />
</div>
</div>
<div id="hierarchical-clustering" class="section level3 hasAnchor" number="1.2.3">
<h3><span class="header-section-number">1.2.3</span> Hierarchical clustering<a href="week-one.html#hierarchical-clustering" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Clustering is a method by which we group together similar observations while separating out the dissimilar ones. We will cluster our samples from the cancer dataset to see which samples cluster together or separately. Hierarchical clustering does not generate discrete clusters of datapoints, but rather creates a dendrogram that indicates the magnitude of similitude between samples. Once again is up to the Data Scientist to decide the right amount of clusters.</p>
<div id="determine-the-most-variable-genes" class="section level4 hasAnchor" number="1.2.3.1">
<h4><span class="header-section-number">1.2.3.1</span> Determine the most variable genes<a href="week-one.html#determine-the-most-variable-genes" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>When performing clustering, we usually reduce the number of genes used, as some of them are not informative. For example, genes that show a mostly constant expression across all samples will not be useful to distinguish the samples, right? One simple method is to select genes that show a <strong>high variance</strong> across all samples.</p>
<div class="sourceCode" id="cb147"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb147-1"><a href="week-one.html#cb147-1" tabindex="-1"></a>brca.exp.tibble <span class="ot">=</span> <span class="fu">as_tibble</span>(brca.exp, <span class="at">rownames=</span><span class="cn">NA</span>) <span class="sc">%&gt;%</span></span>
<span id="cb147-2"><a href="week-one.html#cb147-2" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>(<span class="st">&quot;gene&quot;</span>)</span>
<span id="cb147-3"><a href="week-one.html#cb147-3" tabindex="-1"></a></span>
<span id="cb147-4"><a href="week-one.html#cb147-4" tabindex="-1"></a><span class="do">## create a new column with the variance for all genes across all samples</span></span>
<span id="cb147-5"><a href="week-one.html#cb147-5" tabindex="-1"></a>brca.exp.var <span class="ot">=</span> brca.exp.tibble <span class="sc">%&gt;%</span></span>
<span id="cb147-6"><a href="week-one.html#cb147-6" tabindex="-1"></a>  <span class="fu">rowwise</span>() <span class="sc">%&gt;%</span></span>
<span id="cb147-7"><a href="week-one.html#cb147-7" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">variance =</span> <span class="fu">var</span>(<span class="fu">c_across</span>(<span class="fu">starts_with</span>(<span class="st">&quot;TCGA&quot;</span>)))) </span>
<span id="cb147-8"><a href="week-one.html#cb147-8" tabindex="-1"></a><span class="co"># only includes the columns starting with TCGA</span></span></code></pre></div>
<p>We now want to find the top 25% with the highest variance</p>
<div class="sourceCode" id="cb148"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb148-1"><a href="week-one.html#cb148-1" tabindex="-1"></a><span class="do">## what is the 75% quantile of the variance?</span></span>
<span id="cb148-2"><a href="week-one.html#cb148-2" tabindex="-1"></a>q75 <span class="ot">=</span> <span class="fu">quantile</span>(brca.exp.var<span class="sc">$</span>variance, <span class="at">probs =</span> <span class="fl">0.75</span>)</span>
<span id="cb148-3"><a href="week-one.html#cb148-3" tabindex="-1"></a>q75</span></code></pre></div>
<pre><code>##       75% 
## 0.4934196</code></pre>
<p>So let us select all rows (genes) with a variance higher than or equal to <code>q75</code>:</p>
<div class="sourceCode" id="cb150"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb150-1"><a href="week-one.html#cb150-1" tabindex="-1"></a><span class="do">## only select the genes with a variance in the top 25%</span></span>
<span id="cb150-2"><a href="week-one.html#cb150-2" tabindex="-1"></a>topVariantGenes <span class="ot">&lt;-</span> brca.exp.var <span class="sc">%&gt;%</span></span>
<span id="cb150-3"><a href="week-one.html#cb150-3" tabindex="-1"></a>  <span class="fu">filter</span>(variance <span class="sc">&gt;=</span> q75)</span>
<span id="cb150-4"><a href="week-one.html#cb150-4" tabindex="-1"></a></span>
<span id="cb150-5"><a href="week-one.html#cb150-5" tabindex="-1"></a><span class="fu">print</span>(topVariantGenes<span class="sc">$</span>gene)</span></code></pre></div>
<pre><code>##  [1] &quot;TFF1&quot;      &quot;C4orf7&quot;    &quot;AGR3&quot;      &quot;GABRP&quot;     &quot;C1orf64&quot;   &quot;TFF3&quot;     
##  [7] &quot;ABCC11&quot;    &quot;PGR&quot;       &quot;FABP7&quot;     &quot;SERPINA11&quot; &quot;VGLL1&quot;     &quot;A2ML1&quot;    
## [13] &quot;ELF5&quot;      &quot;PROM1&quot;     &quot;CYP4Z2P&quot;   &quot;SLC6A14&quot;   &quot;CAPN8&quot;     &quot;ABCC8&quot;    
## [19] &quot;SYTL5&quot;     &quot;SFRP1&quot;     &quot;ART3&quot;      &quot;GABBR2&quot;    &quot;PPP1R14C&quot;  &quot;HORMAD1&quot;  
## [25] &quot;LOC84740&quot;</code></pre>
</div>
<div id="computing-the-correlation-between-all-patients" class="section level4 hasAnchor" number="1.2.3.2">
<h4><span class="header-section-number">1.2.3.2</span> Computing the correlation between all patients<a href="week-one.html#computing-the-correlation-between-all-patients" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Let us start by filtering for only the highly variable genes. Then we can directly calculate Spearman correlation.</p>
<div class="sourceCode" id="cb152"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb152-1"><a href="week-one.html#cb152-1" tabindex="-1"></a>brca.exp.highvar.cor <span class="ot">=</span> brca.exp.tibble <span class="sc">%&gt;%</span></span>
<span id="cb152-2"><a href="week-one.html#cb152-2" tabindex="-1"></a>  <span class="fu">filter</span>(gene <span class="sc">%in%</span> topVariantGenes<span class="sc">$</span>gene) <span class="sc">%&gt;%</span>        <span class="co"># from the whole list, select only high variable</span></span>
<span id="cb152-3"><a href="week-one.html#cb152-3" tabindex="-1"></a>  <span class="fu">select</span>(<span class="fu">where</span>(is.numeric)) <span class="sc">%&gt;%</span>                     <span class="co"># get only numerical columns</span></span>
<span id="cb152-4"><a href="week-one.html#cb152-4" tabindex="-1"></a>  <span class="fu">cor</span>(<span class="at">method=</span><span class="st">&quot;spearman&quot;</span>)                            <span class="co"># create correlation-based distance matrix</span></span></code></pre></div>
<p>If we want to display the correlation matrix as a heatmap, we can use the <code>pheatmap</code> function as before:</p>
<div class="sourceCode" id="cb153"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb153-1"><a href="week-one.html#cb153-1" tabindex="-1"></a><span class="fu">library</span>(pheatmap)</span>
<span id="cb153-2"><a href="week-one.html#cb153-2" tabindex="-1"></a><span class="fu">pheatmap</span>(brca.exp.highvar.cor, </span>
<span id="cb153-3"><a href="week-one.html#cb153-3" tabindex="-1"></a>         <span class="at">show_rownames =</span> <span class="cn">FALSE</span>, </span>
<span id="cb153-4"><a href="week-one.html#cb153-4" tabindex="-1"></a>         <span class="at">show_colnames =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-79-1.png" width="672" /></p>
<p>Each cell of this heatmap represents the correlation value between the sample in the row and the sample in the column. The correlation of a sample to itself is always 1 (red diagonal).</p>
<p>The function automatically determines the clustering trees of the rows and columns (which are identical, since the correlation matrix is symmetrical!)</p>
</div>
<div id="including-clinical-annotations-in-the-heatmap" class="section level4 hasAnchor" number="1.2.3.3">
<h4><span class="header-section-number">1.2.3.3</span> Including clinical annotations in the heatmap<a href="week-one.html#including-clinical-annotations-in-the-heatmap" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>This is a nice representation, but in order to interpret this clustering, we need to add some additional (clinical) information to interpret the clustering structure. To do this, we use an annotation data frame containing as columns a number of clinical features.</p>
<p>The clinical annotation is stored in the <code>brca.anno</code> data frame.
We can now plot again the heatmap, using the annotation dataframe to add additional information</p>
<div class="sourceCode" id="cb154"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb154-1"><a href="week-one.html#cb154-1" tabindex="-1"></a><span class="fu">pheatmap</span>(brca.exp.highvar.cor,</span>
<span id="cb154-2"><a href="week-one.html#cb154-2" tabindex="-1"></a>         <span class="at">annotation_row =</span> brca.anno,</span>
<span id="cb154-3"><a href="week-one.html#cb154-3" tabindex="-1"></a>         <span class="at">show_rownames =</span> <span class="cn">FALSE</span>, </span>
<span id="cb154-4"><a href="week-one.html#cb154-4" tabindex="-1"></a>         <span class="at">show_colnames =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-80-1.png" width="768" /></p>
<blockquote>
<p>How would you interpret this dendrogram? Do the clusters you observe make any sense? What are the parameters by which the samples cluster together? How many meaningful clusters can you observe? Do you see any relation between the distribution of the data and your clusters ?</p>
</blockquote>
<blockquote>
<p>The function <code>pheatmap</code> accepts a parameter clustering_method to indicate alternative linkage methods; try other linkage methods (check which are available with the pheatmap help page, which can be accessed by typing <code>?pheatmap</code> in the console!)</p>
</blockquote>
<hr />
</div>
</div>
<div id="principal-component-analysis" class="section level3 hasAnchor" number="1.2.4">
<h3><span class="header-section-number">1.2.4</span> Principal component analysis<a href="week-one.html#principal-component-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We will now use principal component analysis to explore the same dataset, and identify directions (i.e. principal components) with maximal variance. Principal components analysis finds n-dimensional vectors (Principal Components) in the direction of the largest variance, thereby allowing you to describe an n-dimensional dataset with just a few dimensions.</p>
<div class="sourceCode" id="cb155"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb155-1"><a href="week-one.html#cb155-1" tabindex="-1"></a>pca <span class="ot">=</span> topVariantGenes <span class="sc">%&gt;%</span></span>
<span id="cb155-2"><a href="week-one.html#cb155-2" tabindex="-1"></a>  <span class="fu">select</span>(<span class="fu">where</span>(is.numeric)) <span class="sc">%&gt;%</span></span>
<span id="cb155-3"><a href="week-one.html#cb155-3" tabindex="-1"></a>  <span class="fu">t</span>() <span class="sc">%&gt;%</span> <span class="co"># do not forget to transpose the data!</span></span>
<span id="cb155-4"><a href="week-one.html#cb155-4" tabindex="-1"></a>  <span class="fu">prcomp</span>(<span class="at">center =</span> <span class="cn">FALSE</span>, <span class="at">scale =</span> <span class="cn">FALSE</span>) <span class="co"># We set these as false as we have already scaled our data</span></span>
<span id="cb155-5"><a href="week-one.html#cb155-5" tabindex="-1"></a><span class="fu">summary</span>(pca)</span></code></pre></div>
<pre><code>## Importance of components:
##                          PC1    PC2     PC3     PC4     PC5     PC6     PC7
## Standard deviation     3.719 2.1886 0.89796 0.86195 0.69326 0.62793 0.61645
## Proportion of Variance 0.559 0.1936 0.03259 0.03003 0.01942 0.01593 0.01536
## Cumulative Proportion  0.559 0.7526 0.78514 0.81516 0.83458 0.85052 0.86588
##                            PC8     PC9    PC10    PC11    PC12   PC13    PC14
## Standard deviation     0.59717 0.55556 0.53665 0.52007 0.50060 0.4900 0.45974
## Proportion of Variance 0.01441 0.01247 0.01164 0.01093 0.01013 0.0097 0.00854
## Cumulative Proportion  0.88029 0.89276 0.90440 0.91533 0.92546 0.9352 0.94370
##                           PC15    PC16   PC17    PC18    PC19    PC20    PC21
## Standard deviation     0.44696 0.43284 0.4009 0.39542 0.38101 0.35781 0.34271
## Proportion of Variance 0.00807 0.00757 0.0065 0.00632 0.00587 0.00517 0.00475
## Cumulative Proportion  0.95178 0.95935 0.9658 0.97216 0.97803 0.98320 0.98795
##                           PC22    PC23    PC24    PC25
## Standard deviation     0.30347 0.28452 0.26505 0.23424
## Proportion of Variance 0.00372 0.00327 0.00284 0.00222
## Cumulative Proportion  0.99167 0.99494 0.99778 1.00000</code></pre>
<blockquote>
<p>How many principal components do you obtain? Compare this to the dimension of the matrix using the <code>dim()</code> function!</p>
</blockquote>
<blockquote>
<p>What would happen if you would not transpose the matrix with t(…) in the prcomp function?</p>
</blockquote>
<p>Principal components are ranked by the amount of variance that they explain. This can be visualized using a <strong>scree plot</strong>, indicating how much variance each PC explains: the <strong>standard deviation</strong> explained by each principal component is contained in the <code>pca$sdev</code> vector:</p>
<div class="sourceCode" id="cb157"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb157-1"><a href="week-one.html#cb157-1" tabindex="-1"></a>pca<span class="sc">$</span>sdev</span></code></pre></div>
<pre><code>##  [1] 3.7190077 2.1886234 0.8979609 0.8619466 0.6932587 0.6279258 0.6164472
##  [8] 0.5971745 0.5555603 0.5366514 0.5200662 0.5006037 0.4899632 0.4597383
## [15] 0.4469622 0.4328397 0.4008966 0.3954205 0.3810141 0.3578084 0.3427089
## [22] 0.3034732 0.2845180 0.2650462 0.2342352</code></pre>
<p>We see that the standard deviation is indeed going down! Let us now plot the proportion of total variance explained by each PC</p>
<div class="sourceCode" id="cb159"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb159-1"><a href="week-one.html#cb159-1" tabindex="-1"></a>variance <span class="ot">=</span> (pca<span class="sc">$</span>sdev)<span class="sc">^</span><span class="dv">2</span></span>
<span id="cb159-2"><a href="week-one.html#cb159-2" tabindex="-1"></a>prop.variance <span class="ot">=</span> variance<span class="sc">/</span><span class="fu">sum</span>(variance)</span>
<span id="cb159-3"><a href="week-one.html#cb159-3" tabindex="-1"></a><span class="fu">names</span>(prop.variance) <span class="ot">=</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(prop.variance)</span>
<span id="cb159-4"><a href="week-one.html#cb159-4" tabindex="-1"></a></span>
<span id="cb159-5"><a href="week-one.html#cb159-5" tabindex="-1"></a><span class="co"># We make a data.frame from the prop.variance and the PC it corresponds to </span></span>
<span id="cb159-6"><a href="week-one.html#cb159-6" tabindex="-1"></a><span class="co"># we can obtain the PCs using names()</span></span>
<span id="cb159-7"><a href="week-one.html#cb159-7" tabindex="-1"></a><span class="fu">data.frame</span>(<span class="at">proportion =</span> prop.variance, </span>
<span id="cb159-8"><a href="week-one.html#cb159-8" tabindex="-1"></a>           <span class="at">PCs =</span> <span class="fu">as.numeric</span>(<span class="fu">names</span>(prop.variance))) <span class="sc">%&gt;%</span></span>
<span id="cb159-9"><a href="week-one.html#cb159-9" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> PCs, <span class="at">y =</span> proportion)) <span class="sc">+</span></span>
<span id="cb159-10"><a href="week-one.html#cb159-10" tabindex="-1"></a>  <span class="fu">geom_col</span>() <span class="sc">+</span>                               <span class="co"># to make the barplot</span></span>
<span id="cb159-11"><a href="week-one.html#cb159-11" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">y=</span><span class="st">&#39;Proportion of variance&#39;</span>)           <span class="co"># we only plot the first 20 PCs</span></span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-83-1.png" width="672" /></p>
<p>Principal component analysis represents each data point (here: patient) in a new space in which the coordinates are principal components. Check the following output:</p>
<div class="sourceCode" id="cb160"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb160-1"><a href="week-one.html#cb160-1" tabindex="-1"></a><span class="fu">head</span>(pca<span class="sc">$</span>x)</span></code></pre></div>
<p>We can now display the data points (i.e. patients) in the first two principal components. In addition, we can color the dots according to certain clinical parameters:</p>
<div class="sourceCode" id="cb161"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb161-1"><a href="week-one.html#cb161-1" tabindex="-1"></a><span class="co"># We start by creating a dataframe and combining it with the annotation</span></span>
<span id="cb161-2"><a href="week-one.html#cb161-2" tabindex="-1"></a>pca_with_annot <span class="ot">=</span> <span class="fu">as.data.frame</span>(pca<span class="sc">$</span>x) <span class="sc">%&gt;%</span></span>
<span id="cb161-3"><a href="week-one.html#cb161-3" tabindex="-1"></a>  <span class="fu">merge</span>(brca.anno, <span class="at">by =</span> <span class="dv">0</span>) <span class="co"># by = 0 makes use of the rownames as common information</span></span>
<span id="cb161-4"><a href="week-one.html#cb161-4" tabindex="-1"></a></span>
<span id="cb161-5"><a href="week-one.html#cb161-5" tabindex="-1"></a><span class="do">## Now the object is in a ggplot2 friendly format</span></span>
<span id="cb161-6"><a href="week-one.html#cb161-6" tabindex="-1"></a><span class="fu">ggplot</span>(pca_with_annot,</span>
<span id="cb161-7"><a href="week-one.html#cb161-7" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> PC1, <span class="at">y =</span> PC2, <span class="at">colour =</span> ER_status)) <span class="sc">+</span></span>
<span id="cb161-8"><a href="week-one.html#cb161-8" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb161-9"><a href="week-one.html#cb161-9" tabindex="-1"></a>  <span class="fu">scale_colour_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">&quot;grey&quot;</span>, <span class="st">&quot;red&quot;</span>, <span class="st">&quot;navy&quot;</span>)) <span class="co"># scale_colour_manual can be used to change colours</span></span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-85-1.png" width="576" /></p>
<blockquote>
<p>Choose different PCs for this plot. Can you still observe the two clusters corresponding to the ER_status of the patients?</p>
</blockquote>
<hr />
</div>
<div id="exercises-1" class="section level3 hasAnchor" number="1.2.5">
<h3><span class="header-section-number">1.2.5</span> EXERCISES<a href="week-one.html#exercises-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="exercise-1-variance" class="section level4 hasAnchor" number="1.2.5.1">
<h4><span class="header-section-number">1.2.5.1</span> Exercise 1: Variance<a href="week-one.html#exercise-1-variance" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p><em>For this exercise set it is given that all the data cleanup steps have been taken, you don’t need to put them in the results.</em></p>
<ol style="list-style-type: decimal">
<li><p>Make a heatmap of the reduced matrix “topVariantGenes” using the <code>pheatmap()</code> function of the <code>pheatmap</code> library (do not forget to select only for numerical columns). Check for parameters that might change the style of the heatmap (column names, row names, etc..). How is this heatmap different from the heatmap in section 2?</p></li>
<li><p>Repeat the selection of top variable genes (apply the same quantile used to generate “topVariantGenes”), but using the median absolute deviation (or MAD) using the <code>mad()</code> function instead of the <code>sd()</code> function, and store into as <code>brca.topMAD</code></p></li>
<li><p>Extract the gene names of <code>topVariantGenes</code> and <code>brca.topMAD</code> and check how many overlap using the <code>intersect()</code> function.</p></li>
</ol>
</div>
<div id="exercise-2-hierarchical-clustering" class="section level4 hasAnchor" number="1.2.5.2">
<h4><span class="header-section-number">1.2.5.2</span> Exercise 2: Hierarchical clustering<a href="week-one.html#exercise-2-hierarchical-clustering" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>In section 2, we have computed a correlation matrix, and used this matrix to build a clustering tree.</p>
<p>Try different linkage methods using the <code>clustering_method</code> parameter to see if the topology of the dendrogram changes!</p>
<ol start="2" style="list-style-type: decimal">
<li>Try building a distance matrix which would lead to different topologies of the dendrogram, depending on which linkage method is used! Show the dendrograms built with different linkage methods!</li>
</ol>
</div>
<div id="exercise-3-pca" class="section level4 hasAnchor" number="1.2.5.3">
<h4><span class="header-section-number">1.2.5.3</span> Exercise 3: PCA<a href="week-one.html#exercise-3-pca" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<ol style="list-style-type: decimal">
<li><p>Display the patients in the first two principal components (available in <code>pca_with_annot</code>) using <code>geom_point()</code>. Color the patients in the PCA plot according to <code>HER2_status</code>.</p></li>
<li><p>(optional) Color the patients in the PCA plot according to <code>Classification</code>; you will probably need to define some more colors… You can check available colors <a href="http://www.stat.columbia.edu/~tzheng/files/Rcolor.pdf">here</a></p></li>
</ol>
</div>
<div id="going-further-expert-1" class="section level4 hasAnchor" number="1.2.5.4">
<h4><span class="header-section-number">1.2.5.4</span> Going further <em>(expert)</em><a href="week-one.html#going-further-expert-1" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Instead of performing the k-means on the whole gene expression matrix, we can run k-means on the space in which each patient is represented by the first k principal components.</p>
<ol style="list-style-type: decimal">
<li><p>Run k-means with different numbers of clusters (1-10) on the patients using the first 2, 4, 6,… principal components (i.e. the first columns of <code>pca_with_annot</code>). Use the elbow method to evaluate how the within sum of squares (WSS) evolves. What is the optimal number of clusters?</p></li>
<li><p>Represent the patients in the PCA plot as previously, but color them according to the cluster they belong to! Run kmeans with two clusters for this and merge the k-means results.</p></li>
</ol>
</div>
</div>
</div>
<div id="day-3-probability-distributions" class="section level2 hasAnchor" number="1.3">
<h2><span class="header-section-number">1.3</span> Day 3: Probability distributions<a href="week-one.html#day-3-probability-distributions" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="probability-distributions" class="section level3 hasAnchor" number="1.3.1">
<h3><span class="header-section-number">1.3.1</span> Probability distributions<a href="week-one.html#probability-distributions" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>In the previous Exercise Sheet we have learnt more about unsupervised learning, like hierarchical clustering and especially PCA. These are among the fundamental methods of Data Analysis. Today we will learn more about another milestone of statistics: <strong>probability distributions</strong>.</p>
<div class="float">
<img src="figures/distrib.png" alt="Various kinds of distributions and the relations among them." />
<div class="figcaption">Various kinds of distributions and the relations among them.</div>
</div>
<p><a href="https://medium.com/@srowen/common-probability-distributions-347e6b945ce4">Figure source</a></p>
<p>Distributions represented by sparse lines represent outcomes which will be discrete (for example the <em>roll of dice</em> will always have discrete integer values from 1 to 6). Distributions represented by dense lines represent outcomes which can be continuous i.e real numbers (for example the height of people living in Heidelberg).</p>
<p>R has in-built functions for almost all probability distributions</p>
<div class="sourceCode" id="cb162"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb162-1"><a href="week-one.html#cb162-1" tabindex="-1"></a><span class="co"># Get the list of all probability distribution related functions</span></span>
<span id="cb162-2"><a href="week-one.html#cb162-2" tabindex="-1"></a><span class="fu">help</span>(distributions)</span></code></pre></div>
<p>All of these functions for probability distributions follow the same common scheme, the <strong>root name</strong> of the function prefixed by either of <strong>p</strong>, <strong>d</strong>, <strong>q</strong> and <strong>r</strong>. For example for the Normal distribution we have the four variants of function available - <strong>pnorm</strong>, <strong>dnorm</strong>, <strong>qnorm</strong> and <strong>rnorm</strong>.</p>
<p>Here, you can find some specific help on these functions:</p>
<div class="sourceCode" id="cb163"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb163-1"><a href="week-one.html#cb163-1" tabindex="-1"></a>?rnorm</span>
<span id="cb163-2"><a href="week-one.html#cb163-2" tabindex="-1"></a><span class="co"># or</span></span>
<span id="cb163-3"><a href="week-one.html#cb163-3" tabindex="-1"></a>?rpois</span>
<span id="cb163-4"><a href="week-one.html#cb163-4" tabindex="-1"></a><span class="co"># etc ...</span></span></code></pre></div>
<p>In the respective help documentation, you will find details on each of the functions.<br />
Probably, the most difficult to distinguish or to remember are <strong>pnorm()</strong> and <strong>qnorm()</strong> (respectively <strong>ppois()</strong> and <strong>qpois()</strong>, etc …).<br />
We are going to look at them more deeply in the following.</p>
<hr />
<div id="getting-to-know-the-various-functions" class="section level4 hasAnchor" number="1.3.1.1">
<h4><span class="header-section-number">1.3.1.1</span> Getting to know the various functions<a href="week-one.html#getting-to-know-the-various-functions" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Let us get a grasp of what these functions actually do. You should be familiar with the cumulative distribution function, let’s take a look at that and its inverse first. We will work with a Normal distribution. We calculate the <strong>cumulative probability</strong> for the values 1,2,3,4 in three different distributions, using one of the functions described previously.<br />
Short hint: <strong>p</strong> like <strong>cumulative P-robability</strong>. Which function are you going to use?</p>
<div class="sourceCode" id="cb164"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb164-1"><a href="week-one.html#cb164-1" tabindex="-1"></a><span class="co"># Distribution 1: Mean = 2, Standard Deviation = 1</span></span>
<span id="cb164-2"><a href="week-one.html#cb164-2" tabindex="-1"></a><span class="fu">pnorm</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>,<span class="at">mean=</span><span class="dv">2</span>,<span class="at">sd=</span><span class="dv">1</span>)</span></code></pre></div>
<pre><code>## [1] 0.1586553 0.5000000 0.8413447 0.9772499</code></pre>
<div class="sourceCode" id="cb166"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb166-1"><a href="week-one.html#cb166-1" tabindex="-1"></a><span class="co"># Distribution 2: Mean = 2, Standard Deviation = 2</span></span>
<span id="cb166-2"><a href="week-one.html#cb166-2" tabindex="-1"></a><span class="fu">pnorm</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>,<span class="at">mean=</span><span class="dv">2</span>,<span class="at">sd=</span><span class="dv">2</span>)</span></code></pre></div>
<pre><code>## [1] 0.3085375 0.5000000 0.6914625 0.8413447</code></pre>
<div class="sourceCode" id="cb168"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb168-1"><a href="week-one.html#cb168-1" tabindex="-1"></a><span class="co"># Distribution 2: Mean = 4, Standard Deviation = 1</span></span>
<span id="cb168-2"><a href="week-one.html#cb168-2" tabindex="-1"></a><span class="fu">pnorm</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>,<span class="at">mean=</span><span class="dv">4</span>,<span class="at">sd=</span><span class="dv">1</span>)</span></code></pre></div>
<pre><code>## [1] 0.001349898 0.022750132 0.158655254 0.500000000</code></pre>
<blockquote>
<p>Do you understand why the cumulative distribution functions change the way they do?</p>
</blockquote>
<p>Now, on the same distributions, we calculate the <strong>inverse cdf</strong> (inverse cumulative distribution function) for the cumulative probabilities of 25%, 50% and 75%. We use the <code>qnorm()</code> function for this (<strong>q- like quantile</strong>):</p>
<div class="sourceCode" id="cb170"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb170-1"><a href="week-one.html#cb170-1" tabindex="-1"></a><span class="co"># Distribution 1: Mean = 2, Standard Deviation = 1</span></span>
<span id="cb170-2"><a href="week-one.html#cb170-2" tabindex="-1"></a><span class="fu">qnorm</span>(<span class="fu">c</span>(<span class="fl">0.25</span>,<span class="fl">0.5</span>,<span class="fl">0.75</span>),<span class="at">mean=</span><span class="dv">2</span>,<span class="at">sd=</span><span class="dv">1</span>)</span></code></pre></div>
<pre><code>## [1] 1.32551 2.00000 2.67449</code></pre>
<div class="sourceCode" id="cb172"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb172-1"><a href="week-one.html#cb172-1" tabindex="-1"></a><span class="co"># Distribution 2: Mean = 2, Standard Deviation = 2</span></span>
<span id="cb172-2"><a href="week-one.html#cb172-2" tabindex="-1"></a><span class="fu">qnorm</span>(<span class="fu">c</span>(<span class="fl">0.25</span>,<span class="fl">0.5</span>,<span class="fl">0.75</span>),<span class="at">mean=</span><span class="dv">2</span>,<span class="at">sd=</span><span class="dv">2</span>)</span></code></pre></div>
<pre><code>## [1] 0.6510205 2.0000000 3.3489795</code></pre>
<div class="sourceCode" id="cb174"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb174-1"><a href="week-one.html#cb174-1" tabindex="-1"></a><span class="co"># Distribution 3: Mean = 4, Standard Deviation = 1</span></span>
<span id="cb174-2"><a href="week-one.html#cb174-2" tabindex="-1"></a><span class="fu">qnorm</span>(<span class="fu">c</span>(<span class="fl">0.25</span>,<span class="fl">0.5</span>,<span class="fl">0.75</span>),<span class="at">mean=</span><span class="dv">4</span>,<span class="at">sd=</span><span class="dv">1</span>)</span></code></pre></div>
<pre><code>## [1] 3.32551 4.00000 4.67449</code></pre>
<blockquote>
<p>Try with 100% on any of the distributions. Can you explain this result? Do you expect the result to be different in the other ones?</p>
</blockquote>
<p>Now that you know the output of the p- and q- functions, let’s look at <strong>d- like density probability</strong> functions. For any continuous distribution, the value of the function at any specific value is 0. This is why this probability function is used in discrete distribution such as the binomial distribution (function <strong>dbinom()</strong>). We first calculate the probability of getting 5 events out of 5 in a binomial distribution with a probability of 0.5. Then, we calculate the odds of <strong>not</strong> getting 5 out of 5.</p>
<div class="sourceCode" id="cb176"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb176-1"><a href="week-one.html#cb176-1" tabindex="-1"></a><span class="co"># probability of 5 out of 5</span></span>
<span id="cb176-2"><a href="week-one.html#cb176-2" tabindex="-1"></a><span class="fu">dbinom</span>(<span class="dv">5</span>, <span class="at">size=</span><span class="dv">5</span>, <span class="at">prob=</span><span class="fl">0.5</span>) <span class="co"># size = number of trials; prob = probability of success on each trial</span></span></code></pre></div>
<pre><code>## [1] 0.03125</code></pre>
<div class="sourceCode" id="cb178"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb178-1"><a href="week-one.html#cb178-1" tabindex="-1"></a><span class="co"># probability of NOT getting 5 out of 5</span></span>
<span id="cb178-2"><a href="week-one.html#cb178-2" tabindex="-1"></a><span class="dv">1</span><span class="sc">-</span><span class="fu">dbinom</span>(<span class="dv">5</span>, <span class="at">size=</span><span class="dv">5</span>, <span class="at">prob=</span><span class="fl">0.5</span>) </span></code></pre></div>
<pre><code>## [1] 0.96875</code></pre>
<div class="sourceCode" id="cb180"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb180-1"><a href="week-one.html#cb180-1" tabindex="-1"></a><span class="co"># or</span></span>
<span id="cb180-2"><a href="week-one.html#cb180-2" tabindex="-1"></a><span class="fu">pbinom</span>(<span class="dv">4</span>, <span class="at">size=</span><span class="dv">5</span>, <span class="at">prob=</span><span class="fl">0.5</span>) <span class="co"># Remember, pbinom() returns the &quot;cumulative probability&quot; of ...</span></span></code></pre></div>
<pre><code>## [1] 0.96875</code></pre>
<blockquote>
<p>What is the probability of getting 5 out of 10? And NOT getting 5 out of 10?</p>
</blockquote>
<p>Suppose that the distribution of body height is described by a <strong>normal distribution</strong> with <strong>mean = 1.75 m</strong> and standard deviation <strong>sd = 0.15</strong>. What is the probability that someone is taller than 1.9 m? We can use the parameter <strong>lower.tail</strong> of the function <code>pnorm()</code> for this.<br />
What is the probability that someone is smaller than 1.60m?</p>
<div class="sourceCode" id="cb182"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb182-1"><a href="week-one.html#cb182-1" tabindex="-1"></a><span class="do">## taller than 1.9</span></span>
<span id="cb182-2"><a href="week-one.html#cb182-2" tabindex="-1"></a><span class="fu">pnorm</span>(<span class="fl">1.9</span>, <span class="at">mean=</span><span class="fl">1.75</span>, <span class="at">sd=</span><span class="fl">0.15</span>, <span class="at">lower.tail=</span><span class="cn">FALSE</span>)</span></code></pre></div>
<pre><code>## [1] 0.1586553</code></pre>
<div class="sourceCode" id="cb184"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb184-1"><a href="week-one.html#cb184-1" tabindex="-1"></a><span class="co"># or</span></span>
<span id="cb184-2"><a href="week-one.html#cb184-2" tabindex="-1"></a><span class="dv">1</span><span class="sc">-</span><span class="fu">pnorm</span>(<span class="fl">1.9</span>, <span class="at">mean=</span><span class="fl">1.75</span>, <span class="at">sd=</span><span class="fl">0.15</span>)</span></code></pre></div>
<pre><code>## [1] 0.1586553</code></pre>
<div class="sourceCode" id="cb186"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb186-1"><a href="week-one.html#cb186-1" tabindex="-1"></a><span class="do">## smaller than 1.6</span></span>
<span id="cb186-2"><a href="week-one.html#cb186-2" tabindex="-1"></a><span class="fu">pnorm</span>(<span class="fl">1.6</span>, <span class="at">mean=</span><span class="fl">1.75</span>, <span class="at">sd=</span><span class="fl">0.15</span>, <span class="at">lower.tail=</span><span class="cn">TRUE</span>)</span></code></pre></div>
<pre><code>## [1] 0.1586553</code></pre>
<div class="sourceCode" id="cb188"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb188-1"><a href="week-one.html#cb188-1" tabindex="-1"></a><span class="co"># equivalent to </span></span>
<span id="cb188-2"><a href="week-one.html#cb188-2" tabindex="-1"></a><span class="fu">pnorm</span>(<span class="fl">1.6</span>, <span class="at">mean=</span><span class="fl">1.75</span>, <span class="at">sd=</span><span class="fl">0.15</span>) <span class="co"># lower.tail is set to TRUE by default (check the help page using `?pnorm`)</span></span></code></pre></div>
<pre><code>## [1] 0.1586553</code></pre>
<p>Let’s have a look at this distribution using the <strong>dnorm()</strong> function:</p>
<div class="sourceCode" id="cb190"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb190-1"><a href="week-one.html#cb190-1" tabindex="-1"></a>x <span class="ot">=</span> <span class="fu">seq</span>(<span class="dv">1</span>, <span class="fl">2.5</span>, <span class="at">by=</span><span class="fl">0.01</span>)</span>
<span id="cb190-2"><a href="week-one.html#cb190-2" tabindex="-1"></a>y <span class="ot">=</span> <span class="fu">dnorm</span>(x, <span class="at">mean=</span><span class="fl">1.75</span>, <span class="at">sd=</span><span class="fl">0.15</span>)</span>
<span id="cb190-3"><a href="week-one.html#cb190-3" tabindex="-1"></a></span>
<span id="cb190-4"><a href="week-one.html#cb190-4" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb190-5"><a href="week-one.html#cb190-5" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">x =</span> x, <span class="at">y =</span> y),</span>
<span id="cb190-6"><a href="week-one.html#cb190-6" tabindex="-1"></a>            <span class="at">colour =</span> <span class="st">&quot;red&quot;</span>, <span class="at">linewidth =</span> <span class="fl">0.8</span>) <span class="sc">+</span></span>
<span id="cb190-7"><a href="week-one.html#cb190-7" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">c</span>(<span class="fl">1.6</span>, <span class="fl">1.9</span>),</span>
<span id="cb190-8"><a href="week-one.html#cb190-8" tabindex="-1"></a>             <span class="at">linetype =</span> <span class="st">&#39;dashed&#39;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-92-1.png" width="672" /></p>
<p>Let’s finally look at <strong>r- like random</strong> functions. These, unlike the others, don’t return single probabilities or values but rather generate a random distribution of values. We can use this to generate a normal distribution with mean = 10, sd = 5.</p>
<div class="sourceCode" id="cb191"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb191-1"><a href="week-one.html#cb191-1" tabindex="-1"></a><span class="do">## normal distribution</span></span>
<span id="cb191-2"><a href="week-one.html#cb191-2" tabindex="-1"></a>x <span class="ot">=</span> <span class="fu">rnorm</span>(<span class="at">n=</span><span class="dv">1000</span>,<span class="at">mean=</span><span class="dv">10</span>,<span class="at">sd=</span><span class="dv">5</span>)</span>
<span id="cb191-3"><a href="week-one.html#cb191-3" tabindex="-1"></a></span>
<span id="cb191-4"><a href="week-one.html#cb191-4" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb191-5"><a href="week-one.html#cb191-5" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="fu">aes</span>(<span class="at">x=</span>x), <span class="at">bins =</span> <span class="dv">20</span>,</span>
<span id="cb191-6"><a href="week-one.html#cb191-6" tabindex="-1"></a>                 <span class="at">fill =</span> <span class="st">&quot;orange&quot;</span>) <span class="sc">+</span></span>
<span id="cb191-7"><a href="week-one.html#cb191-7" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">y =</span> <span class="st">&quot;Frequency&quot;</span>) </span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-93-1.png" width="576" /></p>
<blockquote>
<p>Can you generate a Poisson distribution and a binomial distribution?</p>
</blockquote>
<hr />
</div>
<div id="normalgaussian-distribution" class="section level4 hasAnchor" number="1.3.1.2">
<h4><span class="header-section-number">1.3.1.2</span> Normal/Gaussian distribution<a href="week-one.html#normalgaussian-distribution" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>The normal or the Gaussian distribution is given as:</p>
<p><span class="math display">\[P(x) = \frac{1}{{\sigma \sqrt {2\pi } }} \cdot e ^ \frac{-(x- \mu)^2}{{2\sigma ^2 }} \]</span> where <span class="math inline">\(\mu\)</span> is the mean of the distribution and <span class="math inline">\(\sigma\)</span> is the standard deviation.</p>
<p>The <strong>standard normal distribution</strong> is a special case of <strong>normal distribution</strong> where the values for <span class="math inline">\(\mu = 0\)</span> and <span class="math inline">\(\sigma = 1\)</span>. Thus, the above equation for the Normal distribution simplifies to:</p>
<p><span class="math display">\[P(x) = \frac{1}{{\sqrt {2\pi } }} \cdot e ^ \frac{-x^2}{2} \]</span> Now for any <span class="math inline">\(x\)</span> we can easily solve this equation since <span class="math inline">\(\pi\)</span> and <span class="math inline">\(e\)</span> are known constants.</p>
<hr />
</div>
<div id="visualization" class="section level4 hasAnchor" number="1.3.1.3">
<h4><span class="header-section-number">1.3.1.3</span> Visualization<a href="week-one.html#visualization" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Let’s generate three random normal distributions with different means and standard deviations and visualize them together</p>
<div class="sourceCode" id="cb192"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb192-1"><a href="week-one.html#cb192-1" tabindex="-1"></a><span class="do">## Use the function `dnorm()` to plot the density distribution</span></span>
<span id="cb192-2"><a href="week-one.html#cb192-2" tabindex="-1"></a>x <span class="ot">=</span> <span class="fu">seq</span>(<span class="sc">-</span><span class="dv">10</span>, <span class="dv">30</span>, <span class="at">by=</span>.<span class="dv">1</span>)</span>
<span id="cb192-3"><a href="week-one.html#cb192-3" tabindex="-1"></a>d1 <span class="ot">=</span> <span class="fu">dnorm</span>(x, <span class="at">mean=</span><span class="dv">0</span>, <span class="at">sd=</span><span class="dv">1</span>)</span>
<span id="cb192-4"><a href="week-one.html#cb192-4" tabindex="-1"></a>d2 <span class="ot">=</span> <span class="fu">dnorm</span>(x, <span class="at">mean=</span><span class="dv">10</span>, <span class="at">sd=</span><span class="dv">1</span>)</span>
<span id="cb192-5"><a href="week-one.html#cb192-5" tabindex="-1"></a>d3 <span class="ot">=</span> <span class="fu">dnorm</span>(x, <span class="at">mean=</span><span class="dv">20</span>, <span class="at">sd=</span><span class="dv">1</span>)</span>
<span id="cb192-6"><a href="week-one.html#cb192-6" tabindex="-1"></a></span>
<span id="cb192-7"><a href="week-one.html#cb192-7" tabindex="-1"></a><span class="co"># Compare with the histogram build from 1000 random number drawn from the standard normal distribution</span></span>
<span id="cb192-8"><a href="week-one.html#cb192-8" tabindex="-1"></a>r1 <span class="ot">=</span> <span class="fu">rnorm</span>(<span class="at">n=</span><span class="dv">1000</span>, <span class="at">mean=</span><span class="dv">0</span>, <span class="at">sd =</span><span class="dv">1</span>)  <span class="co"># random distributions of values</span></span>
<span id="cb192-9"><a href="week-one.html#cb192-9" tabindex="-1"></a>r2 <span class="ot">=</span> <span class="fu">rnorm</span>(<span class="at">n=</span><span class="dv">1000</span>, <span class="at">mean=</span><span class="dv">10</span>, <span class="at">sd=</span><span class="dv">1</span>)</span>
<span id="cb192-10"><a href="week-one.html#cb192-10" tabindex="-1"></a>r3 <span class="ot">=</span> <span class="fu">rnorm</span>(<span class="at">n=</span><span class="dv">1000</span>, <span class="at">mean=</span><span class="dv">20</span>, <span class="at">sd=</span><span class="dv">1</span>)</span>
<span id="cb192-11"><a href="week-one.html#cb192-11" tabindex="-1"></a></span>
<span id="cb192-12"><a href="week-one.html#cb192-12" tabindex="-1"></a><span class="co"># Histogram visualization</span></span>
<span id="cb192-13"><a href="week-one.html#cb192-13" tabindex="-1"></a><span class="co"># We will start by converting the rnorm outputs into one single df</span></span>
<span id="cb192-14"><a href="week-one.html#cb192-14" tabindex="-1"></a><span class="fu">data.frame</span>(<span class="at">vals =</span> <span class="fu">c</span>(r1, r2, r3), <span class="co"># column vals is actual numbers</span></span>
<span id="cb192-15"><a href="week-one.html#cb192-15" tabindex="-1"></a>           <span class="at">Specs =</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">&quot;mean=0, sd=1&quot;</span>, <span class="dv">1000</span>), <span class="co"># column Specs is mean/sd </span></span>
<span id="cb192-16"><a href="week-one.html#cb192-16" tabindex="-1"></a>                     <span class="fu">rep</span>(<span class="st">&quot;mean=10, sd=1&quot;</span>, <span class="dv">1000</span>),</span>
<span id="cb192-17"><a href="week-one.html#cb192-17" tabindex="-1"></a>                     <span class="fu">rep</span>(<span class="st">&quot;mean=20, sd=1&quot;</span>, <span class="dv">1000</span>))</span>
<span id="cb192-18"><a href="week-one.html#cb192-18" tabindex="-1"></a>             ) <span class="sc">%&gt;%</span></span>
<span id="cb192-19"><a href="week-one.html#cb192-19" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> vals, </span>
<span id="cb192-20"><a href="week-one.html#cb192-20" tabindex="-1"></a>             <span class="at">colour =</span> Specs)) <span class="sc">+</span></span>
<span id="cb192-21"><a href="week-one.html#cb192-21" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins =</span> <span class="dv">50</span>) <span class="sc">+</span> </span>
<span id="cb192-22"><a href="week-one.html#cb192-22" tabindex="-1"></a>  <span class="fu">geom_freqpoly</span>(<span class="at">bins =</span> <span class="dv">50</span>) <span class="co"># geom_freqpoly adds in the line on top of the histogram</span></span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-94-1.png" width="672" /></p>
<blockquote>
<p>Play with the mean and sd parameters and visualize the distributions (plain lines) as well as the corresponding histograms.</p>
</blockquote>
<hr />
</div>
<div id="application-on-a-real-dataset" class="section level4 hasAnchor" number="1.3.1.4">
<h4><span class="header-section-number">1.3.1.4</span> Application on a real dataset<a href="week-one.html#application-on-a-real-dataset" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Now we will use the Normal distribution to make predictions about gene expression of TP53 in lung cancer. TP53 is the most commonly mutated gene across multiple cancer types especially in lung cancers. We will read a table (import) containing measurements of TP53 expression levels in 586 patients.</p>
<div class="sourceCode" id="cb193"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb193-1"><a href="week-one.html#cb193-1" tabindex="-1"></a>tp53.exp <span class="ot">=</span> <span class="fu">read.table</span>(<span class="st">&quot;https://www.dropbox.com/s/rwopdr8ycmdg8bd/TP53_expression_LungAdeno.txt?dl=1&quot;</span>, </span>
<span id="cb193-2"><a href="week-one.html#cb193-2" tabindex="-1"></a>                      <span class="at">header=</span>T, <span class="at">sep=</span><span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>)[,<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]</span>
<span id="cb193-3"><a href="week-one.html#cb193-3" tabindex="-1"></a><span class="fu">summary</span>(tp53.exp)</span></code></pre></div>
<pre><code>##    Samples          TP53_expression 
##  Length:586         Min.   :  92.6  
##  Class :character   1st Qu.: 911.8  
##  Mode  :character   Median :1313.3  
##                     Mean   :1380.8  
##                     3rd Qu.:1778.1  
##                     Max.   :4703.9  
##                     NA&#39;s   :69</code></pre>
<div id="data-cleaning-and-central-values" class="section level5 hasAnchor" number="1.3.1.4.1">
<h5><span class="header-section-number">1.3.1.4.1</span> Data cleaning and central values<a href="week-one.html#data-cleaning-and-central-values" class="anchor-section" aria-label="Anchor link to header"></a></h5>
<p>We will remove all the missing values and calculate the mean and standard deviation for the TP53 gene expression.</p>
<div class="sourceCode" id="cb195"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb195-1"><a href="week-one.html#cb195-1" tabindex="-1"></a>tp53.exp <span class="ot">=</span> tp53.exp <span class="sc">%&gt;%</span></span>
<span id="cb195-2"><a href="week-one.html#cb195-2" tabindex="-1"></a>  <span class="fu">na.omit</span>() </span>
<span id="cb195-3"><a href="week-one.html#cb195-3" tabindex="-1"></a></span>
<span id="cb195-4"><a href="week-one.html#cb195-4" tabindex="-1"></a>m.tp53 <span class="ot">=</span> <span class="fu">mean</span>(tp53.exp<span class="sc">$</span>TP53_expression) <span class="co"># mean</span></span>
<span id="cb195-5"><a href="week-one.html#cb195-5" tabindex="-1"></a>m.tp53</span></code></pre></div>
<pre><code>## [1] 1380.822</code></pre>
<div class="sourceCode" id="cb197"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb197-1"><a href="week-one.html#cb197-1" tabindex="-1"></a>s.tp53 <span class="ot">=</span> <span class="fu">sd</span>(tp53.exp<span class="sc">$</span>TP53_expression) <span class="co"># standard deviation</span></span>
<span id="cb197-2"><a href="week-one.html#cb197-2" tabindex="-1"></a>s.tp53</span></code></pre></div>
<pre><code>## [1] 719.5934</code></pre>
</div>
<div id="modeling-using-a-normal-distribution" class="section level5 hasAnchor" number="1.3.1.4.2">
<h5><span class="header-section-number">1.3.1.4.2</span> Modeling using a normal distribution<a href="week-one.html#modeling-using-a-normal-distribution" class="anchor-section" aria-label="Anchor link to header"></a></h5>
<p>Let’s see how well a normal distribution with <span class="math inline">\(\mu = 1380.822\)</span> (m.tp53) and <span class="math inline">\(\sigma = 719.5934\)</span> (s.tp53) can approximate the <strong>real</strong> distribution of TP53 expression.<br />
We assume that the population mean and standard deviation is similar as calculated above since we cannot measure the expression of TP53 in each and every lung cancer patient in the world.</p>
<div class="sourceCode" id="cb199"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb199-1"><a href="week-one.html#cb199-1" tabindex="-1"></a><span class="co"># distribution of the measured data</span></span>
<span id="cb199-2"><a href="week-one.html#cb199-2" tabindex="-1"></a><span class="fu">ggplot</span>(tp53.exp,</span>
<span id="cb199-3"><a href="week-one.html#cb199-3" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> TP53_expression)) <span class="sc">+</span></span>
<span id="cb199-4"><a href="week-one.html#cb199-4" tabindex="-1"></a>  <span class="fu">geom_density</span>() <span class="sc">+</span> </span>
<span id="cb199-5"><a href="week-one.html#cb199-5" tabindex="-1"></a>  <span class="fu">xlim</span>(<span class="sc">-</span><span class="dv">1500</span>, <span class="dv">6000</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-97-1.png" width="672" /></p>
<p>Make a normal distribution with the above parameters</p>
<div class="sourceCode" id="cb200"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb200-1"><a href="week-one.html#cb200-1" tabindex="-1"></a>x <span class="ot">=</span> <span class="fu">seq</span>(<span class="dv">0</span>,<span class="dv">5000</span>,<span class="at">by=</span><span class="dv">5</span>)</span>
<span id="cb200-2"><a href="week-one.html#cb200-2" tabindex="-1"></a>d.pred <span class="ot">=</span> <span class="fu">dnorm</span>(x, <span class="at">mean =</span> m.tp53, <span class="at">sd =</span> s.tp53)</span>
<span id="cb200-3"><a href="week-one.html#cb200-3" tabindex="-1"></a></span>
<span id="cb200-4"><a href="week-one.html#cb200-4" tabindex="-1"></a><span class="co"># Now plot both, predicted and measured data</span></span>
<span id="cb200-5"><a href="week-one.html#cb200-5" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb200-6"><a href="week-one.html#cb200-6" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="fu">aes</span>(<span class="at">x =</span> tp53.exp<span class="sc">$</span>TP53_expression)) <span class="sc">+</span></span>
<span id="cb200-7"><a href="week-one.html#cb200-7" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">x =</span> x, </span>
<span id="cb200-8"><a href="week-one.html#cb200-8" tabindex="-1"></a>                <span class="at">y =</span> d.pred),</span>
<span id="cb200-9"><a href="week-one.html#cb200-9" tabindex="-1"></a>                <span class="at">colour =</span> <span class="st">&quot;red&quot;</span>) <span class="sc">+</span></span>
<span id="cb200-10"><a href="week-one.html#cb200-10" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">c</span>(<span class="fu">quantile</span>(tp53.exp<span class="sc">$</span>TP53_expression, <span class="at">probs =</span> <span class="fu">c</span>(<span class="fl">0.1</span>, <span class="fl">0.9</span>)))) <span class="sc">+</span></span>
<span id="cb200-11"><a href="week-one.html#cb200-11" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">c</span>(<span class="fu">qnorm</span>(<span class="at">p =</span> <span class="fl">0.1</span>, <span class="at">mean =</span> m.tp53, <span class="at">sd =</span> s.tp53), </span>
<span id="cb200-12"><a href="week-one.html#cb200-12" tabindex="-1"></a>                            <span class="fu">qnorm</span>(<span class="at">p =</span> <span class="fl">0.9</span>, <span class="at">mean =</span> m.tp53, <span class="at">sd =</span> s.tp53)),</span>
<span id="cb200-13"><a href="week-one.html#cb200-13" tabindex="-1"></a>             <span class="at">colour =</span> <span class="st">&#39;red&#39;</span>, <span class="at">linetype =</span> <span class="st">&#39;dotted&#39;</span>) </span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-98-1.png" width="672" /></p>
</div>
<div id="data-prediction-using-the-normal-distribution-model" class="section level5 hasAnchor" number="1.3.1.4.3">
<h5><span class="header-section-number">1.3.1.4.3</span> Data prediction using the normal distribution model<a href="week-one.html#data-prediction-using-the-normal-distribution-model" class="anchor-section" aria-label="Anchor link to header"></a></h5>
<p>Using a normal distribution with <span class="math inline">\(\mu = 1380.822\)</span> (m.tp53) and <span class="math inline">\(\sigma = 719.5934\)</span> (s.tp53), we will ask the following questions -</p>
<p>– <strong>(Q1)</strong> What is the probability of observing the expression of TP53 to be <strong>less</strong> than 1000?</p>
<div class="sourceCode" id="cb201"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb201-1"><a href="week-one.html#cb201-1" tabindex="-1"></a><span class="fu">pnorm</span>(<span class="at">q=</span><span class="dv">1000</span>, <span class="at">mean =</span>m.tp53, <span class="at">sd =</span> s.tp53) <span class="co"># returns the cumulative probability</span></span></code></pre></div>
<pre><code>## [1] 0.2983271</code></pre>
<p>– <strong>(Q2)</strong> What is the probability of observing the expression of TP53 to be <strong>greater</strong> than 1000?</p>
<div class="sourceCode" id="cb203"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb203-1"><a href="week-one.html#cb203-1" tabindex="-1"></a><span class="dv">1</span> <span class="sc">-</span> <span class="fu">pnorm</span>(<span class="at">q=</span><span class="dv">1000</span>, <span class="at">mean =</span>m.tp53, <span class="at">sd =</span> s.tp53) </span></code></pre></div>
<pre><code>## [1] 0.7016729</code></pre>
<div class="sourceCode" id="cb205"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb205-1"><a href="week-one.html#cb205-1" tabindex="-1"></a><span class="co"># is same as</span></span>
<span id="cb205-2"><a href="week-one.html#cb205-2" tabindex="-1"></a><span class="fu">pnorm</span>(<span class="at">q=</span><span class="dv">1000</span>, <span class="at">mean =</span>m.tp53, <span class="at">sd =</span> s.tp53, <span class="at">lower.tail =</span> <span class="cn">FALSE</span>) <span class="co"># or pnorm(q=1000, mean =m.tp53, sd = s.tp53, lower.tail = F)</span></span></code></pre></div>
<pre><code>## [1] 0.7016729</code></pre>
</div>
<div id="evaluating-the-quality-of-the-predictions" class="section level5 hasAnchor" number="1.3.1.4.4">
<h5><span class="header-section-number">1.3.1.4.4</span> Evaluating the quality of the predictions<a href="week-one.html#evaluating-the-quality-of-the-predictions" class="anchor-section" aria-label="Anchor link to header"></a></h5>
<p>Let’s check how good these predictions are compared to <strong>real</strong> data.
– <strong>(Q1)</strong> What is the probability of observing the expression of TP53 to be <strong>less</strong> than 1000?</p>
<div class="sourceCode" id="cb207"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb207-1"><a href="week-one.html#cb207-1" tabindex="-1"></a><span class="fu">sum</span>(tp53.exp<span class="sc">$</span>TP53_expression <span class="sc">&lt;</span> <span class="dv">1000</span>)<span class="sc">/</span><span class="fu">nrow</span>(tp53.exp)</span></code></pre></div>
<pre><code>## [1] 0.2978723</code></pre>
<p>– <strong>(Q2)</strong> What is the probability of observing the expression of TP53 to be <strong>greater</strong> than 1000?</p>
<div class="sourceCode" id="cb209"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb209-1"><a href="week-one.html#cb209-1" tabindex="-1"></a><span class="fu">sum</span>(tp53.exp<span class="sc">$</span>TP53_expression <span class="sc">&gt;</span> <span class="dv">1000</span>)<span class="sc">/</span><span class="fu">nrow</span>(tp53.exp)</span></code></pre></div>
<pre><code>## [1] 0.7021277</code></pre>
<blockquote>
<p>I would say those predictions are pretty good !! Now, let’s try to break this model. Re-execute the code above with different <span class="math inline">\(q\)</span> values <code>q=100, q=500, q=4000, q=4500</code> etc. At what values do you think the model would not perform well. HINT: Look at the tails of the distribution!</p>
</blockquote>
<div class="sourceCode" id="cb211"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb211-1"><a href="week-one.html#cb211-1" tabindex="-1"></a><span class="co"># What is the probability of observing the expression of TP53 to be less than q?</span></span>
<span id="cb211-2"><a href="week-one.html#cb211-2" tabindex="-1"></a>q <span class="ot">=</span> <span class="fu">c</span>(<span class="dv">100</span>,<span class="dv">500</span>,<span class="dv">1000</span>,<span class="dv">4000</span>,<span class="dv">4500</span>)</span>
<span id="cb211-3"><a href="week-one.html#cb211-3" tabindex="-1"></a></span>
<span id="cb211-4"><a href="week-one.html#cb211-4" tabindex="-1"></a><span class="co"># for loop used to calculate and store the predicted (a) and real (b) values</span></span>
<span id="cb211-5"><a href="week-one.html#cb211-5" tabindex="-1"></a>pred <span class="ot">=</span> <span class="fu">numeric</span>()</span>
<span id="cb211-6"><a href="week-one.html#cb211-6" tabindex="-1"></a>meas <span class="ot">=</span> <span class="fu">numeric</span>()</span>
<span id="cb211-7"><a href="week-one.html#cb211-7" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(q)){</span>
<span id="cb211-8"><a href="week-one.html#cb211-8" tabindex="-1"></a>  pred[i] <span class="ot">=</span> <span class="fu">pnorm</span>(<span class="at">q=</span>q[i], <span class="at">mean =</span> m.tp53, <span class="at">sd =</span> s.tp53) </span>
<span id="cb211-9"><a href="week-one.html#cb211-9" tabindex="-1"></a>  meas[i] <span class="ot">=</span> <span class="fu">sum</span>(tp53.exp<span class="sc">$</span>TP53_expression <span class="sc">&lt;</span> q[i])<span class="sc">/</span><span class="fu">nrow</span>(tp53.exp)</span>
<span id="cb211-10"><a href="week-one.html#cb211-10" tabindex="-1"></a>}</span>
<span id="cb211-11"><a href="week-one.html#cb211-11" tabindex="-1"></a></span>
<span id="cb211-12"><a href="week-one.html#cb211-12" tabindex="-1"></a><span class="co"># Change into a dataframe </span></span>
<span id="cb211-13"><a href="week-one.html#cb211-13" tabindex="-1"></a>model_mat <span class="ot">=</span> <span class="fu">data.frame</span>(pred, meas, q)</span>
<span id="cb211-14"><a href="week-one.html#cb211-14" tabindex="-1"></a>model_mat</span></code></pre></div>
<pre><code>##         pred        meas    q
## 1 0.03754417 0.001934236  100
## 2 0.11046576 0.104448743  500
## 3 0.29832708 0.297872340 1000
## 4 0.99986358 0.992263056 4000
## 5 0.99999270 0.996131528 4500</code></pre>
<p>Again, using a normal distribution with <span class="math inline">\(\mu = 1380.822\)</span> and <span class="math inline">\(\sigma = 719.5934\)</span>, what if we ask what is the value of TP53 expression at the 10% and 90% quantiles:</p>
<div class="sourceCode" id="cb213"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb213-1"><a href="week-one.html#cb213-1" tabindex="-1"></a><span class="fu">qnorm</span>(<span class="at">p =</span> <span class="fl">0.1</span>, <span class="at">mean =</span> m.tp53, <span class="at">sd =</span> s.tp53)</span></code></pre></div>
<pre><code>## [1] 458.6258</code></pre>
<div class="sourceCode" id="cb215"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb215-1"><a href="week-one.html#cb215-1" tabindex="-1"></a><span class="fu">qnorm</span>(<span class="at">p =</span> <span class="fl">0.9</span>, <span class="at">mean =</span> m.tp53, <span class="at">sd =</span> s.tp53)</span></code></pre></div>
<pre><code>## [1] 2303.018</code></pre>
<p>Let’s check how good these predictions are compared to our real data.</p>
<div class="sourceCode" id="cb217"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb217-1"><a href="week-one.html#cb217-1" tabindex="-1"></a><span class="fu">quantile</span>(tp53.exp<span class="sc">$</span>TP53_expression, </span>
<span id="cb217-2"><a href="week-one.html#cb217-2" tabindex="-1"></a>         <span class="at">probs =</span> <span class="fu">c</span>(<span class="fl">0.1</span>, <span class="fl">0.9</span>))</span></code></pre></div>
<pre><code>##       10%       90% 
##  478.2844 2240.1913</code></pre>
<p>Again the predictions are pretty good!</p>
</div>
<div id="visualization-1" class="section level5 hasAnchor" number="1.3.1.4.5">
<h5><span class="header-section-number">1.3.1.4.5</span> Visualization<a href="week-one.html#visualization-1" class="anchor-section" aria-label="Anchor link to header"></a></h5>
<p>We can also visualize all of this on a simple graph:</p>
<div class="sourceCode" id="cb219"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb219-1"><a href="week-one.html#cb219-1" tabindex="-1"></a><span class="co"># Model prediction:</span></span>
<span id="cb219-2"><a href="week-one.html#cb219-2" tabindex="-1"></a>x <span class="ot">=</span> <span class="fu">seq</span>(<span class="dv">0</span>,<span class="dv">5000</span>,<span class="at">by=</span><span class="dv">5</span>)</span>
<span id="cb219-3"><a href="week-one.html#cb219-3" tabindex="-1"></a>d.pred <span class="ot">=</span> <span class="fu">dnorm</span>(x,<span class="at">mean =</span> m.tp53, <span class="at">sd =</span> s.tp53)</span>
<span id="cb219-4"><a href="week-one.html#cb219-4" tabindex="-1"></a></span>
<span id="cb219-5"><a href="week-one.html#cb219-5" tabindex="-1"></a><span class="co"># Model and measured data and predicted versus measured 0.1 and 0.9 quantiles:</span></span>
<span id="cb219-6"><a href="week-one.html#cb219-6" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb219-7"><a href="week-one.html#cb219-7" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">x =</span> x, </span>
<span id="cb219-8"><a href="week-one.html#cb219-8" tabindex="-1"></a>                <span class="at">y =</span> d.pred),</span>
<span id="cb219-9"><a href="week-one.html#cb219-9" tabindex="-1"></a>            <span class="at">colour =</span> <span class="st">&quot;red&quot;</span>) <span class="sc">+</span></span>
<span id="cb219-10"><a href="week-one.html#cb219-10" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="fu">aes</span>(<span class="at">x =</span> tp53.exp<span class="sc">$</span>TP53_expression)) </span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-106-1.png" width="672" /></p>
<p>Compare the black and red vertical lines (real vs predicted).</p>
<blockquote>
<p>Re-execute the code above with <code>p=0.25, p=0.5, p=0.75 etc</code> and check how good the predictions are.</p>
</blockquote>
</div>
<div id="visualization-using-a-q-q-plot" class="section level5 hasAnchor" number="1.3.1.4.6">
<h5><span class="header-section-number">1.3.1.4.6</span> Visualization using a Q-Q plot<a href="week-one.html#visualization-using-a-q-q-plot" class="anchor-section" aria-label="Anchor link to header"></a></h5>
<p>Now, let’s plot the sample quantiles against theoretical quantiles to check the similarity between the two. This is called a quantile - quantile plot or a Q-Q plot, which you are familiar with (see Exercises Sheet 1).</p>
<div class="sourceCode" id="cb220"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb220-1"><a href="week-one.html#cb220-1" tabindex="-1"></a>q <span class="ot">=</span> <span class="fu">seq</span>(<span class="dv">0</span>,<span class="dv">1</span>,<span class="fl">0.01</span>) <span class="co"># Creating a vector of quantiles</span></span>
<span id="cb220-2"><a href="week-one.html#cb220-2" tabindex="-1"></a></span>
<span id="cb220-3"><a href="week-one.html#cb220-3" tabindex="-1"></a><span class="co"># Find values corresponding to these quantiles in the real data</span></span>
<span id="cb220-4"><a href="week-one.html#cb220-4" tabindex="-1"></a>q.observed <span class="ot">=</span> <span class="fu">quantile</span>(tp53.exp<span class="sc">$</span>TP53_expression, <span class="at">probs =</span> q)</span>
<span id="cb220-5"><a href="week-one.html#cb220-5" tabindex="-1"></a></span>
<span id="cb220-6"><a href="week-one.html#cb220-6" tabindex="-1"></a><span class="co"># Find values corresponding to these quantiles in the theoretical normal distribution</span></span>
<span id="cb220-7"><a href="week-one.html#cb220-7" tabindex="-1"></a>q.theoretical <span class="ot">=</span> <span class="fu">qnorm</span>(<span class="at">p =</span> q, <span class="at">mean =</span> m.tp53, <span class="at">sd =</span> s.tp53)</span>
<span id="cb220-8"><a href="week-one.html#cb220-8" tabindex="-1"></a></span>
<span id="cb220-9"><a href="week-one.html#cb220-9" tabindex="-1"></a><span class="co"># # Correlate the above two values</span></span>
<span id="cb220-10"><a href="week-one.html#cb220-10" tabindex="-1"></a><span class="fu">ggplot</span>(tp53.exp,</span>
<span id="cb220-11"><a href="week-one.html#cb220-11" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">sample =</span> TP53_expression)) <span class="sc">+</span></span>
<span id="cb220-12"><a href="week-one.html#cb220-12" tabindex="-1"></a>  <span class="fu">geom_qq</span>() <span class="sc">+</span></span>
<span id="cb220-13"><a href="week-one.html#cb220-13" tabindex="-1"></a>  <span class="fu">geom_qq_line</span>(<span class="at">colour =</span> <span class="st">&#39;red&#39;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-107-1.png" width="480" /></p>
<div class="sourceCode" id="cb221"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb221-1"><a href="week-one.html#cb221-1" tabindex="-1"></a><span class="do">## Would be the same as:</span></span>
<span id="cb221-2"><a href="week-one.html#cb221-2" tabindex="-1"></a><span class="co"># ggplot() +</span></span>
<span id="cb221-3"><a href="week-one.html#cb221-3" tabindex="-1"></a><span class="co">#   geom_point(aes(x = q.theoretical, </span></span>
<span id="cb221-4"><a href="week-one.html#cb221-4" tabindex="-1"></a><span class="co">#                  y = q.observed), size = 1) +</span></span>
<span id="cb221-5"><a href="week-one.html#cb221-5" tabindex="-1"></a><span class="co">#   geom_abline(intercept = 0, slope = 1, </span></span>
<span id="cb221-6"><a href="week-one.html#cb221-6" tabindex="-1"></a><span class="co">#               size = 1, color = &quot;red&quot;) </span></span></code></pre></div>
<hr />
</div>
</div>
<div id="binomial-distribution" class="section level4 hasAnchor" number="1.3.1.5">
<h4><span class="header-section-number">1.3.1.5</span> Binomial distribution<a href="week-one.html#binomial-distribution" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>A binomial distribution can be defined as -</p>
<p><span class="math display">\[P(x) = \frac{n!}{x!(n-x)!}\cdot p^x \cdot (1-p)^{n-x}\]</span></p>
<p>Where <span class="math inline">\(x\)</span> is the number of successes out of <span class="math inline">\(n\)</span> experiments and <span class="math inline">\(p\)</span> is the probability of success.</p>
<ul>
<li><span class="math inline">\(mean = n \cdot p\)</span></li>
<li><span class="math inline">\(variance = np \cdot (1 - p)\)</span></li>
<li><span class="math inline">\(sd = \sqrt{np \cdot (1 - p)}\)</span></li>
</ul>
<p>The design of the experiment is as follows -</p>
<ul>
<li>The experiment is repeated and are independent of one another</li>
<li>Each experiment has just two outcomes</li>
<li>The probability of success is constant and does not change with experiments</li>
</ul>
<p>We can for example compute the probability of having 7 heads in a series of 10 throws of a coin:</p>
<div class="sourceCode" id="cb222"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb222-1"><a href="week-one.html#cb222-1" tabindex="-1"></a><span class="co"># x = number of successes; size = number of trials; prob = probability of success on each trial</span></span>
<span id="cb222-2"><a href="week-one.html#cb222-2" tabindex="-1"></a><span class="fu">dbinom</span>(<span class="at">x=</span><span class="dv">7</span>, <span class="at">size=</span><span class="dv">10</span>, <span class="at">prob =</span> <span class="fl">0.5</span>) </span></code></pre></div>
<pre><code>## [1] 0.1171875</code></pre>
<p>Or we can compute what the probability is to get 7 or more heads using the function <code>pbinom()</code>. Remember that the parameter “lower.tail” is used to specify whether to calculate the probability of observing x or fewer successes (if <code>lower.tail = TRUE</code>) or the probability of observing more than x successes (<code>lower.tail = FALSE</code>):</p>
<div class="sourceCode" id="cb224"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb224-1"><a href="week-one.html#cb224-1" tabindex="-1"></a><span class="fu">pbinom</span>(<span class="dv">6</span>, <span class="at">size=</span><span class="dv">10</span>,</span>
<span id="cb224-2"><a href="week-one.html#cb224-2" tabindex="-1"></a>       <span class="at">prob=</span><span class="fl">0.5</span>,</span>
<span id="cb224-3"><a href="week-one.html#cb224-3" tabindex="-1"></a>       <span class="at">lower.tail=</span><span class="cn">FALSE</span>) </span></code></pre></div>
<pre><code>## [1] 0.171875</code></pre>
<p>Beware that this syntax means <strong>strictly more than 6</strong>, i.e. 7 or more!!</p>
<blockquote>
<p>How would you compute the probability to get less than 5? What would <code>qbinom(0.3,size=10,prob=0.5,lower.tail=FALSE)</code> represent?</p>
</blockquote>
<hr />
</div>
</div>
<div id="confidence-interval" class="section level3 hasAnchor" number="1.3.2">
<h3><span class="header-section-number">1.3.2</span> Confidence interval<a href="week-one.html#confidence-interval" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The <strong>confidence interval</strong> describes the interval containing the (unknown) expectation value of a distribution with 95% confidence. This means that out of 100 random realizations of this random variable, the true expectation value <span class="math inline">\(\mu\)</span> will indeed be in this interval.</p>
<p>Let us try a simulation: we consider a random variable distributed according to a Poisson distribution <span class="math display">\[P(x) = \frac{{e^{ - \lambda } \lambda ^x }}{{x!}}\]</span> Here, <em>we know the true value of the expectation value</em>. We want to get an estimate for <span class="math inline">\(\lambda\)</span>, and check if the confidence interval contains the true expectation value.</p>
<p>For example, a farmer expects to collect 75 eggs from his hens per hour.</p>
<div class="sourceCode" id="cb226"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb226-1"><a href="week-one.html#cb226-1" tabindex="-1"></a>lambda <span class="ot">=</span> <span class="dv">75</span></span></code></pre></div>
<p>He now collects during 100 days the eggs <span class="math inline">\(N=8\)</span> times a day (each time during one hour). We want to compute the mean <span class="math inline">\(m_N\)</span> over these <span class="math inline">\(N=8\)</span> realizations and determine the 95% confidence interval, and check, how often the expectation value <span class="math inline">\(\mu\)</span> is inside the confidence interval.</p>
<p>Remember that the 95% CI is given by <span class="math display">\[[m_N-t_{95,N-1}\frac{\sigma}{\sqrt{N}},m_N+t_{95,N-1}\frac{\sigma}{\sqrt{N}}]\]</span> where <span class="math inline">\(t_{95,N-1}\)</span> is the critical value for the <span class="math inline">\(t\)</span>-distribution with <span class="math inline">\(n-1\)</span> degrees of freedom.</p>
<p>Let’s start by creating our samples:</p>
<div class="sourceCode" id="cb227"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb227-1"><a href="week-one.html#cb227-1" tabindex="-1"></a><span class="co"># size of the sample</span></span>
<span id="cb227-2"><a href="week-one.html#cb227-2" tabindex="-1"></a>N <span class="ot">=</span> <span class="dv">8</span></span>
<span id="cb227-3"><a href="week-one.html#cb227-3" tabindex="-1"></a></span>
<span id="cb227-4"><a href="week-one.html#cb227-4" tabindex="-1"></a><span class="co"># we now draw 100 times samples of size N=8</span></span>
<span id="cb227-5"><a href="week-one.html#cb227-5" tabindex="-1"></a><span class="do">## rpois is the function used to generate the Poisson distribution</span></span>
<span id="cb227-6"><a href="week-one.html#cb227-6" tabindex="-1"></a>X <span class="ot">=</span> <span class="fu">lapply</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">100</span>, <span class="cf">function</span>(i) {</span>
<span id="cb227-7"><a href="week-one.html#cb227-7" tabindex="-1"></a>  <span class="fu">rpois</span>(N, <span class="at">lambda =</span> lambda)</span>
<span id="cb227-8"><a href="week-one.html#cb227-8" tabindex="-1"></a>  })</span></code></pre></div>
<p><code>lapply()</code> is a function in R that stands for “list apply”. It is used to <em>apply</em> a function to each element of a list or vector and produces a list with the same length as an output.</p>
<p>In this previous example, the input is a vector (<code>1:100</code>).
Then, we <code>lapply(v, function(i))</code>:
+ v: The list or vector you want to apply the function to.
+ function: The function you want to apply where <code>i</code> is each element in X. In this case, we are applying <code>rpois</code> 100 times.</p>
<p>As an output, <code>lapply</code> returns a list where each element has had the specified function applied to it.</p>
<blockquote>
<p>Run View(X) to see how this object looks like. Try using lapply to obtain the mean of all elements in each X using <code>lapply(X, function(i){mean(i)})</code> or <code>lapply(X, mean)</code>.</p>
</blockquote>
<p>Now, we calculate the mean and the standard deviation of the respective samples:</p>
<div class="sourceCode" id="cb228"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb228-1"><a href="week-one.html#cb228-1" tabindex="-1"></a><span class="co"># we compute the sample means</span></span>
<span id="cb228-2"><a href="week-one.html#cb228-2" tabindex="-1"></a>Xm <span class="ot">=</span> <span class="fu">sapply</span>(X,mean)</span>
<span id="cb228-3"><a href="week-one.html#cb228-3" tabindex="-1"></a><span class="co"># and the sample standard deviations</span></span>
<span id="cb228-4"><a href="week-one.html#cb228-4" tabindex="-1"></a>Xsd <span class="ot">=</span> <span class="fu">sapply</span>(X,sd) </span></code></pre></div>
<p>Next, we determine the upper and lower bounds of the 95% CI. Remember that the confidence interval is based on a <span class="math inline">\(t\)</span>-distribution. The degrees of freedom of this distribution is the sample size -1 (<span class="math inline">\(N\)</span>-1=7 in this case)</p>
<div class="sourceCode" id="cb229"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb229-1"><a href="week-one.html#cb229-1" tabindex="-1"></a>df <span class="ot">=</span> N<span class="dv">-1</span></span>
<span id="cb229-2"><a href="week-one.html#cb229-2" tabindex="-1"></a>tc <span class="ot">=</span> <span class="fu">qt</span>(<span class="fu">c</span>(<span class="fl">0.975</span>),df) <span class="co"># this is the critical value for the t-distribution for df = N-1 degrees of freedom and 95% CI</span></span>
<span id="cb229-3"><a href="week-one.html#cb229-3" tabindex="-1"></a></span>
<span id="cb229-4"><a href="week-one.html#cb229-4" tabindex="-1"></a>Xl <span class="ot">=</span> Xm<span class="sc">-</span>tc<span class="sc">*</span>Xsd<span class="sc">/</span><span class="fu">sqrt</span>(N) <span class="co"># upper bound of the 95% CI</span></span>
<span id="cb229-5"><a href="week-one.html#cb229-5" tabindex="-1"></a>Xh <span class="ot">=</span> Xm<span class="sc">+</span>tc<span class="sc">*</span>Xsd<span class="sc">/</span><span class="fu">sqrt</span>(N) <span class="co"># lower bound of the 95% CI</span></span></code></pre></div>
<p>Finally, we determine whether each sample mean is found within the 95% CI or not:</p>
<div class="sourceCode" id="cb230"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb230-1"><a href="week-one.html#cb230-1" tabindex="-1"></a><span class="do">## vector of TRUE/FALSE if the real expectation value lambda is inside the interval</span></span>
<span id="cb230-2"><a href="week-one.html#cb230-2" tabindex="-1"></a>i.ok <span class="ot">=</span>  <span class="fu">as.factor</span>(Xl <span class="sc">&lt;</span> lambda <span class="sc">&amp;</span> Xh <span class="sc">&gt;</span> lambda)</span>
<span id="cb230-3"><a href="week-one.html#cb230-3" tabindex="-1"></a></span>
<span id="cb230-4"><a href="week-one.html#cb230-4" tabindex="-1"></a>plot_data <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb230-5"><a href="week-one.html#cb230-5" tabindex="-1"></a>  <span class="at">n =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">100</span>,</span>
<span id="cb230-6"><a href="week-one.html#cb230-6" tabindex="-1"></a>  <span class="at">Xm =</span> Xm,</span>
<span id="cb230-7"><a href="week-one.html#cb230-7" tabindex="-1"></a>  <span class="at">Xl =</span> Xl,</span>
<span id="cb230-8"><a href="week-one.html#cb230-8" tabindex="-1"></a>  <span class="at">Xh =</span> Xh,</span>
<span id="cb230-9"><a href="week-one.html#cb230-9" tabindex="-1"></a>  <span class="at">i.ok =</span> i.ok</span>
<span id="cb230-10"><a href="week-one.html#cb230-10" tabindex="-1"></a>) </span>
<span id="cb230-11"><a href="week-one.html#cb230-11" tabindex="-1"></a></span>
<span id="cb230-12"><a href="week-one.html#cb230-12" tabindex="-1"></a><span class="co"># Plot using ggplot2</span></span>
<span id="cb230-13"><a href="week-one.html#cb230-13" tabindex="-1"></a><span class="fu">ggplot</span>(plot_data, </span>
<span id="cb230-14"><a href="week-one.html#cb230-14" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> n, <span class="at">y =</span> Xm, </span>
<span id="cb230-15"><a href="week-one.html#cb230-15" tabindex="-1"></a>           <span class="at">color =</span> <span class="st">`</span><span class="at">i.ok</span><span class="st">`</span>)) <span class="sc">+</span></span>
<span id="cb230-16"><a href="week-one.html#cb230-16" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb230-17"><a href="week-one.html#cb230-17" tabindex="-1"></a>  <span class="fu">geom_errorbar</span>(<span class="fu">aes</span>(<span class="at">ymin =</span> Xl, <span class="at">ymax =</span> Xh, <span class="at">color =</span> i.ok), <span class="at">width =</span> <span class="fl">0.2</span>) <span class="sc">+</span></span>
<span id="cb230-18"><a href="week-one.html#cb230-18" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> lambda, <span class="at">linetype =</span> <span class="st">&quot;dashed&quot;</span>, <span class="at">color =</span> <span class="st">&quot;black&quot;</span>) <span class="sc">+</span></span>
<span id="cb230-19"><a href="week-one.html#cb230-19" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">&quot;grey&quot;</span>, <span class="st">&quot;orange&quot;</span>)) <span class="sc">+</span></span>
<span id="cb230-20"><a href="week-one.html#cb230-20" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">y =</span> <span class="st">&quot;Mean values&quot;</span>,</span>
<span id="cb230-21"><a href="week-one.html#cb230-21" tabindex="-1"></a>       <span class="at">title =</span> <span class="fu">paste</span>(<span class="st">&quot;Mean values and confidence intervals, N=&quot;</span>,N),</span>
<span id="cb230-22"><a href="week-one.html#cb230-22" tabindex="-1"></a>       <span class="at">color =</span> <span class="st">&quot;95% CI Status&quot;</span>) </span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-114-1.png" width="768" /></p>
<p>Here, the orange/grey bars represent the confidence interval, the dot is the mean of the sample values, and the dotted line at <code>\lambda</code> represents the true expectation value. Whenever the true expectation value is within the CI, the bar is orange, if not, the bar is grey. How often is the true expectation value outside the CI? Count the grey bars!</p>
<p>It happens 7 times, which fits pretty well with the expected 5%.</p>
<blockquote>
<p>Repeat this simulation, but now with samples of <span class="math inline">\(N=24\)</span> (again 100 times)<br />
What do you observe?<br />
How often is the true expectation value outside the CI? Change to 90% CI and check if that works!</p>
</blockquote>
<hr />
</div>
<div id="exercises-2" class="section level3 hasAnchor" number="1.3.3">
<h3><span class="header-section-number">1.3.3</span> EXERCISES<a href="week-one.html#exercises-2" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="exercise-1-probability-distributions" class="section level4 hasAnchor" number="1.3.3.1">
<h4><span class="header-section-number">1.3.3.1</span> Exercise 1: Probability distributions<a href="week-one.html#exercise-1-probability-distributions" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<ol style="list-style-type: decimal">
<li><p>What is the expression of TP53 observed at 10th percentile? What is the expression of TP53 observed at 90th percentile?</p></li>
<li><p>(optional) Other distribution types can become very similar to the normal distribution under certain conditions. Plot the histogram of 1000 random numbers drawn from the Poisson distribution with lambda = 1, 10, 100, 1000. What do you observe?</p></li>
</ol>
</div>
<div id="exercise-2-confidence-intervals" class="section level4 hasAnchor" number="1.3.3.2">
<h4><span class="header-section-number">1.3.3.2</span> Exercise 2: Confidence intervals<a href="week-one.html#exercise-2-confidence-intervals" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>You are buying 10 packs of gummy bears. You particularly like the red ones and the green ones. A pack contains 6 different colors and you expect them to be equally distributed. There are 84 pieces per 200g pack.</p>
<ol style="list-style-type: decimal">
<li><p>What is the expected amount of red or green gummy bears?</p></li>
<li><p>You selected your 10 packs according to the colors you could see in the pack. At home, you counted the following bears per pack:</p></li>
</ol>
<ul>
<li>for the red ones: 12 16 17 12 16 13 11 18 13 19<br />
</li>
<li>for the green ones: 11 10 15 16 12 14 13 10 13 17<br />
Was your selection procedure a success? In other words, is the expected value bellow (congrats!), within or above (bad luck!) the 95% CI?</li>
</ul>
</div>
</div>
</div>
<div id="day-4-hypothesis-testing" class="section level2 hasAnchor" number="1.4">
<h2><span class="header-section-number">1.4</span> Day 4: Hypothesis testing<a href="week-one.html#day-4-hypothesis-testing" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>On this section, we will go through hypothesis testing. You will start to see how to <em>formulate hypotheses</em> and how to <em>test</em> them.
In addition, we want to learn how to use and interpret hypothesis tests.</p>
<p>We will work again with the diabetes dataset that we used previously.</p>
<div class="sourceCode" id="cb231"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb231-1"><a href="week-one.html#cb231-1" tabindex="-1"></a>dat <span class="ot">=</span> <span class="fu">read.delim</span>(<span class="st">&#39;https://tinyurl.com/y4fark9g&#39;</span>)</span>
<span id="cb231-2"><a href="week-one.html#cb231-2" tabindex="-1"></a></span>
<span id="cb231-3"><a href="week-one.html#cb231-3" tabindex="-1"></a><span class="co"># set the row names using the column id</span></span>
<span id="cb231-4"><a href="week-one.html#cb231-4" tabindex="-1"></a><span class="fu">rownames</span>(dat) <span class="ot">=</span> dat<span class="sc">$</span>id</span></code></pre></div>
<p>Load the required packages</p>
<div class="sourceCode" id="cb232"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb232-1"><a href="week-one.html#cb232-1" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb232-2"><a href="week-one.html#cb232-2" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb232-3"><a href="week-one.html#cb232-3" tabindex="-1"></a><span class="fu">library</span>(tibble)</span></code></pre></div>
<p>Check out the content of the dataset using the summary function</p>
<div class="sourceCode" id="cb233"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb233-1"><a href="week-one.html#cb233-1" tabindex="-1"></a><span class="fu">summary</span>(dat)</span></code></pre></div>
<pre><code>##        id             chol          stab.glu          hdl        
##  Min.   : 1000   Min.   : 78.0   Min.   : 48.0   Min.   : 12.00  
##  1st Qu.: 4792   1st Qu.:179.0   1st Qu.: 81.0   1st Qu.: 38.00  
##  Median :15766   Median :204.0   Median : 89.0   Median : 46.00  
##  Mean   :15978   Mean   :207.8   Mean   :106.7   Mean   : 50.45  
##  3rd Qu.:20336   3rd Qu.:230.0   3rd Qu.:106.0   3rd Qu.: 59.00  
##  Max.   :41756   Max.   :443.0   Max.   :385.0   Max.   :120.00  
##                  NA&#39;s   :1                       NA&#39;s   :1       
##      ratio            glyhb         location              age       
##  Min.   : 1.500   Min.   : 2.68   Length:403         Min.   :19.00  
##  1st Qu.: 3.200   1st Qu.: 4.38   Class :character   1st Qu.:34.00  
##  Median : 4.200   Median : 4.84   Mode  :character   Median :45.00  
##  Mean   : 4.522   Mean   : 5.59                      Mean   :46.85  
##  3rd Qu.: 5.400   3rd Qu.: 5.60                      3rd Qu.:60.00  
##  Max.   :19.300   Max.   :16.11                      Max.   :92.00  
##  NA&#39;s   :1        NA&#39;s   :13                                        
##     gender              height          weight         frame          
##  Length:403         Min.   :52.00   Min.   : 99.0   Length:403        
##  Class :character   1st Qu.:63.00   1st Qu.:151.0   Class :character  
##  Mode  :character   Median :66.00   Median :172.5   Mode  :character  
##                     Mean   :66.02   Mean   :177.6                     
##                     3rd Qu.:69.00   3rd Qu.:200.0                     
##                     Max.   :76.00   Max.   :325.0                     
##                     NA&#39;s   :5       NA&#39;s   :1                         
##      bp.1s           bp.1d            bp.2s           bp.2d       
##  Min.   : 90.0   Min.   : 48.00   Min.   :110.0   Min.   : 60.00  
##  1st Qu.:121.2   1st Qu.: 75.00   1st Qu.:138.0   1st Qu.: 84.00  
##  Median :136.0   Median : 82.00   Median :149.0   Median : 92.00  
##  Mean   :136.9   Mean   : 83.32   Mean   :152.4   Mean   : 92.52  
##  3rd Qu.:146.8   3rd Qu.: 90.00   3rd Qu.:161.0   3rd Qu.:100.00  
##  Max.   :250.0   Max.   :124.00   Max.   :238.0   Max.   :124.00  
##  NA&#39;s   :5       NA&#39;s   :5        NA&#39;s   :262     NA&#39;s   :262     
##      waist           hip           time.ppn     
##  Min.   :26.0   Min.   :30.00   Min.   :   5.0  
##  1st Qu.:33.0   1st Qu.:39.00   1st Qu.:  90.0  
##  Median :37.0   Median :42.00   Median : 240.0  
##  Mean   :37.9   Mean   :43.04   Mean   : 341.2  
##  3rd Qu.:41.0   3rd Qu.:46.00   3rd Qu.: 517.5  
##  Max.   :56.0   Max.   :64.00   Max.   :1560.0  
##  NA&#39;s   :2      NA&#39;s   :2       NA&#39;s   :3</code></pre>
<p>How can we inspect the differences between the weight of men and women? We can start by ploting two histograms or density plots representing the weight by biological sex.</p>
<div class="sourceCode" id="cb235"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb235-1"><a href="week-one.html#cb235-1" tabindex="-1"></a><span class="fu">ggplot</span>(dat,</span>
<span id="cb235-2"><a href="week-one.html#cb235-2" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> weight, </span>
<span id="cb235-3"><a href="week-one.html#cb235-3" tabindex="-1"></a>           <span class="at">fill =</span> gender)) <span class="sc">+</span></span>
<span id="cb235-4"><a href="week-one.html#cb235-4" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="at">alpha =</span> <span class="fl">0.5</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-118-1.png" width="672" /></p>
<p>The distributions look different in shape, right? Where do you think the mean would be located in the plot?</p>
<div class="sourceCode" id="cb236"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb236-1"><a href="week-one.html#cb236-1" tabindex="-1"></a><span class="co"># We can use &quot;filter()&quot; to filter the cholesterol values for men and women</span></span>
<span id="cb236-2"><a href="week-one.html#cb236-2" tabindex="-1"></a>dat.male <span class="ot">=</span> dat <span class="sc">%&gt;%</span></span>
<span id="cb236-3"><a href="week-one.html#cb236-3" tabindex="-1"></a>  <span class="fu">filter</span>(gender <span class="sc">==</span> <span class="st">&#39;male&#39;</span>)</span>
<span id="cb236-4"><a href="week-one.html#cb236-4" tabindex="-1"></a></span>
<span id="cb236-5"><a href="week-one.html#cb236-5" tabindex="-1"></a>dat.female <span class="ot">=</span> dat <span class="sc">%&gt;%</span></span>
<span id="cb236-6"><a href="week-one.html#cb236-6" tabindex="-1"></a>  <span class="fu">filter</span>(gender <span class="sc">==</span> <span class="st">&#39;female&#39;</span>)</span>
<span id="cb236-7"><a href="week-one.html#cb236-7" tabindex="-1"></a></span>
<span id="cb236-8"><a href="week-one.html#cb236-8" tabindex="-1"></a><span class="co"># we will calculate the mean weight by sex.</span></span>
<span id="cb236-9"><a href="week-one.html#cb236-9" tabindex="-1"></a>mean.men <span class="ot">&lt;-</span> <span class="fu">mean</span>(dat.male<span class="sc">$</span>weight, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb236-10"><a href="week-one.html#cb236-10" tabindex="-1"></a>mean.women <span class="ot">&lt;-</span> <span class="fu">mean</span>(dat.female<span class="sc">$</span>weight, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p>And we can then add in the means to the plot as vertical lines.</p>
<div class="sourceCode" id="cb237"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb237-1"><a href="week-one.html#cb237-1" tabindex="-1"></a><span class="fu">ggplot</span>(dat,</span>
<span id="cb237-2"><a href="week-one.html#cb237-2" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> weight, </span>
<span id="cb237-3"><a href="week-one.html#cb237-3" tabindex="-1"></a>           <span class="at">fill =</span> gender)) <span class="sc">+</span></span>
<span id="cb237-4"><a href="week-one.html#cb237-4" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb237-5"><a href="week-one.html#cb237-5" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> mean.men, <span class="at">colour =</span> <span class="st">&quot;cyan&quot;</span>) <span class="sc">+</span></span>
<span id="cb237-6"><a href="week-one.html#cb237-6" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> mean.women, <span class="at">colour =</span> <span class="st">&quot;salmon&quot;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-120-1.png" width="672" /></p>
<blockquote>
<p>Do you think that the mean of the male weight is lower than 180?
Do you think that the mean of the female weight is <strong>really</strong> different from the mean of the male weights?
Do you think that the mean of the male weight is higher than that of the females?</p>
</blockquote>
<hr />
<div id="tests-of-mean" class="section level3 hasAnchor" number="1.4.1">
<h3><span class="header-section-number">1.4.1</span> Tests of mean<a href="week-one.html#tests-of-mean" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Now, we want to use a statistical test to check if,</p>
<ul>
<li>males have a mean weight that is <strong>significantly</strong> lower than a specific value (one-sample, one-tailed test)</li>
<li>there is a <strong>significant</strong> difference in the mean of the weights between the two groups (two-sample, two-sided test)</li>
<li>male have a <strong>significantly</strong> higher mean weight than females (two-sample, one-sided test)</li>
</ul>
<p>Note the use of the word <strong>significant</strong> in the previous statements!</p>
<p>This is exactly what <strong>mean tests</strong> such as the t-test (or the Wilcoxon test) are designed for!
We will perform here a <strong>t-test</strong>.</p>
<hr />
<div id="one-sample-t-test" class="section level4 hasAnchor" number="1.4.1.1">
<h4><span class="header-section-number">1.4.1.1</span> One sample t-test<a href="week-one.html#one-sample-t-test" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Using a one-sample t-test, we can check if values differ significantly from a target value.
For example, you could sample 10 chocolate bars, and test if they significantly differ from the expected weight of 100 g:</p>
<blockquote>
<p>Should we perform a one- or two-sided test?</p>
</blockquote>
<div class="sourceCode" id="cb238"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb238-1"><a href="week-one.html#cb238-1" tabindex="-1"></a>bars <span class="ot">=</span> <span class="fu">c</span>(<span class="dv">103</span>,<span class="dv">103</span>,<span class="dv">97</span>,<span class="fl">102.5</span>,<span class="fl">100.5</span>,<span class="dv">103</span>,<span class="fl">101.3</span>,<span class="fl">99.5</span>,<span class="dv">101</span>,<span class="dv">104</span>) <span class="co"># weights of 10 chocolate bars</span></span>
<span id="cb238-2"><a href="week-one.html#cb238-2" tabindex="-1"></a>chocbar.mean <span class="ot">=</span> <span class="dv">100</span> <span class="co"># expected weight</span></span></code></pre></div>
<p>The function <code>t.test()</code> offers three <strong>alternative</strong> options: <em>two.sided</em>, <em>less</em> and <em>greater</em>.
Here, if we want to test whether the mean weight of the 10 chocolate bars is <strong>different</strong> from the expected weight of 100 g, we want to perform a <em>two sided</em> test and use the alternative <em>two.sided</em>.</p>
<p>It is essential to <strong>clearly formulate the H0 and H1 hypothesis</strong>.
There are two alternative but equivalent ways to do so.
Either:</p>
<ul>
<li>H0: the expectation value of the random variable “Weight of a chocolate bar” is <strong>equal</strong> to 100 g.<br />
</li>
<li>H1: the expectation value of the random variable “Weight of a chocolate bar” is <strong>different</strong> from 100 g.</li>
</ul>
<p>or</p>
<ul>
<li>H0: the mean weight of a chocolate bar is not significantly different from 100 g.<br />
</li>
<li>H1: the mean weight of a chocolate bar is significantly different from 100 g.</li>
</ul>
<p>Note the difference between these two formulations, and ask for help if you have questions about this!</p>
<div class="sourceCode" id="cb239"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb239-1"><a href="week-one.html#cb239-1" tabindex="-1"></a><span class="fu">t.test</span>(<span class="at">x =</span> bars, <span class="at">mu =</span> chocbar.mean, <span class="at">alternative =</span> <span class="st">&quot;two.sided&quot;</span>)</span></code></pre></div>
<pre><code>## 
##  One Sample t-test
## 
## data:  bars
## t = 2.233, df = 9, p-value = 0.05244
## alternative hypothesis: true mean is not equal to 100
## 95 percent confidence interval:
##   99.98067 102.97933
## sample estimates:
## mean of x 
##    101.48</code></pre>
<blockquote>
<p>How would you interpret this result?
Can you reject the H0 hypothesis?</p>
</blockquote>
<p>Using <span class="math inline">\(\alpha=0.05\)</span>, the H0 hypothesis can not be rejected as the p-value is 0.05244 (p-value &gt;= 0.05).
With <span class="math inline">\(\alpha=0.05\)</span>, the mean weight of the chocolate bars is not significantly different from 100 g.</p>
<p>Using <span class="math inline">\(\alpha=0.1\)</span>, the H0 hypothesis can be rejected (p-value &lt; 0.1).
With <span class="math inline">\(\alpha=0.1\)</span>, the mean weight of the chocolate bars is significantly different from 100 g.</p>
<p><strong>BUT</strong> … It does not mean that <span class="math inline">\(\alpha\)</span> should be chosen with respect to the results of the t.test!!!</p>
<p>Before running a t.test, <strong>formulate the null hypothesis H0 and the alternative hypothesis H1</strong> and <strong>decide about the alpha</strong> value.
Remember, <strong>alpha</strong> represent the <strong>false positive rate</strong>: under the H0 hypothesis (test of two identical distributions), this is the proportion of tests that will detect a difference between the two groups (p-value &lt; <span class="math inline">\(\alpha\)</span>).</p>
<p><strong>Beware not to get confused between one-/two-sample tests, and one-/two-sided tests!</strong></p>
<p>Regarding the mean weight of the males, we would like to check whether males have a mean weight that is significantly <strong>lower</strong> than 180.
Here as well, we will perform a <strong>one-sample test</strong>, with mu = 180.
However, we will perform a <strong>one-sided</strong> test using the alternative option <strong>less</strong>.</p>
<p>The hypotheses can be formulated as:</p>
<ul>
<li><p>H0: the expectation value of the random variable “Weight of male patients” is <strong>equal or greater</strong> 180.<br />
</p></li>
<li><p>H1: the expectation value of the random variable “Weight of male patients” is <strong>less</strong> than 180.</p></li>
</ul>
<div class="sourceCode" id="cb241"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb241-1"><a href="week-one.html#cb241-1" tabindex="-1"></a><span class="fu">t.test</span>(dat.male<span class="sc">$</span>weight, <span class="at">mu =</span> <span class="dv">180</span>, <span class="at">alternative =</span> <span class="st">&quot;less&quot;</span>)</span></code></pre></div>
<pre><code>## 
##  One Sample t-test
## 
## data:  dat.male$weight
## t = 0.64094, df = 167, p-value = 0.7388
## alternative hypothesis: true mean is less than 180
## 95 percent confidence interval:
##      -Inf 186.8629
## sample estimates:
## mean of x 
##  181.9167</code></pre>
<blockquote>
<p>How would you interpret the result with <span class="math inline">\(\alpha=0.05\)</span>?
Can you reject the H0 hypothesis?</p>
</blockquote>
<hr />
</div>
<div id="two-sample-t-test-2-sided" class="section level4 hasAnchor" number="1.4.1.2">
<h4><span class="header-section-number">1.4.1.2</span> Two-sample t-test (2-sided)<a href="week-one.html#two-sample-t-test-2-sided" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Now, we will compare the mean of the weights between males and females.<br />
</p>
<p>According to the previous histogram, females have a different mean weight as males.
This can be tested using a <em>two-sample and two.tailed</em> t.test.</p>
<div class="sourceCode" id="cb243"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb243-1"><a href="week-one.html#cb243-1" tabindex="-1"></a><span class="fu">ggplot</span>(dat,</span>
<span id="cb243-2"><a href="week-one.html#cb243-2" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> weight, </span>
<span id="cb243-3"><a href="week-one.html#cb243-3" tabindex="-1"></a>           <span class="at">fill =</span> gender)) <span class="sc">+</span></span>
<span id="cb243-4"><a href="week-one.html#cb243-4" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb243-5"><a href="week-one.html#cb243-5" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> mean.men, <span class="at">colour =</span> <span class="st">&quot;cyan&quot;</span>) <span class="sc">+</span></span>
<span id="cb243-6"><a href="week-one.html#cb243-6" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> mean.women, <span class="at">colour =</span> <span class="st">&quot;salmon&quot;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-124-1.png" width="672" /></p>
<p>The hypotheses can be formulated as:</p>
<ul>
<li><p>H0: the expectation value of the random variable “Weight of male patients” is <strong>equal</strong> to the expectation value of the random variable “Weight of female patients”.<br />
</p></li>
<li><p>H1: the expectation value of the random variable “Weight of male patients” is <strong>different</strong> from the expectation value of the random variable “Weight of female patients”.</p></li>
</ul>
<div class="sourceCode" id="cb244"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb244-1"><a href="week-one.html#cb244-1" tabindex="-1"></a><span class="fu">t.test</span>(dat.male<span class="sc">$</span>weight, </span>
<span id="cb244-2"><a href="week-one.html#cb244-2" tabindex="-1"></a>       dat.female<span class="sc">$</span>weight) <span class="co"># remember, alternative = &quot;two.sided&quot; per default.</span></span></code></pre></div>
<pre><code>## 
##  Welch Two Sample t-test
## 
## data:  dat.male$weight and dat.female$weight
## t = 1.8453, df = 372.45, p-value = 0.06579
## alternative hypothesis: true difference in means is not equal to 0
## 95 percent confidence interval:
##  -0.4875828 15.3465572
## sample estimates:
## mean of x mean of y 
##  181.9167  174.4872</code></pre>
<blockquote>
<p>How would you interpret the result with <span class="math inline">\(\alpha=0.05\)</span>?
Can you reject the H0 hypothesis?</p>
</blockquote>
<hr />
</div>
<div id="two-sample-t-test-1-sided" class="section level4 hasAnchor" number="1.4.1.3">
<h4><span class="header-section-number">1.4.1.3</span> Two-sample t-test (1-sided)<a href="week-one.html#two-sample-t-test-1-sided" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Looking at the histogram, other observers could in principle see a difference between the mean values of the weights and formulate the following hypotheses:</p>
<ul>
<li><p>H0: the expectation value of the random variable “Weight of male patients” is <strong>equal or lower</strong> to the expectation value of the random variable “Weight of female patients”.<br />
</p></li>
<li><p>H1: the expectation value of the random variable “Weight of male patients” is <strong>higher</strong> than the expectation value of the random variable “Weight of female patients”.</p></li>
</ul>
<div class="sourceCode" id="cb246"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb246-1"><a href="week-one.html#cb246-1" tabindex="-1"></a><span class="fu">t.test</span>(dat.male<span class="sc">$</span>weight, </span>
<span id="cb246-2"><a href="week-one.html#cb246-2" tabindex="-1"></a>       dat.female<span class="sc">$</span>weight, <span class="at">alternative =</span> <span class="st">&quot;greater&quot;</span>)</span></code></pre></div>
<pre><code>## 
##  Welch Two Sample t-test
## 
## data:  dat.male$weight and dat.female$weight
## t = 1.8453, df = 372.45, p-value = 0.0329
## alternative hypothesis: true difference in means is greater than 0
## 95 percent confidence interval:
##  0.7903498       Inf
## sample estimates:
## mean of x mean of y 
##  181.9167  174.4872</code></pre>
<blockquote>
<p>How would you interpret the result with <span class="math inline">\(\alpha=0.05\)</span>?</p>
</blockquote>
<blockquote>
<p>Can you explain why the p-value of the one-tailed t.test is lower than the p-value of the two-tailed t.test?
What is the relation between these two values?</p>
</blockquote>
<div class="sourceCode" id="cb248"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb248-1"><a href="week-one.html#cb248-1" tabindex="-1"></a><span class="co"># p-value of the two-sided t-test versus p-value of the one-sided t-test: </span></span>
<span id="cb248-2"><a href="week-one.html#cb248-2" tabindex="-1"></a><span class="fu">t.test</span>(dat.male<span class="sc">$</span>weight, </span>
<span id="cb248-3"><a href="week-one.html#cb248-3" tabindex="-1"></a>       dat.female<span class="sc">$</span>weight, <span class="at">alternative =</span> <span class="st">&quot;two.sided&quot;</span>)<span class="sc">$</span>p.value <span class="co"># two-tailed</span></span></code></pre></div>
<pre><code>## [1] 0.06579432</code></pre>
<div class="sourceCode" id="cb250"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb250-1"><a href="week-one.html#cb250-1" tabindex="-1"></a><span class="fu">t.test</span>(dat.male<span class="sc">$</span>weight, </span>
<span id="cb250-2"><a href="week-one.html#cb250-2" tabindex="-1"></a>       dat.female<span class="sc">$</span>weight, <span class="at">alternative =</span> <span class="st">&quot;greater&quot;</span>)<span class="sc">$</span>p.value <span class="co"># one-tailed</span></span></code></pre></div>
<pre><code>## [1] 0.03289716</code></pre>
<p>This can be visualized using the t-distribution.
Here (see above result of the t.test), <em>t = 1.8453</em> and <em>df = 372.45</em>.</p>
<p>In the <strong>one-tailed t.test</strong>, the p-value is the area under the curve for t &gt; 1.8453 (alternative greater) <strong>OR</strong> for t &lt; -1.8453 (alternative less).</p>
<p><img src="_main_files/figure-html/unnamed-chunk-128-1.png" width="480" /></p>
<p>In the <strong>two-tailed t.test</strong>, the p-value is the area under the curve for t &gt; 1.8453 (alternative greater) <strong>AND</strong> for t &lt; -1.8453 (alternative less).
It is two times the p-value of the one-sided t.test!</p>
<p><img src="_main_files/figure-html/unnamed-chunk-129-1.png" width="480" /></p>
<p>IMPORTANTLY, a t-test can only be performed if the data is <strong>normally distributed</strong>!
If the data is not normally distributed, you will need to use a non-parametric test, like Wilcoxon test.</p>
<hr />
</div>
<div id="wilcoxon-test" class="section level4 hasAnchor" number="1.4.1.4">
<h4><span class="header-section-number">1.4.1.4</span> Wilcoxon test<a href="week-one.html#wilcoxon-test" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>What if the data is not normally distributed?
In that case, we are not supposed to use the t-test for testing differences between mean values!
Let us us see an example.</p>
<p>Let us start by loading the data.</p>
<div class="sourceCode" id="cb252"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb252-1"><a href="week-one.html#cb252-1" tabindex="-1"></a>all.aml <span class="ot">=</span> <span class="fu">read.delim</span>(<span class="st">&#39;http://bioinfo.ipmb.uni-heidelberg.de/crg/datascience3fs/practicals/data/all.aml.cleaned.csv&#39;</span>,</span>
<span id="cb252-2"><a href="week-one.html#cb252-2" tabindex="-1"></a>                     <span class="at">header=</span><span class="cn">TRUE</span>)</span>
<span id="cb252-3"><a href="week-one.html#cb252-3" tabindex="-1"></a>all.aml.anno <span class="ot">=</span> <span class="fu">read.delim</span>(<span class="st">&quot;http://bioinfo.ipmb.uni-heidelberg.de/crg/datascience3fs/practicals/data/all.aml.anno.cleaned.csv&quot;</span>,</span>
<span id="cb252-4"><a href="week-one.html#cb252-4" tabindex="-1"></a>                          <span class="at">header=</span><span class="cn">TRUE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb252-5"><a href="week-one.html#cb252-5" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">id =</span> <span class="fu">paste0</span>(<span class="st">&quot;pat&quot;</span>, Samples))</span></code></pre></div>
<p>We check whether the distribution of the expression values for the gene <em>SOX4</em> corresponds to a normal distribution:</p>
<div class="sourceCode" id="cb253"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb253-1"><a href="week-one.html#cb253-1" tabindex="-1"></a>expression <span class="ot">=</span> all.aml <span class="sc">%&gt;%</span></span>
<span id="cb253-2"><a href="week-one.html#cb253-2" tabindex="-1"></a>  <span class="fu">t</span>() <span class="sc">%&gt;%</span></span>
<span id="cb253-3"><a href="week-one.html#cb253-3" tabindex="-1"></a>  <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb253-4"><a href="week-one.html#cb253-4" tabindex="-1"></a>  <span class="fu">na.omit</span>()</span>
<span id="cb253-5"><a href="week-one.html#cb253-5" tabindex="-1"></a></span>
<span id="cb253-6"><a href="week-one.html#cb253-6" tabindex="-1"></a><span class="co"># Plot histogram </span></span>
<span id="cb253-7"><a href="week-one.html#cb253-7" tabindex="-1"></a><span class="fu">ggplot</span>(expression,</span>
<span id="cb253-8"><a href="week-one.html#cb253-8" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> SOX4)) <span class="sc">+</span></span>
<span id="cb253-9"><a href="week-one.html#cb253-9" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins =</span> <span class="dv">20</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-131-1.png" width="528" /></p>
<div class="sourceCode" id="cb254"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb254-1"><a href="week-one.html#cb254-1" tabindex="-1"></a><span class="co"># Check if it&#39;s normally distributed using a QQ-plot </span></span>
<span id="cb254-2"><a href="week-one.html#cb254-2" tabindex="-1"></a><span class="fu">ggplot</span>(expression,</span>
<span id="cb254-3"><a href="week-one.html#cb254-3" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">sample =</span> SOX4)) <span class="sc">+</span></span>
<span id="cb254-4"><a href="week-one.html#cb254-4" tabindex="-1"></a>  <span class="fu">geom_qq</span>() <span class="sc">+</span></span>
<span id="cb254-5"><a href="week-one.html#cb254-5" tabindex="-1"></a>  <span class="fu">geom_qq_line</span>(<span class="at">colour =</span> <span class="st">&#39;red&#39;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-131-2.png" width="528" /></p>
<p>This looks everything but normal!
In that case, we cannot apply the t.test, but need to apply a <strong>non-parametric test</strong> called the <em>Wilcoxon test</em>.
This test is performed not on the <em>values</em> (like the t-test) but on the <em>ranks</em> of these values (remember the difference between the Pearson’s and the Spearman’s correlations!)</p>
<div class="sourceCode" id="cb255"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb255-1"><a href="week-one.html#cb255-1" tabindex="-1"></a><span class="co"># divide the gene expression data in two groups according to ALL or AML patients:</span></span>
<span id="cb255-2"><a href="week-one.html#cb255-2" tabindex="-1"></a><span class="co"># obtain the AML and rest</span></span>
<span id="cb255-3"><a href="week-one.html#cb255-3" tabindex="-1"></a>aml.patient.id <span class="ot">=</span> all.aml.anno <span class="sc">%&gt;%</span></span>
<span id="cb255-4"><a href="week-one.html#cb255-4" tabindex="-1"></a>  <span class="fu">filter</span>(ALL.AML <span class="sc">==</span> <span class="st">&quot;AML&quot;</span>)</span>
<span id="cb255-5"><a href="week-one.html#cb255-5" tabindex="-1"></a>other.id <span class="ot">=</span> all.aml.anno <span class="sc">%&gt;%</span></span>
<span id="cb255-6"><a href="week-one.html#cb255-6" tabindex="-1"></a>  <span class="fu">filter</span>(ALL.AML <span class="sc">!=</span> <span class="st">&quot;AML&quot;</span>) <span class="co"># filters for those which are not labeled AML</span></span>
<span id="cb255-7"><a href="week-one.html#cb255-7" tabindex="-1"></a></span>
<span id="cb255-8"><a href="week-one.html#cb255-8" tabindex="-1"></a>gene.all <span class="ot">=</span> expression <span class="sc">%&gt;%</span></span>
<span id="cb255-9"><a href="week-one.html#cb255-9" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>(<span class="st">&quot;id&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb255-10"><a href="week-one.html#cb255-10" tabindex="-1"></a>  <span class="fu">filter</span>(id <span class="sc">%in%</span> other.id<span class="sc">$</span>id)</span>
<span id="cb255-11"><a href="week-one.html#cb255-11" tabindex="-1"></a></span>
<span id="cb255-12"><a href="week-one.html#cb255-12" tabindex="-1"></a>gene.aml <span class="ot">=</span> expression <span class="sc">%&gt;%</span></span>
<span id="cb255-13"><a href="week-one.html#cb255-13" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>(<span class="st">&quot;id&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb255-14"><a href="week-one.html#cb255-14" tabindex="-1"></a>  <span class="fu">filter</span>(id <span class="sc">%in%</span> aml.patient.id<span class="sc">$</span>id)</span>
<span id="cb255-15"><a href="week-one.html#cb255-15" tabindex="-1"></a>  </span>
<span id="cb255-16"><a href="week-one.html#cb255-16" tabindex="-1"></a><span class="co"># test for a difference in the mean expression values using the Wilcoxon test:</span></span>
<span id="cb255-17"><a href="week-one.html#cb255-17" tabindex="-1"></a><span class="fu">wilcox.test</span>(gene.aml<span class="sc">$</span>SOX4, gene.all<span class="sc">$</span>SOX4)</span></code></pre></div>
<pre><code>## 
##  Wilcoxon rank sum test with continuity correction
## 
## data:  gene.aml$SOX4 and gene.all$SOX4
## W = 260.5, p-value = 0.0001125
## alternative hypothesis: true location shift is not equal to 0</code></pre>
<p>Compare the obtained p-value with the p-value obtained if we would have used the t-test:</p>
<div class="sourceCode" id="cb257"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb257-1"><a href="week-one.html#cb257-1" tabindex="-1"></a><span class="fu">t.test</span>(gene.aml<span class="sc">$</span>SOX4, gene.all<span class="sc">$</span>SOX4)</span></code></pre></div>
<pre><code>## 
##  Welch Two Sample t-test
## 
## data:  gene.aml$SOX4 and gene.all$SOX4
## t = -3.6216, df = 54.611, p-value = 0.000642
## alternative hypothesis: true difference in means is not equal to 0
## 95 percent confidence interval:
##  -2605.3197  -748.9356
## sample estimates:
## mean of x mean of y 
##  1653.000  3330.128</code></pre>
<p>The p-values are very different!!
So is the difference of expression between ALL and AML patients for this gene significant or not taking <span class="math inline">\(\alpha=0.05\)</span>?</p>
<p>Here, we <strong>cannot</strong> trust the t-test due to the non-normality of the data!
Hence, the correct p-value is the one from the Wilcoxon test.</p>
<hr />
</div>
</div>
<div id="proportion-tests" class="section level3 hasAnchor" number="1.4.2">
<h3><span class="header-section-number">1.4.2</span> Proportion tests<a href="week-one.html#proportion-tests" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The t-test and Wilcoxon tests are <strong>tests of the mean</strong>, meaning that we are <strong>comparing the means of two samples and looking for significant differences</strong>.</p>
<p>But there are other hypothesis that one might want to test, related to the relationship between two <strong>categorical</strong> variables:</p>
<ul>
<li>is the proportion of men <strong>significantly</strong> higher in the patients from Louisa compared to the ones from Buckingham?</li>
<li>is the proportion of smokers under 18 in Germany <strong>significantly</strong> higher than in other European countries?</li>
</ul>
<p>The proportion test (<strong>Fisher Exact Test</strong> or <strong>chi-squared test</strong>) are used to investigate the relationship between 2 categorical variables, starting from a <strong>contingency table</strong>.
We will use a dataset with clinical information about breast cancer patients.</p>
<div class="sourceCode" id="cb259"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb259-1"><a href="week-one.html#cb259-1" tabindex="-1"></a>dat.brca <span class="ot">=</span> <span class="fu">read.delim</span>(<span class="st">&#39;http://bioinfo.ipmb.uni-heidelberg.de/crg/datascience3fs/practicals/data/gbsg_ba_ca.dat&#39;</span>, </span>
<span id="cb259-2"><a href="week-one.html#cb259-2" tabindex="-1"></a>                      <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<blockquote>
<p>Check which variables in this dataset are categorical/ordinal/numerical.</p>
</blockquote>
<p>We can now check if there is a significant relationship between some variables.
For example, we can verify if the choice of treatment with tamoxifen (variable <code>hormon</code>) is related to the pre-/post-menopausal status (variable <code>meno</code>)</p>
<p>First, we can build the contingency table for these 2 variables:</p>
<div class="sourceCode" id="cb260"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb260-1"><a href="week-one.html#cb260-1" tabindex="-1"></a><span class="do">## build contingency table</span></span>
<span id="cb260-2"><a href="week-one.html#cb260-2" tabindex="-1"></a><span class="fu">table</span>(dat.brca<span class="sc">$</span>meno, dat.brca<span class="sc">$</span>hormon)</span></code></pre></div>
<pre><code>##                 
##                  had tamoxifen no tamoxifen
##   Postmenopausal           187          209
##   premenopausal             59          231</code></pre>
<p>We can compute the <strong>odds-ratio (OR)</strong> for these two variables.
This metric measures the strength of the association between two events, which in this case would be treatment with tamoxifen and menopausal status.</p>
<p>Here’s how to interpret the odds ratio (OR) value:</p>
<ul>
<li><strong>OR = 1</strong>: events are independent<br />
</li>
<li><strong>OR &gt; 1</strong>: events are positively correlated<br />
</li>
<li><strong>OR &lt; 1</strong>: events are negatively correlated</li>
</ul>
<div class="sourceCode" id="cb262"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb262-1"><a href="week-one.html#cb262-1" tabindex="-1"></a>CT <span class="ot">=</span> <span class="fu">table</span>(dat.brca<span class="sc">$</span>meno,dat.brca<span class="sc">$</span>hormon)</span>
<span id="cb262-2"><a href="week-one.html#cb262-2" tabindex="-1"></a></span>
<span id="cb262-3"><a href="week-one.html#cb262-3" tabindex="-1"></a>OR <span class="ot">=</span> (CT[<span class="dv">1</span>,<span class="dv">1</span>]<span class="sc">/</span>CT[<span class="dv">1</span>,<span class="dv">2</span>])<span class="sc">/</span>(CT[<span class="dv">2</span>,<span class="dv">1</span>]<span class="sc">/</span>CT[<span class="dv">2</span>,<span class="dv">2</span>])</span>
<span id="cb262-4"><a href="week-one.html#cb262-4" tabindex="-1"></a>OR</span></code></pre></div>
<pre><code>## [1] 3.503122</code></pre>
<blockquote>
<p>How would the odds-ratio look like if you would transpose the matrix?</p>
</blockquote>
<p>Now we can run the one-sided <strong>Fisher Exact Test</strong> (FET).
The H0/H1 hypothesis are:</p>
<ul>
<li>H0: the odds-ratio is not significantly larger than one</li>
<li>H1: the odds-ratio is significantly larger than one</li>
</ul>
<div class="sourceCode" id="cb264"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb264-1"><a href="week-one.html#cb264-1" tabindex="-1"></a><span class="do">## build contingency table</span></span>
<span id="cb264-2"><a href="week-one.html#cb264-2" tabindex="-1"></a>tab <span class="ot">=</span> <span class="fu">table</span>(dat.brca<span class="sc">$</span>meno,dat.brca<span class="sc">$</span>hormon)</span>
<span id="cb264-3"><a href="week-one.html#cb264-3" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb264-4"><a href="week-one.html#cb264-4" tabindex="-1"></a><span class="do">## run the FET</span></span>
<span id="cb264-5"><a href="week-one.html#cb264-5" tabindex="-1"></a><span class="fu">fisher.test</span>(tab, <span class="at">alternative =</span> <span class="st">&#39;greater&#39;</span>)</span></code></pre></div>
<pre><code>## 
##  Fisher&#39;s Exact Test for Count Data
## 
## data:  tab
## p-value = 1.388e-13
## alternative hypothesis: true odds ratio is greater than 1
## 95 percent confidence interval:
##  2.580638      Inf
## sample estimates:
## odds ratio 
##   3.496639</code></pre>
<p>Check if your computation of the odds-ratio is right!
Compute also the <strong>two-sided</strong> test:</p>
<blockquote>
<p>Formulate the H0/H1 hypothesis!</p>
</blockquote>
<div class="sourceCode" id="cb266"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb266-1"><a href="week-one.html#cb266-1" tabindex="-1"></a><span class="fu">fisher.test</span>(tab)</span></code></pre></div>
<pre><code>## 
##  Fisher&#39;s Exact Test for Count Data
## 
## data:  tab
## p-value = 2.475e-13
## alternative hypothesis: true odds ratio is not equal to 1
## 95 percent confidence interval:
##  2.444410 5.049548
## sample estimates:
## odds ratio 
##   3.496639</code></pre>
<p>We can also use the <strong>chi-square test</strong> to answer the same question.
The Chi-square test compares the <strong>observed</strong> number of occurrences in the contingency table to the <strong>expected</strong> number of occurrences if there was no relationship between the variables.</p>
<ul>
<li>H0: the observed and expected occurrences are not significantly different</li>
<li>H1: the observed and expected occurrences are significantly different</li>
</ul>
<div class="sourceCode" id="cb268"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb268-1"><a href="week-one.html#cb268-1" tabindex="-1"></a><span class="fu">chisq.test</span>(tab)</span></code></pre></div>
<pre><code>## 
##  Pearson&#39;s Chi-squared test with Yates&#39; continuity correction
## 
## data:  tab
## X-squared = 51.416, df = 1, p-value = 7.473e-13</code></pre>
<p>Now we want to verify the impact of age on the grade of the tumor.
We categorize the patients in under and over 40 year groups, and perform a chi-squared test:</p>
<div class="sourceCode" id="cb270"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb270-1"><a href="week-one.html#cb270-1" tabindex="-1"></a><span class="do">## contingency table</span></span>
<span id="cb270-2"><a href="week-one.html#cb270-2" tabindex="-1"></a>tab <span class="ot">=</span> <span class="fu">table</span>(dat.brca<span class="sc">$</span>age<span class="sc">&gt;</span><span class="dv">40</span>,</span>
<span id="cb270-3"><a href="week-one.html#cb270-3" tabindex="-1"></a>            dat.brca<span class="sc">$</span>grade)</span>
<span id="cb270-4"><a href="week-one.html#cb270-4" tabindex="-1"></a>tab <span class="co"># tumor grades 1,2,3; age under 40 (FALSE) and over 40 (TRUE)</span></span></code></pre></div>
<pre><code>##        
##           1   2   3
##   FALSE   6  41  26
##   TRUE   75 403 135</code></pre>
<div class="sourceCode" id="cb272"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb272-1"><a href="week-one.html#cb272-1" tabindex="-1"></a><span class="do">##</span></span>
<span id="cb272-2"><a href="week-one.html#cb272-2" tabindex="-1"></a><span class="fu">chisq.test</span>(tab)</span></code></pre></div>
<pre><code>## 
##  Pearson&#39;s Chi-squared test
## 
## data:  tab
## X-squared = 6.9515, df = 2, p-value = 0.03094</code></pre>
<p>We can determine the table of expected counts using the <code>apply()</code> function.
This works similarly to the <code>lapply</code>, but it can be used to <em>apply</em> a function by rows (if 1), columns (if 2), or both:</p>
<div class="sourceCode" id="cb274"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb274-1"><a href="week-one.html#cb274-1" tabindex="-1"></a>tot <span class="ot">=</span> <span class="fu">apply</span>(tab,<span class="dv">2</span>,sum) <span class="co"># this is the total number of occurrences in the 3 categories of &quot;grade&quot;, independently of &quot;age&quot;</span></span>
<span id="cb274-2"><a href="week-one.html#cb274-2" tabindex="-1"></a>tot</span></code></pre></div>
<pre><code>##   1   2   3 
##  81 444 161</code></pre>
<div class="sourceCode" id="cb276"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb276-1"><a href="week-one.html#cb276-1" tabindex="-1"></a>age <span class="ot">=</span> <span class="fu">apply</span>(tab,<span class="dv">1</span>,sum) <span class="co"># this is the total number of persons above/below 40, independently of &quot;grade&quot;</span></span>
<span id="cb276-2"><a href="week-one.html#cb276-2" tabindex="-1"></a>age</span></code></pre></div>
<pre><code>## FALSE  TRUE 
##    73   613</code></pre>
<p>On the other hand, <code>sapply</code> is commonly used for dataframe or other tabular formats:</p>
<div class="sourceCode" id="cb278"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb278-1"><a href="week-one.html#cb278-1" tabindex="-1"></a>tot.proportions <span class="ot">=</span> tot<span class="sc">/</span><span class="fu">sum</span>(tot) <span class="co"># HO proportions (the proportions of occurrences in the three categories of grade, independently of age)</span></span>
<span id="cb278-2"><a href="week-one.html#cb278-2" tabindex="-1"></a></span>
<span id="cb278-3"><a href="week-one.html#cb278-3" tabindex="-1"></a>tab.exp <span class="ot">=</span> <span class="fu">sapply</span>(tot.proportions,<span class="cf">function</span>(x) {x<span class="sc">*</span>age}) <span class="co"># expected counts under H0</span></span>
<span id="cb278-4"><a href="week-one.html#cb278-4" tabindex="-1"></a>tab.exp</span></code></pre></div>
<pre><code>##               1         2         3
## FALSE  8.619534  47.24781  17.13265
## TRUE  72.380466 396.75219 143.86735</code></pre>
<blockquote>
<p>How would you compute the chi-square test statistic using the <code>tab</code> and <code>tab.exp</code> tables?</p>
</blockquote>
<hr />
</div>
<div id="exercises-3" class="section level3 hasAnchor" number="1.4.3">
<h3><span class="header-section-number">1.4.3</span> EXERCISES<a href="week-one.html#exercises-3" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="exercise-1-one-sided-t-test" class="section level4 hasAnchor" number="1.4.3.1">
<h4><span class="header-section-number">1.4.3.1</span> Exercise 1: One-sided t-test<a href="week-one.html#exercise-1-one-sided-t-test" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<ol style="list-style-type: decimal">
<li>Consider the following graph, formulate the hypotheses H0 and H1 and perform a (one-sided) t-test.
Interpret the result using <span class="math inline">\(\alpha=0.05\)</span>.</li>
</ol>
<p><img src="_main_files/figure-html/unnamed-chunk-143-1.png" width="672" /></p>
<ol start="2" style="list-style-type: decimal">
<li>Calculate the mean age of the men.</li>
<li>Compare it to age = 50. Formulate the hypotheses H0 and H1 and perform a (one-sided) t-test. Interpret the result using <span class="math inline">\(\alpha=0.05\)</span>.</li>
</ol>
</div>
<div id="exercice-2-two-sided-t-test" class="section level4 hasAnchor" number="1.4.3.2">
<h4><span class="header-section-number">1.4.3.2</span> Exercice 2: Two-sided t-test<a href="week-one.html#exercice-2-two-sided-t-test" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Can you find interesting differences in the mean values of parameters of the dataset <strong>dat</strong> for the two groups defined by the <strong>location</strong>?
Follow these steps:</p>
<ol style="list-style-type: decimal">
<li><p>Select the two groups according to the location.
To do so, check the result of <code>distinct(location)</code> and create two tibbles corresponding to each of possible the locations.</p></li>
<li><p>Calculate the mean values of the numerical parameters for each group (ex: age, height, weight, waist, hip, …).
<em>Hint</em>: create first a dataframe with numeric columns only (use <code>select(which(is.numerical))</code>).
Select the rows corresponding to the two groups and use an <code>summarise()</code> and <code>group_by</code> loop to calculate the mean values (grouped by location).</p></li>
<li><p>Select one of these, formulate the H0 and H1 hypotheses and perform a (two-sided) t-test.
Interpret the result (<span class="math inline">\(\alpha=0.05\)</span>).</p></li>
</ol>
</div>
<div id="exercise-3-mean-and-proportion-testing" class="section level4 hasAnchor" number="1.4.3.3">
<h4><span class="header-section-number">1.4.3.3</span> Exercise 3: Mean and proportion testing<a href="week-one.html#exercise-3-mean-and-proportion-testing" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>What test would you use for the following questions?</p>
<ul>
<li>A lotion company has to figure out whether their last product is more likely to give acne to men rather than women.</li>
<li>The department of education wants to find out whether social science students have higher grades than science students.</li>
<li>A biologist needs to find out whether a specific gene is more likely to be silenced in lactose intolerant people.</li>
</ul>
</div>
<div id="going-further-checking-the-normality-of-the-distribution" class="section level4 hasAnchor" number="1.4.3.4">
<h4><span class="header-section-number">1.4.3.4</span> Going further: Checking the normality of the distribution<a href="week-one.html#going-further-checking-the-normality-of-the-distribution" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>In principle, t-tests require that the data is approximately normally distributed.
If not, we can use <strong>non-parametric</strong> tests (see next lecture).</p>
<p>In order to check whether the data is normally distributed or not, it is possible to perform a <strong>Shapiro-Wilk</strong> normality test (see lecture).
This statistical test is implemented in R in the function <code>shapiro.test()</code>. Try it out with any of the datasets we used before.</p>
<p>A p-value inferior to the chosen <span class="math inline">\(\alpha\)</span> level here (we will use 0.05), means that the null hypothesis can be rejected. Therefore, the data distribution tested is not normal and you need to use a non-parametric test.</p>
</div>
</div>
</div>
<div id="day-5-multiple-testing-and-regression" class="section level2 hasAnchor" number="1.5">
<h2><span class="header-section-number">1.5</span> Day 5: Multiple testing and regression<a href="week-one.html#day-5-multiple-testing-and-regression" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="multiple-testing" class="section level3 hasAnchor" number="1.5.1">
<h3><span class="header-section-number">1.5.1</span> Multiple testing<a href="week-one.html#multiple-testing" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>In the previous examples we were only testing for one single hypothesis. However, it is common in biology that we have more than one hypothesis being tested.<br />
The table below is a reminder of which errors can be made when testing hypotheses:</p>
<table>
<thead>
<tr class="header">
<th align="left"></th>
<th align="left">H0 is true</th>
<th align="left">H0 is false</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">Decision: Reject H0</td>
<td align="left">Type I Error</td>
<td align="left">Correct Decision</td>
</tr>
<tr class="even">
<td align="left">Decision: Do not reject H0</td>
<td align="left">Correct Decision</td>
<td align="left">Type II Error</td>
</tr>
</tbody>
</table>
<p>Decision is associated with <span class="math inline">\(\alpha\)</span>, so if the p-value is under 0.05, we reject the null hypothesis. However, when testing multiple hypothesis, we are prone to erroneously rejecting the 1st, 2nd… or Nth null hypothesis. I.e., if we test 1000 hypotheses simultaneously, we expect to erroneously reject 50 (5%) just by chance!</p>
<p>The <strong>family wise error rate (FWER)</strong> is the probability of making at least one Type I error and we can control this probability using a p-value correction.</p>
<div id="example" class="section level4 hasAnchor" number="1.5.1.1">
<h4><span class="header-section-number">1.5.1.1</span> Example<a href="week-one.html#example" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Imagine that you are trying to test the effect of a treatment A on a cell line in the lab. You run a gene expression experiment to compare treatment with control (untreated) for a panel of 20 genes you pre-selected.</p>
<p>The comparison between treatment and control for the panel gives you a set p-values of the 20 genes</p>
<div class="sourceCode" id="cb280"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb280-1"><a href="week-one.html#cb280-1" tabindex="-1"></a>p.vals <span class="ot">=</span> <span class="fu">c</span>(<span class="fl">0.1</span>, <span class="fl">0.0001</span>, <span class="fl">0.04</span>, <span class="fl">1.2e-20</span>, <span class="fl">0.002</span>, <span class="fl">0.01</span>, <span class="fl">0.5</span>, <span class="dv">1</span>, <span class="fl">0.02</span>, <span class="fl">1.9e-5</span>,</span>
<span id="cb280-2"><a href="week-one.html#cb280-2" tabindex="-1"></a>           <span class="fl">1.1e-6</span>, <span class="fl">0.03</span>, <span class="fl">0.7</span>, <span class="fl">0.06</span>, <span class="fl">0.01</span>, <span class="fl">0.3</span>, <span class="fl">0.05</span>, <span class="fl">1e-13</span>, <span class="fl">0.032</span>, <span class="fl">0.004</span>)</span></code></pre></div>
<p>But since you ran 20 different tests to generate these, there is a need to correct the p-values. You can do this on the p-values directly using the <code>p.adjust()</code> function. Run <code>?p.adjust()</code> to see which methods can be used.</p>
<p>We will use the false discovery rate (FDR) correction on this example.</p>
<div class="sourceCode" id="cb281"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb281-1"><a href="week-one.html#cb281-1" tabindex="-1"></a>p.vals_FDR <span class="ot">=</span> <span class="fu">p.adjust</span>(p.vals, <span class="at">method=</span><span class="st">&quot;fdr&quot;</span>)</span>
<span id="cb281-2"><a href="week-one.html#cb281-2" tabindex="-1"></a>p.vals_FDR</span></code></pre></div>
<pre><code>##  [1] 1.250000e-01 4.000000e-04 6.153846e-02 2.400000e-19 6.666667e-03
##  [6] 2.222222e-02 5.555556e-01 1.000000e+00 4.000000e-02 9.500000e-05
## [11] 7.333333e-06 5.333333e-02 7.368421e-01 8.000000e-02 2.222222e-02
## [16] 3.529412e-01 7.142857e-02 1.000000e-12 5.333333e-02 1.142857e-02</code></pre>
<p>Check how the p-values change!</p>
<p><img src="_main_files/figure-html/unnamed-chunk-147-1.png" width="50%" /><img src="_main_files/figure-html/unnamed-chunk-147-2.png" width="50%" /></p>
<blockquote>
<p>How many genes change from being significant to non-significant? How does this impact your experiment?</p>
</blockquote>
</div>
</div>
<div id="regression" class="section level3 hasAnchor" number="1.5.2">
<h3><span class="header-section-number">1.5.2</span> Regression<a href="week-one.html#regression" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>In this last part, we want to learn how to build a <strong>regression model</strong> in order to make predictions on certain quantitative variables using other quantitative variables. The important steps here are:</p>
<ul>
<li><strong>learning</strong> the model</li>
<li><strong>testing</strong> the model to evaluate its performances, and also check that the assumptions of the linearity are given</li>
<li><strong>predict</strong> values based on new data points</li>
</ul>
<p>We will use again the diabetes data set, and build a simple linear regression models with a single explanatory variable.
We will focus on predicting the <strong>cholesterol level</strong>.</p>
<p>Load the data and perform some basic inspection on the data:</p>
<div class="sourceCode" id="cb283"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb283-1"><a href="week-one.html#cb283-1" tabindex="-1"></a>tmp <span class="ot">=</span> <span class="fu">read.table</span>(<span class="st">&#39;https://www.dropbox.com/s/zviurze7c85quyw/diabetes_full.csv?dl=1&#39;</span>,<span class="at">header=</span><span class="cn">TRUE</span>,<span class="at">sep=</span><span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>) </span></code></pre></div>
<p>We will limit the dataset to the numerical variables</p>
<div class="sourceCode" id="cb284"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb284-1"><a href="week-one.html#cb284-1" tabindex="-1"></a>dat <span class="ot">=</span> tmp <span class="sc">%&gt;%</span></span>
<span id="cb284-2"><a href="week-one.html#cb284-2" tabindex="-1"></a>  <span class="fu">select</span>(</span>
<span id="cb284-3"><a href="week-one.html#cb284-3" tabindex="-1"></a>    <span class="fu">where</span>(is.numeric)</span>
<span id="cb284-4"><a href="week-one.html#cb284-4" tabindex="-1"></a>  )</span>
<span id="cb284-5"><a href="week-one.html#cb284-5" tabindex="-1"></a><span class="fu">head</span>(dat)</span></code></pre></div>
<pre><code>##     id chol stab.glu hdl ratio glyhb age height weight bp.1s bp.1d bp.2s bp.2d
## 1 1000  203       82  56   3.6  4.31  46     62    121   118    59    NA    NA
## 2 1001  165       97  24   6.9  4.44  29     64    218   112    68    NA    NA
## 3 1002  228       92  37   6.2  4.64  58     61    256   190    92   185    92
## 4 1003   78       93  12   6.5  4.63  67     67    119   110    50    NA    NA
## 5 1005  249       90  28   8.9  7.72  64     68    183   138    80    NA    NA
## 6 1008  248       94  69   3.6  4.81  34     71    190   132    86    NA    NA
##   waist hip time.ppn
## 1    29  38      720
## 2    46  48      360
## 3    49  57      180
## 4    33  38      480
## 5    44  41      300
## 6    36  42      195</code></pre>
<p>First, we have to do some data cleaning and remove all patients with at least 1 “NA”. Use <code>na.omit()</code>.</p>
<div class="sourceCode" id="cb286"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb286-1"><a href="week-one.html#cb286-1" tabindex="-1"></a><span class="co"># Select the patients without NAs</span></span>
<span id="cb286-2"><a href="week-one.html#cb286-2" tabindex="-1"></a>dat <span class="ot">=</span> dat <span class="sc">%&gt;%</span></span>
<span id="cb286-3"><a href="week-one.html#cb286-3" tabindex="-1"></a>  <span class="fu">na.omit</span>()</span></code></pre></div>
<p>Then, we can make a heatmap and visualize the correlation values between each of the variables in the dataset.</p>
<div class="sourceCode" id="cb287"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb287-1"><a href="week-one.html#cb287-1" tabindex="-1"></a><span class="fu">library</span>(pheatmap)</span>
<span id="cb287-2"><a href="week-one.html#cb287-2" tabindex="-1"></a>cor.vals <span class="ot">=</span> <span class="fu">cor</span>(dat, <span class="at">method =</span> <span class="st">&quot;spearman&quot;</span>)</span>
<span id="cb287-3"><a href="week-one.html#cb287-3" tabindex="-1"></a></span>
<span id="cb287-4"><a href="week-one.html#cb287-4" tabindex="-1"></a><span class="fu">pheatmap</span>(cor.vals,</span>
<span id="cb287-5"><a href="week-one.html#cb287-5" tabindex="-1"></a>         <span class="at">cluster_cols =</span> <span class="cn">FALSE</span>,</span>
<span id="cb287-6"><a href="week-one.html#cb287-6" tabindex="-1"></a>         <span class="at">cluster_rows =</span> <span class="cn">FALSE</span>,</span>
<span id="cb287-7"><a href="week-one.html#cb287-7" tabindex="-1"></a>         <span class="at">display_numbers =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-151-1.png" width="672" /></p>
<blockquote>
<p>What are the strongest correlations? Do they make sense?</p>
</blockquote>
<p>Apart from displaying the raw correlation values, we can test if any of those are <strong>statistically significant</strong>, i.e. if they are significantly positive (one-sided test), negative (one-sided test), or non zero (two-sided test). To do so, we will use <code>cor.test</code> function.</p>
<p>For example, there seems to be a positive correlation between <em>stabilized glucose</em> (stab.glu) and <em>hip circumference</em> (hip).</p>
<div class="sourceCode" id="cb288"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb288-1"><a href="week-one.html#cb288-1" tabindex="-1"></a><span class="do">## compute correlation</span></span>
<span id="cb288-2"><a href="week-one.html#cb288-2" tabindex="-1"></a><span class="fu">cor</span>(dat<span class="sc">$</span>stab.glu,dat<span class="sc">$</span>hip)</span></code></pre></div>
<pre><code>## [1] 0.03528725</code></pre>
<div class="sourceCode" id="cb290"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb290-1"><a href="week-one.html#cb290-1" tabindex="-1"></a><span class="do">##</span></span>
<span id="cb290-2"><a href="week-one.html#cb290-2" tabindex="-1"></a><span class="do">## test for significance</span></span>
<span id="cb290-3"><a href="week-one.html#cb290-3" tabindex="-1"></a><span class="fu">cor.test</span>(dat<span class="sc">$</span>stab.glu, dat<span class="sc">$</span>hip)</span></code></pre></div>
<pre><code>## 
##  Pearson&#39;s product-moment correlation
## 
## data:  dat$stab.glu and dat$hip
## t = 0.40873, df = 134, p-value = 0.6834
## alternative hypothesis: true correlation is not equal to 0
## 95 percent confidence interval:
##  -0.1338406  0.2024178
## sample estimates:
##        cor 
## 0.03528725</code></pre>
<blockquote>
<p>Read carefully this output, and make sure you understand it. Check other pairs!</p>
</blockquote>
<div id="univariate-linear-regression" class="section level4 hasAnchor" number="1.5.2.1">
<h4><span class="header-section-number">1.5.2.1</span> Univariate linear regression<a href="week-one.html#univariate-linear-regression" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Next, we will assess what is the most promising variable to predict cholesterol level. Go back to the correlation heatmap to see if there are variables highly correlated with it.</p>
<p>We will use glycosilated hemoglobin (<code>glyhb</code>) as a predictor of the cholesterol level and the function <code>lm()</code>.</p>
<div class="sourceCode" id="cb292"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb292-1"><a href="week-one.html#cb292-1" tabindex="-1"></a>l.g <span class="ot">=</span> <span class="fu">lm</span>(chol <span class="sc">~</span> glyhb, <span class="at">data=</span>dat)</span>
<span id="cb292-2"><a href="week-one.html#cb292-2" tabindex="-1"></a><span class="fu">summary</span>(l.g)</span></code></pre></div>
<pre><code>## 
## Call:
## lm(formula = chol ~ glyhb, data = dat)
## 
## Residuals:
##     Min      1Q  Median      3Q     Max 
## -84.458 -32.011  -7.436  20.584 156.407 
## 
## Coefficients:
##             Estimate Std. Error t value Pr(&gt;|t|)    
## (Intercept)  169.514      9.774  17.343  &lt; 2e-16 ***
## glyhb          8.182      1.517   5.393 3.04e-07 ***
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## Residual standard error: 44.02 on 134 degrees of freedom
## Multiple R-squared:  0.1783, Adjusted R-squared:  0.1722 
## F-statistic: 29.08 on 1 and 134 DF,  p-value: 3.045e-07</code></pre>
<p>For a simple linear regression (with only one explanatory variable), the p-value of the t-test for the slope is identical with the p-value of the F-test for the global model. <em>This will no longer be the case when including several explanatory variables!</em></p>
<p>By the way, have we checked that a linear regression makes sense in this case? Remember that we have to check that:</p>
<ul>
<li>the residuals are <strong>normally distributed</strong></li>
<li>there is <strong>no correlation</strong> between the residuals and the explanatory variable</li>
</ul>
<div class="sourceCode" id="cb294"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb294-1"><a href="week-one.html#cb294-1" tabindex="-1"></a><span class="co"># normal distribution of residuals?</span></span>
<span id="cb294-2"><a href="week-one.html#cb294-2" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb294-3"><a href="week-one.html#cb294-3" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="fu">aes</span>(<span class="at">x =</span> l.g<span class="sc">$</span>residuals)) </span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-154-1.png" width="384" /></p>
<div class="sourceCode" id="cb295"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb295-1"><a href="week-one.html#cb295-1" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb295-2"><a href="week-one.html#cb295-2" tabindex="-1"></a>  <span class="fu">geom_qq</span>(<span class="fu">aes</span>(<span class="at">sample =</span> l.g<span class="sc">$</span>residuals)) <span class="sc">+</span></span>
<span id="cb295-3"><a href="week-one.html#cb295-3" tabindex="-1"></a>  <span class="fu">geom_qq_line</span>(<span class="fu">aes</span>(<span class="at">sample =</span> l.g<span class="sc">$</span>residuals),</span>
<span id="cb295-4"><a href="week-one.html#cb295-4" tabindex="-1"></a>               <span class="at">colour =</span> <span class="st">&#39;red&#39;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-154-2.png" width="384" /></p>
<div class="sourceCode" id="cb296"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb296-1"><a href="week-one.html#cb296-1" tabindex="-1"></a><span class="do">## correlation residuals x-values?</span></span>
<span id="cb296-2"><a href="week-one.html#cb296-2" tabindex="-1"></a>corr.gly_resi <span class="ot">=</span> <span class="fu">cor</span>(dat<span class="sc">$</span>glyhb,l.g<span class="sc">$</span>residuals)</span>
<span id="cb296-3"><a href="week-one.html#cb296-3" tabindex="-1"></a></span>
<span id="cb296-4"><a href="week-one.html#cb296-4" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb296-5"><a href="week-one.html#cb296-5" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> dat<span class="sc">$</span>glyhb, <span class="at">y =</span> l.g<span class="sc">$</span>residuals))</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-154-3.png" width="384" /></p>
<blockquote>
<p>What is your overall opinion about the validity of the regression model here?</p>
</blockquote>
<p>We can now use the model to predict the cholesterol values, and compare them to the real cholesterol values, since we have the information in this dataset.</p>
<div class="sourceCode" id="cb297"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb297-1"><a href="week-one.html#cb297-1" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb297-2"><a href="week-one.html#cb297-2" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> dat<span class="sc">$</span>chol, </span>
<span id="cb297-3"><a href="week-one.html#cb297-3" tabindex="-1"></a>                 <span class="at">y =</span> l.g<span class="sc">$</span>fitted.values),</span>
<span id="cb297-4"><a href="week-one.html#cb297-4" tabindex="-1"></a>             <span class="at">colour =</span> <span class="st">&quot;navy&quot;</span>) <span class="sc">+</span></span>
<span id="cb297-5"><a href="week-one.html#cb297-5" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&#39;Real values&#39;</span>,</span>
<span id="cb297-6"><a href="week-one.html#cb297-6" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">&#39;Predicted values&#39;</span>) <span class="sc">+</span></span>
<span id="cb297-7"><a href="week-one.html#cb297-7" tabindex="-1"></a>  <span class="fu">geom_abline</span>(<span class="at">intercept =</span> <span class="dv">0</span>, <span class="at">slope =</span> <span class="dv">1</span>, </span>
<span id="cb297-8"><a href="week-one.html#cb297-8" tabindex="-1"></a>              <span class="at">color =</span> <span class="st">&#39;red&#39;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-155-1.png" width="384" /></p>
<p>Not really super convincing, right? Let’s put more information into the model!</p>
</div>
<div id="multiple-regression-model" class="section level4 hasAnchor" number="1.5.2.2">
<h4><span class="header-section-number">1.5.2.2</span> Multiple regression model<a href="week-one.html#multiple-regression-model" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Let us include all the information to try to predict cholesterol level:</p>
<div class="sourceCode" id="cb298"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb298-1"><a href="week-one.html#cb298-1" tabindex="-1"></a>l.all <span class="ot">=</span> <span class="fu">lm</span>(chol <span class="sc">~</span> .,<span class="at">data=</span>dat)</span>
<span id="cb298-2"><a href="week-one.html#cb298-2" tabindex="-1"></a><span class="fu">summary</span>(l.all)</span></code></pre></div>
<pre><code>## 
## Call:
## lm(formula = chol ~ ., data = dat)
## 
## Residuals:
##    Min     1Q Median     3Q    Max 
## -77.81 -10.15  -0.88  10.96  55.28 
## 
## Coefficients:
##               Estimate Std. Error t value Pr(&gt;|t|)    
## (Intercept)  2.662e+01  5.724e+01   0.465   0.6428    
## id          -2.722e-05  1.410e-04  -0.193   0.8472    
## stab.glu     6.908e-02  5.848e-02   1.181   0.2398    
## hdl          2.249e+00  1.500e-01  14.988   &lt;2e-16 ***
## ratio        2.520e+01  1.354e+00  18.611   &lt;2e-16 ***
## glyhb        4.451e-01  1.491e+00   0.299   0.7658    
## age          1.566e-01  1.654e-01   0.947   0.3457    
## height      -1.294e+00  6.645e-01  -1.947   0.0539 .  
## weight      -6.610e-02  1.284e-01  -0.515   0.6077    
## bp.1s        9.021e-02  2.053e-01   0.440   0.6611    
## bp.1d       -1.778e-01  2.945e-01  -0.604   0.5472    
## bp.2s       -1.940e-01  2.004e-01  -0.968   0.3350    
## bp.2d        5.934e-01  2.983e-01   1.989   0.0490 *  
## waist        1.013e+00  7.203e-01   1.406   0.1623    
## hip         -5.150e-01  8.398e-01  -0.613   0.5409    
## time.ppn    -9.561e-03  7.239e-03  -1.321   0.1891    
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## Residual standard error: 21.78 on 120 degrees of freedom
## Multiple R-squared:  0.8199, Adjusted R-squared:  0.7974 
## F-statistic: 36.42 on 15 and 120 DF,  p-value: &lt; 2.2e-16</code></pre>
<blockquote>
<p>Do you note something unexpected in this report?</p>
</blockquote>
<p>We can see that the inclusion of several explanatory variables improves the regression. Check the <span class="math inline">\(R^2\)</span> values for example!</p>
<p>We can see that the variables <code>weight</code>, <code>waist</code> and <code>hip</code> do not reach the significance level. Two explanations are possible</p>
<ol style="list-style-type: decimal">
<li>either these variables are indeed non-informative regarding the prediction of the cholesterol level</li>
<li>or the mutual correlation between these 3 variables interferes with the model.</li>
</ol>
<p>We can remove waist and hip for example, and redo the regression</p>
<div class="sourceCode" id="cb300"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb300-1"><a href="week-one.html#cb300-1" tabindex="-1"></a>l.less <span class="ot">=</span> <span class="fu">lm</span>(chol <span class="sc">~</span> stab.glu <span class="sc">+</span> hdl <span class="sc">+</span> glyhb <span class="sc">+</span> age <span class="sc">+</span> height <span class="sc">+</span> weight  <span class="sc">+</span> bp<span class="fl">.1</span>s <span class="sc">+</span> bp<span class="fl">.1</span>d,<span class="at">data=</span>dat)</span>
<span id="cb300-2"><a href="week-one.html#cb300-2" tabindex="-1"></a><span class="fu">summary</span>(l.less)</span></code></pre></div>
<pre><code>## 
## Call:
## lm(formula = chol ~ stab.glu + hdl + glyhb + age + height + weight + 
##     bp.1s + bp.1d, data = dat)
## 
## Residuals:
##     Min      1Q  Median      3Q     Max 
## -71.763 -29.004  -7.782  26.894 166.868 
## 
## Coefficients:
##               Estimate Std. Error t value Pr(&gt;|t|)    
## (Intercept) 137.126166  75.534099   1.815 0.071819 .  
## stab.glu     -0.055663   0.109905  -0.506 0.613406    
## hdl           0.607084   0.227293   2.671 0.008555 ** 
## glyhb         9.914922   2.634210   3.764 0.000254 ***
## age           0.242536   0.312447   0.776 0.439046    
## height       -0.632074   0.985494  -0.641 0.522433    
## weight       -0.058539   0.100419  -0.583 0.560962    
## bp.1s         0.007349   0.215115   0.034 0.972801    
## bp.1d         0.387031   0.373868   1.035 0.302539    
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## Residual standard error: 42.44 on 127 degrees of freedom
## Multiple R-squared:  0.2759, Adjusted R-squared:  0.2303 
## F-statistic: 6.049 on 8 and 127 DF,  p-value: 1.39e-06</code></pre>
<p>We can see how the <code>weight</code> variable seems indeed to contribute to the prediction of the cholesterol. The removal of the strongly correlated variables has increased its significance. It now almost reaches significance at the 5% level!</p>
<blockquote>
<p>Check the result of the F-test to compare both models!</p>
</blockquote>
<p>We can now check if the prediction are better that with the univariate model</p>
<div class="sourceCode" id="cb302"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb302-1"><a href="week-one.html#cb302-1" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb302-2"><a href="week-one.html#cb302-2" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> dat<span class="sc">$</span>chol, </span>
<span id="cb302-3"><a href="week-one.html#cb302-3" tabindex="-1"></a>                 <span class="at">y =</span> l.less<span class="sc">$</span>fitted.values),</span>
<span id="cb302-4"><a href="week-one.html#cb302-4" tabindex="-1"></a>             <span class="at">colour =</span> <span class="st">&quot;navy&quot;</span>) <span class="sc">+</span></span>
<span id="cb302-5"><a href="week-one.html#cb302-5" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&#39;Real values&#39;</span>,</span>
<span id="cb302-6"><a href="week-one.html#cb302-6" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">&#39;Predicted values&#39;</span>) <span class="sc">+</span></span>
<span id="cb302-7"><a href="week-one.html#cb302-7" tabindex="-1"></a>  <span class="fu">geom_abline</span>(<span class="at">intercept =</span> <span class="dv">0</span>, <span class="at">slope =</span> <span class="dv">1</span>, </span>
<span id="cb302-8"><a href="week-one.html#cb302-8" tabindex="-1"></a>              <span class="at">color =</span> <span class="st">&#39;red&#39;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-158-1.png" width="384" /></p>
<p>Better? I would say so… To determine the accuracy, we can compute the so called <strong>root mean squared error (RMSE)</strong>:</p>
<p><span class="math display">\[
RMSE = \sqrt{\frac{1}{n}\sum_{i=1}^n (x_i-\hat{x_i})^2}
\]</span></p>
<div class="sourceCode" id="cb303"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb303-1"><a href="week-one.html#cb303-1" tabindex="-1"></a>n <span class="ot">=</span> <span class="fu">nrow</span>(dat)</span>
<span id="cb303-2"><a href="week-one.html#cb303-2" tabindex="-1"></a>rmse <span class="ot">=</span> <span class="fu">sqrt</span>(<span class="dv">1</span><span class="sc">/</span>n<span class="sc">*</span><span class="fu">sum</span>(l.less<span class="sc">$</span>residuals<span class="sc">^</span><span class="dv">2</span>))</span>
<span id="cb303-3"><a href="week-one.html#cb303-3" tabindex="-1"></a>rmse</span></code></pre></div>
<pre><code>## [1] 41.01479</code></pre>
<p>Of course, this is cheating, right? We are predicting the values on <strong>exactly</strong> the same data we used to learn the model. In real machine learning, we need to perform <strong>cross-validation</strong>, i.e. learn the model on one part of the data (<em>training set</em>) and validate it on another set (<em>test set</em>).</p>
<p>Let us split the dataset in a test and training set randomly:</p>
<div class="sourceCode" id="cb305"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb305-1"><a href="week-one.html#cb305-1" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1234</span>)</span>
<span id="cb305-2"><a href="week-one.html#cb305-2" tabindex="-1"></a><span class="do">## take 200 random patients to form the training set</span></span>
<span id="cb305-3"><a href="week-one.html#cb305-3" tabindex="-1"></a>i.train <span class="ot">=</span> <span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(dat),<span class="dv">100</span>)</span>
<span id="cb305-4"><a href="week-one.html#cb305-4" tabindex="-1"></a><span class="do">##</span></span>
<span id="cb305-5"><a href="week-one.html#cb305-5" tabindex="-1"></a>dat.train <span class="ot">=</span> dat[i.train,]</span>
<span id="cb305-6"><a href="week-one.html#cb305-6" tabindex="-1"></a>dat.test <span class="ot">=</span> dat[<span class="sc">-</span>i.train,]</span></code></pre></div>
<p>We now learn a new model on the train dataset:</p>
<div class="sourceCode" id="cb306"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb306-1"><a href="week-one.html#cb306-1" tabindex="-1"></a>l.train <span class="ot">=</span> <span class="fu">lm</span>(chol <span class="sc">~</span> stab.glu <span class="sc">+</span> hdl <span class="sc">+</span> glyhb <span class="sc">+</span> age <span class="sc">+</span> height <span class="sc">+</span> weight  <span class="sc">+</span> bp<span class="fl">.1</span>s <span class="sc">+</span> bp<span class="fl">.1</span>d,<span class="at">data=</span>dat.train)</span>
<span id="cb306-2"><a href="week-one.html#cb306-2" tabindex="-1"></a><span class="fu">summary</span>(l.train)</span></code></pre></div>
<pre><code>## 
## Call:
## lm(formula = chol ~ stab.glu + hdl + glyhb + age + height + weight + 
##     bp.1s + bp.1d, data = dat.train)
## 
## Residuals:
##     Min      1Q  Median      3Q     Max 
## -76.806 -28.259  -6.705  22.934 158.094 
## 
## Coefficients:
##              Estimate Std. Error t value Pr(&gt;|t|)   
## (Intercept) 214.29240   95.18482   2.251  0.02677 * 
## stab.glu     -0.04772    0.13938  -0.342  0.73288   
## hdl           0.24900    0.31841   0.782  0.43624   
## glyhb        10.14338    3.22686   3.143  0.00225 **
## age           0.41957    0.39691   1.057  0.29327   
## height       -1.11670    1.26775  -0.881  0.38072   
## weight       -0.11006    0.12823  -0.858  0.39298   
## bp.1s        -0.12595    0.28635  -0.440  0.66110   
## bp.1d         0.31732    0.50740   0.625  0.53328   
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## Residual standard error: 46 on 91 degrees of freedom
## Multiple R-squared:  0.2749, Adjusted R-squared:  0.2111 
## F-statistic: 4.312 on 8 and 91 DF,  p-value: 0.0001965</code></pre>
<p>We can compute the RMSE for the training dataset</p>
<div class="sourceCode" id="cb308"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb308-1"><a href="week-one.html#cb308-1" tabindex="-1"></a>n.train <span class="ot">=</span> <span class="fu">nrow</span>(dat.train)</span>
<span id="cb308-2"><a href="week-one.html#cb308-2" tabindex="-1"></a>rmse.train <span class="ot">=</span> <span class="fu">sqrt</span>(<span class="dv">1</span><span class="sc">/</span>n.train<span class="sc">*</span><span class="fu">sum</span>(l.train<span class="sc">$</span>residuals<span class="sc">^</span><span class="dv">2</span>))</span>
<span id="cb308-3"><a href="week-one.html#cb308-3" tabindex="-1"></a>rmse.train</span></code></pre></div>
<pre><code>## [1] 43.88137</code></pre>
<p>Let use use that model to predict cholesterol values for the left out test set</p>
<div class="sourceCode" id="cb310"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb310-1"><a href="week-one.html#cb310-1" tabindex="-1"></a>pred <span class="ot">=</span> <span class="fu">predict</span>(l.train, <span class="at">newdata =</span> dat.test)</span></code></pre></div>
<p>and compute the rmse:</p>
<div class="sourceCode" id="cb311"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb311-1"><a href="week-one.html#cb311-1" tabindex="-1"></a>n.test <span class="ot">=</span> <span class="fu">nrow</span>(dat.test)</span>
<span id="cb311-2"><a href="week-one.html#cb311-2" tabindex="-1"></a>residuals <span class="ot">=</span> dat.test<span class="sc">$</span>chol <span class="sc">-</span> pred</span>
<span id="cb311-3"><a href="week-one.html#cb311-3" tabindex="-1"></a>rmse.test <span class="ot">=</span> <span class="fu">sqrt</span>(<span class="dv">1</span><span class="sc">/</span>n.test<span class="sc">*</span><span class="fu">sum</span>(residuals<span class="sc">^</span><span class="dv">2</span>))</span>
<span id="cb311-4"><a href="week-one.html#cb311-4" tabindex="-1"></a>rmse.test</span></code></pre></div>
<pre><code>## [1] 34.75285</code></pre>
<p>Of course, the RMSE is higher on the test dataset, since this does not include the data used for the establishment of the regression model; however, this is a more realistic estimation of the validity of the model, as it indicates how well the model could be extended to novel, independent data!</p>
<p>An important topic here is <strong>feature selection</strong>, i.e. finding the optimal and minimal set of explanatory variables that allow to predict well the output variable.</p>
<p>We now repeat the train/test split 10 times with each time a different random split; plot the 10 <code>rmse.train</code> and <code>rmse.test</code> values!</p>
<div class="sourceCode" id="cb313"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb313-1"><a href="week-one.html#cb313-1" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">345</span>)</span>
<span id="cb313-2"><a href="week-one.html#cb313-2" tabindex="-1"></a>RMSE <span class="ot">&lt;-</span> <span class="fu">sapply</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="cf">function</span>(x) {</span>
<span id="cb313-3"><a href="week-one.html#cb313-3" tabindex="-1"></a>  i.train <span class="ot">=</span> <span class="fu">sample</span> (<span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(dat),<span class="dv">100</span>)</span>
<span id="cb313-4"><a href="week-one.html#cb313-4" tabindex="-1"></a>  <span class="do">##</span></span>
<span id="cb313-5"><a href="week-one.html#cb313-5" tabindex="-1"></a>  dat.train <span class="ot">=</span> dat[i.train,]</span>
<span id="cb313-6"><a href="week-one.html#cb313-6" tabindex="-1"></a>  dat.test <span class="ot">=</span> dat[<span class="sc">-</span>i.train,]</span>
<span id="cb313-7"><a href="week-one.html#cb313-7" tabindex="-1"></a>  <span class="do">##</span></span>
<span id="cb313-8"><a href="week-one.html#cb313-8" tabindex="-1"></a>  l.train <span class="ot">=</span> <span class="fu">lm</span>(chol <span class="sc">~</span> stab.glu <span class="sc">+</span> hdl <span class="sc">+</span> glyhb <span class="sc">+</span> age <span class="sc">+</span> height <span class="sc">+</span> weight  <span class="sc">+</span> bp<span class="fl">.1</span>s <span class="sc">+</span> bp<span class="fl">.1</span>d,<span class="at">data=</span>dat.train)</span>
<span id="cb313-9"><a href="week-one.html#cb313-9" tabindex="-1"></a>  <span class="do">##</span></span>
<span id="cb313-10"><a href="week-one.html#cb313-10" tabindex="-1"></a>  n.train <span class="ot">=</span> <span class="fu">nrow</span>(dat.train)</span>
<span id="cb313-11"><a href="week-one.html#cb313-11" tabindex="-1"></a>  rmse.train <span class="ot">=</span> <span class="fu">sqrt</span>(<span class="dv">1</span><span class="sc">/</span>n.train<span class="sc">*</span><span class="fu">sum</span>(l.train<span class="sc">$</span>residuals<span class="sc">^</span><span class="dv">2</span>))</span>
<span id="cb313-12"><a href="week-one.html#cb313-12" tabindex="-1"></a>  <span class="do">##</span></span>
<span id="cb313-13"><a href="week-one.html#cb313-13" tabindex="-1"></a>  pred <span class="ot">=</span> <span class="fu">predict</span>(l.train,<span class="at">newdata =</span> dat.test)</span>
<span id="cb313-14"><a href="week-one.html#cb313-14" tabindex="-1"></a>  <span class="do">##</span></span>
<span id="cb313-15"><a href="week-one.html#cb313-15" tabindex="-1"></a>  n.test <span class="ot">=</span> <span class="fu">nrow</span>(dat.test)</span>
<span id="cb313-16"><a href="week-one.html#cb313-16" tabindex="-1"></a>  residuals <span class="ot">=</span> dat.test<span class="sc">$</span>chol <span class="sc">-</span> pred</span>
<span id="cb313-17"><a href="week-one.html#cb313-17" tabindex="-1"></a>  rmse.test <span class="ot">=</span> <span class="fu">sqrt</span>(<span class="dv">1</span><span class="sc">/</span>n.test<span class="sc">*</span><span class="fu">sum</span>(residuals<span class="sc">^</span><span class="dv">2</span>))</span>
<span id="cb313-18"><a href="week-one.html#cb313-18" tabindex="-1"></a>  RMSE <span class="ot">&lt;-</span> <span class="fu">c</span>(rmse.train,rmse.test)</span>
<span id="cb313-19"><a href="week-one.html#cb313-19" tabindex="-1"></a>  RMSE</span>
<span id="cb313-20"><a href="week-one.html#cb313-20" tabindex="-1"></a>})</span>
<span id="cb313-21"><a href="week-one.html#cb313-21" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb313-22"><a href="week-one.html#cb313-22" tabindex="-1"></a></span>
<span id="cb313-23"><a href="week-one.html#cb313-23" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb313-24"><a href="week-one.html#cb313-24" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">y =</span> RMSE[<span class="dv">1</span>,],</span>
<span id="cb313-25"><a href="week-one.html#cb313-25" tabindex="-1"></a>                 <span class="at">x =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="at">colour =</span> <span class="st">&quot;gold&quot;</span>)) <span class="sc">+</span></span>
<span id="cb313-26"><a href="week-one.html#cb313-26" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">y =</span> RMSE[<span class="dv">2</span>,],</span>
<span id="cb313-27"><a href="week-one.html#cb313-27" tabindex="-1"></a>                 <span class="at">x =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="at">colour =</span> <span class="st">&quot;steelblue&quot;</span>)) <span class="sc">+</span></span>
<span id="cb313-28"><a href="week-one.html#cb313-28" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">breaks=</span><span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">4</span>,<span class="dv">6</span>,<span class="dv">8</span>,<span class="dv">10</span>)) <span class="sc">+</span></span>
<span id="cb313-29"><a href="week-one.html#cb313-29" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="fu">mean</span>(RMSE[<span class="dv">1</span>,]),</span>
<span id="cb313-30"><a href="week-one.html#cb313-30" tabindex="-1"></a>             <span class="at">colour =</span> <span class="st">&quot;gold&quot;</span>) <span class="sc">+</span></span>
<span id="cb313-31"><a href="week-one.html#cb313-31" tabindex="-1"></a>  <span class="fu">ylim</span>(<span class="fu">min</span>(RMSE), <span class="fu">max</span>(RMSE <span class="sc">+</span><span class="dv">2</span>)) <span class="sc">+</span></span>
<span id="cb313-32"><a href="week-one.html#cb313-32" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="fu">mean</span>(RMSE[<span class="dv">2</span>,]),</span>
<span id="cb313-33"><a href="week-one.html#cb313-33" tabindex="-1"></a>             <span class="at">colour =</span> <span class="st">&quot;steelblue&quot;</span>) <span class="sc">+</span></span>
<span id="cb313-34"><a href="week-one.html#cb313-34" tabindex="-1"></a>  <span class="fu">scale_colour_manual</span>(<span class="at">name =</span> <span class="st">&quot;RMSE&quot;</span>, <span class="at">guide =</span> <span class="st">&#39;legend&#39;</span>,</span>
<span id="cb313-35"><a href="week-one.html#cb313-35" tabindex="-1"></a>                      <span class="at">values =</span><span class="fu">c</span>(<span class="st">&#39;gold&#39;</span><span class="ot">=</span><span class="st">&#39;gold&#39;</span>,<span class="st">&#39;steelblue&#39;</span><span class="ot">=</span><span class="st">&#39;steelblue&#39;</span>), </span>
<span id="cb313-36"><a href="week-one.html#cb313-36" tabindex="-1"></a>                      <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">&#39;train&#39;</span>,<span class="st">&#39;test&#39;</span>)) <span class="sc">+</span></span>
<span id="cb313-37"><a href="week-one.html#cb313-37" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&quot;Iteration&quot;</span>, <span class="at">y =</span> <span class="st">&quot;RMSE values&quot;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-165-1.png" width="672" /></p>

</div>
</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="index.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="week-two.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/USERNAME/REPO/edit/BRANCH/01-week.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["_main.pdf", "_main.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
